// @webiny/app-headless-cms/admin/plugins/fieldRenderers/boolean/booleanSwitch.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import get from "lodash/get";
import { Switch } from "@webiny/ui/Switch";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/fields/boolean");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-boolean",
  renderer: {
    rendererName: "boolean-input",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Boolean Input"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders a simple switch button."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "boolean" && !field.multipleValues && !get(field, "predefinedValues.enabled");
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind;
      var Bind = getBind();
      return /*#__PURE__*/React.createElement(Bind, null, function (bindProps) {
        return /*#__PURE__*/React.createElement(Switch, Object.assign({}, bindProps, {
          label: field.label,
          description: field.helpText
        }));
      });
    }
  }
};
export default plugin;
