// @webiny/app-form-builder/admin/components/FormEditor/Tabs/EditTab/EditFieldDialog/FieldTypeSelector.js
import React from "react";
import { Elevation } from "@webiny/ui/Elevation";
import { Typography } from "@webiny/ui/Typography";
import { Icon } from "@webiny/ui/Icon";
import { css } from "emotion";
var fieldTypeBox = /*#__PURE__*/css({
  width: 150,
  height: 150,
  textAlign: "center",
  margin: 20,
  padding: 15,
  display: "flex",
  justifyContent: "flex-start",
  alignItems: "center",
  cursor: "pointer",
  transition: "opacity 225ms",
  flexDirection: "column",
  backgroundColor: "var(--mdc-theme-surface) !important",
  ".webiny-ui-icon": {
    color: "var(--mdc-theme-secondary)",
    height: 50,
    transition: "color 225ms"
  },
  "&:hover": {
    opacity: 0.8,
    ".webiny-ui-icon": {
      color: "var(--mdc-theme-primary)"
    }
  }
}, "label:fieldTypeBox;");

var FbFormModelFieldSelector = function FbFormModelFieldSelector(_ref) {
  var fieldType = _ref.fieldType,
      onClick = _ref.onClick;
  return /*#__PURE__*/React.createElement("span", {
    onClick: onClick,
    style: {
      display: "inline-block"
    }
  }, /*#__PURE__*/React.createElement(Elevation, {
    z: 2,
    className: fieldTypeBox
  }, /*#__PURE__*/React.createElement(Icon, {
    icon: fieldType.icon
  }), /*#__PURE__*/React.createElement(Typography, {
    use: "headline5"
  }, fieldType.label), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement(Typography, {
    use: "caption"
  }, fieldType.description)));
};

export default FbFormModelFieldSelector;
