// @webiny/app-form-builder/admin/components/FormEditor/Tabs/EditTab/EditFieldDialog/GeneralTab.js
import _camelCase from "lodash/camelCase";
import React, { useCallback } from "react";
import { Input } from "@webiny/ui/Input";
import { Grid, Cell } from "@webiny/ui/Grid";
import { useFormEditor } from "../../../Context";
import { validation } from "@webiny/validation";

var GeneralTab = function GeneralTab(_ref) {
  var field = _ref.field,
      form = _ref.form;
  var Bind = form.Bind,
      setValue = form.setValue;

  var _useFormEditor = useFormEditor(),
      getField = _useFormEditor.getField,
      getFieldPlugin = _useFormEditor.getFieldPlugin;

  var afterChangeLabel = useCallback(function (value) {
    setValue("fieldId", _camelCase(value));
  }, []);
  var uniqueFieldIdValidator = useCallback(function (fieldId) {
    var existingField = getField({
      fieldId: fieldId
    });

    if (!existingField) {
      return;
    }

    if (existingField._id === field._id) {
      return true;
    }

    throw new Error("Please enter a unique Field ID");
  }, undefined);
  var fieldPlugin = getFieldPlugin({
    name: field.name
  });
  var additionalSettings = null;

  if (typeof fieldPlugin.field.renderSettings === "function") {
    additionalSettings = fieldPlugin.field.renderSettings({
      form: form,
      afterChangeLabel: afterChangeLabel,
      uniqueFieldIdValidator: uniqueFieldIdValidator
    });
  }

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
    span: 6
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "label",
    validators: validation.create("required"),
    afterChange: afterChangeLabel
  }, /*#__PURE__*/React.createElement(Input, {
    label: "Label",
    autoFocus: true
  }))), /*#__PURE__*/React.createElement(Cell, {
    span: 6
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "fieldId",
    validators: [validation.create("required"), uniqueFieldIdValidator]
  }, /*#__PURE__*/React.createElement(Input, {
    label: "Field ID"
  }))), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "helpText"
  }, /*#__PURE__*/React.createElement(Input, {
    label: "Help text",
    description: "Help text (optional)"
  })))), additionalSettings);
};

export default GeneralTab;
