// @webiny/app-page-builder/editor/plugins/toolbar/onboarding/slides/Slide5.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import { css } from "emotion";
import { keyframes } from "emotion";
import classnames from "classnames";
import SlideOutline from "./SlideOutline";
import { ReactComponent as BlocksIcon } from "@svgr/webpack!../assets/slide-5-icon.svg";
import { ReactComponent as HandIcon } from "@svgr/webpack!../assets/hand-icon.svg";
import { ReactComponent as BlockMock } from "@svgr/webpack!../assets/block-mock.svg";
import { ReactComponent as Hover } from "@svgr/webpack!../assets/slide-4-hover.svg";
import { ReactComponent as ElementList } from "@svgr/webpack!../assets/slide-5-element-list.svg";
var handPulsateSlide5 = /*#__PURE__*/keyframes("0%{display:none;visibility:hidden;opacity:0;}50%{display:none;visibility:hidden;opacity:0;}100%{display:block;visibility:block;opacity:1;}label:handPulsateSlide5;");
var elementListAppearSlide5 = /*#__PURE__*/keyframes("0%{opacity:0;}90%{opacity:0;transform:translateX(-25px);}100%{transform:translateX(0px);opacity:1;}label:elementListAppearSlide5;");
var svgElement = /*#__PURE__*/css({
  display: "none",
  position: "absolute"
}, "label:svgElement;");
var handStylesSlide5 = /*#__PURE__*/css({
  display: "block",
  left: 38,
  top: 36,
  width: 25,
  height: "auto",
  zIndex: 10,
  animationName: handPulsateSlide5,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "2s"
}, "label:handStylesSlide5;");
var blockMockStyleSlide5 = /*#__PURE__*/css({
  position: "absolute",
  display: "block",
  left: 115,
  top: 85
}, "label:blockMockStyleSlide5;");
var hoverStyleSlide5 = /*#__PURE__*/css({
  position: "absolute",
  left: 110,
  top: 81,
  display: "block"
}, "label:hoverStyleSlide5;");
var elementListStyleSlide5 = /*#__PURE__*/css({
  left: 43,
  top: 32,
  display: "block",
  animationName: elementListAppearSlide5,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "2s"
}, "label:elementListStyleSlide5;");

var Slide5 = /*#__PURE__*/function (_React$Component) {
  _inherits(Slide5, _React$Component);

  var _super = _createSuper(Slide5);

  function Slide5() {
    var _this;

    _classCallCheck(this, Slide5);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "slideNumber", 5);

    _defineProperty(_assertThisInitialized(_this), "editorContent", function () {
      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(HandIcon, {
        className: classnames(svgElement, _defineProperty({}, handStylesSlide5, _this.props.currentSlide === _this.slideNumber))
      }), /*#__PURE__*/React.createElement(BlockMock, {
        className: blockMockStyleSlide5
      }), /*#__PURE__*/React.createElement(Hover, {
        className: hoverStyleSlide5
      }), /*#__PURE__*/React.createElement(ElementList, {
        className: classnames(svgElement, _defineProperty({}, elementListStyleSlide5, _this.props.currentSlide === _this.slideNumber))
      }));
    });

    return _this;
  }

  _createClass(Slide5, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(SlideOutline, {
        title: "Content",
        description: "Content resides inside columns. You can add different types of content, such as text, \n" + "images, videos, buttons, and many more. Using those basic elements you can \n" + "create a whole plethora of different content.",
        learnMoreText: "Click here to learn more about Content.",
        learnMoreLink: "https://docs.webiny.com/cms/content",
        icon: /*#__PURE__*/React.createElement(BlocksIcon, null),
        editorContent: this.editorContent()
      });
    }
  }]);

  return Slide5;
}(React.Component);

export default Slide5;
