// @webiny/app-page-builder/editor/plugins/toolbar/onboarding/slides/Slide2.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import { css } from "emotion";
import { keyframes } from "emotion";
import classnames from "classnames";
import SlideOutline from "./SlideOutline";
import { ReactComponent as BlocksIcon } from "@svgr/webpack!../assets/slide-2-icon.svg";
import { ReactComponent as HandIcon } from "@svgr/webpack!../assets/hand-icon.svg";
import { ReactComponent as BlockMock } from "@svgr/webpack!../assets/block-mock.svg";
import { ReactComponent as Hover } from "@svgr/webpack!../assets/slide-2-hover.svg";
var handPulsateSlide2 = /*#__PURE__*/keyframes("0%{display:none;visibility:hidden;opacity:0;}100%{display:block;visibility:block;opacity:1;}label:handPulsateSlide2;");
var blockAppearSlide2 = /*#__PURE__*/keyframes("0%{transform:translateY(-100px);opacity:0;}100%{transform:translateY(0px);opacity:1;}label:blockAppearSlide2;");
var hoverAppearSlide2 = /*#__PURE__*/keyframes("0%{opacity:0;}66%{opacity:0;}100%{opacity:1;}label:hoverAppearSlide2;");
var svgElement = /*#__PURE__*/css({
  display: "none",
  position: "absolute"
}, "label:svgElement;");
var handStylesSlide2 = /*#__PURE__*/css({
  display: "block",
  right: 49,
  bottom: -14,
  width: 50,
  height: "auto",
  zIndex: 10,
  animationName: handPulsateSlide2,
  animationIterationCount: "infinite",
  animationTimingFunction: "ease-in",
  animationDuration: "3s"
}, "label:handStylesSlide2;");
var blockMockStyleSlide2 = /*#__PURE__*/css({
  display: "block",
  left: 115,
  top: 85,
  animationName: blockAppearSlide2,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "1s"
}, "label:blockMockStyleSlide2;");
var hoverStyleSlide2 = /*#__PURE__*/css({
  left: 75,
  top: 55,
  display: "block",
  animationName: hoverAppearSlide2,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "2s"
}, "label:hoverStyleSlide2;");

var Slide2 = /*#__PURE__*/function (_React$Component) {
  _inherits(Slide2, _React$Component);

  var _super = _createSuper(Slide2);

  function Slide2() {
    var _this;

    _classCallCheck(this, Slide2);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "slideNumber", 2);

    _defineProperty(_assertThisInitialized(_this), "editorContent", function () {
      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(HandIcon, {
        className: classnames(svgElement, _defineProperty({}, handStylesSlide2, _this.props.currentSlide === _this.slideNumber))
      }), /*#__PURE__*/React.createElement(BlockMock, {
        className: classnames(svgElement, _defineProperty({}, blockMockStyleSlide2, _this.props.currentSlide === _this.slideNumber))
      }), /*#__PURE__*/React.createElement(Hover, {
        className: classnames(svgElement, _defineProperty({}, hoverStyleSlide2, _this.props.currentSlide === _this.slideNumber))
      }));
    });

    return _this;
  }

  _createClass(Slide2, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(SlideOutline, {
        title: "Blocks",
        description: "Blocks, are the main sections of your page inside which all content is placed.\n" + "To add a block, click the green button in the lower right corner.\n" + "You can start with an empty block. or use one of the pre-built ones.",
        learnMoreText: "Click here to learn more about Blocks.",
        learnMoreLink: "https://docs.webiny.com/cms/blocks",
        icon: /*#__PURE__*/React.createElement(BlocksIcon, null),
        editorContent: this.editorContent()
      });
    }
  }]);

  return Slide2;
}(React.Component);

export default Slide2;
