// @webiny/app-page-builder/editor/plugins/toolbar/onboarding/slides/Slide3.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import { css } from "emotion";
import { keyframes } from "emotion";
import classnames from "classnames";
import SlideOutline from "./SlideOutline";
import { ReactComponent as BlocksIcon } from "@svgr/webpack!../assets/slide-3-icon.svg";
import { ReactComponent as HandIcon } from "@svgr/webpack!../assets/hand-icon.svg";
import { ReactComponent as BlockMock } from "@svgr/webpack!../assets/block-mock.svg";
import { ReactComponent as Hover } from "@svgr/webpack!../assets/slide-3-hover.svg";
import { ReactComponent as ElementList } from "@svgr/webpack!../assets/slide-3-element-list.svg";
var handPulsateSlide3 = /*#__PURE__*/keyframes("0%{display:none;visibility:hidden;opacity:0;}50%{display:none;visibility:hidden;opacity:0;}100%{display:block;visibility:block;opacity:1;}label:handPulsateSlide3;");
var elementListAppearSlide3 = /*#__PURE__*/keyframes("0%{opacity:0;}90%{opacity:0;transform:translateX(-25px);}100%{transform:translateX(0px);opacity:1;}label:elementListAppearSlide3;");
var hoverAppearSlide3 = /*#__PURE__*/keyframes("0%{opacity:0;}66%{opacity:0;}100%{opacity:1;}label:hoverAppearSlide3;");
var svgElement = /*#__PURE__*/css({
  display: "none",
  position: "absolute"
}, "label:svgElement;");
var handStylesSlide3 = /*#__PURE__*/css({
  display: "block",
  left: 38,
  top: 36,
  width: 25,
  height: "auto",
  zIndex: 10,
  animationName: handPulsateSlide3,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "4s"
}, "label:handStylesSlide3;");
var blockMockStyleSlide3 = /*#__PURE__*/css({
  position: "absolute",
  display: "block",
  left: 115,
  top: 85
}, "label:blockMockStyleSlide3;");
var hoverStyleSlide3 = /*#__PURE__*/css({
  left: 100,
  top: 70,
  display: "block",
  animationName: hoverAppearSlide3,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "2s"
}, "label:hoverStyleSlide3;");
var elementListStyleSlide3 = /*#__PURE__*/css({
  left: 43,
  top: 32,
  display: "block",
  animationName: elementListAppearSlide3,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "4s"
}, "label:elementListStyleSlide3;");

var Slide3 = /*#__PURE__*/function (_React$Component) {
  _inherits(Slide3, _React$Component);

  var _super = _createSuper(Slide3);

  function Slide3() {
    var _this;

    _classCallCheck(this, Slide3);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "slideNumber", 3);

    _defineProperty(_assertThisInitialized(_this), "editorContent", function () {
      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(HandIcon, {
        className: classnames(svgElement, _defineProperty({}, handStylesSlide3, _this.props.currentSlide === _this.slideNumber))
      }), /*#__PURE__*/React.createElement(BlockMock, {
        className: blockMockStyleSlide3
      }), /*#__PURE__*/React.createElement(Hover, {
        className: classnames(svgElement, _defineProperty({}, hoverStyleSlide3, _this.props.currentSlide === _this.slideNumber))
      }), /*#__PURE__*/React.createElement(ElementList, {
        className: classnames(svgElement, _defineProperty({}, elementListStyleSlide3, _this.props.currentSlide === _this.slideNumber))
      }));
    });

    return _this;
  }

  _createClass(Slide3, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(SlideOutline, {
        title: "Rows",
        description: "Inside each block, you have one or more  rows. Rows are used to spread content\n" + "vertically. To add a new row, click the plus icon in the top left corner, \n" + "you will find rows under Layouts section.",
        learnMoreText: "Click here to learn more about Rows.",
        learnMoreLink: "https://docs.webiny.com/cms/rows",
        icon: /*#__PURE__*/React.createElement(BlocksIcon, null),
        editorContent: this.editorContent()
      });
    }
  }]);

  return Slide3;
}(React.Component);

export default Slide3;
