// @webiny/app-page-builder/editor/plugins/toolbar/onboarding/slides/Slide6.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import { css } from "emotion";
import { keyframes } from "emotion";
import classnames from "classnames";
import SlideOutline from "./SlideOutline";
import { ReactComponent as BlocksIcon } from "@svgr/webpack!../assets/slide-6-icon.svg";
import { ReactComponent as HandIcon } from "@svgr/webpack!../assets/hand-icon.svg";
import { ReactComponent as BlockMock } from "@svgr/webpack!../assets/block-mock.svg";
import { ReactComponent as Hover } from "@svgr/webpack!../assets/slide-6-hover.svg";
import { ReactComponent as Dialog } from "@svgr/webpack!../assets/slide-6-dialog.svg";
import { ReactComponent as Toolbar } from "@svgr/webpack!../assets/slide-6-toolbar.svg";
var handPulsateSlide6 = /*#__PURE__*/keyframes("0%{display:none;visibility:hidden;opacity:0;}50%{display:none;visibility:hidden;opacity:0;}100%{display:block;visibility:block;opacity:1;}label:handPulsateSlide6;");
var toolbarListAppearSlide6 = /*#__PURE__*/keyframes("0%{opacity:0;}50%{opacity:0;transform:translateY(-25px);}100%{transform:translateY(0px);opacity:1;}label:toolbarListAppearSlide6;");
var hoverHightlightSlide6 = /*#__PURE__*/keyframes("0%{opacity:0;}50%{opacity:0;}100%{opacity:1;}label:hoverHightlightSlide6;");
var svgElementSlide6 = /*#__PURE__*/css({
  display: "none",
  position: "absolute"
}, "label:svgElementSlide6;");
var handStylesSlide6 = /*#__PURE__*/css({
  display: "block",
  left: 490,
  top: 170,
  width: 25,
  height: "auto",
  zIndex: 10,
  animationName: handPulsateSlide6,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "2s"
}, "label:handStylesSlide6;");
var blockMockStyleSlide6 = /*#__PURE__*/css({
  position: "absolute",
  display: "block",
  left: 115,
  top: 85
}, "label:blockMockStyleSlide6;");
var hoverStyleSlide6 = /*#__PURE__*/css({
  position: "absolute",
  left: 110,
  top: 81,
  display: "block"
}, "label:hoverStyleSlide6;");
var hoverStyleSlide6AnimInit = /*#__PURE__*/css({
  opacity: 0
}, "label:hoverStyleSlide6AnimInit;");
var hoverStyleSlide6Anim = /*#__PURE__*/css({
  ".highlighted-layer": {
    animationName: hoverHightlightSlide6,
    animationIterationCount: 1,
    animationTimingFunction: "ease-in",
    animationDuration: "1s"
  }
}, "label:hoverStyleSlide6Anim;");
var toolbarStyleSlide6 = /*#__PURE__*/css({
  left: 40,
  top: 1,
  display: "block",
  position: "absolute",
  animationName: toolbarListAppearSlide6,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "3s"
}, "label:toolbarStyleSlide6;");
var dialogStyleSlide6 = /*#__PURE__*/css({
  left: 394,
  top: 30,
  display: "block",
  position: "absolute",
  animationName: toolbarListAppearSlide6,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "4s"
}, "label:dialogStyleSlide6;");

var Slide6 = /*#__PURE__*/function (_React$Component) {
  _inherits(Slide6, _React$Component);

  var _super = _createSuper(Slide6);

  function Slide6() {
    var _this;

    _classCallCheck(this, Slide6);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "slideNumber", 6);

    _defineProperty(_assertThisInitialized(_this), "editorContent", function () {
      var _classnames2;

      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(HandIcon, {
        className: classnames(svgElementSlide6, _defineProperty({}, handStylesSlide6, _this.props.currentSlide === _this.slideNumber))
      }), /*#__PURE__*/React.createElement(BlockMock, {
        className: blockMockStyleSlide6
      }), /*#__PURE__*/React.createElement(Hover, {
        className: classnames(hoverStyleSlide6, (_classnames2 = {}, _defineProperty(_classnames2, hoverStyleSlide6Anim, _this.props.currentSlide === _this.slideNumber), _defineProperty(_classnames2, hoverStyleSlide6AnimInit, _this.props.currentSlide !== _this.slideNumber), _classnames2))
      }), /*#__PURE__*/React.createElement(Toolbar, {
        className: classnames(svgElementSlide6, _defineProperty({}, toolbarStyleSlide6, _this.props.currentSlide === _this.slideNumber))
      }), /*#__PURE__*/React.createElement(Dialog, {
        className: classnames(svgElementSlide6, _defineProperty({}, dialogStyleSlide6, _this.props.currentSlide === _this.slideNumber))
      }));
    });

    return _this;
  }

  _createClass(Slide6, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(SlideOutline, {
        title: "Style & Positioning",
        description: "When you hover over a layer, click on it to select it. On the top of the page, you will see \n" + "additional options to control the visual styles of that layer.\n" + "Layers can also be dragged and dropped, to change their position on the page.",
        learnMoreText: "Click here to learn more about Style & Positioning.",
        learnMoreLink: "https://docs.webiny.com/cms/style-and-positioning",
        icon: /*#__PURE__*/React.createElement(BlocksIcon, null),
        editorContent: this.editorContent()
      });
    }
  }]);

  return Slide6;
}(React.Component);

export default Slide6;
