// @webiny/app-page-builder/editor/plugins/toolbar/onboarding/slides/Slide4.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import { css } from "emotion";
import { keyframes } from "emotion";
import classnames from "classnames";
import SlideOutline from "./SlideOutline";
import { ReactComponent as BlocksIcon } from "@svgr/webpack!../assets/slide-4-icon.svg";
import { ReactComponent as BlockMock } from "@svgr/webpack!../assets/block-mock.svg";
import { ReactComponent as Hover } from "@svgr/webpack!../assets/slide-4-hover.svg";
var hoverAppearSlide4 = /*#__PURE__*/keyframes("0%{opacity:0;}66%{opacity:0;}100%{opacity:1;}label:hoverAppearSlide4;");
var svgElement = /*#__PURE__*/css({
  display: "none",
  position: "absolute"
}, "label:svgElement;");
var blockMockStyleSlide4 = /*#__PURE__*/css({
  position: "absolute",
  display: "block",
  left: 115,
  top: 85
}, "label:blockMockStyleSlide4;");
var hoverStyleSlide4 = /*#__PURE__*/css({
  left: 110,
  top: 81,
  display: "block",
  animationName: hoverAppearSlide4,
  animationIterationCount: 1,
  animationTimingFunction: "ease-in",
  animationDuration: "2s"
}, "label:hoverStyleSlide4;");

var Slide4 = /*#__PURE__*/function (_React$Component) {
  _inherits(Slide4, _React$Component);

  var _super = _createSuper(Slide4);

  function Slide4() {
    var _this;

    _classCallCheck(this, Slide4);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "slideNumber", 4);

    _defineProperty(_assertThisInitialized(_this), "editorContent", function () {
      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(BlockMock, {
        className: blockMockStyleSlide4
      }), /*#__PURE__*/React.createElement(Hover, {
        className: classnames(svgElement, _defineProperty({}, hoverStyleSlide4, _this.props.currentSlide === _this.slideNumber))
      }));
    });

    return _this;
  }

  _createClass(Slide4, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(SlideOutline, {
        title: "Columns",
        description: "Inside each row, you can have one or more columns stacked horizontally. You will \n" + "find columns under the same section a Rows.",
        learnMoreText: "Click here to learn more about Columns.",
        learnMoreLink: "https://docs.webiny.com/cms/columns",
        icon: /*#__PURE__*/React.createElement(BlocksIcon, null),
        editorContent: this.editorContent()
      });
    }
  }]);

  return Slide4;
}(React.Component);

export default Slide4;
