// @webiny/app-page-builder/editor/plugins/elements/social/twitter/index.js
import React from "react";
import styled from "@emotion/styled";
import kebabCase from "lodash/kebabCase";
import { Typography } from "@webiny/ui/Typography";
import { validation } from "@webiny/validation";
import { createEmbedPlugin, createEmbedSettingsPlugin } from "../../utils/oembed";
import placeholder from "./placeholder.jpg";
import { ReactComponent as LogoIcon } from "@svgr/webpack!./twitter-brands.svg";
import Accordion from "../../../elementSettings/components/Accordion";
import InputField from "../../../elementSettings/components/InputField";
import { ButtonContainer, SimpleButton } from "../../../elementSettings/components/StyledComponents";
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "e1v0vpu10",
  label: "PreviewBox"
})({
  textAlign: "center",
  height: 50,
  svg: {
    height: 50,
    width: 50
  }
});
export default (function () {
  var args = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var elementType = kebabCase(args.elementType || "twitter");
  var defaultToolbar = {
    title: "Tweet",
    group: "pb-editor-element-group-social",
    preview: function preview() {
      return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(LogoIcon, null));
    }
  };
  return [createEmbedPlugin({
    type: elementType,
    toolbar: typeof args.toolbar === "function" ? args.toolbar(defaultToolbar) : defaultToolbar,
    create: args.create,
    settings: args.settings,
    oembed: {
      global: "twttr",
      sdk: "https://platform.twitter.com/widgets.js",
      init: function init(_ref) {
        var node = _ref.node;
        window.twttr.widgets.load(node);
      }
    },
    renderElementPreview: function renderElementPreview(_ref2) {
      var width = _ref2.width,
          height = _ref2.height;
      return /*#__PURE__*/React.createElement("img", {
        style: {
          width: width,
          height: height
        },
        src: placeholder,
        alt: "Tweet"
      });
    }
  }), createEmbedSettingsPlugin({
    type: elementType,
    render: function render(_ref3) {
      var Bind = _ref3.Bind,
          submit = _ref3.submit;
      return /*#__PURE__*/React.createElement(Accordion, {
        title: "Twitter",
        defaultValue: true
      }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Bind, {
        name: "source.url",
        validators: validation.create("required,url")
      }, /*#__PURE__*/React.createElement(InputField, {
        placeholder: "https://twitter.com/",
        description: "Enter a Tweet URL"
      })), /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
        onClick: submit
      }, /*#__PURE__*/React.createElement(Typography, {
        use: "caption"
      }, "Save")))));
    }
  })];
});
