// @webiny/app-page-builder/admin/plugins/pageDetails/header/Header.js
import React from "react";
import { css } from "emotion";
import { renderPlugins } from "@webiny/app/plugins";
import { Typography } from "@webiny/ui/Typography";
import { Grid, Cell } from "@webiny/ui/Grid";
var headerTitle = /*#__PURE__*/css({
  "&.mdc-layout-grid": {
    borderBottom: "1px solid var(--mdc-theme-on-background)",
    color: "var(--mdc-theme-text-primary-on-background)",
    background: "var(--mdc-theme-surface)",
    paddingTop: 10,
    paddingBottom: 9,
    ".mdc-layout-grid__inner": {
      alignItems: "center"
    }
  }
}, "label:headerTitle;");
var pageTitle = /*#__PURE__*/css({
  whiteSpace: "nowrap",
  overflow: "hidden",
  textOverflow: "ellipsis"
}, "label:pageTitle;");
var headerActions = /*#__PURE__*/css({
  justifyContent: "flex-end",
  marginRight: "-15px",
  display: "flex",
  alignItems: "center"
}, "label:headerActions;");

var Header = function Header(props) {
  var page = props.page;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Grid, {
    className: headerTitle
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 8,
    className: pageTitle
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "headline5"
  }, page.title)), /*#__PURE__*/React.createElement(Cell, {
    span: 4,
    className: headerActions
  }, renderPlugins("pb-page-details-header-left", props), renderPlugins("pb-page-details-header-right", props))));
};

export default Header;
