// @webiny/app-form-builder/admin/plugins/editor/formFieldValidators/gte.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Input } from "@webiny/ui/Input";
import { validation } from "@webiny/validation";
var plugin = {
  type: "form-editor-field-validator",
  name: "form-editor-field-validator-gte",
  validator: {
    name: "gte",
    label: "Greater or equal",
    description: "Entered value must be equal or greater than the provided max value.",
    defaultMessage: "Value is too small.",
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.Bind;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "settings.value",
        validators: validation.create("required,numeric")
      }, /*#__PURE__*/React.createElement(Input, {
        type: "number",
        label: "Value",
        description: "This is the greatest value that will be allowed"
      }))));
    }
  }
};
export default plugin;
