// @webiny/app-page-builder/editor/plugins/elements/code/codesandbox/index.js
import React from "react";
import styled from "@emotion/styled";
import { validation } from "@webiny/validation";
import { Typography } from "@webiny/ui/Typography";
import { createEmbedPlugin, createEmbedSettingsPlugin } from "./../../utils/oembed/createEmbedPlugin";
import { ReactComponent as LogoIcon } from "@svgr/webpack!./codesandbox-logo.svg";
import Accordion from "../../../elementSettings/components/Accordion";
import InputField from "../../../elementSettings/components/InputField";
import { ButtonContainer, SimpleButton } from "../../../elementSettings/components/StyledComponents";
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "e18465i10",
  label: "PreviewBox"
})({
  textAlign: "center",
  height: 50,
  svg: {
    height: 50,
    width: 50
  }
});
export default (function () {
  return [createEmbedPlugin({
    type: "codesandbox",
    toolbar: {
      title: "CodeSandbox",
      group: "pb-editor-element-group-code",
      preview: function preview() {
        return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(LogoIcon, null));
      }
    },
    oembed: {
      onData: function onData(data) {
        data.html = data.html.replace(/1000px/g, "100%").replace(/1000/g, "100%");
        return data;
      }
    }
  }), createEmbedSettingsPlugin({
    type: "codesandbox",
    render: function render(_ref) {
      var Bind = _ref.Bind,
          submit = _ref.submit;
      return /*#__PURE__*/React.createElement(Accordion, {
        title: "CodeSandbox",
        defaultValue: true
      }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Bind, {
        name: "source.url",
        validators: validation.create("required,url")
      }, /*#__PURE__*/React.createElement(InputField, {
        placeholder: "https://codesandbox.io/",
        description: "Enter a CodeSandbox URL"
      })), /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
        onClick: submit
      }, /*#__PURE__*/React.createElement(Typography, {
        use: "caption"
      }, "Save")))));
    }
  })];
});
