// @webiny/app-admin/components/EmptyView.js
import React from "react";
import styled from "@emotion/styled";
import { Typography } from "@webiny/ui/Typography";
import { ReactComponent as TouchIcon } from "@svgr/webpack!../assets/icons/touch_app.svg";
var EmptyViewWrapper = /*#__PURE__*/styled("div", {
  target: "eu81oh40",
  label: "EmptyViewWrapper"
})({
  width: "100%",
  height: "80%",
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",
  "& .media": {
    "& svg": {
      width: 75,
      height: 75,
      fill: "var(--mdc-theme-text-icon-on-background)"
    },
    marginBottom: 30
  },
  "& .title__container": {
    maxWidth: 276,
    textAlign: "center",
    marginBottom: 24,
    "& .title": {
      color: "var(--mdc-theme-text-secondary-on-background)"
    }
  },
  "& .action__container": {
    "& .mdc-button": {
      padding: "0px 16px",
      backgroundColor: "var(--mdc-theme-surface)"
    }
  }
});

var EmptyView = function EmptyView(_ref) {
  var _ref$icon = _ref.icon,
      icon = _ref$icon === void 0 ? /*#__PURE__*/React.createElement(TouchIcon, null) : _ref$icon,
      title = _ref.title,
      action = _ref.action;
  return /*#__PURE__*/React.createElement(EmptyViewWrapper, null, /*#__PURE__*/React.createElement("div", {
    className: "media"
  }, icon), /*#__PURE__*/React.createElement("div", {
    className: "title__container"
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "subtitle1",
    className: "title"
  }, title)), /*#__PURE__*/React.createElement("div", {
    className: "action__container"
  }, action));
};

export default EmptyView;
