// @webiny/app-admin/components/SearchUI.js
import React from "react";
import styled from "@emotion/styled";
import InputField from "./SimpleUI/InputField";
import { ReactComponent as SearchIcon } from "@svgr/webpack!../assets/icons/search-24px.svg";
var SearchWrapper = /*#__PURE__*/styled("div", {
  target: "epa3kz70",
  label: "SearchWrapper"
})({
  width: "100%",
  height: "100%",
  display: "flex",
  position: "relative",
  "& .search__icon": {
    position: "absolute",
    top: 0,
    left: 0,
    bottom: 0,
    pointerEvents: "none",
    display: "flex",
    alignItems: "center",
    paddingLeft: "0.25rem",
    color: "var(--mdc-theme-text-secondary-on-background)"
  },
  "& .search__input": {
    width: "100%",
    padding: "0.5rem 0.5rem 0.5rem 2.25rem",
    outline: "none",
    border: "none",
    color: "var(--mdc-theme-text-primary-on-background)",
    "&:focus": {
      outline: "none"
    },
    "&::placeholder": {
      textTransform: "uppercase",
      letterSpacing: "0.05em",
      color: "var(--mdc-theme-text-secondary-on-background)"
    }
  }
});

var Search = function Search(_ref) {
  var value = _ref.value,
      onChange = _ref.onChange,
      _ref$inputPlaceholder = _ref.inputPlaceholder,
      inputPlaceholder = _ref$inputPlaceholder === void 0 ? "Search..." : _ref$inputPlaceholder;
  return /*#__PURE__*/React.createElement(SearchWrapper, {
    "data-testid": "default-data-list.search"
  }, /*#__PURE__*/React.createElement("div", {
    className: "search__icon"
  }, /*#__PURE__*/React.createElement(SearchIcon, null)), /*#__PURE__*/React.createElement(InputField, {
    className: "search__input",
    placeholder: inputPlaceholder,
    value: value,
    onChange: onChange,
    autoComplete: "off"
  }));
};

export default Search;
