// @webiny/api-security-admin-users/graphql/base.gql.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _plugins = require("@webiny/handler-graphql/plugins");

const emptyResolver = () => ({});

var _default = new _plugins.GraphQLSchemaPlugin({
  typeDefs: `
            type SecurityQuery {
                _empty: String
            }

            type SecurityMutation {
                _empty: String
            }

            extend type Query {
                security: SecurityQuery
            }

            extend type Mutation {
                security: SecurityMutation
            }
            
            type SecurityCreatedBy {
                id: ID
                displayName: String
            }

            type SecurityError {
                code: String
                message: String
                data: JSON
            }

            type SecurityBooleanResponse {
                data: Boolean
                error: SecurityError
            }
        `,
  resolvers: {
    Query: {
      security: emptyResolver
    },
    Mutation: {
      security: emptyResolver
    }
  }
});

exports.default = _default;
