// @webiny/api-security-admin-users/graphql/install.gql.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _responses = require("@webiny/handler-graphql/responses");

var _GraphQLSchemaPlugin = require("@webiny/handler-graphql/plugins/GraphQLSchemaPlugin");

var _default = new _GraphQLSchemaPlugin.GraphQLSchemaPlugin({
  typeDefs:
  /* GraphQL */
  `
        input SecurityInstallInput {
            firstName: String!
            lastName: String!
            login: String!
        }

        extend type SecurityQuery {
            "Get installed version"
            version: String
        }

        extend type SecurityMutation {
            "Install Security"
            install(data: SecurityInstallInput!): SecurityBooleanResponse
        }
    `,
  resolvers: {
    SecurityQuery: {
      version: async (root, args, context) => {
        return await context.security.system.getVersion();
      }
    },
    SecurityMutation: {
      install: async (root, args, context) => {
        try {
          await context.security.system.install(args.data);
          return new _responses.Response(true);
        } catch (e) {
          return new _responses.ErrorResponse(e);
        }
      }
    }
  }
});

exports.default = _default;
