// @webiny/app-page-builder/editor/plugins/elements/utils/oembed/createEmbedPlugin.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["content"];
import * as React from "react";
import OEmbed from "../../../../components/OEmbed";
import { DisplayMode } from "../../../../../types";
import { createInitialPerDeviceSettingValue } from "../../../elementSettings/elementSettingsUtils";
export var createEmbedPlugin = function createEmbedPlugin(config) {
  var defaultSettings = ["pb-editor-page-element-settings-delete"];
  return {
    name: "pb-editor-page-element-" + config.type,
    type: "pb-editor-page-element",
    elementType: config.type,
    toolbar: config.toolbar,
    settings: typeof config.settings === "function" ? config.settings(defaultSettings) : defaultSettings,
    target: config.target || ["cell", "block", "list-item"],
    // eslint-disable-next-line
    create: function create(_ref) {
      var _ref$content = _ref.content,
          content = _ref$content === void 0 ? {} : _ref$content,
          options = _objectWithoutProperties(_ref, _excluded);

      var defaultValue = _objectSpread({
        type: config.type,
        elements: [],
        data: {
          settings: {
            margin: createInitialPerDeviceSettingValue({
              all: "0px"
            }, DisplayMode.DESKTOP),
            padding: createInitialPerDeviceSettingValue({
              all: "0px"
            }, DisplayMode.DESKTOP)
          }
        }
      }, options);

      return typeof config.create === "function" ? config.create(defaultValue) : defaultValue;
    },
    render: function render(props) {
      if (config.render) {
        return config.render(props);
      }

      return /*#__PURE__*/React.createElement(OEmbed, Object.assign({
        element: props.element
      }, config.oembed || {}));
    },
    onCreate: config.onCreate || "open-settings",
    renderElementPreview: config.renderElementPreview
  };
};
export var createEmbedSettingsPlugin = function createEmbedSettingsPlugin(_ref2) {
  var type = _ref2.type,
      render = _ref2.render;
  return {
    name: "pb-editor-page-element-advanced-settings-" + type,
    type: "pb-editor-page-element-advanced-settings",
    elementType: type,
    render: render
  };
};
