// @webiny/app-headless-cms/admin/plugins/fields/boolean.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import React from "react";
import { ReactComponent as BooleanIcon } from "@svgr/webpack!./icons/toggle_on-black-24px.svg";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/fields");
var plugin = {
  type: "cms-editor-field-type",
  name: "cms-editor-field-type-boolean",
  field: {
    type: "boolean",
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Boolean"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Store boolean (\"yes\" or \"no\" ) values."]))),
    icon: /*#__PURE__*/React.createElement(BooleanIcon, null),
    allowMultipleValues: false,
    allowPredefinedValues: false,
    multipleValuesLabel: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Use as a list of booleans"]))),
    createField: function createField() {
      return {
        type: this.type,
        validation: [],
        renderer: {
          name: ""
        }
      };
    }
  }
};
export default plugin;
