// @webiny/app-headless-cms/admin/plugins/fields/file.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import React from "react";
import { ReactComponent as FileIcon } from "@svgr/webpack!./icons/round_insert_drive_file-24px.svg";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/fields");
var plugin = {
  type: "cms-editor-field-type",
  name: "cms-editor-field-type-file",
  field: {
    type: "file",
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Files"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Images, videos and other files."]))),
    icon: /*#__PURE__*/React.createElement(FileIcon, null),
    validators: ["required"],
    allowMultipleValues: true,
    allowPredefinedValues: false,
    multipleValuesLabel: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Use as a list of files or an image gallery"]))),
    createField: function createField() {
      return {
        type: this.type,
        validation: [],
        renderer: {
          name: ""
        }
      };
    }
  }
};
export default plugin;
