// @webiny/app-headless-cms/admin/plugins/fields/longText.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

import React from "react";
import { ReactComponent as LongTextIcon } from "@svgr/webpack!./icons/round-notes.svg";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Input } from "@webiny/ui/Input";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/fields");
var plugin = {
  type: "cms-editor-field-type",
  name: "cms-editor-field-type-long-text",
  field: {
    type: "long-text",
    validators: ["required", "minLength", "maxLength", "pattern"],
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Long text"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Long comments, notes, multi line values."]))),
    icon: /*#__PURE__*/React.createElement(LongTextIcon, null),
    allowMultipleValues: true,
    allowPredefinedValues: false,
    // TODO: implement "renderPredefinedValues" and set to true.
    multipleValuesLabel: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Use as a list of long texts"]))),
    createField: function createField() {
      return {
        type: this.type,
        validation: [],
        renderer: {
          name: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.form.Bind;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "placeholderText"
      }, /*#__PURE__*/React.createElement(Input, {
        label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Placeholder text"]))),
        description: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Placeholder text (optional)"])))
      }))));
    }
  }
};
export default plugin;
