// @webiny/app-headless-cms/admin/plugins/fields/number.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6;

import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { i18n } from "@webiny/app/i18n";
import { Input } from "@webiny/ui/Input";
import PredefinedValuesDynamicFieldset from "./PredefinedValuesDynamicFieldset";
import { ReactComponent as FloatIcon } from "@svgr/webpack!./icons/round-looks_3-24px.svg";
var t = i18n.ns("app-headless-cms/admin/fields");
var plugin = {
  type: "cms-editor-field-type",
  name: "cms-editor-field-type-number",
  field: {
    type: "number",
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Number"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Store numbers."]))),
    icon: /*#__PURE__*/React.createElement(FloatIcon, null),
    validators: ["required", "gte", "lte"],
    allowMultipleValues: true,
    allowPredefinedValues: true,
    multipleValuesLabel: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Use as a list of numbers"]))),
    createField: function createField() {
      return {
        type: this.type,
        validation: [],
        renderer: {
          name: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.form.Bind;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "placeholderText"
      }, /*#__PURE__*/React.createElement(Input, {
        label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Placeholder text"]))),
        description: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Placeholder text (optional)"])))
      }))));
    },
    renderPredefinedValues: function renderPredefinedValues(props) {
      return /*#__PURE__*/React.createElement(PredefinedValuesDynamicFieldset, Object.assign({}, props, {
        renderValueInput: function renderValueInput(Bind) {
          return /*#__PURE__*/React.createElement(Bind, {
            name: "value"
          }, function (bind) {
            return /*#__PURE__*/React.createElement(Input, Object.assign({}, bind, {
              type: "number",
              label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Value"]))),
              onChange: function onChange(value) {
                return bind.onChange(parseFloat(value));
              }
            }));
          });
        }
      }));
    }
  }
};
export default plugin;
