// @webiny/app-headless-cms/admin/plugins/fields/object.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import React from "react";
import { ReactComponent as ObjectIcon } from "@svgr/webpack!./icons/ballot_black_24dp.svg";
import { i18n } from "@webiny/app/i18n";
import { ObjectFields } from "./object/ObjectFields";
import { createFieldsList } from "../../graphql/createFieldsList";
var t = i18n.ns("app-headless-cms/admin/fields");
var plugin = {
  type: "cms-editor-field-type",
  name: "cms-editor-field-type-object",
  field: {
    type: "object",
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Object"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Store nested data structures."]))),
    icon: /*#__PURE__*/React.createElement(ObjectIcon, null),
    allowMultipleValues: true,
    allowPredefinedValues: false,
    multipleValuesLabel: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Use as a repeatable object"]))),
    createField: function createField() {
      return {
        type: this.type,
        validation: [],
        settings: {
          fields: [],
          layout: []
        },
        renderer: {
          name: ""
        }
      };
    },
    render: function render(props) {
      return /*#__PURE__*/React.createElement(ObjectFields, props);
    },
    graphql: {
      queryField: function queryField(_ref) {
        var field = _ref.field;
        return "{ ".concat(createFieldsList(field.settings.fields), " }");
      }
    }
  }
};
export default plugin;
