// @webiny/app-headless-cms/admin/plugins/fields/richText.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

import React from "react";
import { i18n } from "@webiny/app/i18n";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ReactComponent as NotesIcon } from "@svgr/webpack!./icons/round-text_fields-24px.svg";
import { Input } from "@webiny/ui/Input";
var t = i18n.ns("app-headless-cms/admin/fields");
var plugin = {
  type: "cms-editor-field-type",
  name: "cms-editor-field-type-richText",
  field: {
    type: "rich-text",
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Rich text"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Text formatting with references and media."]))),
    icon: /*#__PURE__*/React.createElement(NotesIcon, null),
    allowMultipleValues: true,
    allowPredefinedValues: false,
    multipleValuesLabel: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Use as a list of rich texts"]))),
    createField: function createField() {
      return {
        type: this.type,
        name: this.name,
        validation: [],
        renderer: {
          name: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.form.Bind;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "placeholderText"
      }, /*#__PURE__*/React.createElement(Input, {
        label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Placeholder text"]))),
        description: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Placeholder text (optional)"])))
      }))));
    }
  }
};
export default plugin;
