// @webiny/app-headless-cms/admin/plugins/fields/text.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6;

import React from "react";
import { ReactComponent as TextIcon } from "@svgr/webpack!./icons/round-text_fields-24px.svg";
import { Grid, Cell } from "@webiny/ui/Grid";
import { i18n } from "@webiny/app/i18n";
import { Input } from "@webiny/ui/Input";
import PredefinedValuesDynamicFieldset from "./PredefinedValuesDynamicFieldset";
var t = i18n.ns("app-headless-cms/admin/fields");
var plugin = {
  type: "cms-editor-field-type",
  name: "cms-editor-field-type-text",
  field: {
    type: "text",
    validators: ["required", "minLength", "maxLength", "pattern"],
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Text"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Titles, names, single line values."]))),
    icon: /*#__PURE__*/React.createElement(TextIcon, null),
    allowMultipleValues: true,
    allowPredefinedValues: true,
    multipleValuesLabel: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Use as a list of texts"]))),
    createField: function createField() {
      return {
        type: this.type,
        validation: [],
        renderer: {
          name: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.form.Bind;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "placeholderText"
      }, /*#__PURE__*/React.createElement(Input, {
        label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Placeholder text"]))),
        description: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Placeholder text (optional)"])))
      }))));
    },
    renderPredefinedValues: function renderPredefinedValues(props) {
      return /*#__PURE__*/React.createElement(PredefinedValuesDynamicFieldset, props, /*#__PURE__*/React.createElement(Input, {
        label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Value"])))
      }));
    }
  }
};
export default plugin;
