// @webiny/api-page-builder/utils/models/DefaultSettings.model.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fields = require("@commodo/fields");

var _commodoFieldsObject = require("commodo-fields-object");

var _trimEnd = _interopRequireDefault(require("lodash/trimEnd"));

// We don't want trailing slashes in Page Builder app's important URLs (website URL, website preview URL, app URL).
const trimTrailingSlashes = value => (0, _trimEnd.default)(value, "/"); // `pid` (page ID) is an ID consisting only of the unique page ID, without the version suffix (e.g. #0002).


const extractPid = value => {
  if (typeof value !== "string") {
    return null;
  }

  const [pid] = value.split("#");
  return pid || null;
};

var _default = (0, _fields.withFields)({
  name: (0, _fields.string)({
    validation: "required,maxLength:500"
  }),
  websiteUrl: (0, _fields.onSet)(trimTrailingSlashes)((0, _fields.string)({
    validation: "url,maxLength:500"
  })),
  websitePreviewUrl: (0, _fields.onSet)(trimTrailingSlashes)((0, _fields.string)({
    validation: "url,maxLength:500"
  })),
  favicon: (0, _commodoFieldsObject.object)({}),
  logo: (0, _commodoFieldsObject.object)({}),
  prerendering: (0, _fields.fields)({
    instanceOf: (0, _fields.withFields)({
      app: (0, _fields.fields)({
        instanceOf: (0, _fields.withFields)({
          url: (0, _fields.onSet)(trimTrailingSlashes)((0, _fields.string)({
            validation: "url"
          }))
        })()
      }),
      storage: (0, _fields.fields)({
        instanceOf: (0, _fields.withFields)({
          name: (0, _fields.string)({
            validation: "maxLength:500"
          })
        })()
      }),
      meta: (0, _commodoFieldsObject.object)()
    })()
  }),
  social: (0, _fields.fields)({
    value: {},
    instanceOf: (0, _fields.withFields)({
      facebook: (0, _fields.string)({
        validation: "url,maxLength:500"
      }),
      twitter: (0, _fields.string)({
        validation: "url,maxLength:500"
      }),
      instagram: (0, _fields.string)({
        validation: "url,maxLength:500"
      }),
      image: (0, _commodoFieldsObject.object)({})
    })()
  }),
  pages: (0, _fields.fields)({
    value: {},
    instanceOf: (0, _fields.withFields)({
      home: (0, _fields.onSet)(extractPid)((0, _fields.string)()),
      notFound: (0, _fields.onSet)(extractPid)((0, _fields.string)())
    })()
  })
})();

exports.default = _default;
