// @webiny/app-page-builder/editor/plugins/elements/social/pinterest/index.js
import React from "react";
import styled from "@emotion/styled";
import kebabCase from "lodash/kebabCase";
import { Typography } from "@webiny/ui/Typography";
import { createEmbedPlugin, createEmbedSettingsPlugin } from "../../utils/oembed";
import PinterestEmbed from "./PinterestEmbed";
import { validation } from "@webiny/validation";
import { ReactComponent as LogoIcon } from "@svgr/webpack!./pinterest-brands.svg";
import Accordion from "../../../elementSettings/components/Accordion";
import Wrapper from "../../../elementSettings/components/Wrapper";
import InputField from "../../../elementSettings/components/InputField";
import SelectField from "../../../elementSettings/components/SelectField";
import { ButtonContainer, SimpleButton, classes } from "../../../elementSettings/components/StyledComponents";
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "e6sbjxh0",
  label: "PreviewBox"
})({
  textAlign: "center",
  height: 50,
  svg: {
    height: 50,
    width: 50
  }
});
export default (function () {
  var args = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var elementType = kebabCase(args.elementType || "pinterest");
  var defaultToolbar = {
    title: "Pinterest post",
    group: "pb-editor-element-group-social",
    preview: function preview() {
      return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(LogoIcon, null));
    }
  };
  return [createEmbedPlugin({
    type: elementType,
    toolbar: typeof args.toolbar === "function" ? args.toolbar(defaultToolbar) : defaultToolbar,
    create: args.create,
    settings: args.settings,
    render: function render(_ref) {
      var element = _ref.element;
      return /*#__PURE__*/React.createElement(PinterestEmbed, {
        element: element
      });
    }
  }), createEmbedSettingsPlugin({
    type: elementType,
    render: function render(_ref2) {
      var Bind = _ref2.Bind,
          submit = _ref2.submit;
      return /*#__PURE__*/React.createElement(Accordion, {
        title: "Pinterest",
        defaultValue: true
      }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Wrapper, {
        label: "URL",
        containerClassName: classes.simpleGrid,
        leftCellSpan: 3,
        rightCellSpan: 9
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "source.url",
        validators: validation.create("required,url")
      }, /*#__PURE__*/React.createElement(InputField, {
        placeholder: "https://pinterest.com/pin/823666219335767857/",
        description: "Enter a Pinterest URL"
      }))), /*#__PURE__*/React.createElement(Wrapper, {
        label: "Size",
        containerClassName: classes.simpleGrid,
        leftCellSpan: 3,
        rightCellSpan: 9
      }, /*#__PURE__*/React.createElement(Bind, {
        defaultValue: "small",
        name: "source.size"
      }, function (_ref3) {
        var value = _ref3.value,
            onChange = _ref3.onChange;
        return /*#__PURE__*/React.createElement(SelectField, {
          value: value,
          onChange: onChange
        }, /*#__PURE__*/React.createElement("option", {
          value: "small"
        }, "Small"), /*#__PURE__*/React.createElement("option", {
          value: "medium"
        }, "Medium"), /*#__PURE__*/React.createElement("option", {
          value: "large"
        }, "Large"));
      })), /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
        onClick: submit
      }, /*#__PURE__*/React.createElement(Typography, {
        use: "caption"
      }, "Save")))));
    }
  })];
});
