// @webiny/ui/Carousel/Carousel.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import * as React from "react";
import NukaCarousel from "nuka-carousel";
import pick from "lodash/pick";

var Carousel = /*#__PURE__*/function (_React$Component) {
  _inherits(Carousel, _React$Component);

  var _super = _createSuper(Carousel);

  function Carousel() {
    _classCallCheck(this, Carousel);

    return _super.apply(this, arguments);
  }

  _createClass(Carousel, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/React.createElement(NukaCarousel, Object.assign({}, pick(this.props, Carousel.nukaProps), this.props.renderPreviousSlide ? {
        renderCenterLeftControls: this.props.renderPreviousSlide
      } : {}, this.props.renderNextSlide ? {
        renderCenterRightControls: this.props.renderNextSlide
      } : {}, this.props.renderBottomNav ? {
        renderBottomCenterControls: this.props.renderBottomNav
      } : {}), this.props.children);
    }
  }]);

  return Carousel;
}(React.Component);

_defineProperty(Carousel, "defaultProps", {
  swiping: true,
  dragging: false,
  heightMode: "first",
  easing: "easeExpInOut"
});

_defineProperty(Carousel, "nukaProps", ["children", "afterSlide", "autoplay", "autoplayInterval", "beforeSlide", "dragging", "easing", "heightMode", "pauseOnHover", "slideIndex", "speed", "swiping", "transitionMode"]);

export default Carousel;
