// @webiny/app-page-builder/editor/plugins/elements/cell/Cell.js
import React from "react";
import styled from "@emotion/styled";
import DropZone from "../../../components/DropZone";
import Element from "../../../components/Element";
var CellStyle = /*#__PURE__*/styled("div", {
  target: "epof9910",
  label: "CellStyle"
})({
  position: "relative"
});

var Cell = function Cell(_ref) {
  var id = _ref.id,
      dropElement = _ref.dropElement,
      index = _ref.index,
      _ref$isLast = _ref.isLast,
      isLast = _ref$isLast === void 0 ? false : _ref$isLast,
      type = _ref.type;
  return /*#__PURE__*/React.createElement(CellStyle, null, /*#__PURE__*/React.createElement(DropZone.Above, {
    type: type,
    onDrop: function onDrop(source) {
      return dropElement(source, index);
    }
  }), /*#__PURE__*/React.createElement(Element, {
    id: id
  }), isLast && /*#__PURE__*/React.createElement(DropZone.Below, {
    type: type,
    onDrop: function onDrop(source) {
      return dropElement(source, index + 1);
    }
  }));
};

export default /*#__PURE__*/React.memo(Cell);
