// @webiny/app-page-builder/editor/plugins/elements/social/instagram/index.js
import React from "react";
import styled from "@emotion/styled";
import { validation } from "@webiny/validation";
import { Typography } from "@webiny/ui/Typography";
import { createEmbedPlugin, createEmbedSettingsPlugin } from "./../../utils/oembed/createEmbedPlugin";
import placeholder from "./placeholder.jpg";
import { ReactComponent as LogoIcon } from "@svgr/webpack!./instagram-brands.svg";
import Accordion from "../../../elementSettings/components/Accordion";
import InputField from "../../../elementSettings/components/InputField";
import { ButtonContainer, SimpleButton } from "../../../elementSettings/components/StyledComponents";
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "ed58sfq0",
  label: "PreviewBox"
})({
  textAlign: "center",
  height: 50,
  svg: {
    height: 50,
    width: 50
  }
});
export default (function () {
  return [createEmbedPlugin({
    type: "instagram",
    toolbar: {
      title: "Instagram Post",
      group: "pb-editor-element-group-social",
      preview: function preview() {
        return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(LogoIcon, null));
      }
    },
    oembed: {
      global: "instgrm",
      sdk: "https://www.instagram.com/embed.js",
      init: function init(_ref) {
        var node = _ref.node;
        // @ts-ignore
        window.instgrm.Embeds.process(node.firstChild);
      }
    },
    renderElementPreview: function renderElementPreview(_ref2) {
      var width = _ref2.width,
          height = _ref2.height;
      return /*#__PURE__*/React.createElement("img", {
        style: {
          width: width,
          height: height
        },
        src: placeholder,
        alt: "Instagram"
      });
    }
  }), createEmbedSettingsPlugin({
    type: "instagram",
    render: function render(_ref3) {
      var Bind = _ref3.Bind,
          submit = _ref3.submit;
      return /*#__PURE__*/React.createElement(Accordion, {
        title: "Instagram",
        defaultValue: true
      }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Bind, {
        name: "source.url",
        validators: validation.create("required,url")
      }, /*#__PURE__*/React.createElement(InputField, {
        placeholder: "https://www.instagram.com/marvel/",
        description: "Enter a Instagram URL"
      })), /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
        onClick: submit
      }, /*#__PURE__*/React.createElement(Typography, {
        use: "caption"
      }, "Save")))));
    }
  })];
});
