// @webiny/app-page-builder/editor/plugins/elementSettings/advanced/advancedSettingsEditorAction.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import { plugins } from "@webiny/plugins";
export var advancedSettingsEditorAction = function advancedSettingsEditorAction(state, meta, _ref) {
  var element = _ref.element,
      source = _ref.source;
  // Check the source of the element (could be `saved` element which behaves differently from other elements)
  var sourcePlugin = plugins.byType("pb-editor-page-element").find(function (pl) {
    return pl.elementType === source.type;
  });

  if (!sourcePlugin) {
    return {};
  }

  var sourceOnCreate = sourcePlugin.onCreate;

  if (!sourceOnCreate || sourceOnCreate !== "skip") {
    // If source element does not define a specific `onCreate` behavior - continue with the actual element plugin
    var plugin = plugins.byType("pb-editor-page-element").find(function (pl) {
      return pl.elementType === element.type;
    });

    if (!plugin) {
      return {};
    }

    var onCreate = plugin.onCreate;

    if (onCreate && onCreate === "open-settings") {
      return {
        state: _objectSpread(_objectSpread({}, state), {}, {
          activeElement: element.id,
          sidebar: {
            // Mark "Element" settings tab active in sidebar.
            activeTabIndex: 1,
            // Highlight "Element" settings tab in sidebar.
            highlightTab: true
          }
        })
      };
    }
  }

  return {};
};
