// @webiny/app-admin/plugins/globalSearch/index.js
import * as React from "react";
import SearchBar from "./SearchBar";
import { GenericElement } from "../../ui/elements/GenericElement";
import { UIViewPlugin } from "../../ui/UIView";
import { AdminView } from "../../ui/views/AdminView"; // !EXAMPLE!
// This demonstrates how you can create view-specific plugin classes.
//
// class AdminViewPlugin extends ViewPlugin<AdminView> {
//     constructor(apply: ApplyFunction<AdminView>) {
//         super(AdminView, apply);
//     }
// }

export var globalSearch = new UIViewPlugin(AdminView, function (view) {
  view.getHeaderElement().getCenterSection().addElement(new GenericElement("searchBar", function () {
    return /*#__PURE__*/React.createElement(SearchBar, null);
  }));
});
export var globalSearchHotkey = {
  type: "admin-global-search-prevent-hotkey",
  name: "admin-global-search-prevent-hotkey-input",
  preventOpen: function preventOpen(e) {
    // Define a list of all node types we want to prevent the event from.
    var ignoreNodes = ["INPUT", "TEXTAREA"];

    if (ignoreNodes.includes(e.target.nodeName)) {
      return true;
    }
  }
};
