// @webiny/app-admin/plugins/globalSearch/styled.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
// TODO remove
// @ts-nocheck
import { css } from "emotion";
import styled from "@emotion/styled";
export var SearchShortcut = /*#__PURE__*/styled("div", {
  target: "eh6vkyz0",
  label: "SearchShortcut"
})({
  border: "1px solid var(--mdc-theme-surface)",
  color: "var(--mdc-theme-surface)",
  borderRadius: 5,
  width: 20,
  height: 24,
  marginTop: 1,
  textAlign: "center",
  fontSize: "12px",
  paddingTop: "3px",
  boxSizing: "border-box",
  cursor: "default",
  opacity: "0.8"
});
export var SearchBarWrapper = /*#__PURE__*/styled("div", {
  target: "eh6vkyz1",
  label: "SearchBarWrapper"
})({
  display: "flex",
  width: "100%"
});
export var SearchBarInputWrapper = /*#__PURE__*/styled("div", {
  target: "eh6vkyz2",
  label: "SearchBarInputWrapper"
})({
  display: "flex",
  width: "100%",
  position: "relative"
});
export var searchBarInput = /*#__PURE__*/css({
  width: "100%",
  lineHeight: 1,
  border: "none",
  outline: "none",
  backgroundColor: "transparent",
  "&.mdc-text-field__input": {
    paddingTop: "5px !important",
    paddingLeft: "10px !important",
    borderBottom: "none !important",
    height: "25px !important",
    color: "var(--mdc-theme-surface)",
    "&::placeholder": {
      color: "var(--mdc-theme-surface) !important",
      opacity: "1 !important"
    },
    "&:focus::placeholder": {
      color: "var(--mdc-theme-text-secondary-on-background) !important"
    }
  }
}, "label:searchBarInput;");
export var searchBarDropdown = /*#__PURE__*/css({
  position: "absolute",
  background: "white",
  left: -1,
  width: "calc(100% + 2px)",
  top: 45,
  zIndex: 2,
  color: "var(--mdc-theme-on-surface)",
  borderRadius: "0 0 5px 5px",
  ".mdc-list": {
    padding: 0,
    ".mdc-list-item": {
      "&.selected": {
        fontWeight: "bold"
      },
      "&.highlighted": {
        backgroundColor: "var(--mdc-theme-on-background)"
      }
    }
  }
}, "label:searchBarDropdown;");
export var icon = /*#__PURE__*/css({
  "&.mdc-button__icon": {
    width: "auto !important",
    paddingRight: 5,
    opacity: 0.75,
    marginTop: 2
  }
}, "label:icon;");
export var iconSearchType = /*#__PURE__*/css({
  cursor: "pointer",
  paddingLeft: 5
}, "label:iconSearchType;");
export var searchWrapper = /*#__PURE__*/css({
  backgroundColor: "rgba(0, 0, 0, 0.1)",
  display: "flex",
  padding: "10px 20px",
  position: "relative",
  borderRadius: 4,
  transition: "background 100ms ease-in,width 100ms ease-out",
  color: "var(--mdc-theme-surface)",
  border: "1px solid transparent",
  "&.active": _defineProperty({
    color: "var(--mdc-theme-on-surface)",
    border: "1px solid var(--mdc-theme-text-hint-on-dark)",
    background: "var(--mdc-theme-surface)",
    boxShadow: "0 1px 1px var(--mdc-theme-text-hint-on-dark)",
    input: {
      color: "var(--mdc-theme-on-surface)"
    }
  }, SearchShortcut, {
    display: "none"
  }),
  input: {
    color: "var(--mdc-theme-surface)"
  }
}, "label:searchWrapper;");
