// @webiny/app-form-builder/admin/components/FormEditor/DropZone/Horizontal.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import styled from "@emotion/styled";
import Droppable from "../Droppable";
var InnerDiv = /*#__PURE__*/styled("div", {
  target: "e5k6cvq0",
  label: "InnerDiv"
})({
  height: 15,
  width: "100%",
  zIndex: 3,
  borderRadius: 0,
  boxSizing: "border-box",
  display: "none",
  border: "1px dashed black",
  borderSpacing: 5
});
var BackgroundColorDiv = /*#__PURE__*/styled("div", {
  target: "e5k6cvq1",
  label: "BackgroundColorDiv"
})({
  width: "100%",
  height: "100%"
});
var OuterDiv = /*#__PURE__*/styled("div", {
  target: "e5k6cvq2",
  label: "OuterDiv"
})({
  margin: 0,
  padding: 0,
  width: "calc(100% + 2px)",
  zIndex: 10,
  backgroundColor: "transparent",
  position: "absolute",
  display: "flex",
  justifyContent: "center"
}, function (props) {
  var _ref;

  return _ref = {}, _defineProperty(_ref, props.last ? "bottom" : "top", -15), _defineProperty(_ref, InnerDiv, _defineProperty({
    borderColor: props.isOver ? "var(--mdc-theme-primary)" : "var(--mdc-theme-secondary)",
    display: props.isDragging ? "block" : "none"
  }, BackgroundColorDiv, {
    opacity: 0.5,
    backgroundColor: props.isOver ? "var(--mdc-theme-primary)" : "var(--mdc-theme-secondary)"
  })), _ref;
});

var Horizontal = function Horizontal(_ref2) {
  var last = _ref2.last,
      onDrop = _ref2.onDrop,
      isVisible = _ref2.isVisible;
  return /*#__PURE__*/React.createElement(Droppable, {
    onDrop: onDrop,
    isVisible: isVisible
  }, function (_ref3) {
    var _ref4;

    var isOver = _ref3.isOver,
        isDragging = _ref3.isDragging,
        drop = _ref3.drop;
    return /*#__PURE__*/React.createElement("div", {
      ref: drop,
      style: (_ref4 = {
        height: "25px",
        width: "100%",
        position: "absolute"
      }, _defineProperty(_ref4, last ? "bottom" : "top", 0), _defineProperty(_ref4, "left", 0), _defineProperty(_ref4, "zIndex", isDragging ? 1000 : -1), _ref4),
      "data-testid": last ? "fb.editor.dropzone.horizontal-last" : ""
    }, /*#__PURE__*/React.createElement(OuterDiv, {
      isOver: isOver,
      isDragging: isDragging,
      last: last
    }, /*#__PURE__*/React.createElement(InnerDiv, null, /*#__PURE__*/React.createElement(BackgroundColorDiv, null))));
  });
};

export default Horizontal;
