// @webiny/app-form-builder/admin/components/FormEditor/DropZone/Vertical.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import styled from "@emotion/styled";
import Droppable from "../Droppable";
var InnerDivVertical = /*#__PURE__*/styled("div", {
  target: "e1dmppbp0",
  label: "InnerDivVertical"
})({
  position: "absolute",
  width: 10,
  height: "100%",
  zIndex: 3,
  borderRadius: 0,
  display: "none",
  boxSizing: "border-box",
  border: "1px dashed black",
  borderSpacing: 5
});
var BackgroundColorDiv = /*#__PURE__*/styled("div", {
  target: "e1dmppbp1",
  label: "BackgroundColorDiv"
})({
  width: "100%",
  height: "100%"
});
var OuterDivVertical = /*#__PURE__*/styled("div", {
  target: "e1dmppbp2",
  label: "OuterDivVertical"
})({
  position: "absolute",
  width: "30%",
  top: 0,
  height: "100%",
  zIndex: 10,
  backgroundColor: "transparent"
}, function (props) {
  var _InnerDivVertical, _ref;

  return _ref = {}, _defineProperty(_ref, props.last ? "right" : "left", -9), _defineProperty(_ref, "textAlign", props.last ? "right" : "left"), _defineProperty(_ref, InnerDivVertical, (_InnerDivVertical = {
    borderColor: props.isOver ? "var(--mdc-theme-primary)" : "var(--mdc-theme-secondary)"
  }, _defineProperty(_InnerDivVertical, props.last ? "right" : "left", -2), _defineProperty(_InnerDivVertical, "display", props.isDragging ? "block" : "none"), _defineProperty(_InnerDivVertical, BackgroundColorDiv, {
    opacity: 0.5,
    backgroundColor: props.isOver ? "var(--mdc-theme-primary)" : "var(--mdc-theme-secondary)"
  }), _InnerDivVertical)), _ref;
});

var Vertical = function Vertical(_ref2) {
  var last = _ref2.last,
      onDrop = _ref2.onDrop,
      isVisible = _ref2.isVisible;
  return /*#__PURE__*/React.createElement(Droppable, {
    onDrop: onDrop,
    isVisible: isVisible
  }, function (_ref3) {
    var _ref4;

    var isOver = _ref3.isOver,
        isDragging = _ref3.isDragging,
        drop = _ref3.drop;
    return /*#__PURE__*/React.createElement("div", {
      ref: drop,
      style: (_ref4 = {
        width: "30%",
        maxWidth: "100px",
        height: "100%",
        position: "absolute",
        top: 0
      }, _defineProperty(_ref4, last ? "right" : "left", 0), _defineProperty(_ref4, "zIndex", isDragging ? 1000 : -1), _ref4)
    }, /*#__PURE__*/React.createElement(OuterDivVertical, {
      isOver: isOver,
      isDragging: isDragging,
      last: last
    }, /*#__PURE__*/React.createElement(InnerDivVertical, null, /*#__PURE__*/React.createElement(BackgroundColorDiv, null))));
  });
};

export default Vertical;
