// @webiny/ui/ImageEditor/toolbar/flip.js
import React from "react";
import { ReactComponent as FlipIcon } from "@svgr/webpack!./icons/flip.svg";
import Cropper from "cropperjs";
import "cropperjs/dist/cropper.css";
import { IconButton, ButtonDefault } from "../../Button";
import { Tooltip } from "../../Tooltip";
var cropper = null;
var flipped = {
  x: 1,
  y: 1
};

var renderForm = function renderForm() {
  return /*#__PURE__*/React.createElement("div", {
    style: {
      textAlign: "center"
    }
  }, /*#__PURE__*/React.createElement(ButtonDefault, {
    onClick: function onClick() {
      if (!cropper) {
        return;
      }

      flipped.x = flipped.x === 1 ? -1 : 1;
      cropper.scaleX(flipped.x);
    }
  }, "FlipX"), /*#__PURE__*/React.createElement(ButtonDefault, {
    onClick: function onClick() {
      if (!cropper) {
        return;
      }

      flipped.y = flipped.y === 1 ? -1 : 1;
      cropper.scaleY(flipped.y);
    }
  }, "FlipY"));
};

var tool = {
  name: "flip",
  icon: function icon(_ref) {
    var activateTool = _ref.activateTool;
    return /*#__PURE__*/React.createElement(Tooltip, {
      placement: "bottom",
      content: "Flip"
    }, /*#__PURE__*/React.createElement(IconButton, {
      icon: /*#__PURE__*/React.createElement(FlipIcon, null),
      onClick: activateTool
    }));
  },
  renderForm: renderForm,
  cancel: function cancel() {
    return cropper && cropper.destroy();
  },
  onActivate: function onActivate(_ref2) {
    var canvas = _ref2.canvas;
    cropper = new Cropper(canvas.current, {
      background: false,
      modal: false,
      guides: false,
      dragMode: "none",
      highlight: false,
      autoCrop: false
    });
  },
  apply: function apply(_ref3) {
    var canvas = _ref3.canvas;
    return new Promise(function (resolve) {
      if (!cropper) {
        resolve();
        return;
      }

      var current = canvas.current;
      var src = cropper.getCroppedCanvas().toDataURL();

      if (current) {
        var image = new window.Image();
        var ctx = current.getContext("2d");

        image.onload = function () {
          ctx.drawImage(image, 0, 0);
          current.width = image.width;
          current.height = image.height;
          ctx.drawImage(image, 0, 0);
          resolve();
        };

        image.src = src;
      }

      cropper.destroy();
    });
  }
};
export default tool;
