// @webiny/app-admin/plugins/menu/renderers/MenuSectionItemRenderer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { css } from "emotion";
import { Link } from "@webiny/react-router";
import { List, ListItem } from "@webiny/ui/List";
import { UIRenderer } from "../../../ui/UIRenderer";
var linkStyle = /*#__PURE__*/css({
  color: "var(--mdc-theme-text-primary-on-background)",
  textDecoration: "none",
  display: "flex",
  alignItems: "center",
  width: "100%",
  height: "100%",
  outline: "none",
  paddingLeft: 65,
  "&:hover": {
    textDecoration: "none"
  }
}, "label:linkStyle;");
var submenuItems = /*#__PURE__*/css({
  ".mdc-drawer &.mdc-list-item": {
    paddingLeft: 0
  }
}, "label:submenuItems;");
var submenuList = /*#__PURE__*/css({
  "&.mdc-list": {
    padding: 0
  }
}, "label:submenuList;");
export var MenuSectionItemRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(MenuSectionItemRenderer, _UIRenderer);

  var _super = _createSuper(MenuSectionItemRenderer);

  function MenuSectionItemRenderer() {
    _classCallCheck(this, MenuSectionItemRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(MenuSectionItemRenderer, [{
    key: "canRender",
    value: function canRender(element) {
      return [2, 3].includes(element.depth) && Boolean(element.config.path);
    }
  }, {
    key: "render",
    value: function render(_ref) {
      var element = _ref.element,
          props = _ref.props;
      var defaultOnClick = element.getView().getNavigationHook().hideMenu;
      var onClick = element.config.onClick || defaultOnClick;
      return /*#__PURE__*/React.createElement(List, {
        className: submenuList
      }, /*#__PURE__*/React.createElement(ListItem, {
        className: submenuItems,
        "data-testid": element.config.testId
      }, element.config.path ? /*#__PURE__*/React.createElement(Link, {
        className: linkStyle,
        to: element.config.path,
        onClick: onClick ? function () {
          return onClick(props);
        } : null
      }, element.config.label) : /*#__PURE__*/React.createElement("span", {
        onClick: onClick ? function () {
          return onClick(props);
        } : null,
        className: linkStyle
      }, element.config.label)));
    }
  }]);

  return MenuSectionItemRenderer;
}(UIRenderer);
