// @webiny/app-admin/plugins/menu/renderers/MenuLinkRenderer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { Link } from "@webiny/react-router";
import { ListItem, ListItemGraphic } from "@webiny/ui/List";
import { UIRenderer } from "../../../ui/UIRenderer";
import { Icon } from "@webiny/ui/Icon";
import { FooterElement } from "../../../ui/views/NavigationView/FooterElement";
export var MenuLinkRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(MenuLinkRenderer, _UIRenderer);

  var _super = _createSuper(MenuLinkRenderer);

  function MenuLinkRenderer() {
    _classCallCheck(this, MenuLinkRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(MenuLinkRenderer, [{
    key: "canRender",
    value: function canRender(element) {
      var isInFooter = Boolean(element.getParentByType(FooterElement));
      return element.depth === 1 && isInFooter;
    }
  }, {
    key: "render",
    value: function render(_ref) {
      var element = _ref.element,
          props = _ref.props;
      var defaultOnClick = element.getView().getNavigationHook().hideMenu;
      var onClick = element.config.onClick || defaultOnClick;
      return /*#__PURE__*/React.createElement(Link, {
        to: element.config.path,
        "data-testid": element.config.testId,
        onClick: onClick ? function () {
          return onClick(props);
        } : null
      }, /*#__PURE__*/React.createElement(ListItem, {
        ripple: false
      }, element.config.icon ? /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
        icon: element.config.icon
      })) : null, element.config.label));
    }
  }]);

  return MenuLinkRenderer;
}(UIRenderer);
