// @webiny/app-admin/plugins/menu/renderers/MenuSectionRenderer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React, { Fragment } from "react";
import { css } from "emotion";
import { Typography } from "@webiny/ui/Typography";
import { UIRenderer } from "../../../ui/UIRenderer";
var menuSectionTitle = /*#__PURE__*/css({
  marginLeft: 20,
  display: "flex",
  alignItems: "center",
  color: "var(--mdc-theme-on-surface)"
}, "label:menuSectionTitle;");
var iconWrapper = /*#__PURE__*/css({
  marginRight: 5,
  color: "var(--mdc-theme-on-surface)"
}, "label:iconWrapper;");
export var MenuSectionRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(MenuSectionRenderer, _UIRenderer);

  var _super = _createSuper(MenuSectionRenderer);

  function MenuSectionRenderer() {
    _classCallCheck(this, MenuSectionRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(MenuSectionRenderer, [{
    key: "canRender",
    value: function canRender(element) {
      return element.depth === 2;
    }
  }, {
    key: "render",
    value: function render(_ref) {
      var element = _ref.element,
          next = _ref.next;
      return /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement("div", {
        className: menuSectionTitle
      }, /*#__PURE__*/React.createElement("div", {
        className: iconWrapper
      }, element.config.icon), /*#__PURE__*/React.createElement(Typography, {
        use: "overline"
      }, element.config.label)), next());
    }
  }]);

  return MenuSectionRenderer;
}(UIRenderer);
