// @webiny/app-page-builder/editor/recoil/actions/deleteElement/action.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import { UpdateElementTreeActionEvent, SaveRevisionActionEvent } from "..";

var removeElementFromParent = function removeElementFromParent(parent, id) {
  return _objectSpread(_objectSpread({}, parent), {}, {
    elements: parent.elements.filter(function (child) {
      return child !== id;
    })
  });
};

var getElementParentById = /*#__PURE__*/function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(state, id) {
    var element;
    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return state.getElementById(id);

          case 2:
            element = _context.sent;
            _context.next = 5;
            return state.getElementById(element.parent);

          case 5:
            return _context.abrupt("return", _context.sent);

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function getElementParentById(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

export var deleteElementAction = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(state, meta, args) {
    var element, parent, newParent;
    return _regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            element = args.element;
            _context2.next = 3;
            return getElementParentById(state, element.id);

          case 3:
            parent = _context2.sent;
            newParent = removeElementFromParent(parent, element.id);
            return _context2.abrupt("return", {
              state: _objectSpread(_objectSpread({}, state), {}, {
                elements: _objectSpread(_objectSpread({}, state.elements), {}, _defineProperty({}, newParent.id, newParent)),
                activeElement: null,
                highlightElement: null
              }),
              actions: [new SaveRevisionActionEvent(), new UpdateElementTreeActionEvent()]
            });

          case 6:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));

  return function deleteElementAction(_x3, _x4, _x5) {
    return _ref2.apply(this, arguments);
  };
}();
