// @webiny/app-page-builder/editor/plugins/blockEditing/AddBlock.js
import React from "react";
import styled from "@emotion/styled";
import { useEventActionHandler } from "../../hooks/useEventActionHandler";
import { TogglePluginActionEvent } from "../../recoil/actions";
import { ButtonFloating } from "@webiny/ui/Button";
import { ReactComponent as AddIcon } from "@svgr/webpack!../../assets/icons/add.svg";
var SIDEBAR_WIDTH = 300;
var BottomRight = /*#__PURE__*/styled("div", {
  target: "e1iz1lpp0",
  label: "BottomRight"
})({
  position: "fixed",
  zIndex: 25,
  bottom: 20,
  right: 20 + SIDEBAR_WIDTH
});

var AddBlock = function AddBlock() {
  var handler = useEventActionHandler();

  var onClickHandler = function onClickHandler() {
    handler.trigger(new TogglePluginActionEvent({
      name: "pb-editor-search-blocks-bar"
    }));
  };

  return /*#__PURE__*/React.createElement(BottomRight, null, /*#__PURE__*/React.createElement(ButtonFloating, {
    onClick: onClickHandler,
    icon: /*#__PURE__*/React.createElement(AddIcon, null)
  }));
};

export default /*#__PURE__*/React.memo(AddBlock);
