// @webiny/app-page-builder/editor/plugins/blockEditing/AddContent.js
import React from "react";
import { useRecoilValue } from "recoil";
import { keyframes } from "emotion";
import styled from "@emotion/styled";
import { Elevation } from "@webiny/ui/Elevation";
import { ButtonFloating } from "@webiny/ui/Button";
import { useEventActionHandler } from "../../hooks/useEventActionHandler";
import { ReactComponent as AddIcon } from "@svgr/webpack!../../assets/icons/add.svg";
import { TogglePluginActionEvent } from "../../recoil/actions";
import { uiAtom } from "../../recoil/modules";
import { elementsInContentTotalSelector } from "../../recoil/modules/page/selectors/elementsInContentTotalSelector";
var pulse = /*#__PURE__*/keyframes("0%{box-shadow:0 0 0 0 rgba(0,204,176,0.4);}70%{box-shadow:0 0 0 30px rgba(0,204,176,0);}100%{box-shadow:0 0 0 0 rgba(0,204,176,0);}label:pulse;");
var AddBlockContainer = /*#__PURE__*/styled("div", {
  target: "e1d6k8pe0",
  label: "AddBlockContainer"
})(function (_ref) {
  var displayMode = _ref.displayMode;
  var marginLeft = displayMode === "desktop" ? 54 : 0;
  return {
    position: "absolute",
    zIndex: 11,
    top: "50%",
    left: "50%",
    transform: "translate(-50%,-50%)",
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center",
    width: "150px",
    height: "100px",
    borderRadius: 2,
    color: "var(--mdc-theme-on-surface)",
    marginLeft: marginLeft,
    "& .elevation": {
      backgroundColor: "var(--mdc-theme-surface)",
      padding: displayMode === "mobile-portrait" ? "16px 8px" : "30px 20px",
      transform: "translateY(-50%)",
      borderRadius: 2
    }
  };
});
var AddBlockContent = /*#__PURE__*/styled("div", {
  target: "e1d6k8pe1",
  label: "AddBlockContent"
})(function (_ref2) {
  var displayMode = _ref2.displayMode;
  return {
    width: displayMode === "mobile-portrait" ? 280 : 300,
    margin: 5,
    textAlign: "center",
    display: "flex",
    alignItems: "center"
  };
});

var AddContent = function AddContent() {
  var _useRecoilValue = useRecoilValue(uiAtom),
      displayMode = _useRecoilValue.displayMode;

  var totalElements = useRecoilValue(elementsInContentTotalSelector);
  var eventActionHandler = useEventActionHandler();

  if (totalElements) {
    return null;
  }

  var onClickHandler = function onClickHandler() {
    eventActionHandler.trigger(new TogglePluginActionEvent({
      name: "pb-editor-search-blocks-bar"
    }));
  };

  return /*#__PURE__*/React.createElement(AddBlockContainer, {
    "data-type": "container",
    displayMode: displayMode
  }, /*#__PURE__*/React.createElement(Elevation, {
    z: 4,
    className: "elevation"
  }, /*#__PURE__*/React.createElement(AddBlockContent, {
    displayMode: displayMode
  }, "Click the", /*#__PURE__*/React.createElement(ButtonFloating, {
    "data-testid": "pb-content-add-block-button",
    style: {
      animation: pulse + " 3s ease infinite",
      margin: "0 10px"
    },
    small: true,
    icon: /*#__PURE__*/React.createElement(AddIcon, null),
    onClick: onClickHandler
  }), "to start adding content")));
};

export default AddContent;
