// @webiny/app-headless-cms/admin/plugins/routes.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

var _excluded = ["children"];
import React, { Suspense, lazy } from "react";
import Helmet from "react-helmet";
import { SecureRoute } from "@webiny/app-security/components";
import { CircularProgress } from "@webiny/ui/Progress";
import { Route } from "@webiny/react-router";
import { AdminLayout } from "@webiny/app-admin/components/AdminLayout";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/routes");

var Loader = function Loader(_ref) {
  var children = _ref.children,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(Suspense, {
    fallback: /*#__PURE__*/React.createElement(CircularProgress, null)
  }, /*#__PURE__*/React.cloneElement(children, props));
};

var ContentModelEditor = /*#__PURE__*/lazy(function () {
  return import("../views/contentModels/ContentModelEditor");
});
var ContentModelsView = /*#__PURE__*/lazy(function () {
  return import("../views/contentModels/ContentModels");
});
var ContentModelGroupsView = /*#__PURE__*/lazy(function () {
  return import("../views/contentModelGroups/ContentModelGroups");
});
var ContentEntiesView = /*#__PURE__*/lazy(function () {
  return import("../views/contentEntries/ContentEntries");
});
var plugins = [{
  name: "route-cms-content-models-groups",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/cms/content-model-groups",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: "cms.contentModelGroup"
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, null, /*#__PURE__*/React.createElement("title", null, t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Content Model Groups"]))))), /*#__PURE__*/React.createElement(Loader, null, /*#__PURE__*/React.createElement(ContentModelGroupsView, null))));
    }
  })
}, {
  name: "route-cms-content-models-manage",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/cms/content-entries/:modelId",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: "cms.contentModel"
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, null, /*#__PURE__*/React.createElement("title", null, t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Content"]))))), /*#__PURE__*/React.createElement(Loader, null, /*#__PURE__*/React.createElement(ContentEntiesView, null))));
    }
  })
}, {
  name: "route-cms-content-models-editor",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/cms/content-models/:modelId",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: "cms.contentModel"
      }, /*#__PURE__*/React.createElement(Helmet, null, /*#__PURE__*/React.createElement("title", null, t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Edit Content Model"]))))), /*#__PURE__*/React.createElement(Loader, null, /*#__PURE__*/React.createElement(ContentModelEditor, null)));
    }
  })
}, {
  name: "route-cms-content-models",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/cms/content-models",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: "cms.contentModel"
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, {
        title: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Content Models"])))
      }), /*#__PURE__*/React.createElement(Loader, null, /*#__PURE__*/React.createElement(ContentModelsView, null))));
    }
  })
}];
export default plugins;
