// @webiny/app-admin/ui/elements/AccordionElement.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _get from "@babel/runtime/helpers/get";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
var _excluded = ["open"];
import React from "react";
import { UIElement } from "../UIElement";
import { Accordion, AccordionItem } from "@webiny/ui/Accordion";
export var AccordionItemElement = /*#__PURE__*/function (_UIElement) {
  _inherits(AccordionItemElement, _UIElement);

  var _super = _createSuper(AccordionItemElement);

  function AccordionItemElement(id, config) {
    _classCallCheck(this, AccordionItemElement);

    return _super.call(this, id, config);
  }

  _createClass(AccordionItemElement, [{
    key: "setTitle",
    value: function setTitle(title) {
      this.config.title = title;
    }
  }, {
    key: "setDescription",
    value: function setDescription(description) {
      this.config.description = description;
    }
  }, {
    key: "render",
    value: function render(props) {
      var _this$config = this.config,
          open = _this$config.open,
          rest = _objectWithoutProperties(_this$config, _excluded);

      var isOpened = typeof open === "function" ? open(props) : open;
      return /*#__PURE__*/React.createElement(AccordionItem, Object.assign({
        open: isOpened
      }, rest), _get(_getPrototypeOf(AccordionItemElement.prototype), "render", this).call(this, props));
    }
  }]);

  return AccordionItemElement;
}(UIElement);
export var AccordionElement = /*#__PURE__*/function (_UIElement2) {
  _inherits(AccordionElement, _UIElement2);

  var _super2 = _createSuper(AccordionElement);

  function AccordionElement(id, config) {
    var _this;

    _classCallCheck(this, AccordionElement);

    _this = _super2.call(this, id, config);

    _this.useGrid(false);

    _this.config.items.forEach(function (item) {
      _this.addElement(new AccordionItemElement(item.id, item));
    });

    return _this;
  }

  _createClass(AccordionElement, [{
    key: "getAccordionItemElement",
    value: function getAccordionItemElement(id) {
      return this.getElement(id);
    }
  }, {
    key: "render",
    value: function render(props) {
      // @ts-ignore
      return /*#__PURE__*/React.createElement(Accordion, {
        elevation: 0
      }, _get(_getPrototypeOf(AccordionElement.prototype), "render", this).call(this, props));
    }
  }]);

  return AccordionElement;
}(UIElement);
