// @webiny/app-admin/ui/elements/ButtonGroupElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _get from "@babel/runtime/helpers/get";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { UIElement } from "../UIElement";
import styled from "@emotion/styled";
var ButtonGroup = /*#__PURE__*/styled("div", {
  target: "eztnaqv0",
  label: "ButtonGroup"
})({
  display: "flex",
  height: "100%",
  width: "100%",
  alignItems: "center",
  justifyContent: "center",
  "> button": {
    marginRight: 15
  }
});
export var ButtonGroupElement = /*#__PURE__*/function (_UIElement) {
  _inherits(ButtonGroupElement, _UIElement);

  var _super = _createSuper(ButtonGroupElement);

  function ButtonGroupElement(id) {
    var _this;

    _classCallCheck(this, ButtonGroupElement);

    _this = _super.call(this, id);

    _this.useGrid(false);

    return _this;
  }

  _createClass(ButtonGroupElement, [{
    key: "render",
    value: function render(props) {
      return /*#__PURE__*/React.createElement(ButtonGroup, null, _get(_getPrototypeOf(ButtonGroupElement.prototype), "render", this).call(this, props));
    }
  }]);

  return ButtonGroupElement;
}(UIElement);
