// @webiny/app-admin/ui/elements/SmallButtonElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { ButtonDefault, ButtonPrimary, ButtonSecondary } from "@webiny/ui/Button";
import { ButtonElement } from "./ButtonElement";
var BUTTONS = {
  default: ButtonDefault,
  primary: ButtonPrimary,
  secondary: ButtonSecondary
};
export var SmallButtonElement = /*#__PURE__*/function (_ButtonElement) {
  _inherits(SmallButtonElement, _ButtonElement);

  var _super = _createSuper(SmallButtonElement);

  function SmallButtonElement(id, config) {
    var _this;

    _classCallCheck(this, SmallButtonElement);

    _this = _super.call(this, id, config);

    _this.applyPlugins(SmallButtonElement);

    return _this;
  }

  _createClass(SmallButtonElement, [{
    key: "render",
    value: function render(props) {
      var Component = BUTTONS[this.getType()];

      var _onClick = this.getOnClick();

      return /*#__PURE__*/React.createElement(Component, {
        small: true,
        onClick: function onClick() {
          return _onClick(props);
        }
      }, this.getLabel());
    }
  }]);

  return SmallButtonElement;
}(ButtonElement);
