// @webiny/app-page-builder/render/plugins/elements/embeds/pinterest/PinterestEmbed.js
import _get from "lodash/get";
import React, { useEffect } from "react";
import { css } from "emotion";
import { useHandler } from "@webiny/app/hooks/useHandler";

function appendSDK(props) {
  var element = props.element;

  var _ref = _get(element, "data.source") || {},
      url = _ref.url;

  if (!url || window["PinUtils"]) {
    return Promise.resolve();
  }

  return new Promise(function (resolve) {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = encodeURI("https://assets.pinterest.com/js/pinit.js");
    script.setAttribute("async", "");
    script.setAttribute("charset", "utf-8");
    script.onload = resolve;
    document.body.appendChild(script);
  });
}

function initEmbed(props) {
  var element = props.element;
  var node = document.getElementById(element.id); // @ts-ignore

  if (node && window.PinUtils) {
    // @ts-ignore
    window.PinUtils.build();
  }
}

var centerAlign = /*#__PURE__*/css({
  textAlign: "center"
}, "label:centerAlign;");
export default (function (props) {
  var _ref2 = _get(props.element, "data.source") || {},
      url = _ref2.url;

  useEffect(function () {
    appendSDK(props).then(function () {
      return initEmbed(props);
    });
  }, []);
  var renderEmbed = useHandler(props, function (_ref3) {
    var element = _ref3.element;
    return function renderEmbed() {
      var data = _get(element, "data.source");

      return /*#__PURE__*/React.createElement("div", {
        id: element.id,
        className: centerAlign
      }, /*#__PURE__*/React.createElement("a", {
        "data-pin-do": "embedPin",
        "data-pin-width": data.size || "small",
        href: data.url
      }));
    };
  });
  return url ? renderEmbed() : null;
});
