// @webiny/app-page-builder/editor/plugins/elements/pagesList/index.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React from "react";
import styled from "@emotion/styled";
import kebabCase from "lodash/kebabCase";
import { DisplayMode } from "../../../../types";
import { createInitialPerDeviceSettingValue } from "../../elementSettings/elementSettingsUtils";
import { ReactComponent as PageListIcon } from "@svgr/webpack!./page-list-icon.svg";
import PagesList from "./PagesList";
import PagesListFilterSettings from "./PagesListFilterSettings";
import PagesListDesignSettings from "./PagesListDesignSettings";
export default (function () {
  var args = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var PreviewBox = /*#__PURE__*/styled("div", {
    target: "er2nsm10",
    label: "PreviewBox"
  })({
    textAlign: "center",
    margin: "0 auto",
    width: 100,
    svg: {
      width: 100
    }
  });
  var elementType = kebabCase(args.elementType || "pages-list");
  var defaultToolbar = {
    title: "List of pages",
    group: "pb-editor-element-group-basic",
    preview: function preview() {
      return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(PageListIcon, null));
    }
  };
  var defaultSettings = ["pb-editor-page-element-settings-delete"];
  return [{
    name: "pb-editor-page-element-".concat(elementType),
    type: "pb-editor-page-element",
    elementType: elementType,
    toolbar: typeof args.toolbar === "function" ? args.toolbar(defaultToolbar) : defaultToolbar,
    settings: typeof args.settings === "function" ? args.settings(defaultSettings) : defaultSettings,
    target: ["cell", "block"],
    onCreate: "open-settings",
    create: function create() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var defaultValue = _objectSpread({
        type: this.elementType,
        data: {
          resultsPerPage: 10,
          component: "default",
          settings: {
            margin: createInitialPerDeviceSettingValue({
              all: "0px"
            }, DisplayMode.DESKTOP),
            padding: createInitialPerDeviceSettingValue({
              all: "0px"
            }, DisplayMode.DESKTOP)
          }
        }
      }, options);

      return typeof args.create === "function" ? args.create(defaultValue) : defaultValue;
    },
    render: function render(_ref) {
      var element = _ref.element;
      return /*#__PURE__*/React.createElement(PagesList, {
        data: element.data
      });
    }
  }, {
    name: "pb-editor-page-element-advanced-settings-pages-list-filter",
    type: "pb-editor-page-element-advanced-settings",
    elementType: "pages-list",
    render: function render(props) {
      return /*#__PURE__*/React.createElement(PagesListFilterSettings, props);
    }
  }, {
    name: "pb-editor-page-element-advanced-settings-pages-list-design",
    type: "pb-editor-page-element-advanced-settings",
    elementType: "pages-list",
    render: function render(props) {
      return /*#__PURE__*/React.createElement(PagesListDesignSettings, props);
    }
  }];
});
