// @webiny/app-page-builder/editor/plugins/elements/pagesList/PagesListFilterSettings.js
import * as React from "react";
import { css } from "emotion";
import { TagsMultiAutocomplete } from "../../../../admin/components/TagsMultiAutocomplete";
import { CategoriesAutocomplete } from "../../../../admin/components/CategoriesAutocomplete";
import Accordion from "../../elementSettings/components/Accordion";
import Wrapper from "../../elementSettings/components/Wrapper";
import SelectField from "../../elementSettings/components/SelectField";
import { ButtonContainer, classes, COLORS, SimpleButton } from "../../elementSettings/components/StyledComponents";
import { Cell, Grid } from "@webiny/ui/Grid";
var autoCompleteStyle = /*#__PURE__*/css({
  "& .mdc-text-field": {
    height: "30px !important",
    padding: "0px !important"
  },
  "& .mdc-text-field__input": {
    padding: "4px 8px !important",
    border: "none !important",
    backgroundColor: "".concat(COLORS.lightGray, " !important"),
    caretColor: "inherit !important"
  },
  "& .mdc-floating-label": {
    display: "none"
  },
  "& .mdc-line-ripple": {
    display: "none"
  },
  "& .mdc-elevation--z1": {
    top: "30px !important"
  }
}, "label:autoCompleteStyle;");

var PagesListFilterSettings = function PagesListFilterSettings(_ref) {
  var Bind = _ref.Bind,
      submit = _ref.submit;
  return /*#__PURE__*/React.createElement(Accordion, {
    title: "Filter",
    defaultValue: true
  }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Wrapper, {
    label: "Category",
    containerClassName: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "category"
  }, /*#__PURE__*/React.createElement(CategoriesAutocomplete, {
    label: "Category",
    className: autoCompleteStyle
  }))), /*#__PURE__*/React.createElement(Wrapper, {
    label: "Sort By",
    containerClassName: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "sortBy",
    defaultValue: "publishedOn"
  }, function (_ref2) {
    var value = _ref2.value,
        onChange = _ref2.onChange;
    return /*#__PURE__*/React.createElement(SelectField, {
      value: value,
      onChange: onChange
    }, /*#__PURE__*/React.createElement("option", {
      value: "publishedOn"
    }, "Publishing date"), /*#__PURE__*/React.createElement("option", {
      value: "title"
    }, "Title"));
  })), /*#__PURE__*/React.createElement(Wrapper, {
    label: "Sort Direction",
    containerClassName: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "sortDirection",
    defaultValue: "desc"
  }, function (_ref3) {
    var value = _ref3.value,
        onChange = _ref3.onChange;
    return /*#__PURE__*/React.createElement(SelectField, {
      value: value,
      onChange: onChange
    }, /*#__PURE__*/React.createElement("option", {
      value: "desc"
    }, "Descending"), /*#__PURE__*/React.createElement("option", {
      value: "asc"
    }, "Ascending"));
  })), /*#__PURE__*/React.createElement(Wrapper, {
    label: "Tags",
    containerClassName: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "tags"
  }, /*#__PURE__*/React.createElement(TagsMultiAutocomplete, {
    className: autoCompleteStyle,
    label: "Filter by tags",
    description: "Enter tags to filter pages"
  }))), /*#__PURE__*/React.createElement(Wrapper, {
    label: "Filter by tags rule",
    containerClassName: classes.simpleGrid,
    leftCellSpan: 12,
    rightCellSpan: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "tagsRule",
    defaultValue: "all"
  }, function (_ref4) {
    var value = _ref4.value,
        onChange = _ref4.onChange;
    return /*#__PURE__*/React.createElement(SelectField, {
      value: value,
      onChange: onChange
    }, /*#__PURE__*/React.createElement("option", {
      value: "all"
    }, "Page must include all tags"), /*#__PURE__*/React.createElement("option", {
      value: "any"
    }, "Page must include any of the tags"));
  })), /*#__PURE__*/React.createElement(Grid, {
    className: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
    onClick: submit
  }, "Save"))))));
};

export default PagesListFilterSettings;
