// @webiny/app-page-builder/editor/plugins/elements/pagesList/PagesListDesignSettings.js
import * as React from "react";
import { plugins } from "@webiny/plugins";
import { validation } from "@webiny/validation";
import Accordion from "../../elementSettings/components/Accordion";
import Wrapper from "../../elementSettings/components/Wrapper";
import InputField from "../../elementSettings/components/InputField";
import SelectField from "../../elementSettings/components/SelectField";
import { ButtonContainer, classes, SimpleButton } from "../../elementSettings/components/StyledComponents";
import { Cell, Grid } from "@webiny/ui/Grid";

var PagesListDesignSettings = function PagesListDesignSettings(_ref) {
  var Bind = _ref.Bind,
      submit = _ref.submit;
  var components = plugins.byType("pb-page-element-pages-list-component");
  return /*#__PURE__*/React.createElement(Accordion, {
    title: "Design",
    defaultValue: true
  }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Wrapper, {
    label: "Component",
    containerClassName: classes.simpleGrid,
    leftCellSpan: 5,
    rightCellSpan: 7
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "component",
    defaultValue: components[0] ? components[0].componentName : null
  }, function (_ref2) {
    var value = _ref2.value,
        onChange = _ref2.onChange;
    return /*#__PURE__*/React.createElement(SelectField, {
      value: value,
      onChange: onChange,
      description: "Select a component to render the list"
    }, components.map(function (cmp) {
      return /*#__PURE__*/React.createElement("option", {
        key: cmp.name,
        value: cmp.componentName
      }, cmp.title);
    }));
  })), /*#__PURE__*/React.createElement(Wrapper, {
    label: "Results per page",
    containerClassName: classes.simpleGrid,
    leftCellSpan: 5,
    rightCellSpan: 7
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "resultsPerPage",
    validators: validation.create("numeric")
  }, /*#__PURE__*/React.createElement(InputField, null))), /*#__PURE__*/React.createElement(Grid, {
    className: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
    onClick: submit
  }, "Save"))))));
};

export default PagesListDesignSettings;
