// @webiny/app-page-builder/admin/plugins/menuItems/page/PageForm.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import { css as _css } from "emotion";
var _excluded = ["onChange"];
import * as React from "react";
import { Form } from "@webiny/form";
import { Input } from "@webiny/ui/Input";
import { Typography } from "@webiny/ui/Typography";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ButtonSecondary, ButtonPrimary } from "@webiny/ui/Button";
import { PagesAutocomplete } from "../../../components/PagesAutocomplete";
import { Elevation } from "@webiny/ui/Elevation";
import { validation } from "@webiny/validation";
var menuPageFormStyle = {
  color: "var(--mdc-theme-on-surface)",
  backgroundColor: "var(--mdc-theme-background) !important"
};

var LinkForm = function LinkForm(_ref) {
  var data = _ref.data,
      onSubmit = _ref.onSubmit,
      onCancel = _ref.onCancel;
  return /*#__PURE__*/React.createElement(Elevation, {
    z: 4,
    className: _css(menuPageFormStyle)
  }, /*#__PURE__*/React.createElement(Form, {
    data: data,
    onSubmit: onSubmit
  }, function (_ref2) {
    var submit = _ref2.submit,
        Bind = _ref2.Bind,
        data = _ref2.data,
        form = _ref2.form;
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Typography, {
      use: "overline"
    }, "Page menu item"))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "page",
      validators: validation.create("required")
    }, function (_ref3) {
      var _onChange = _ref3.onChange,
          rest = _objectWithoutProperties(_ref3, _excluded);

      return /*#__PURE__*/React.createElement(PagesAutocomplete, Object.assign({}, rest, {
        onChange: function onChange(value, selection) {
          _onChange(value);

          if (!data.title) {
            form.setState(function (state) {
              if (selection) {
                state.data.title = selection.title;
              }
            });
          }
        },
        label: "Page"
      }));
    }))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "title",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      label: "Title"
    })))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(ButtonSecondary, {
      onClick: onCancel
    }, "Cancel"), /*#__PURE__*/React.createElement(ButtonPrimary, {
      onClick: submit,
      style: {
        float: "right"
      }
    }, "Save menu item"))));
  }));
};

export default LinkForm;
