// @webiny/app-admin/components/SimpleForm/SimpleForm.js
import * as React from "react";
import styled from "@emotion/styled";
import { css } from "emotion";
import classNames from "classnames";
import { Cell, Grid } from "@webiny/ui/Grid";
import { Typography } from "@webiny/ui/Typography";
import { Elevation } from "@webiny/ui/Elevation";
import { Icon } from "@webiny/ui/Icon";
var SimpleFormContainer = /*#__PURE__*/styled("div", {
  target: "ep5jo8l0",
  label: "SimpleFormContainer"
})({
  position: "relative",
  margin: "17px 50px"
});
var header = /*#__PURE__*/css({
  borderBottom: "1px solid var(--mdc-theme-on-background)",
  color: "var(--mdc-theme-text-primary-on-background)"
}, "label:header;");
var title = /*#__PURE__*/css({
  display: "flex",
  alignItems: "center"
}, "label:title;");
var actions = /*#__PURE__*/css({
  display: "flex",
  justifyContent: "flex-end",
  alignItems: "center"
}, "label:actions;");
var icon = /*#__PURE__*/css({
  marginRight: 15,
  color: "var(--mdc-theme-text-primary-on-background)"
}, "label:icon;");
var footer = /*#__PURE__*/css({
  borderTop: "1px solid var(--mdc-theme-on-background)",
  color: "var(--mdc-theme-text-primary-on-background)",
  textAlign: "right",
  "&.mdc-layout-grid": {
    padding: 25 // "25px 50px"

  }
}, "label:footer;");
export var SimpleForm = function SimpleForm(props) {
  return /*#__PURE__*/React.createElement(SimpleFormContainer, {
    className: classNames("webiny-data-list", props.className),
    "data-testid": props["data-testid"]
  }, props.noElevation ? props.children : /*#__PURE__*/React.createElement(Elevation, {
    z: 1
  }, props.children));
};
export var SimpleFormHeader = function SimpleFormHeader(props) {
  return /*#__PURE__*/React.createElement(Grid, {
    className: header
  }, /*#__PURE__*/React.createElement(Cell, {
    span: props.children ? 6 : 12,
    className: title
  }, /*#__PURE__*/React.createElement(React.Fragment, null, props.icon && /*#__PURE__*/React.createElement(Icon, {
    className: icon,
    icon: props.icon
  }), /*#__PURE__*/React.createElement(Typography, {
    use: "headline5"
  }, props.title))), props.children && /*#__PURE__*/React.createElement(Cell, {
    span: 6,
    className: actions
  }, props.children));
};
export var SimpleFormFooter = function SimpleFormFooter(props) {
  return /*#__PURE__*/React.createElement(Grid, {
    className: footer
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, props.children));
};
export var SimpleFormContent = function SimpleFormContent(props) {
  return props.children;
};
