// @webiny/app-page-builder/editor/ui/views/SEOSettingsView.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import trim from "lodash/trim";
import { InputElement } from "@webiny/app-admin/ui/elements/form/InputElement";
import { PageSettingsFormView } from "./PageSettingsView/PageSettingsFormView";
import { validation } from "@webiny/validation";
import { SmallButtonElement } from "@webiny/app-admin/ui/elements/SmallButtonElement";
import { TypographyElement } from "@webiny/app-admin/ui/elements/TypographyElement";
import { LabelElement } from "@webiny/app-admin/ui/elements/LabelElement";
import { ButtonElement } from "@webiny/app-admin/ui/elements/ButtonElement";
import { DynamicFieldsetElement } from "@webiny/app-admin/ui/elements/form/DynamicFieldsetElement";
import { ButtonGroupElement } from "@webiny/app-admin/ui/elements/ButtonGroupElement";
export var SEOSettingsView = /*#__PURE__*/function (_PageSettingsFormView) {
  _inherits(SEOSettingsView, _PageSettingsFormView);

  var _super = _createSuper(SEOSettingsView);

  function SEOSettingsView() {
    var _this;

    _classCallCheck(this, SEOSettingsView);

    _this = _super.call(this, "SEOSettingsView");

    _this.setTitle("SEO Media");

    _this.addField(new InputElement("title", {
      name: "settings.seo.title",
      label: "Title",
      description: "SEO title."
    }));

    _this.addField(new InputElement("description", {
      name: "settings.seo.description",
      label: "Description",
      description: "SEO description."
    }));

    var metaTags = new DynamicFieldsetElement("metaTags", {
      name: "settings.seo.meta"
    });
    var header = new TypographyElement("metaTags.header", {
      typography: "button"
    });
    header.addElement(new LabelElement("metaTags.header.label", {
      text: "Meta tags"
    }));
    metaTags.setHeaderElement(header);
    metaTags.setCreateEmptyElement(function (_ref) {
      var actions = _ref.actions;
      var emptyElement = new TypographyElement("metaTags.empty", {
        typography: "button"
      });
      emptyElement.addElement(new LabelElement("metaTags.empty.label", {
        text: "To add other meta tags, click "
      }));
      emptyElement.addElement(new ButtonElement("metaTags.empty.buttonCreate", {
        label: "Add meta tag",
        type: "primary",
        onClick: actions.add()
      }));
      return emptyElement;
    });
    metaTags.setCreateRowElement(function (_ref2) {
      var row = _ref2.row,
          actions = _ref2.actions,
          index = _ref2.index;
      var nameElement = row.addElement(new InputElement("metaTags.".concat(index, ".name"), {
        name: "settings.seo.meta.".concat(index, ".name"),
        label: "Name",
        validators: function validators() {
          return validation.create("required");
        },
        beforeChange: function beforeChange(tag, cb) {
          return cb(trim(tag));
        }
      }));
      var contentElement = new InputElement("metaTags.".concat(index, ".content"), {
        name: "settings.seo.meta.".concat(index, ".content"),
        label: "Content",
        validators: function validators() {
          return validation.create("required");
        },
        beforeChange: function beforeChange(tag, cb) {
          return cb(trim(tag));
        }
      });
      contentElement.moveAfter(nameElement);
      var buttons = new ButtonGroupElement("metaTags.".concat(index, ".buttons"));
      buttons.addElement(new SmallButtonElement("metaTags.".concat(index, ".buttonAdd"), {
        type: "primary",
        label: "+",
        onClick: actions.add(index)
      }));
      buttons.addElement(new SmallButtonElement("metaTags.".concat(index, ".buttonRemove"), {
        type: "secondary",
        label: "-",
        onClick: actions.remove(index)
      }));
      buttons.moveAfter(contentElement);
      return row;
    });

    _this.addField(metaTags);

    _this.applyPlugins(SEOSettingsView);

    return _this;
  }

  return SEOSettingsView;
}(PageSettingsFormView);
