// @webiny/app-admin/components/FileManager/BottomInfoBar/UploadStatus.js
import React from "react";
import styled from "@emotion/styled";
import { CircularProgress } from "@webiny/ui/Progress";
var StatusWrapper = /*#__PURE__*/styled("div", {
  target: "e8hlyp40",
  label: "StatusWrapper"
})({
  color: "var(--mdc-theme-primary)",
  position: "absolute",
  right: 0,
  bottom: 10,
  marginRight: 10,
  display: "flex",
  alignItems: "center",
  "> div": {
    display: "inline-block"
  }
});
var CircularProgressHolder = /*#__PURE__*/styled("div", {
  target: "e8hlyp41",
  label: "CircularProgressHolder"
})({
  position: "relative",
  height: 12,
  width: 12
});
var UploadingLabel = /*#__PURE__*/styled("div", {
  target: "e8hlyp42",
  label: "UploadingLabel"
})({
  marginRight: 5
});

var UploadStatus = function UploadStatus(_ref) {
  var uploading = _ref.uploading;

  if (!uploading) {
    return null;
  }

  return /*#__PURE__*/React.createElement(StatusWrapper, null, /*#__PURE__*/React.createElement(UploadingLabel, null, "Uploading..."), /*#__PURE__*/React.createElement(CircularProgressHolder, null, /*#__PURE__*/React.createElement(CircularProgress, {
    size: 10,
    spinnerWidth: 1
  })));
};

export default UploadStatus;
