// @webiny/api-headless-cms-ddb-es/operations/settings/CmsSettingsDynamoElastic.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _configurations = _interopRequireDefault(require("../../configurations"));

var _error = _interopRequireDefault(require("@webiny/error"));

var _utils = require("../../utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

const convertToDbData = data => {
  return _objectSpread(_objectSpread({}, data), {}, {
    contentModelLastChange: data.contentModelLastChange.toISOString()
  });
};

const convertFromDbData = data => {
  if (!data) {
    return null;
  }

  let contentModelLastChange;

  try {
    contentModelLastChange = new Date(data.contentModelLastChange);
  } catch {
    contentModelLastChange = new Date();
  }

  return _objectSpread(_objectSpread({}, data), {}, {
    contentModelLastChange
  });
};

const SETTINGS_SECONDARY_KEY = "settings";

class CmsSettingsDynamoElastic {
  get context() {
    return this._context;
  }

  get partitionKey() {
    return `${(0, _utils.createBasePartitionKey)(this.context)}#SETTINGS`;
  }

  constructor({
    context
  }) {
    (0, _defineProperty2.default)(this, "_context", void 0);
    this._context = context;
  }

  async get() {
    const {
      db
    } = this.context;
    const [[settings]] = await db.read(_objectSpread(_objectSpread({}, _configurations.default.db()), {}, {
      query: {
        PK: this.partitionKey,
        SK: SETTINGS_SECONDARY_KEY
      }
    }));

    if (!settings) {
      return null;
    }

    return convertFromDbData(settings);
  }

  async create(data) {
    const {
      db
    } = this.context;
    const dbData = convertToDbData(data);

    try {
      await db.create(_objectSpread(_objectSpread({}, _configurations.default.db()), {}, {
        data: _objectSpread({
          PK: this.partitionKey,
          SK: SETTINGS_SECONDARY_KEY
        }, dbData)
      }));
    } catch (ex) {
      throw new _error.default(ex.message || "Could not create settings.", ex.code || "CREATE_SETTINGS_ERROR", {
        error: ex,
        data
      });
    }
  }

  async update(data) {
    const {
      db
    } = this.context;

    try {
      await db.update(_objectSpread(_objectSpread({}, _configurations.default.db()), {}, {
        query: {
          PK: this.partitionKey,
          SK: SETTINGS_SECONDARY_KEY
        },
        data: convertToDbData(data)
      }));
    } catch (ex) {
      throw new _error.default(ex.message || "Could not update settings.", ex.code || "UPDATE_SETTINGS_ERROR", {
        error: ex,
        data
      });
    }
  }

}

exports.default = CmsSettingsDynamoElastic;
