// @webiny/app-page-builder/editor/components/EmptyElementGroupView.js
import React from "react";
import styled from "@emotion/styled";
import { Typography } from "@webiny/ui/Typography";
var EmptyViewWrapper = /*#__PURE__*/styled("div", {
  target: "ekdncod0",
  label: "EmptyViewWrapper"
})({
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  padding: "48px 16px 16px",
  "& .media": {
    "& svg": {
      width: 75,
      height: 75,
      fill: "var(--mdc-theme-text-icon-on-background)"
    },
    marginBottom: 24
  },
  "& .title": {
    color: "var(--mdc-theme-text-primary-on-background)",
    marginBottom: 16
  },
  "& .body": {
    color: "var(--mdc-theme-text-secondary-on-background)",
    textAlign: "center"
  }
});

var EmptyElementGroupView = function EmptyElementGroupView(_ref) {
  var icon = _ref.icon,
      title = _ref.title,
      body = _ref.body;
  return /*#__PURE__*/React.createElement(EmptyViewWrapper, null, /*#__PURE__*/React.createElement("div", {
    className: "media"
  }, icon), /*#__PURE__*/React.createElement(Typography, {
    use: "headline6",
    className: "title"
  }, title), /*#__PURE__*/React.createElement(Typography, {
    use: "body1",
    className: "body"
  }, body));
};

export default EmptyElementGroupView;
