// @webiny/app-form-builder/admin/plugins/editor/triggers/components/WebhooksRequestsDynamicFieldset.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import React from "react";
import { DynamicFieldset } from "@webiny/ui/DynamicFieldset";
import { Input } from "@webiny/ui/Input";
import styled from "@emotion/styled";
import { Typography } from "@webiny/ui/Typography";
import { ButtonSecondary, ButtonDefault } from "@webiny/ui/Button";
import { validation } from "@webiny/validation";
var Fieldset = /*#__PURE__*/styled("div", {
  target: "ef1gvu10",
  label: "Fieldset"
})({
  position: "relative",
  width: "100%",
  marginBottom: 15,
  ".webiny-ui-button": {
    position: "absolute",
    display: "block",
    right: 10,
    top: 13
  }
});
var Header = /*#__PURE__*/styled("div", {
  target: "ef1gvu11",
  label: "Header"
})({
  display: "flex",
  justifyContent: "space-between",
  marginBottom: 15
});

var WebhooksRequestsDynamicFieldset = function WebhooksRequestsDynamicFieldset(props) {
  var onChange = props.onChange,
      value = props.value,
      Bind = props.Bind;

  var addUrl = function addUrl() {
    var newValue = Array.isArray(value) ? _toConsumableArray(value) : [];
    newValue.push("");
    onChange(newValue);
  };

  return /*#__PURE__*/React.createElement(DynamicFieldset, {
    value: value,
    onChange: onChange
  }, function (_ref) {
    var actions = _ref.actions,
        header = _ref.header,
        row = _ref.row,
        empty = _ref.empty;
    return /*#__PURE__*/React.createElement(React.Fragment, null, row(function (_ref2) {
      var index = _ref2.index;
      return /*#__PURE__*/React.createElement(Fieldset, null, /*#__PURE__*/React.createElement(Bind, {
        validators: validation.create("required,url"),
        name: "urls.".concat(index)
      }, /*#__PURE__*/React.createElement(Input, {
        label: props.inputLabel
      })), /*#__PURE__*/React.createElement(ButtonSecondary, {
        small: true,
        onClick: actions.remove(index)
      }, "remove"));
    }), empty(function () {
      return /*#__PURE__*/React.createElement(Header, null, /*#__PURE__*/React.createElement(Typography, {
        use: "overline"
      }, props.title), /*#__PURE__*/React.createElement(ButtonDefault, {
        onClick: addUrl
      }, props.addButtonLabel));
    }), header(function () {
      return /*#__PURE__*/React.createElement(Header, null, /*#__PURE__*/React.createElement(Typography, {
        use: "overline"
      }, props.title), /*#__PURE__*/React.createElement(ButtonDefault, {
        onClick: addUrl
      }, props.addButtonLabel));
    }));
  });
};

export default WebhooksRequestsDynamicFieldset;
