// @webiny/app-page-builder/admin/views/Categories/CategoriesDialog.js
import React from "react";
import { css } from "emotion";
import { useRouter } from "@webiny/react-router";
import { Query } from "@apollo/react-components";
import { Dialog, DialogTitle, DialogContent, DialogActions } from "@webiny/ui/Dialog";
import { List, ListItem, ListItemText, ListItemTextPrimary, ListItemTextSecondary } from "@webiny/ui/List";
import { ButtonDefault } from "@webiny/ui/Button";
import { LIST_CATEGORIES } from "./graphql";
var narrowDialog = /*#__PURE__*/css({
  ".mdc-dialog__surface": {
    width: 400,
    minWidth: 400
  }
}, "label:narrowDialog;");

var CategoriesDialog = function CategoriesDialog(_ref) {
  var open = _ref.open,
      onClose = _ref.onClose,
      onSelect = _ref.onSelect,
      children = _ref.children;

  var _useRouter = useRouter(),
      history = _useRouter.history;

  return /*#__PURE__*/React.createElement(Dialog, {
    open: open,
    onClose: onClose,
    className: narrowDialog,
    "data-testid": "pb-new-page-category-modal"
  }, children, /*#__PURE__*/React.createElement(DialogTitle, null, "Select a category"), /*#__PURE__*/React.createElement(DialogContent, null, /*#__PURE__*/React.createElement(List, {
    twoLine: true
  }, /*#__PURE__*/React.createElement(Query, {
    query: LIST_CATEGORIES
  }, function (_ref2) {
    var _data$pageBuilder, _data$pageBuilder$lis;

    var data = _ref2.data,
        loading = _ref2.loading;

    if (loading) {
      return /*#__PURE__*/React.createElement("span", null, "Loading categories...");
    }

    return /*#__PURE__*/React.createElement(React.Fragment, null, data === null || data === void 0 ? void 0 : (_data$pageBuilder = data.pageBuilder) === null || _data$pageBuilder === void 0 ? void 0 : (_data$pageBuilder$lis = _data$pageBuilder.listCategories) === null || _data$pageBuilder$lis === void 0 ? void 0 : _data$pageBuilder$lis.data.map(function (item) {
      return /*#__PURE__*/React.createElement(ListItem, {
        key: item.slug,
        onClick: function onClick() {
          return onSelect(item);
        }
      }, /*#__PURE__*/React.createElement(ListItemText, null, /*#__PURE__*/React.createElement(ListItemTextPrimary, null, item.name), /*#__PURE__*/React.createElement(ListItemTextSecondary, null, item.url)));
    }));
  }))), /*#__PURE__*/React.createElement(DialogActions, null, /*#__PURE__*/React.createElement(ButtonDefault, {
    onClick: function onClick() {
      return history.push("/page-builder/categories");
    }
  }, "+ Create new category")));
};

export default CategoriesDialog;
