// @webiny/app-page-builder/admin/views/Categories/CategoriesForm.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useCallback, useMemo } from "react";
import styled from "@emotion/styled";
import { i18n } from "@webiny/app/i18n";
import { Form } from "@webiny/form";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ButtonDefault, ButtonIcon, ButtonPrimary } from "@webiny/ui/Button";
import { CircularProgress } from "@webiny/ui/Progress";
import { useMutation, useQuery } from "@apollo/react-hooks";
import { SimpleForm, SimpleFormFooter, SimpleFormContent, SimpleFormHeader } from "@webiny/app-admin/components/SimpleForm";
import { validation } from "@webiny/validation";
import { GET_CATEGORY, CREATE_CATEGORY, UPDATE_CATEGORY, LIST_CATEGORIES } from "./graphql";
import { useRouter } from "@webiny/react-router";
import { useSnackbar } from "@webiny/app-admin/hooks/useSnackbar";
import { Input } from "@webiny/ui/Input";
import { categoryUrlValidator } from "./validators";
import { plugins } from "@webiny/plugins";
import { Select } from "@webiny/ui/Select";
import { useSecurity } from "@webiny/app-security";
import pick from "object.pick";
import get from "lodash/get";
import set from "lodash/set";
import isEmpty from "lodash/isEmpty";
import EmptyView from "@webiny/app-admin/components/EmptyView";
import { ReactComponent as AddIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/add-18px.svg";
var t = i18n.ns("app-page-builder/admin/categories/form");
var ButtonWrapper = /*#__PURE__*/styled("div", {
  target: "ec6in3r0",
  label: "ButtonWrapper"
})({
  display: "flex",
  justifyContent: "space-between"
});

var CategoriesForm = function CategoriesForm(_ref) {
  var _getQuery$data, _getQuery$data$pageBu, _getQuery$data$pageBu2;

  var canCreate = _ref.canCreate;

  var _useRouter = useRouter(),
      location = _useRouter.location,
      history = _useRouter.history;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var layouts = React.useMemo(function () {
    return plugins.byType("pb-page-layout").map(function (pl) {
      return pl.layout;
    });
  }, []);
  var newEntry = new URLSearchParams(location.search).get("new") === "true";
  var slug = new URLSearchParams(location.search).get("slug");
  var getQuery = useQuery(GET_CATEGORY, {
    variables: {
      slug: slug
    },
    skip: !slug,
    onCompleted: function onCompleted(data) {
      var _data$pageBuilder, _data$pageBuilder$get;

      var error = data === null || data === void 0 ? void 0 : (_data$pageBuilder = data.pageBuilder) === null || _data$pageBuilder === void 0 ? void 0 : (_data$pageBuilder$get = _data$pageBuilder.getCategory) === null || _data$pageBuilder$get === void 0 ? void 0 : _data$pageBuilder$get.error;

      if (error) {
        history.push("/page-builder/categories");
        showSnackbar(error.message);
      }
    }
  });
  var loadedCategory = ((_getQuery$data = getQuery.data) === null || _getQuery$data === void 0 ? void 0 : (_getQuery$data$pageBu = _getQuery$data.pageBuilder) === null || _getQuery$data$pageBu === void 0 ? void 0 : (_getQuery$data$pageBu2 = _getQuery$data$pageBu.getCategory) === null || _getQuery$data$pageBu2 === void 0 ? void 0 : _getQuery$data$pageBu2.data) || {};

  var _useMutation = useMutation(CREATE_CATEGORY, {
    refetchQueries: [{
      query: LIST_CATEGORIES
    }]
  }),
      _useMutation2 = _slicedToArray(_useMutation, 2),
      create = _useMutation2[0],
      createMutation = _useMutation2[1];

  var _useMutation3 = useMutation(UPDATE_CATEGORY, {
    refetchQueries: [{
      query: LIST_CATEGORIES
    }],
    update: function update(cache, _ref2) {
      var data = _ref2.data;
      var categoryDataFromCache = cache.readQuery({
        query: GET_CATEGORY,
        variables: {
          slug: slug
        }
      });
      var updatedCategoryData = get(data, "pageBuilder.category.data");

      if (updatedCategoryData) {
        cache.writeQuery({
          query: GET_CATEGORY,
          data: set(categoryDataFromCache, "pageBuilder.getCategory.data", updatedCategoryData)
        });
      }
    }
  }),
      _useMutation4 = _slicedToArray(_useMutation3, 2),
      update = _useMutation4[0],
      updateMutation = _useMutation4[1];

  var loading = [getQuery, createMutation, updateMutation].find(function (item) {
    return item.loading;
  });
  var onSubmit = useCallback( /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(formData) {
      var _response$data, _response$data$pageBu, _response$data$pageBu2;

      var isUpdate, data, _ref4, _ref5, operation, args, response, error;

      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              isUpdate = loadedCategory.slug;
              data = pick(formData, ["slug", "name", "url", "layout"]);
              _ref4 = isUpdate ? [update, {
                variables: {
                  slug: formData.slug,
                  data: data
                }
              }] : [create, {
                variables: {
                  data: data
                }
              }], _ref5 = _slicedToArray(_ref4, 2), operation = _ref5[0], args = _ref5[1];
              _context.next = 5;
              return operation(args);

            case 5:
              response = _context.sent;
              error = response === null || response === void 0 ? void 0 : (_response$data = response.data) === null || _response$data === void 0 ? void 0 : (_response$data$pageBu = _response$data.pageBuilder) === null || _response$data$pageBu === void 0 ? void 0 : (_response$data$pageBu2 = _response$data$pageBu.category) === null || _response$data$pageBu2 === void 0 ? void 0 : _response$data$pageBu2.error;

              if (!error) {
                _context.next = 9;
                break;
              }

              return _context.abrupt("return", showSnackbar(error.message));

            case 9:
              !isUpdate && history.push("/page-builder/categories?slug=".concat(formData.slug));
              showSnackbar(t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Category saved successfully."]))));

            case 11:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x) {
      return _ref3.apply(this, arguments);
    };
  }(), [loadedCategory.slug]);
  var data = useMemo(function () {
    var _getQuery$data2, _getQuery$data2$pageB;

    return ((_getQuery$data2 = getQuery.data) === null || _getQuery$data2 === void 0 ? void 0 : (_getQuery$data2$pageB = _getQuery$data2.pageBuilder) === null || _getQuery$data2$pageB === void 0 ? void 0 : _getQuery$data2$pageB.getCategory.data) || {};
  }, [loadedCategory.slug]);

  var _useSecurity = useSecurity(),
      identity = _useSecurity.identity;

  var pbMenuPermission = useMemo(function () {
    return identity.getPermission("pb.category");
  }, []);
  var canSave = useMemo(function () {
    // User should be able to save the form
    // if it's a new entry and user has the "own" permission set.
    if (!loadedCategory.slug && pbMenuPermission.own) {
      return true;
    }

    if (pbMenuPermission.own) {
      var _loadedCategory$creat;

      return (loadedCategory === null || loadedCategory === void 0 ? void 0 : (_loadedCategory$creat = loadedCategory.createdBy) === null || _loadedCategory$creat === void 0 ? void 0 : _loadedCategory$creat.id) === identity.login;
    }

    if (typeof pbMenuPermission.rwd === "string") {
      return pbMenuPermission.rwd.includes("w");
    }

    return true;
  }, [loadedCategory.slug]);
  var showEmptyView = !newEntry && !loading && isEmpty(data); // Render "No content selected" view.

  if (showEmptyView) {
    return /*#__PURE__*/React.createElement(EmptyView, {
      title: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Click on the left side list to display category details {message}"])))({
        message: canCreate ? "or create a..." : ""
      }),
      action: canCreate ? /*#__PURE__*/React.createElement(ButtonDefault, {
        "data-testid": "new-record-button",
        onClick: function onClick() {
          return history.push("/page-builder/categories?new=true");
        }
      }, /*#__PURE__*/React.createElement(ButtonIcon, {
        icon: /*#__PURE__*/React.createElement(AddIcon, null)
      }), " ", t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["New Category"])))) : null
    });
  }

  return /*#__PURE__*/React.createElement(Form, {
    data: data,
    onSubmit: onSubmit
  }, function (_ref6) {
    var data = _ref6.data,
        form = _ref6.form,
        Bind = _ref6.Bind;
    return /*#__PURE__*/React.createElement(SimpleForm, null, loading && /*#__PURE__*/React.createElement(CircularProgress, null), /*#__PURE__*/React.createElement(SimpleFormHeader, {
      title: data.name || t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["New category"])))
    }), /*#__PURE__*/React.createElement(SimpleFormContent, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 6
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "name",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      label: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Name"])))
    }))), /*#__PURE__*/React.createElement(Cell, {
      span: 6
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "slug",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      disabled: data.createdOn,
      label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Slug"])))
    }))), /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "url",
      validators: [validation.create("required"), categoryUrlValidator]
    }, /*#__PURE__*/React.createElement(Input, {
      disabled: data.id,
      label: t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["URL"])))
    }))), /*#__PURE__*/React.createElement(Cell, {
      span: 6
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "layout",
      defaultValue: layouts[0].name
    }, /*#__PURE__*/React.createElement(Select, {
      label: t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["Layout"])))
    }, layouts.map(function (_ref7) {
      var name = _ref7.name,
          title = _ref7.title;
      return /*#__PURE__*/React.createElement("option", {
        key: name,
        value: name
      }, title);
    })))))), /*#__PURE__*/React.createElement(SimpleFormFooter, null, /*#__PURE__*/React.createElement(ButtonWrapper, null, /*#__PURE__*/React.createElement(ButtonDefault, {
      onClick: function onClick() {
        return history.push("/page-builder/categories");
      }
    }, t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Cancel"])))), canSave && /*#__PURE__*/React.createElement(ButtonPrimary, {
      onClick: form.submit
    }, t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["Save category"])))))));
  });
};

export default CategoriesForm;
