// @webiny/app-page-builder/render/plugins/elements/pagesList/components/GridPageList.js
import _get from "lodash/get";
import * as React from "react";
import { Link } from "@webiny/react-router";
import { ReactComponent as PrevIcon } from "@svgr/webpack!./icons/round-navigate_before-24px.svg";
import { ReactComponent as NextIcon } from "@svgr/webpack!./icons/round-navigate_next-24px.svg";

var formatDate = function formatDate(date) {
  var d = new Date(date);
  var year = d.getFullYear();
  var month = d.getMonth();
  var day = d.getDate();
  var monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
  return monthNames[month] + " " + day + ", " + year;
};

var PageItem = function PageItem(_ref) {
  var data = _ref.data,
      className = _ref.className;

  var image = _get(data, "images.general.src");

  return /*#__PURE__*/React.createElement(Link, {
    to: data.url,
    className: "webiny-pb-page-element-page-list__item " + className
  }, image ? /*#__PURE__*/React.createElement("div", {
    className: "webiny-pb-page-element-page-list__media",
    style: {
      backgroundImage: "url(\"".concat(image, "?width=500\")")
    }
  }) : /*#__PURE__*/React.createElement("div", {
    className: "webiny-pb-page-element-page-list__media",
    style: {
      backgroundColor: "lightgray"
    }
  }), /*#__PURE__*/React.createElement("div", {
    className: "webiny-pb-page-element-page-list__content"
  }, /*#__PURE__*/React.createElement("h3", {
    className: "webiny-pb-page-element-page-list__title webiny-pb-typography-h3"
  }, data.title), /*#__PURE__*/React.createElement("p", {
    className: "webiny-pb-page-element-page-list__snippet webiny-pb-typography-description"
  }, data.snippet), /*#__PURE__*/React.createElement("div", {
    className: "webiny-pb-page-element-page-list__date webiny-pb-typography-description"
  }, formatDate(data.publishedOn))));
};

var GridPageList = function GridPageList(_ref2) {
  var data = _ref2.data,
      nextPage = _ref2.nextPage,
      prevPage = _ref2.prevPage;
  return /*#__PURE__*/React.createElement("div", {
    className: "webiny-pb-page-element-page-list webiny-pb-page-element-page-list--grid"
  }, /*#__PURE__*/React.createElement("div", {
    className: "webiny-pb-page-element-page-list__items"
  }, data.map(function (page) {
    return /*#__PURE__*/React.createElement(PageItem, {
      key: page.id,
      data: page
    });
  })), /*#__PURE__*/React.createElement("div", {
    className: "webiny-pb-page-element-page-list__navigation"
  }, prevPage && /*#__PURE__*/React.createElement("a", {
    onClick: prevPage
  }, /*#__PURE__*/React.createElement(PrevIcon, null), " Prev page"), nextPage && /*#__PURE__*/React.createElement("a", {
    onClick: nextPage
  }, "Next page ", /*#__PURE__*/React.createElement(NextIcon, null))));
};

export default GridPageList;
