// @webiny/app-page-builder/admin/views/Menus/MenusDataList.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useCallback, useMemo, useState } from "react";
import orderBy from "lodash/orderBy";
import { i18n } from "@webiny/app/i18n";
import { useRouter } from "@webiny/react-router";
import { useQuery, useMutation } from "@apollo/react-hooks";
import { LIST_MENUS, DELETE_MENU } from "./graphql";
import { useSnackbar } from "@webiny/app-admin/hooks/useSnackbar";
import { useConfirmationDialog } from "@webiny/app-admin/hooks/useConfirmationDialog";
import { useSecurity } from "@webiny/app-security";
import { DataList, DataListModalOverlay, DataListModalOverlayAction, ScrollList, ListItem, ListItemText, ListItemMeta, ListActions, ListItemTextSecondary } from "@webiny/ui/List";
import { DeleteIcon } from "@webiny/ui/List/DataList/icons";
import { ButtonIcon, ButtonSecondary } from "@webiny/ui/Button";
import { ReactComponent as AddIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/add-18px.svg";
import { ReactComponent as FilterIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/filter-24px.svg";
import SearchUI from "@webiny/app-admin/components/SearchUI";
import { Cell, Grid } from "@webiny/ui/Grid";
import { Select } from "@webiny/ui/Select";
import { serializeSorters, deserializeSorters } from "../utils";
var t = i18n.ns("app-page-builder/admin/menus/data-list");
var SORTERS = [{
  label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Newest to oldest"]))),
  sorters: {
    createdOn: "desc"
  }
}, {
  label: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Oldest to newest"]))),
  sorters: {
    createdOn: "asc"
  }
}, {
  label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Title A-Z"]))),
  sorters: {
    title: "asc"
  }
}, {
  label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Title Z-A"]))),
  sorters: {
    title: "desc"
  }
}];

var PageBuilderMenusDataList = function PageBuilderMenusDataList(_ref) {
  var _listQuery$data, _listQuery$data$pageB, _listQuery$data$pageB2;

  var canCreate = _ref.canCreate;

  var _useState = useState(""),
      _useState2 = _slicedToArray(_useState, 2),
      filter = _useState2[0],
      setFilter = _useState2[1];

  var _useState3 = useState(serializeSorters(SORTERS[0].sorters)),
      _useState4 = _slicedToArray(_useState3, 2),
      sort = _useState4[0],
      setSort = _useState4[1];

  var _useRouter = useRouter(),
      history = _useRouter.history;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var listQuery = useQuery(LIST_MENUS);

  var _useMutation = useMutation(DELETE_MENU, {
    refetchQueries: [{
      query: LIST_MENUS
    }]
  }),
      _useMutation2 = _slicedToArray(_useMutation, 2),
      deleteIt = _useMutation2[0],
      deleteMutation = _useMutation2[1];

  var _useConfirmationDialo = useConfirmationDialog(),
      showConfirmation = _useConfirmationDialo.showConfirmation;

  var filterMenus = useCallback(function (_ref2) {
    var title = _ref2.title,
        slug = _ref2.slug,
        description = _ref2.description;
    return title.toLowerCase().includes(filter) || slug.toLowerCase().includes(filter) || description && description.toLowerCase().includes(filter);
  }, [filter]);
  var sortMenus = useCallback(function (users) {
    if (!sort) {
      return users;
    }

    var _Object$entries = Object.entries(deserializeSorters(sort)),
        _Object$entries2 = _slicedToArray(_Object$entries, 1),
        _Object$entries2$ = _slicedToArray(_Object$entries2[0], 2),
        key = _Object$entries2$[0],
        value = _Object$entries2$[1];

    return orderBy(users, [key], [value]);
  }, [sort]);
  var data = (listQuery === null || listQuery === void 0 ? void 0 : (_listQuery$data = listQuery.data) === null || _listQuery$data === void 0 ? void 0 : (_listQuery$data$pageB = _listQuery$data.pageBuilder) === null || _listQuery$data$pageB === void 0 ? void 0 : (_listQuery$data$pageB2 = _listQuery$data$pageB.listMenus) === null || _listQuery$data$pageB2 === void 0 ? void 0 : _listQuery$data$pageB2.data) || [];
  var slug = new URLSearchParams(location.search).get("slug");
  var deleteItem = useCallback(function (item) {
    showConfirmation( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var _response$data, _response$data$pageBu, _response$data$pageBu2;

      var response, error;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return deleteIt({
                variables: item
              });

            case 2:
              response = _context.sent;
              error = response === null || response === void 0 ? void 0 : (_response$data = response.data) === null || _response$data === void 0 ? void 0 : (_response$data$pageBu = _response$data.pageBuilder) === null || _response$data$pageBu === void 0 ? void 0 : (_response$data$pageBu2 = _response$data$pageBu.deletePageBuilderMenu) === null || _response$data$pageBu2 === void 0 ? void 0 : _response$data$pageBu2.error;

              if (!error) {
                _context.next = 6;
                break;
              }

              return _context.abrupt("return", showSnackbar(error.message));

            case 6:
              showSnackbar(t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Menu \"{slug}\" deleted."])))({
                slug: item.slug
              }));

              if (slug === item.slug) {
                history.push("/page-builder/menus");
              }

            case 8:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    })));
  }, [slug]);

  var _useSecurity = useSecurity(),
      identity = _useSecurity.identity;

  var pbMenuPermission = useMemo(function () {
    return identity.getPermission("pb.menu");
  }, []);
  var canDelete = useCallback(function (item) {
    if (pbMenuPermission.own) {
      return item.createdBy.id === identity.login;
    }

    if (typeof pbMenuPermission.rwd === "string") {
      return pbMenuPermission.rwd.includes("d");
    }

    return true;
  }, []);
  var menusDataListModalOverlay = useMemo(function () {
    return /*#__PURE__*/React.createElement(DataListModalOverlay, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Select, {
      value: sort,
      onChange: setSort,
      label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Sort by"]))),
      description: "Sort pages by"
    }, SORTERS.map(function (_ref4) {
      var label = _ref4.label,
          sorters = _ref4.sorters;
      return /*#__PURE__*/React.createElement("option", {
        key: label,
        value: serializeSorters(sorters)
      }, label);
    })))));
  }, [sort]);
  var loading = [listQuery, deleteMutation].find(function (item) {
    return item.loading;
  });
  var filteredData = filter === "" ? data : data.filter(filterMenus);
  var menuList = sortMenus(filteredData);
  return /*#__PURE__*/React.createElement(DataList, {
    loading: Boolean(loading),
    data: menuList,
    title: t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Menus"]))),
    actions: canCreate ? /*#__PURE__*/React.createElement(ButtonSecondary, {
      "data-testid": "data-list-new-record-button",
      onClick: function onClick() {
        return history.push("/page-builder/menus?new=true");
      }
    }, /*#__PURE__*/React.createElement(ButtonIcon, {
      icon: /*#__PURE__*/React.createElement(AddIcon, null)
    }), " ", t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["New Menu"])))) : null,
    search: /*#__PURE__*/React.createElement(SearchUI, {
      value: filter,
      onChange: setFilter,
      inputPlaceholder: t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Search menus"])))
    }),
    modalOverlay: menusDataListModalOverlay,
    modalOverlayAction: /*#__PURE__*/React.createElement(DataListModalOverlayAction, {
      icon: /*#__PURE__*/React.createElement(FilterIcon, null),
      "data-testid": "default-data-list.filter"
    })
  }, function (_ref5) {
    var data = _ref5.data;
    return /*#__PURE__*/React.createElement(ScrollList, {
      "data-testid": "default-data-list"
    }, data.map(function (item) {
      return /*#__PURE__*/React.createElement(ListItem, {
        key: item.slug,
        selected: item.slug === slug
      }, /*#__PURE__*/React.createElement(ListItemText, {
        onClick: function onClick() {
          return history.push("/page-builder/menus?slug=".concat(item.slug));
        }
      }, item.title, /*#__PURE__*/React.createElement(ListItemTextSecondary, null, item.description || t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["No description provided."]))))), canDelete(item) && /*#__PURE__*/React.createElement(ListItemMeta, null, /*#__PURE__*/React.createElement(ListActions, null, /*#__PURE__*/React.createElement(DeleteIcon, {
        onClick: function onClick() {
          return deleteItem(item);
        }
      }))));
    }));
  });
};

export default PageBuilderMenusDataList;
