// @webiny/app-page-builder/admin/views/Menus/MenusDialog.js
import React from "react";
import { css } from "emotion";
import { useRouter } from "@webiny/react-router";
import { Query } from "@apollo/react-components";
import { Dialog, DialogTitle, DialogContent, DialogActions } from "@webiny/ui/Dialog";
import { List, ListItem, ListItemText, ListItemTextPrimary, ListItemTextSecondary } from "@webiny/ui/List";
import { ButtonDefault } from "@webiny/ui/Button";
import { LIST_MENUS } from "./graphql";
var narrowDialog = /*#__PURE__*/css({
  ".mdc-dialog__surface": {
    width: 400,
    minWidth: 400
  }
}, "label:narrowDialog;");

var MenusDialog = function MenusDialog(_ref) {
  var open = _ref.open,
      onClose = _ref.onClose,
      onSelect = _ref.onSelect,
      children = _ref.children;

  var _useRouter = useRouter(),
      history = _useRouter.history;

  return /*#__PURE__*/React.createElement(Dialog, {
    open: open,
    onClose: onClose,
    className: narrowDialog,
    "data-testid": "pb-new-page-menu-modal"
  }, children, /*#__PURE__*/React.createElement(DialogTitle, null, "Select a menu"), /*#__PURE__*/React.createElement(DialogContent, null, /*#__PURE__*/React.createElement(List, {
    twoLine: true
  }, /*#__PURE__*/React.createElement(Query, {
    query: LIST_MENUS
  }, function (_ref2) {
    var data = _ref2.data,
        loading = _ref2.loading;

    if (loading) {
      return /*#__PURE__*/React.createElement("span", null, "Loading menus...");
    }

    return /*#__PURE__*/React.createElement(React.Fragment, null, data.pageBuilder.menus.data.map(function (item) {
      return /*#__PURE__*/React.createElement(ListItem, {
        key: item.id,
        onClick: function onClick() {
          return onSelect(item);
        }
      }, /*#__PURE__*/React.createElement(ListItemText, null, /*#__PURE__*/React.createElement(ListItemTextPrimary, null, item.name), /*#__PURE__*/React.createElement(ListItemTextSecondary, null, item.url)));
    }));
  }))), /*#__PURE__*/React.createElement(DialogActions, null, /*#__PURE__*/React.createElement(ButtonDefault, {
    onClick: function onClick() {
      return history.push("/page-builder/menus");
    }
  }, "+ Create new menu")));
};

export default MenusDialog;
