// @webiny/app-page-builder/editor/plugins/elements/button/SimpleEditableText.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React, { useCallback, useRef } from "react";

var SimpleEditableText = function SimpleEditableText(_ref) {
  var _ref$value = _ref.value,
      defaultValue = _ref$value === void 0 ? "" : _ref$value,
      onFocus = _ref.onFocus,
      onBlur = _ref.onBlur,
      onChange = _ref.onChange,
      _ref$options = _ref.options,
      options = _ref$options === void 0 ? {} : _ref$options,
      element = _ref.element,
      className = _ref.className;
  var value = useRef(defaultValue);
  var inputRef = useRef(null);
  var onChangeHandler = useCallback(function (ev) {
    var target = ev.target; // Remove space from HTML

    var elementValue = target.innerHTML.replace(/&nbsp;/gi, "") || "";
    ev.preventDefault();
    ev.stopPropagation();

    if (elementValue === value.current) {
      return false;
    }

    value.current = elementValue;
    onChange(elementValue);
  }, [onChange]);
  var onBlurHandler = useCallback(function () {
    if (!onBlur) {
      return;
    }

    onBlur();
  }, [onBlur]);
  var onFocusHandler = useCallback(function () {
    if (!onFocus) {
      return;
    }

    onFocus();
  }, [onFocus]);
  return /*#__PURE__*/React.createElement(element || "div", _objectSpread({
    className: className,
    contentEditable: true,
    onInput: onChangeHandler,
    onBlur: onBlurHandler,
    onFocus: onFocusHandler,
    dangerouslySetInnerHTML: {
      __html: value.current
    },
    ref: inputRef
  }, options));
};

export default /*#__PURE__*/React.memo(SimpleEditableText);
