// @webiny/app-page-builder/editor/components/Editor/DefaultEditorBar.js
import React from "react";
import { TopAppBar, TopAppBarSection } from "@webiny/ui/TopAppBar";
import { renderPlugins } from "@webiny/app/plugins";
import { css } from "emotion";
var topBar = /*#__PURE__*/css({
  boxShadow: "1px 0px 5px 0px rgba(128,128,128,1)"
}, "label:topBar;");
var centerTopBar = /*#__PURE__*/css({
  "&.mdc-top-app-bar__section": {
    paddingTop: 0,
    paddingBottom: 0
  }
}, "label:centerTopBar;");

var DefaultEditorBar = function DefaultEditorBar() {
  return /*#__PURE__*/React.createElement(TopAppBar, {
    className: topBar,
    fixed: true
  }, /*#__PURE__*/React.createElement(TopAppBarSection, {
    style: {
      width: "33%"
    },
    alignStart: true
  }, renderPlugins("pb-editor-default-bar-left")), /*#__PURE__*/React.createElement(TopAppBarSection, {
    className: centerTopBar
  }, renderPlugins("pb-editor-default-bar-center")), /*#__PURE__*/React.createElement(TopAppBarSection, {
    style: {
      width: "33%"
    },
    alignEnd: true
  }, renderPlugins("pb-editor-default-bar-right")));
};

export default /*#__PURE__*/React.memo(DefaultEditorBar);
