// @webiny/app-headless-cms/admin/plugins/fields/object/ObjectFields.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React, { useCallback } from "react";
import { set } from "dot-prop-immutable";
import { FieldEditor, useFieldEditor } from "../../../components/FieldEditor";
export var ObjectFields = function ObjectFields(_ref) {
  var field = _ref.field;

  var _useFieldEditor = useFieldEditor(),
      getField = _useFieldEditor.getField,
      updateField = _useFieldEditor.updateField;

  var onChange = useCallback(function (_ref2) {
    var fields = _ref2.fields,
        layout = _ref2.layout;
    var currentField = getField({
      id: field.id
    });
    var updatedField = set(currentField, "settings", function (settings) {
      return _objectSpread(_objectSpread({}, settings), {}, {
        fields: fields,
        layout: layout
      });
    });
    updateField(updatedField);
  }, [field]);
  return /*#__PURE__*/React.createElement(FieldEditor, {
    parent: field,
    fields: field.settings.fields,
    layout: field.settings.layout,
    onChange: onChange
  });
};
