// @webiny/app-security-admin-users-cognito/views/RequireNewPassword.js
import * as React from "react";
import { Form } from "@webiny/form";
import { validation } from "@webiny/validation";
import { ButtonPrimary } from "@webiny/ui/Button";
import { Input } from "@webiny/ui/Input";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Typography } from "@webiny/ui/Typography";
import { Elevation } from "@webiny/ui/Elevation";
import { useAuthenticator } from "@webiny/app-security-cognito-authentication/hooks/useAuthenticator";
import { useRequireNewPassword } from "@webiny/app-security-cognito-authentication/hooks/useRequireNewPassword";
import StateContainer from "./StateContainer";
import { alignRight, InnerContent, Title } from "./StyledComponents";

var sentenceCase = function sentenceCase(str) {
  var lower = str.toLowerCase();
  return lower[0].toUpperCase() + lower.substring(1);
};

var RequireNewPassword = function RequireNewPassword() {
  var _useAuthenticator = useAuthenticator(),
      checkingUser = _useAuthenticator.checkingUser;

  var _useRequireNewPasswor = useRequireNewPassword(),
      shouldRender = _useRequireNewPasswor.shouldRender,
      requiredAttributes = _useRequireNewPasswor.requiredAttributes,
      confirm = _useRequireNewPasswor.confirm;

  if (!shouldRender || checkingUser) {
    return null;
  }

  return /*#__PURE__*/React.createElement(StateContainer, null, /*#__PURE__*/React.createElement(Form, {
    onSubmit: function onSubmit(_ref) {
      var password = _ref.password,
          requiredAttributes = _ref.requiredAttributes;
      return confirm({
        password: password,
        requiredAttributes: requiredAttributes
      });
    },
    submitOnEnter: true
  }, function (_ref2) {
    var Bind = _ref2.Bind,
        submit = _ref2.submit;
    return /*#__PURE__*/React.createElement(Elevation, {
      z: 2
    }, /*#__PURE__*/React.createElement(InnerContent, null, /*#__PURE__*/React.createElement(Title, null, /*#__PURE__*/React.createElement(Typography, {
      use: "headline4"
    }, "Set New Password")), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "password",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      type: "password",
      label: "New password",
      outlined: true
    })))), requiredAttributes.length > 0 && /*#__PURE__*/React.createElement(Title, null, /*#__PURE__*/React.createElement(Typography, {
      use: "headline6"
    }, "Please enter additional information")), /*#__PURE__*/React.createElement(Grid, null, requiredAttributes.map(function (name) {
      return /*#__PURE__*/React.createElement(Cell, {
        key: name,
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: name,
        validators: validation.create("required")
      }, /*#__PURE__*/React.createElement(Input, {
        label: sentenceCase(name),
        outlined: true
      })));
    })), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12,
      className: alignRight
    }, /*#__PURE__*/React.createElement(ButtonPrimary, {
      onClick: submit
    }, "Set password")))));
  }));
};

export default RequireNewPassword;
