// @webiny/app-security-admin-users-cognito/views/SignIn.js
import * as React from "react";
import { Form } from "@webiny/form";
import { validation } from "@webiny/validation";
import { ButtonPrimary } from "@webiny/ui/Button";
import { Input } from "@webiny/ui/Input";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Typography } from "@webiny/ui/Typography";
import { Elevation } from "@webiny/ui/Elevation";
import { Alert } from "@webiny/ui/Alert";
import { CircularProgress } from "@webiny/ui/Progress";
import { useAuthenticator } from "@webiny/app-security-cognito-authentication/hooks/useAuthenticator";
import { useSignIn } from "@webiny/app-security-cognito-authentication/hooks/useSignIn";
import StateContainer from "./StateContainer";
import { alignRight, alignCenter, InnerContent, Title, errorMessage } from "./StyledComponents";

var SignIn = function SignIn() {
  var _useAuthenticator = useAuthenticator(),
      message = _useAuthenticator.message,
      changeState = _useAuthenticator.changeState,
      checkingUser = _useAuthenticator.checkingUser;

  var _useSignIn = useSignIn(),
      signIn = _useSignIn.signIn,
      loading = _useSignIn.loading,
      error = _useSignIn.error,
      shouldRender = _useSignIn.shouldRender;

  if (!shouldRender || checkingUser) {
    return null;
  }

  return /*#__PURE__*/React.createElement(StateContainer, null, /*#__PURE__*/React.createElement(Form, {
    onSubmit: signIn,
    submitOnEnter: true
  }, function (_ref) {
    var Bind = _ref.Bind,
        submit = _ref.submit;
    return /*#__PURE__*/React.createElement(Elevation, {
      z: 2
    }, /*#__PURE__*/React.createElement(InnerContent, null, loading && /*#__PURE__*/React.createElement(CircularProgress, null), /*#__PURE__*/React.createElement(Title, null, /*#__PURE__*/React.createElement("h1", null, /*#__PURE__*/React.createElement(Typography, {
      use: "headline4"
    }, "Sign In"))), message && !error && /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Alert, {
      title: message.title,
      type: message.type
    }, message.text))), error && /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12,
      className: errorMessage
    }, /*#__PURE__*/React.createElement(Alert, {
      title: "Authentication error",
      type: "danger"
    }, error.message))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "username",
      validators: validation.create("required,email"),
      beforeChange: function beforeChange(val, cb) {
        return cb(val.toLowerCase());
      }
    }, /*#__PURE__*/React.createElement(Input, {
      label: "Your e-mail"
    }))), /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "password",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      type: "password",
      label: "Your password"
    }))), /*#__PURE__*/React.createElement(Cell, {
      span: 12,
      className: alignRight
    }, /*#__PURE__*/React.createElement(ButtonPrimary, {
      "data-testid": "submit-sign-in-form-button",
      onClick: submit
    }, "Submit")), /*#__PURE__*/React.createElement(Cell, {
      span: 12,
      className: alignCenter
    }, /*#__PURE__*/React.createElement("a", {
      href: "#",
      onClick: function onClick() {
        return changeState("forgotPassword");
      }
    }, "Forgot password?")))));
  }));
};

export default SignIn;
