// @webiny/app-security-admin-users-cognito/views/StyledComponents.js
import styled from "@emotion/styled";
import { css } from "emotion";
export var Wrapper = /*#__PURE__*/styled("section", {
  target: "e1fvc3ia0",
  label: "Wrapper"
})({
  display: "flex",
  flexDirection: "column",
  justifyContent: "center",
  minHeight: "100vh",
  color: "var(--mdc-theme-on-surface)"
});
export var Logo = /*#__PURE__*/styled("img", {
  target: "e1fvc3ia1",
  label: "Logo"
})({
  margin: "0 auto",
  marginBottom: 30,
  width: 125
});
export var LoginContent = /*#__PURE__*/styled("div", {
  target: "e1fvc3ia2",
  label: "LoginContent"
})({
  width: "100%",
  maxWidth: 500,
  margin: "0 auto 25px auto",
  ".mdc-elevation--z2": {
    borderRadius: 4,
    boxShadow: "0 1px 3px 0 rgba(0,0,0,0.15)"
  },
  a: {
    textDecoration: "none",
    color: "var(--mdc-theme-primary)",
    fontWeight: 600,
    "&:hover": {
      textDecoration: "underline"
    }
  }
});
export var InnerContent = /*#__PURE__*/styled("div", {
  target: "e1fvc3ia3",
  label: "InnerContent"
})({
  position: "relative",
  padding: 25
});
export var Footer = /*#__PURE__*/styled("div", {
  target: "e1fvc3ia4",
  label: "Footer"
})({
  textAlign: "center",
  marginBottom: 75,
  a: {
    textDecoration: "none",
    color: "var(--mdc-theme-primary)"
  }
});
export var Title = /*#__PURE__*/styled("div", {
  target: "e1fvc3ia5",
  label: "Title"
})({
  textAlign: "center",
  margin: "10px 25px"
});
export var alignRight = /*#__PURE__*/css({
  textAlign: "right"
}, "label:alignRight;");
export var alignCenter = /*#__PURE__*/css({
  textAlign: "center"
}, "label:alignCenter;");
export var errorMessage = /*#__PURE__*/css({
  color: "red"
}, "label:errorMessage;");
