// @webiny/app-headless-cms/admin/plugins/fieldRenderers/longText/longText.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import get from "lodash/get";
import { Input } from "@webiny/ui/Input";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/fields/text");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-long-text-textarea",
  renderer: {
    rendererName: "long-text-text-area",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Text Area"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders a simple text area, suitable for larger amounts of text."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "long-text" && !field.multipleValues && !get(field, "predefinedValues.enabled");
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind;
      var Bind = getBind();
      return /*#__PURE__*/React.createElement(Bind, null, function (bind) {
        return /*#__PURE__*/React.createElement(Input, Object.assign({}, bind, {
          rows: 5,
          label: field.label,
          placeholder: field.placeholderText,
          description: field.helpText
        }));
      });
    }
  }
};
export default plugin;
