// @webiny/app-form-builder/page-builder/admin/plugins/components/FormElement.js
import React from "react";
import styled from "@emotion/styled";
import { ElementRoot } from "@webiny/app-page-builder/render/components/ElementRoot";
import { Form as FormsForm } from "../../../../components/Form";
import useRenderEmptyEmbed from "@webiny/app-page-builder/editor/plugins/elements/utils/oembed/useRenderEmptyEmbed";
var Overlay = /*#__PURE__*/styled("div", {
  target: "e10tbqra0",
  label: "Overlay"
})({
  background: "black",
  position: "absolute",
  width: "100%",
  height: "100%",
  zIndex: 10,
  opacity: 0.25
});

var renderContent = function renderContent(form) {
  var props = {
    preview: true,
    parentId: form.revision === "latest" ? form.parent : undefined,
    revisionId: form.revision === "latest" ? undefined : form.revision
  };
  return /*#__PURE__*/React.createElement(FormsForm, props);
};

var FormElement = function FormElement(_ref) {
  var _element$data;

  var element = _ref.element,
      isActive = _ref.isActive;

  var _ref2 = ((_element$data = element.data) === null || _element$data === void 0 ? void 0 : _element$data.settings) || {},
      _ref2$form = _ref2.form,
      form = _ref2$form === void 0 ? {} : _ref2$form;

  var renderEmpty = useRenderEmptyEmbed(element);
  return /*#__PURE__*/React.createElement(React.Fragment, null, isActive && /*#__PURE__*/React.createElement(Overlay, null), /*#__PURE__*/React.createElement(ElementRoot, {
    key: "form-".concat(form.parent, "-").concat(form.revision),
    element: element,
    className: "webiny-pb-element-form"
  }, form.revision ? renderContent(form) : renderEmpty()));
};

export default /*#__PURE__*/React.memo(FormElement);
