// @webiny/app-form-builder/admin/plugins/editor/formFields/checkboxes.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ReactComponent as Icon } from "@svgr/webpack!./icons/round-check_box-24px.svg";
import OptionsList from "./components/OptionsList";
var plugin = {
  type: "form-editor-field-type",
  name: "form-editor-field-type-checkbox",
  field: {
    type: "checkbox",
    name: "checkbox",
    validators: ["required"],
    label: "Checkboxes",
    description: "Choose one or more options",
    icon: /*#__PURE__*/React.createElement(Icon, null),
    createField: function createField() {
      return {
        type: this.type,
        name: this.name,
        validation: [],
        settings: {
          defaultValue: []
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var form = _ref.form;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(OptionsList, {
        form: form,
        multiple: true
      })));
    }
  }
};
export default plugin;
