// @webiny/app-form-builder/admin/plugins/editor/formFields/hidden.js
import React from "react";
import { Input } from "@webiny/ui/Input";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ReactComponent as HiddenIcon } from "@svgr/webpack!./icons/round-visibility_off-24px.svg";
var plugin = {
  type: "form-editor-field-type",
  name: "form-editor-field-type-hidden",
  field: {
    type: "hidden",
    name: "hidden",
    label: "Hidden",
    description: "Predefined values, hidden text",
    icon: /*#__PURE__*/React.createElement(HiddenIcon, null),
    createField: function createField() {
      return {
        type: this.type,
        name: this.name,
        validation: [],
        settings: {
          defaultValue: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.form.Bind;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "settings.defaultValue"
      }, /*#__PURE__*/React.createElement(Input, {
        label: "Default value",
        description: "Default value (optional)"
      }))));
    }
  }
};
export default plugin;
