// @webiny/app-form-builder/admin/plugins/editor/formFields/number.js
import React from "react";
import { ReactComponent as NumberIcon } from "@svgr/webpack!./icons/round-looks_3-24px.svg";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Input } from "@webiny/ui/Input";
var plugin = {
  type: "form-editor-field-type",
  name: "form-editor-field-type-number",
  field: {
    type: "number",
    name: "number",
    label: "Number",
    description: "ID, order number, rating, quantity",
    icon: /*#__PURE__*/React.createElement(NumberIcon, null),
    validators: ["required", "gte", "lte", "in"],
    createField: function createField() {
      return {
        type: this.type,
        name: this.name,
        validation: [],
        settings: {
          defaultValue: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.form.Bind;
      // TODO: @ts-adrian: spread Bind komponente na donju komponentu
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "placeholderText"
      }, /*#__PURE__*/React.createElement(Input, {
        label: "Placeholder text",
        description: "Placeholder text (optional)"
      }))));
    }
  }
};
export default plugin;
