// @webiny/app-form-builder/admin/plugins/editor/formFields/radioButtons.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ReactComponent as Icon } from "@svgr/webpack!./icons/round-radio_button_checked-24px.svg";
import OptionsList from "./components/OptionsList";
var plugin = {
  type: "form-editor-field-type",
  name: "form-editor-field-type-radio",
  field: {
    type: "radio",
    name: "radio",
    validators: ["required"],
    label: "Radio",
    description: "Choose a single option",
    icon: /*#__PURE__*/React.createElement(Icon, null),
    createField: function createField() {
      return {
        type: this.type,
        name: this.name,
        validation: [],
        settings: {
          defaultValue: []
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var form = _ref.form;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(OptionsList, {
        form: form
      })));
    }
  }
};
export default plugin;
