// @webiny/app-form-builder/admin/plugins/editor/formFields/select.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ReactComponent as Icon } from "@svgr/webpack!./icons/dropdown-icon.svg";
import OptionsList from "./components/OptionsList";
import { Input } from "@webiny/ui/Input";
var plugin = {
  type: "form-editor-field-type",
  name: "form-editor-field-type-select",
  field: {
    type: "select",
    name: "select",
    validators: ["required"],
    label: "Select",
    description: "Dropdown, select one of the options",
    icon: /*#__PURE__*/React.createElement(Icon, null),
    createField: function createField() {
      return {
        type: this.type,
        name: this.name,
        validation: [],
        settings: {
          defaultValue: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var form = _ref.form;
      var Bind = form.Bind; // TODO: @ts-adrian: spread Bind komponente na donju komponentu

      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "placeholderText"
      }, /*#__PURE__*/React.createElement(Input, {
        label: "Placeholder text",
        description: "Placeholder text (optional)"
      }))), /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(OptionsList, {
        form: form
      })));
    }
  }
};
export default plugin;
