// @webiny/app-form-builder/admin/plugins/editor/formFields/textarea.js
import React from "react";
import { ReactComponent as TextIcon } from "@svgr/webpack!./icons/round-text_fields-24px.svg";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Input } from "@webiny/ui/Input";
var plugin = {
  type: "form-editor-field-type",
  name: "form-editor-field-type-textarea",
  field: {
    name: "textarea",
    type: "textarea",
    validators: ["required", "minLength", "maxLength", "pattern"],
    label: "Long Text",
    description: "Descriptions, comments or paragraphs or text",
    icon: /*#__PURE__*/React.createElement(TextIcon, null),
    createField: function createField() {
      return {
        type: this.type,
        name: this.name,
        validation: [],
        settings: {
          defaultValue: ""
        }
      };
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.form.Bind;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "placeholderText"
      }, /*#__PURE__*/React.createElement(Input, {
        label: "Placeholder text",
        description: "Placeholder text (optional)"
      }))), /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "settings.defaultValue"
      }, /*#__PURE__*/React.createElement(Input, {
        rows: 4,
        label: "Default value",
        description: "Default value (optional)"
      }))), /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "settings.rows"
      }, /*#__PURE__*/React.createElement(Input, {
        type: "number",
        label: "Textarea rows",
        description: "Default value (optional)"
      }))));
    }
  }
};
export default plugin;
