// @webiny/app-admin/components/FileManager/BottomInfoBar.js
import React from "react";
import mime from "mime/lite";
import styled from "@emotion/styled";
import SupportedFileTypes from "./BottomInfoBar/SupportedFileTypes";
import UploadStatus from "./BottomInfoBar/UploadStatus";
mime.define({
  "image/x-icon": ["ico"]
}, true);
mime.define({
  "image/jpg": ["jpg"]
}, true);
mime.define({
  "image/vnd.microsoft.icon": ["ico"]
}, true);
var BottomInfoBarWrapper = /*#__PURE__*/styled("div", {
  target: "el4b0yq0",
  label: "BottomInfoBarWrapper"
})({
  fontSize: "0.8rem",
  position: "sticky",
  bottom: 0,
  height: 30,
  color: "var(--mdc-theme-text-secondary-on-background)",
  borderTop: "1px solid var(--mdc-theme-on-background)",
  backgroundColor: "var(--mdc-theme-surface)",
  width: "100%",
  transform: "translateZ(0)",
  overflow: "hidden",
  display: "flex",
  alignItems: "center",
  zIndex: 1,
  "> div": {
    padding: "0 10px",
    width: "100%"
  }
});

var BottomInfoBar = function BottomInfoBar(props) {
  return /*#__PURE__*/React.createElement(BottomInfoBarWrapper, null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(SupportedFileTypes, props), /*#__PURE__*/React.createElement(UploadStatus, props)));
};

export default BottomInfoBar;
