// @webiny/app-admin/components/FileManager/DropFilesHere.js
import React from "react";
import { css } from "emotion";
import classNames from "classnames";
import { Icon } from "@webiny/ui/Icon";
import { ReactComponent as UploadIcon } from "@svgr/webpack!./icons/round-cloud_upload-24px.svg";
var styles = /*#__PURE__*/css({
  margin: "0 auto",
  paddingTop: 0,
  height: "100%",
  zIndex: 2,
  width: "100%",
  position: "absolute",
  backgroundColor: "var(--mdc-theme-text-hint-on-light)",
  "&.empty": {
    backgroundColor: "transparent",
    "> div": {
      backgroundColor: "var(--mdc-theme-surface)"
    }
  },
  "> div": {
    textAlign: "center",
    width: 300,
    height: 300,
    backgroundColor: "var(--mdc-theme-background)",
    borderRadius: "50%",
    position: "absolute",
    left: "50%",
    top: "50%",
    transform: "translateX(-50%) translateY(-100%)",
    "> div": {
      position: "absolute",
      top: 90,
      width: 300,
      color: "var(--mdc-theme-on-surface)",
      "svg.mdc-button__icon": {
        width: 100,
        display: "inline-block",
        color: "var(--mdc-theme-on-surface)"
      }
    }
  }
}, "label:styles;");
export default function DropFilesHere(_ref) {
  var onDrop = _ref.onDrop,
      onDragLeave = _ref.onDragLeave,
      empty = _ref.empty,
      onClick = _ref.onClick;
  return /*#__PURE__*/React.createElement("div", {
    className: classNames(styles, {
      empty: empty
    }),
    onDrop: onDrop,
    onClick: onClick,
    onDragLeave: onDragLeave
  }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Icon, {
    icon: /*#__PURE__*/React.createElement(UploadIcon, null)
  }), /*#__PURE__*/React.createElement("div", null, "Drop files here"))));
}
