// @webiny/app-admin/components/FileManager/FileManagerView.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useRef, useCallback, useMemo } from "react";
import { css } from "emotion";
import styled from "@emotion/styled";
import Files from "react-butterfiles";
import { ButtonPrimary, ButtonIcon } from "@webiny/ui/Button";
import { Icon } from "@webiny/ui/Icon";
import File from "./File";
import { useQuery, useMutation, useApolloClient } from "@apollo/react-hooks";
import { LIST_FILES, CREATE_FILE, GET_FILE_SETTINGS } from "./graphql";
import getFileTypePlugin from "./getFileTypePlugin";
import get from "lodash/get";
import debounce from "lodash/debounce";
import getFileUploader from "./getFileUploader";
import outputFileSelectionError from "./outputFileSelectionError";
import DropFilesHere from "./DropFilesHere";
import NoResults from "./NoResults";
import FileDetails from "./FileDetails";
import LeftSidebar from "./LeftSidebar";
import BottomInfoBar from "./BottomInfoBar";
import { OverlayLayout } from "../OverlayLayout";
import { useSnackbar } from "../../hooks/useSnackbar";
import { Scrollbar } from "@webiny/ui/Scrollbar";
import { CircularProgress } from "@webiny/ui/Progress";
import { i18n } from "@webiny/app/i18n";
import { useSecurity } from "@webiny/app-security";
import { useHotkeys } from "react-hotkeyz";
import { useFileManager } from "./FileManagerContext";
import { ReactComponent as SearchIcon } from "@svgr/webpack!./icons/round-search-24px.svg";
import { ReactComponent as UploadIcon } from "@svgr/webpack!./icons/round-cloud_upload-24px.svg";
import NoPermissionView from "./NoPermissionView";
var t = i18n.ns("app-admin/file-manager/file-manager-view");
var style = {
  draggingFeedback: /*#__PURE__*/css({
    position: "fixed",
    top: 0,
    left: 0,
    width: "100%",
    height: "100%",
    opacity: 0.5,
    background: "white",
    zIndex: 100
  }, "label:draggingFeedback;"),
  leftDrawer: {
    header: /*#__PURE__*/css({
      textAlign: "center",
      fontSize: 18,
      padding: 10,
      fontWeight: 600,
      color: "var(--mdc-theme-on-surface)"
    }, "label:header;")
  }
};
var InputSearch = /*#__PURE__*/styled("div", {
  target: "ee8m09m0",
  label: "InputSearch"
})({
  backgroundColor: "var(--mdc-theme-on-background)",
  position: "relative",
  height: 32,
  padding: 3,
  width: "100%",
  borderRadius: 2,
  "> input": {
    border: "none",
    fontSize: 14,
    width: "calc(100% - 10px)",
    height: "100%",
    marginLeft: 50,
    backgroundColor: "transparent",
    outline: "none",
    color: "var(--mdc-theme-text-primary-on-background)"
  }
});
var searchIcon = /*#__PURE__*/css({
  "&.mdc-button__icon": {
    color: "var(--mdc-theme-text-secondary-on-background)",
    position: "absolute",
    width: 24,
    height: 24,
    left: 15,
    top: 7
  }
}, "label:searchIcon;");
var FileListWrapper = /*#__PURE__*/styled("div", {
  target: "ee8m09m1",
  label: "FileListWrapper"
})({
  float: "right",
  display: "inline-block",
  width: "calc(100vw - 270px)",
  height: "100%"
});
var FileList = /*#__PURE__*/styled("div", {
  target: "ee8m09m2",
  label: "FileList"
})({
  width: "100%",
  display: "grid",

  /* define the number of grid columns */
  gridTemplateColumns: "repeat( auto-fill, minmax(220px, 1fr) )",
  marginBottom: 95
});

function renderFile(props) {
  var file = props.file;
  var plugin = getFileTypePlugin(file);
  return /*#__PURE__*/React.createElement(File, Object.assign({}, props, {
    key: file.id
  }), plugin.render({
    file: file
  }));
}

var renderEmpty = function renderEmpty(_ref) {
  var hasPreviouslyUploadedFiles = _ref.hasPreviouslyUploadedFiles,
      browseFiles = _ref.browseFiles,
      fmFilePermission = _ref.fmFilePermission;

  if (!fmFilePermission) {
    return /*#__PURE__*/React.createElement(NoPermissionView, null);
  }

  if (hasPreviouslyUploadedFiles) {
    return /*#__PURE__*/React.createElement(NoResults, null);
  }

  return /*#__PURE__*/React.createElement(DropFilesHere, {
    empty: true,
    onClick: browseFiles
  });
};

function FileManagerView(props) {
  var onClose = props.onClose,
      onChange = props.onChange,
      accept = props.accept,
      multiple = props.multiple,
      maxSize = props.maxSize,
      multipleMaxCount = props.multipleMaxCount,
      multipleMaxSize = props.multipleMaxSize;

  var _useFileManager = useFileManager(),
      selected = _useFileManager.selected,
      toggleSelected = _useFileManager.toggleSelected,
      dragging = _useFileManager.dragging,
      setDragging = _useFileManager.setDragging,
      uploading = _useFileManager.uploading,
      setUploading = _useFileManager.setUploading,
      _showFileDetails = _useFileManager.showFileDetails,
      showingFileDetails = _useFileManager.showingFileDetails,
      queryParams = _useFileManager.queryParams,
      setQueryParams = _useFileManager.setQueryParams,
      hasPreviouslyUploadedFiles = _useFileManager.hasPreviouslyUploadedFiles,
      setHasPreviouslyUploadedFiles = _useFileManager.setHasPreviouslyUploadedFiles;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var _useSecurity = useSecurity(),
      identity = _useSecurity.identity;

  var fmFilePermission = useMemo(function () {
    return identity.getPermission("fm.file");
  }, []);
  var canCreate = useMemo(function () {
    // Bail out early if no access
    if (!fmFilePermission) {
      return false;
    }

    if (fmFilePermission.own) {
      return true;
    }

    if (typeof fmFilePermission.rwd === "string") {
      return fmFilePermission.rwd.includes("w");
    }

    return true;
  }, [fmFilePermission]);
  var canEdit = useCallback(function (item) {
    // Bail out early if no access
    if (!fmFilePermission) {
      return false;
    }

    var creatorId = get(item, "createdBy.id");

    if (fmFilePermission.own && creatorId) {
      return creatorId === identity.login;
    }

    if (typeof fmFilePermission.rwd === "string") {
      return fmFilePermission.rwd.includes("w");
    }

    return true;
  }, [fmFilePermission]);
  var searchOnChange = useCallback( // @ts-ignore
  debounce(function (search) {
    return setQueryParams({
      search: search
    });
  }, 500), []);

  var _toggleTag = useCallback( /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(_ref2) {
      var tag, queryParams, finalTags;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              tag = _ref2.tag, queryParams = _ref2.queryParams;
              finalTags = Array.isArray(queryParams.tags) ? _toConsumableArray(queryParams.tags) : [];

              if (finalTags.includes(tag)) {
                finalTags.splice(finalTags.indexOf(tag), 1);
              } else {
                finalTags.push(tag);
              }

              setQueryParams(_objectSpread(_objectSpread({}, queryParams), {}, {
                tags: finalTags
              }));

            case 4:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x) {
      return _ref3.apply(this, arguments);
    };
  }(), []);

  var getFileUploadErrorMessage = useCallback(function (e) {
    if (typeof e === "string") {
      var match = e.match(/Message>(.*?)<\/Message/);

      if (match) {
        var _match = _slicedToArray(match, 2),
            message = _match[1];

        return message;
      }

      return e;
    }

    return e.message;
  }, []);

  var updateCacheAfterCreateFile = function updateCacheAfterCreateFile(cache, newFile) {
    var newFileData = get(newFile, "data.fileManager.createFile.data");
    var data = cache.readQuery({
      query: LIST_FILES,
      variables: queryParams
    });
    cache.writeQuery({
      query: LIST_FILES,
      variables: queryParams,
      data: {
        fileManager: _objectSpread(_objectSpread({}, data.fileManager), {}, {
          listFiles: _objectSpread(_objectSpread({}, data.fileManager.listFiles), {}, {
            data: [newFileData].concat(_toConsumableArray(data.fileManager.listFiles.data || []))
          })
        })
      }
    });
  };

  var getFileDetailsFile = useCallback(function getFileDetailsFile(_ref4) {
    var src = _ref4.src,
        list = _ref4.list;
    return list.find(function (item) {
      return item.src === src;
    });
  }, []);
  useHotkeys({
    zIndex: 50,
    keys: {
      esc: onClose
    }
  });
  var searchInput = useRef();
  var apolloClient = useApolloClient();
  var gqlQuery = useQuery(LIST_FILES, {
    variables: queryParams,
    onCompleted: function onCompleted(response) {
      var list = get(response, "fileManager.listFiles.data") || [];

      if (hasPreviouslyUploadedFiles === null) {
        setHasPreviouslyUploadedFiles(list.length > 0);
      }
    }
  });
  var refreshOnScroll = useCallback(debounce(function (_ref5) {
    var scrollFrame = _ref5.scrollFrame,
        fetchMore = _ref5.fetchMore;

    if (scrollFrame.top > 0.9) {
      var cursor = get(gqlQuery.data, "fileManager.listFiles.meta.cursor");

      if (cursor) {
        fetchMore({
          variables: {
            after: cursor
          },
          updateQuery: function updateQuery(prev, _ref6) {
            var fetchMoreResult = _ref6.fetchMoreResult;

            if (!fetchMoreResult) {
              return prev;
            }

            var next = _objectSpread({}, fetchMoreResult);

            next.fileManager.listFiles.data = [].concat(_toConsumableArray(prev.fileManager.listFiles.data), _toConsumableArray(fetchMoreResult.fileManager.listFiles.data));
            return next;
          }
        });
      }
    }
  }, 500), [gqlQuery]);
  var data = gqlQuery.data,
      fetchMore = gqlQuery.fetchMore,
      loading = gqlQuery.loading;
  var list = get(data, "fileManager.listFiles.data") || [];

  var _useMutation = useMutation(CREATE_FILE, {
    update: updateCacheAfterCreateFile
  }),
      _useMutation2 = _slicedToArray(_useMutation, 1),
      createFile = _useMutation2[0];

  var uploadFile = /*#__PURE__*/function () {
    var _ref7 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3(files) {
      var list, errors;
      return _regeneratorRuntime.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              setUploading(true);
              list = Array.isArray(files) ? files : [files];
              errors = [];
              _context3.next = 5;
              return Promise.all(list.map( /*#__PURE__*/function () {
                var _ref8 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(file) {
                  var response;
                  return _regeneratorRuntime.wrap(function _callee2$(_context2) {
                    while (1) {
                      switch (_context2.prev = _context2.next) {
                        case 0:
                          _context2.prev = 0;
                          _context2.next = 3;
                          return getFileUploader()(file, {
                            apolloClient: apolloClient
                          });

                        case 3:
                          response = _context2.sent;
                          _context2.next = 6;
                          return createFile({
                            variables: {
                              data: response
                            }
                          });

                        case 6:
                          _context2.next = 11;
                          break;

                        case 8:
                          _context2.prev = 8;
                          _context2.t0 = _context2["catch"](0);
                          errors.push({
                            file: file,
                            e: _context2.t0
                          });

                        case 11:
                        case "end":
                          return _context2.stop();
                      }
                    }
                  }, _callee2, null, [[0, 8]]);
                }));

                return function (_x3) {
                  return _ref8.apply(this, arguments);
                };
              }()));

            case 5:
              if (!hasPreviouslyUploadedFiles) {
                setHasPreviouslyUploadedFiles(true);
              }

              setUploading(false);

              if (!(errors.length > 0)) {
                _context3.next = 9;
                break;
              }

              return _context3.abrupt("return", setTimeout(function () {
                showSnackbar( /*#__PURE__*/React.createElement(React.Fragment, null, t(_templateObject || (_templateObject = _taggedTemplateLiteral(["One or more files were not uploaded successfully:"]))), /*#__PURE__*/React.createElement("ol", null, errors.map(function (_ref9) {
                  var file = _ref9.file,
                      e = _ref9.e;
                  return /*#__PURE__*/React.createElement("li", {
                    key: file.name
                  }, /*#__PURE__*/React.createElement("strong", null, file.name), ": ", getFileUploadErrorMessage(e));
                }))));
              }, 750));

            case 9:
              // We wait 750ms, just for everything to settle down a bit.
              setTimeout(function () {
                return showSnackbar(t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["File upload complete."]))));
              }, 750);

            case 10:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3);
    }));

    return function uploadFile(_x2) {
      return _ref7.apply(this, arguments);
    };
  }();

  var renderUploadFileAction = useCallback(function (_ref10) {
    var browseFiles = _ref10.browseFiles;

    if (!canCreate) {
      return null;
    }

    return /*#__PURE__*/React.createElement(ButtonPrimary, {
      onClick: browseFiles,
      disabled: uploading
    }, /*#__PURE__*/React.createElement(ButtonIcon, {
      icon: /*#__PURE__*/React.createElement(UploadIcon, null)
    }), t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Upload..."]))));
  }, [uploading, canCreate]);
  var settingsQuery = useQuery(GET_FILE_SETTINGS);
  var settings = get(settingsQuery.data, "fileManager.getSettings.data") || {};
  return /*#__PURE__*/React.createElement(Files, {
    multiple: true,
    maxSize: settings.uploadMaxFileSize ? settings.uploadMaxFileSize + "b" : maxSize,
    multipleMaxSize: multipleMaxSize,
    multipleMaxCount: multipleMaxCount,
    accept: accept,
    onSuccess: function onSuccess(files) {
      return uploadFile(files.map(function (file) {
        return file.src.file;
      }));
    },
    onError: function onError(errors) {
      var message = outputFileSelectionError(errors);
      showSnackbar(message);
    }
  }, function (_ref11) {
    var getDropZoneProps = _ref11.getDropZoneProps,
        browseFiles = _ref11.browseFiles,
        validateFiles = _ref11.validateFiles;
    return /*#__PURE__*/React.createElement(OverlayLayout, Object.assign({}, getDropZoneProps({
      onDragEnter: function onDragEnter() {
        return hasPreviouslyUploadedFiles && setDragging(true);
      },
      onExited: onClose
    }), {
      barLeft: /*#__PURE__*/React.createElement(InputSearch, null, /*#__PURE__*/React.createElement(Icon, {
        className: searchIcon,
        icon: /*#__PURE__*/React.createElement(SearchIcon, null)
      }), /*#__PURE__*/React.createElement("input", {
        ref: searchInput,
        onChange: function onChange(e) {
          return searchOnChange(e.target.value);
        },
        placeholder: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Search by filename or tags"]))),
        disabled: !fmFilePermission,
        "data-testid": "file-manager.search-input"
      })),
      barRight: selected.length > 0 ? /*#__PURE__*/React.createElement(ButtonPrimary, {
        disabled: uploading,
        onClick: /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee4() {
          return _regeneratorRuntime.wrap(function _callee4$(_context4) {
            while (1) {
              switch (_context4.prev = _context4.next) {
                case 0:
                  _context4.next = 2;
                  return onChange(multiple ? selected : selected[0]);

                case 2:
                  onClose();

                case 3:
                case "end":
                  return _context4.stop();
              }
            }
          }, _callee4);
        }))
      }, t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Select"]))), " ", multiple && "(".concat(selected.length, ")")) : renderUploadFileAction({
        browseFiles: browseFiles
      })
    }), /*#__PURE__*/React.createElement(React.Fragment, null, dragging && hasPreviouslyUploadedFiles && /*#__PURE__*/React.createElement(DropFilesHere // @ts-ignore TODO: @adrian - className is never rendered?!
    , {
      className: style.draggingFeedback,
      onDragLeave: function onDragLeave() {
        return setDragging(false);
      },
      onDrop: function onDrop() {
        return setDragging(false);
      }
    }), /*#__PURE__*/React.createElement(FileDetails, {
      validateFiles: validateFiles,
      uploadFile: uploadFile,
      file: getFileDetailsFile({
        list: list,
        src: showingFileDetails
      }),
      canEdit: canEdit
    }), /*#__PURE__*/React.createElement(LeftSidebar, {
      queryParams: queryParams,
      toggleTag: function toggleTag(tag) {
        return _toggleTag({
          tag: tag,
          queryParams: queryParams
        });
      }
    }), /*#__PURE__*/React.createElement(FileListWrapper, {
      "data-testid": "fm-list-wrapper"
    }, loading && /*#__PURE__*/React.createElement(CircularProgress, {
      label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Loading Files..."]))),
      style: {
        opacity: 1
      }
    }), /*#__PURE__*/React.createElement(Scrollbar, {
      onScrollFrame: function onScrollFrame(scrollFrame) {
        return refreshOnScroll({
          scrollFrame: scrollFrame,
          fetchMore: fetchMore
        });
      }
    }, /*#__PURE__*/React.createElement(FileList, null, list.length ? list.map(function (file) {
      return renderFile({
        uploadFile: uploadFile,
        file: file,
        showFileDetails: function showFileDetails() {
          return _showFileDetails(file.src);
        },
        selected: selected.find(function (current) {
          return current.src === file.src;
        }),
        onSelect: typeof onChange === "undefined" ? undefined : /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee5() {
          return _regeneratorRuntime.wrap(function _callee5$(_context5) {
            while (1) {
              switch (_context5.prev = _context5.next) {
                case 0:
                  if (!multiple) {
                    _context5.next = 3;
                    break;
                  }

                  toggleSelected(file);
                  return _context5.abrupt("return");

                case 3:
                  _context5.next = 5;
                  return onChange(file);

                case 5:
                  onClose();

                case 6:
                case "end":
                  return _context5.stop();
              }
            }
          }, _callee5);
        }))
      });
    }) : renderEmpty({
      hasPreviouslyUploadedFiles: hasPreviouslyUploadedFiles,
      browseFiles: browseFiles,
      fmFilePermission: fmFilePermission
    }))), /*#__PURE__*/React.createElement(BottomInfoBar, {
      accept: accept,
      uploading: uploading
    }))));
  });
}

FileManagerView.defaultProps = {
  multiple: false,
  maxSize: "1000mb",
  multipleMaxSize: "1000mb",
  multipleMaxCount: 100
};
export default FileManagerView;
