// @webiny/app-admin/components/FileManager/LeftSidebar.js
import React from "react";
import { useQuery } from "@apollo/react-hooks";
import get from "lodash/get";
import sortBy from "lodash/sortBy";
import { css } from "emotion";
import styled from "@emotion/styled";
import { Icon } from "@webiny/ui/Icon";
import { ReactComponent as TagIcon } from "@svgr/webpack!./icons/round-label-24px.svg";
import { LIST_TAGS } from "./graphql";
var style = {
  leftDrawer: /*#__PURE__*/css({
    float: "left",
    display: "inline-block",
    width: 250,
    height: "100%",
    backgroundColor: "var(--mdc-theme-surface)",
    padding: 10
  }, "label:leftDrawer;"),
  noTagged: /*#__PURE__*/css({
    paddingTop: 100,
    textAlign: "center",
    color: "var(--mdc-theme-on-surface)"
  }, "label:noTagged;")
};
var TagList = /*#__PURE__*/styled("div", {
  target: "e1tgxgqn0",
  label: "TagList"
})({
  display: "flex",
  flexDirection: "column"
});
var Tag = /*#__PURE__*/styled("div", {
  target: "e1tgxgqn1",
  label: "Tag"
})({
  display: "flex",
  flexDirection: "row",
  height: 40,
  alignItems: "center",
  cursor: "pointer",
  paddingLeft: 15,
  color: "var(--mdc-theme-on-surface)",
  svg: {
    color: "var(--mdc-theme-on-surface)",
    marginRight: 10
  },
  "&:hover": {
    backgroundColor: "var(--mdc-theme-background)"
  },
  "&.active": {
    svg: {
      color: "var(--mdc-theme-secondary)"
    }
  }
});

function LeftSidebar(_ref) {
  var toggleTag = _ref.toggleTag,
      tags = _ref.queryParams.tags;
  var activeTags = Array.isArray(tags) ? tags : [];

  var _useQuery = useQuery(LIST_TAGS),
      data = _useQuery.data;

  var list = get(data, "fileManager.listTags") || [];

  if (list.length === 0) {
    return /*#__PURE__*/React.createElement("div", {
      className: style.leftDrawer,
      "data-testid": "fm.left-drawer.empty-tag"
    }, /*#__PURE__*/React.createElement("div", {
      className: style.noTagged
    }, "Once you tag an image, the tag will be displayed here."));
  } // Sort "tags" list in "ASC" order.


  var tagList = sortBy(list);
  return /*#__PURE__*/React.createElement("div", {
    className: style.leftDrawer
  }, /*#__PURE__*/React.createElement(TagList, {
    "data-testid": "fm.left-drawer.tag-list"
  }, tagList.map(function (item, index) {
    return /*#__PURE__*/React.createElement(Tag, {
      className: activeTags.includes(item) && "active",
      key: item + index,
      onClick: function onClick() {
        return toggleTag(item);
      }
    }, /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(TagIcon, null)
    }), " ", item);
  })));
}

export default LeftSidebar;
