// @webiny/app-form-builder/admin/plugins/editor/defaultBar/Name/Name.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";

var _templateObject, _templateObject2, _templateObject3;

import React, { useCallback, useState } from "react";
import { Input } from "@webiny/ui/Input";
import { Tooltip } from "@webiny/ui/Tooltip";
import { Typography } from "@webiny/ui/Typography";
import { useFormEditor } from "../../../../components/FormEditor/Context";
import { useHotkeys } from "react-hotkeyz";
import { FormMeta, FormName, formNameWrapper, FormVersion, NameInputWrapper, NameWrapper } from "./NameStyled";
import { i18n } from "@webiny/app/i18n";
var t = i18n.namespace("FormEditor.Name");
export var Name = function Name() {
  var _useFormEditor = useFormEditor(),
      state = _useFormEditor.state,
      setData = _useFormEditor.setData;

  var _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      localName = _useState2[0],
      setLocalName = _useState2[1];

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      editingEnabled = _useState4[0],
      setEditing = _useState4[1];

  function cancelChanges() {
    setEditing(false);
  }

  function startEditing() {
    setLocalName(state.data.name);
    setEditing(true);
  }

  var saveTitle = useCallback(function (e) {
    e.preventDefault();
    setData(function (data) {
      data.name = localName;
      return data;
    });
    setEditing(false);
  }, [localName]);
  useHotkeys({
    zIndex: 100,
    keys: {
      "alt+cmd+enter": startEditing
    }
  });
  useHotkeys({
    zIndex: 101,
    disabled: !editingEnabled,
    keys: {
      esc: function esc(e) {
        e.preventDefault();
        cancelChanges();
      },
      enter: saveTitle
    }
  }); // Disable autoFocus because for some reason, blur event would automatically be triggered when clicking
  // on the page title when doing Cypress testing. Not sure if this is RMWC or Cypress related issue.

  var autoFocus = !window.Cypress;
  return editingEnabled ? /*#__PURE__*/React.createElement(NameInputWrapper, null, /*#__PURE__*/React.createElement(Input, {
    autoFocus: autoFocus,
    fullwidth: true,
    value: localName,
    onChange: setLocalName,
    onBlur: saveTitle
  })) : /*#__PURE__*/React.createElement(NameWrapper, null, /*#__PURE__*/React.createElement(FormMeta, null, /*#__PURE__*/React.createElement(Typography, {
    use: "overline"
  }, "status: ".concat(state.data.published ? t(_templateObject || (_templateObject = _taggedTemplateLiteral(["published"]))) : t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["draft"])))))), /*#__PURE__*/React.createElement("div", {
    style: {
      width: "100%",
      display: "flex"
    }
  }, /*#__PURE__*/React.createElement(Tooltip, {
    className: formNameWrapper,
    placement: "bottom",
    content: /*#__PURE__*/React.createElement("span", null, t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["rename"]))))
  }, /*#__PURE__*/React.createElement(FormName, {
    "data-testid": "fb-editor-form-title",
    onClick: startEditing
  }, state.data.name)), /*#__PURE__*/React.createElement(FormVersion, null, "(v".concat(state.data.version, ")"))));
};
