// @webiny/app-headless-cms/admin/components/ContentEntryForm/ContentEntryForm.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["onForm"];
import React, { useCallback, useEffect, useRef } from "react";
import styled from "@emotion/styled";
import { Form } from "@webiny/form";
import { plugins } from "@webiny/plugins";
import { CircularProgress } from "@webiny/ui/Progress";
import RenderFieldElement from "./RenderFieldElement";
import { useContentEntryForm } from "./useContentEntryForm";
import { Fields } from "./Fields";
var FormWrapper = /*#__PURE__*/styled("div", {
  target: "ejw5gwy0",
  label: "FormWrapper"
})({
  height: "calc(100vh - 260px)",
  overflow: "auto"
});
export var ContentEntryForm = function ContentEntryForm(_ref) {
  var onForm = _ref.onForm,
      props = _objectWithoutProperties(_ref, _excluded);

  var contentModel = props.contentModel;

  var _useContentEntryForm = useContentEntryForm(props),
      loading = _useContentEntryForm.loading,
      data = _useContentEntryForm.data,
      onChange = _useContentEntryForm.onChange,
      onSubmit = _useContentEntryForm.onSubmit,
      invalidFields = _useContentEntryForm.invalidFields;

  var ref = useRef(null);
  useEffect(function () {
    typeof onForm === "function" && onForm(ref.current);
  }, []);
  var formRenderer = plugins.byType("cms-content-form-renderer").find(function (pl) {
    return pl.modelId === contentModel.modelId;
  });
  var renderCustomLayout = useCallback(function (formRenderProps) {
    var fields = contentModel.fields.reduce(function (acc, field) {
      acc[field.fieldId] = /*#__PURE__*/React.createElement(RenderFieldElement, {
        field: field,
        Bind: formRenderProps.Bind,
        contentModel: contentModel
      });
      return acc;
    }, {});
    return formRenderer.render(_objectSpread(_objectSpread({}, formRenderProps), {}, {
      contentModel: contentModel,
      fields: fields
    }));
  }, [formRenderer]);
  return /*#__PURE__*/React.createElement(Form, {
    onChange: onChange,
    onSubmit: onSubmit,
    data: data,
    ref: ref,
    invalidFields: invalidFields
  }, function (formProps) {
    return /*#__PURE__*/React.createElement(FormWrapper, {
      "data-testid": "cms-content-form"
    }, loading && /*#__PURE__*/React.createElement(CircularProgress, null), formRenderer ? renderCustomLayout(formProps) : /*#__PURE__*/React.createElement(Fields, Object.assign({
      contentModel: contentModel,
      fields: contentModel.fields,
      layout: contentModel.layout
    }, formProps)));
  });
};
