// @webiny/app-headless-cms/admin/components/ContentEntryForm/Fields.js
import React from "react";
import { Cell, Grid } from "@webiny/ui/Grid";
import RenderFieldElement from "./RenderFieldElement";

var getFieldById = function getFieldById(fields, id) {
  return fields.find(function (field) {
    return field.id === id;
  });
};

export var Fields = function Fields(_ref) {
  var Bind = _ref.Bind,
      fields = _ref.fields,
      layout = _ref.layout,
      contentModel = _ref.contentModel,
      gridClassName = _ref.gridClassName;
  return /*#__PURE__*/React.createElement(Grid, {
    className: gridClassName
  }, layout.map(function (row, rowIndex) {
    return /*#__PURE__*/React.createElement(React.Fragment, {
      key: rowIndex
    }, row.map(function (fieldId) {
      return /*#__PURE__*/React.createElement(Cell, {
        span: Math.floor(12 / row.length),
        key: fieldId
      }, /*#__PURE__*/React.createElement(RenderFieldElement, {
        field: getFieldById(fields, fieldId),
        Bind: Bind,
        contentModel: contentModel
      }));
    }));
  }));
};
