// @webiny/react-router/context/RouterContext.js
import React, { useMemo } from "react";
import { plugins } from "@webiny/plugins";
import { useApolloClient } from "@apollo/react-hooks";
export var RouterContext = /*#__PURE__*/React.createContext(null);
export var RouterProvider = function RouterProvider(_ref) {
  var children = _ref.children;
  var apolloClient = null;

  try {
    apolloClient = useApolloClient();
  } catch (_unused) {// If there is no ApolloProvider, apolloClient will not exist.
  }

  var value = useMemo(function () {
    var onLinkPlugins = plugins.byType("react-router-on-link");
    return {
      onLink: function onLink(link) {
        for (var i = 0; i < onLinkPlugins.length; i++) {
          var _onLink = onLinkPlugins[i].onLink;

          if (typeof _onLink === "function") {
            _onLink({
              link: link,
              apolloClient: apolloClient
            });
          }
        }
      }
    };
  }, []);
  return /*#__PURE__*/React.createElement(RouterContext.Provider, {
    value: value
  }, children);
};
export var RouterConsumer = function RouterConsumer(_ref2) {
  var children = _ref2.children;
  return /*#__PURE__*/React.createElement(RouterContext.Consumer, null, function (props) {
    return /*#__PURE__*/React.cloneElement(children, props);
  });
};
