// @webiny/app-headless-cms/admin/plugins/fieldRenderers/dateTime/dateTimeField.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import get from "lodash/get";
import { i18n } from "@webiny/app/i18n";
import DateTimeWithoutTimezone from "./DateTimeWithoutTimezone";
import DateTimeWithTimezone from "./DateTimeWithTimezone";
import Time from "./Time";
import Input from "./Input";
var t = i18n.ns("app-headless-cms/admin/fields/date-time");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-date-time",
  renderer: {
    rendererName: "date-time-input",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Date/Time Input"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders input for various formats of date and time."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "datetime" && !field.multipleValues && !get(field, "predefinedValues.enabled");
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind;
      var Bind = getBind();
      return /*#__PURE__*/React.createElement(Bind, null, function (bind) {
        if (field.settings.type === "dateTimeWithoutTimezone") {
          return /*#__PURE__*/React.createElement(DateTimeWithoutTimezone, {
            field: field,
            bind: bind
          });
        }

        if (field.settings.type === "dateTimeWithTimezone") {
          return /*#__PURE__*/React.createElement(DateTimeWithTimezone, {
            field: field,
            bind: bind
          });
        }

        if (field.settings.type === "time") {
          return /*#__PURE__*/React.createElement(Time, {
            field: field,
            bind: bind
          });
        }

        return /*#__PURE__*/React.createElement(Input, {
          bind: bind,
          field: field,
          type: field.settings.type
        });
      });
    }
  }
};
export default plugin;
