// @webiny/app-security-admin-users/plugins/userMenu/userInfo.js
import React from "react";
import { css } from "emotion";
import { Link } from "@webiny/react-router";
import { useSecurity } from "@webiny/app-security/hooks/useSecurity";
import { Image } from "@webiny/app/components";
import { ListItem, ListItemGraphic } from "@webiny/ui/List";
import { Typography } from "@webiny/ui/Typography";
import { Avatar } from "@webiny/ui/Avatar";
import { GenericElement } from "@webiny/app-admin/ui/elements/GenericElement";
import { UIViewPlugin } from "@webiny/app-admin/ui/UIView";
import { AdminView } from "@webiny/app-admin/ui/views/AdminView";
var avatarImage = /*#__PURE__*/css({
  height: "40px !important",
  width: "40px !important",
  borderRadius: "50%",
  display: "inline-block",
  ">div": {
    backgroundColor: "var(--mdc-theme-surface)"
  }
}, "label:avatarImage;");
var linkStyles = /*#__PURE__*/css({
  backgroundColor: "var(--mdc-theme-background)",
  marginTop: -25,
  marginBottom: 0,
  paddingTop: 0,
  paddingBottom: 0,
  textDecoration: "none",
  display: "block",
  "&:hover": {
    textDecoration: "none"
  },
  ">.mdc-list-item": {
    display: "flex",
    height: "65px !important",
    marginTop: 15,
    ".avatar": {
      marginTop: 10
    }
  },
  "h3, h3>.mdc-typography--headline6": {
    lineHeight: "1em !important"
  },
  ".mdc-menu &.mdc-list-item": {
    ">.mdc-list-item__text": {
      fontWeight: "bold"
    }
  },
  ".mdc-list-item": {
    height: "auto",
    ".mdc-typography--headline6": {
      overflow: "hidden",
      whiteSpace: "nowrap",
      textOverflow: "ellipsis",
      maxWidth: 180,
      display: "block"
    },
    ".mdc-typography--subtitle2": {
      overflow: "hidden",
      whiteSpace: "nowrap",
      textOverflow: "ellipsis",
      maxWidth: 180
    }
  }
}, "label:linkStyles;");

var UserInfo = function UserInfo() {
  var security = useSecurity();

  if (!security || !security.identity) {
    return null;
  }

  var _security$identity = security.identity,
      login = _security$identity.login,
      firstName = _security$identity.firstName,
      lastName = _security$identity.lastName,
      avatar = _security$identity.avatar,
      gravatar = _security$identity.gravatar;
  var fullName = "".concat(firstName, " ").concat(lastName);
  return /*#__PURE__*/React.createElement(Link, {
    to: "/account",
    className: linkStyles
  }, /*#__PURE__*/React.createElement(ListItem, {
    ripple: false
  }, /*#__PURE__*/React.createElement(ListItemGraphic, {
    className: avatarImage
  }, /*#__PURE__*/React.createElement(Avatar, {
    className: "avatar",
    src: avatar ? avatar.src : gravatar,
    alt: fullName,
    fallbackText: fullName,
    renderImage: function renderImage(props) {
      return /*#__PURE__*/React.createElement(Image, Object.assign({}, props, {
        transform: {
          width: 100
        }
      }));
    }
  })), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("h3", null, /*#__PURE__*/React.createElement(Typography, {
    use: "headline6"
  }, fullName)), /*#__PURE__*/React.createElement(Typography, {
    use: "subtitle2"
  }, login))));
};

export default (function () {
  return new UIViewPlugin(AdminView, function (view) {
    var userMenu = view.getElement("userMenu");

    if (!userMenu) {
      return;
    }

    var userInfo = new GenericElement("userInfo", function () {
      return /*#__PURE__*/React.createElement(UserInfo, null);
    });
    userInfo.moveToTheBeginningOf(userMenu);
    view.refresh();
  });
});
