// @webiny/app-security-admin-users/plugins/userMenu/userImage.js
import React from "react";
import { Avatar } from "@webiny/ui/Avatar";
import { Image } from "@webiny/app/components";
import { useSecurity } from "@webiny/app-security/hooks/useSecurity";
import { GenericElement } from "@webiny/app-admin/ui/elements/GenericElement";
import { UIViewPlugin } from "@webiny/app-admin/ui/UIView";
import { AdminView } from "@webiny/app-admin/ui/views/AdminView";

var UserImage = function UserImage() {
  var _useSecurity = useSecurity(),
      identity = _useSecurity.identity;

  if (!identity) {
    return null;
  }

  var fullName = identity.fullName,
      avatar = identity.avatar,
      gravatar = identity.gravatar;
  return /*#__PURE__*/React.createElement(Avatar, {
    "data-testid": "logged-in-user-menu-avatar",
    src: avatar ? avatar.src : gravatar,
    alt: fullName,
    fallbackText: fullName,
    renderImage: function renderImage(props) {
      return /*#__PURE__*/React.createElement(Image, Object.assign({}, props, {
        transform: {
          width: 100
        }
      }));
    }
  });
};

export default (function () {
  return new UIViewPlugin(AdminView, function (view) {
    var userMenu = view.getElement("userMenu");

    if (userMenu) {
      userMenu.setMenuHandleElement(new GenericElement("handle", function () {
        return /*#__PURE__*/React.createElement(UserImage, null);
      }));
    }
  });
});
