// @webiny/app-page-builder/editor/plugins/responsiveMode/index.js
import React from "react";
import { css } from "emotion";
import { DisplayMode } from "../../../types"; // Icons

import { ReactComponent as DesktopIcon } from "@svgr/webpack!./icons/laptop_mac.svg";
import { ReactComponent as TabletIcon } from "@svgr/webpack!./icons/tablet_mac.svg";
import { ReactComponent as MobileIcon } from "@svgr/webpack!./icons/phone_iphone.svg";
import { ReactComponent as StarIcon } from "@svgr/webpack!./icons/star_rate.svg";
var rotateStyle = /*#__PURE__*/css({
  transform: "rotate(90deg)"
}, "label:rotateStyle;");
export default (function () {
  return [{
    type: "pb-editor-responsive-mode",
    name: "pb-editor-responsive-mode-desktop",
    config: {
      displayMode: DisplayMode.DESKTOP,
      toolTip: {
        title: "Desktop",
        subTitle: "Base breakpoint",
        body: "Desktop styles apply at all breakpoints, unless they're\n                       edited at a smaller breakpoint. Start your styling here.",
        subTitleIcon: /*#__PURE__*/React.createElement(StarIcon, null)
      },
      icon: /*#__PURE__*/React.createElement(DesktopIcon, null)
    }
  }, {
    type: "pb-editor-responsive-mode",
    name: "pb-editor-responsive-mode-tablet",
    config: {
      displayMode: DisplayMode.TABLET,
      toolTip: {
        title: "Tablet",
        subTitle: "991px and down",
        body: "Styles added here will apply at 991px and down, unless they're\n                       edited at a smaller breakpoint."
      },
      icon: /*#__PURE__*/React.createElement(TabletIcon, null)
    }
  }, {
    type: "pb-editor-responsive-mode",
    name: "pb-editor-responsive-mode-mobile-landscape",
    config: {
      displayMode: DisplayMode.MOBILE_LANDSCAPE,
      toolTip: {
        title: "Mobile landscape",
        subTitle: "767px and down",
        body: "Styles added here will apply at 767px and down, unless they're\n                       edited at a smaller breakpoint."
      },
      icon: /*#__PURE__*/React.createElement(MobileIcon, {
        className: rotateStyle
      })
    }
  }, {
    type: "pb-editor-responsive-mode",
    name: "pb-editor-responsive-mode-mobile-portrait",
    config: {
      displayMode: DisplayMode.MOBILE_PORTRAIT,
      toolTip: {
        title: "Mobile portrait",
        subTitle: "478px and down",
        body: "Styles added here will apply at 478px and down."
      },
      icon: /*#__PURE__*/React.createElement(MobileIcon, null)
    }
  }];
});
