// @webiny/app-page-builder/editor/plugins/elementSettings/hooks/useElementStyleSettings.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import { useCallback, useEffect } from "react";
import { useRecoilState, useRecoilValue } from "recoil";
import { plugins } from "@webiny/plugins";
import { useKeyHandler } from "../../../hooks/useKeyHandler";
import { userElementStyleSettingsPlugins } from "../../../helpers";
import { activeElementAtom, elementByIdSelector } from "../../../recoil/modules";

var getElementActions = function getElementActions(plugin) {
  if (!plugin || !plugin.settings) {
    return [];
  }

  var pluginSettings = [].concat(_toConsumableArray(userElementStyleSettingsPlugins(plugin.elementType)), _toConsumableArray(plugin.settings));
  var elementActions = pluginSettings.map(function (pl) {
    if (typeof pl === "string") {
      return {
        plugin: plugins.byName(pl),
        options: {}
      };
    }

    if (Array.isArray(pl)) {
      return {
        plugin: plugins.byName(pl[0]),
        options: pl[1]
      };
    }

    return null;
  });
  return elementActions // Eliminate empty plugins
  .filter(function (pl) {
    return pl && pl.plugin;
  }) // Eliminate plugins other than "PbEditorPageElementStyleSettingsPlugin".
  .filter(function (pl) {
    return pl && pl.plugin && pl.plugin.type === "pb-editor-page-element-style-settings";
  }) // Eliminate duplicate plugins
  .filter(function (pl, index, array) {
    return array.findIndex(function (item) {
      return item.plugin.name === pl.plugin.name;
    }) === index;
  });
};

var useElementStyleSettings = function useElementStyleSettings() {
  var _useRecoilState = useRecoilState(activeElementAtom),
      _useRecoilState2 = _slicedToArray(_useRecoilState, 2),
      activeElement = _useRecoilState2[0],
      setActiveElementAtomValue = _useRecoilState2[1];

  var element = useRecoilValue(elementByIdSelector(activeElement));
  var elementType = element === null || element === void 0 ? void 0 : element.type;
  var deactivateElement = useCallback(function () {
    setActiveElementAtomValue(null);
  }, []);

  var _useKeyHandler = useKeyHandler(),
      addKeyHandler = _useKeyHandler.addKeyHandler,
      removeKeyHandler = _useKeyHandler.removeKeyHandler;

  useEffect(function () {
    addKeyHandler("escape", function (e) {
      e.preventDefault();
      deactivateElement();
    });
    return function () {
      return removeKeyHandler("escape");
    };
  });
  var plugin = plugins.byType("pb-editor-page-element").find(function (pl) {
    return pl.elementType === elementType;
  });
  return getElementActions(plugin);
};

export default useElementStyleSettings;
