// @webiny/app-headless-cms/admin/views/contentEntries/ContentEntry/header/saveAndPublishContent/SaveAndPublishContent.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React from "react";
import { css } from "emotion";
import { i18n } from "@webiny/app/i18n";
import { ButtonPrimary } from "@webiny/ui/Button";
import { useConfirmationDialog } from "@webiny/app-admin/hooks/useConfirmationDialog";
import { useRevision } from "../../useRevision";
import { useContentEntry } from "../../../hooks/useContentEntry";
import usePermission from "../../../../../hooks/usePermission";
var t = i18n.ns("app-headless-cms/admin/plugins/content-details/header/publish-revision");
var buttonStyles = /*#__PURE__*/css({
  marginLeft: 16
}, "label:buttonStyles;");

var SaveAndPublishButton = function SaveAndPublishButton() {
  var _useContentEntry = useContentEntry(),
      form = _useContentEntry.form,
      loading = _useContentEntry.loading,
      entry = _useContentEntry.entry;

  var _useRevision = useRevision({
    revision: entry
  }),
      publishRevision = _useRevision.publishRevision;

  var _useConfirmationDialo = useConfirmationDialog({
    title: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Publish content"]))),
    message: /*#__PURE__*/React.createElement("p", null, t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["You are about to publish a new revision. Are you sure you want to continue?"])))),
    dataTestId: "cms-confirm-save-and-publish"
  }),
      showConfirmation = _useConfirmationDialo.showConfirmation;

  var _usePermission = usePermission(),
      canEdit = _usePermission.canEdit,
      canPublish = _usePermission.canPublish;

  if (!canEdit(entry, "cms.contentEntry") || !canPublish("cms.contentEntry")) {
    return null;
  }

  return /*#__PURE__*/React.createElement(ButtonPrimary, {
    className: buttonStyles,
    onClick: function onClick() {
      showConfirmation( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var entry;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return form.current.submit();

              case 2:
                entry = _context.sent;

                if (entry) {
                  _context.next = 5;
                  break;
                }

                return _context.abrupt("return");

              case 5:
                _context.next = 7;
                return publishRevision(entry.id);

              case 7:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      })));
    },
    disabled: loading
  }, t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Save & Publish"]))));
};

export default SaveAndPublishButton;
