// @webiny/app-admin/ui/views/NavigationView/FooterElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _get from "@babel/runtime/helpers/get";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import { List, ListItem } from "@webiny/ui/List";
import { UIElement } from "../../UIElement";
import { GenericElement } from "../../elements/GenericElement";
import { PlaceholderElement } from "../../elements/PlaceholderElement";
import { MenuFooter, subFooter } from "./Styled";
export var FooterElement = /*#__PURE__*/function (_UIElement) {
  _inherits(FooterElement, _UIElement);

  var _super = _createSuper(FooterElement);

  function FooterElement(id, config) {
    var _this;

    _classCallCheck(this, FooterElement);

    _this = _super.call(this, id, config);

    _defineProperty(_assertThisInitialized(_this), "_footerPlaceholder", void 0);

    _this.useGrid(false);

    _this._footerPlaceholder = _this.addElement(new PlaceholderElement("navigation.footer.placeholder"));
    new GenericElement("webiny.version", function () {
      return /*#__PURE__*/React.createElement(ListItem, {
        ripple: false,
        className: subFooter
      }, "Webiny v", process.env.REACT_APP_WEBINY_VERSION);
    }).moveAfter(_this._footerPlaceholder);
    return _this;
  }

  _createClass(FooterElement, [{
    key: "addMenuElement",
    value: function addMenuElement(element) {
      element.moveBefore(this._footerPlaceholder);
      return element;
    }
  }, {
    key: "render",
    value: function render(props) {
      return /*#__PURE__*/React.createElement(MenuFooter, null, /*#__PURE__*/React.createElement(List, {
        nonInteractive: true
      }, _get(_getPrototypeOf(FooterElement.prototype), "render", this).call(this, props)));
    }
  }]);

  return FooterElement;
}(UIElement);
