// @webiny/app-admin/ui/views/NavigationView/HeaderElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _get from "@babel/runtime/helpers/get";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { UIElement } from "../../UIElement";
import { DrawerHeader } from "@webiny/ui/Drawer";
import { GenericElement } from "../../elements/GenericElement";
import { PlaceholderElement } from "../../elements/PlaceholderElement";
import { TopAppBarTitle } from "@webiny/ui/TopAppBar";
import { MenuHeader, navHeader } from "./Styled";
export var ElementID;

(function (ElementID) {
  ElementID["Logo"] = "headerLogo";
  ElementID["MenuButton"] = "headerMenuButton";
})(ElementID || (ElementID = {}));

export var HeaderElement = /*#__PURE__*/function (_UIElement) {
  _inherits(HeaderElement, _UIElement);

  var _super = _createSuper(HeaderElement);

  function HeaderElement(id, config) {
    var _this;

    _classCallCheck(this, HeaderElement);

    _this = _super.call(this, id, config);

    _this.useGrid(false);

    _this.addElement(new PlaceholderElement(ElementID.MenuButton));

    _this.addElement(new PlaceholderElement(ElementID.Logo));

    return _this;
  }

  _createClass(HeaderElement, [{
    key: "setLogo",
    value: function setLogo(logo) {
      this.getElement(ElementID.Logo).replaceWith(new GenericElement(ElementID.Logo, function () {
        return /*#__PURE__*/React.createElement(TopAppBarTitle, null, logo);
      }));
    }
  }, {
    key: "setMenuButton",
    value: function setMenuButton(menuButton) {
      this.getElement(ElementID.MenuButton).replaceWith(new GenericElement(ElementID.MenuButton, function () {
        return menuButton;
      }));
    }
  }, {
    key: "render",
    value: function render(props) {
      return /*#__PURE__*/React.createElement(DrawerHeader, {
        className: navHeader
      }, /*#__PURE__*/React.createElement(MenuHeader, null, _get(_getPrototypeOf(HeaderElement.prototype), "render", this).call(this, props)));
    }
  }]);

  return HeaderElement;
}(UIElement);
