// @webiny/app-admin/ui/views/NavigationView/Styled.js
import styled from "@emotion/styled";
import { css } from "emotion";
export var MenuHeader = /*#__PURE__*/styled("div", {
  target: "e1b4rzuv0",
  label: "MenuHeader"
})({
  display: "flex",
  alignItems: "center",
  padding: 5,
  backgroundColor: "var(--mdc-theme-surface)",
  borderBottom: "1px solid var(--mdc-theme-on-background)"
});
export var navHeader = /*#__PURE__*/css({
  padding: 0,
  "&.mdc-drawer__header": {
    padding: "0 !important",
    minHeight: 0
  }
}, "label:navHeader;");
export var navContent = /*#__PURE__*/css({
  padding: "0 !important"
}, "label:navContent;");
export var logoStyle = /*#__PURE__*/css({
  "&.mdc-top-app-bar__title": {
    paddingLeft: 15,
    ".webiny-logo": {
      color: "var(--mdc-theme-primary)"
    }
  }
}, "label:logoStyle;");
export var MenuFooter = /*#__PURE__*/styled("div", {
  target: "e1b4rzuv1",
  label: "MenuFooter"
})({
  borderTop: "1px solid var(--mdc-theme-on-background)",
  a: {
    color: "var(--mdc-theme-text-on-primary)",
    textDecoration: "none"
  }
});
export var subFooter = /*#__PURE__*/css({
  ".mdc-drawer &.mdc-list-item": {
    borderTop: "1px solid var(--mdc-theme-on-background)",
    padding: "10px 16px",
    margin: 0
  }
}, "label:subFooter;");
