// @webiny/api-headless-cms-ddb-es/definitions/entry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _dynamodbToolbox = require("dynamodb-toolbox");

/**
 * TODO when saving this entity remove fields:
 * - ignore
 * - version
 * - savedOn
 */
var _default = params => {
  const {
    table
  } = params;
  return new _dynamodbToolbox.Entity({
    name: "ContentEntry",
    table,
    attributes: {
      PK: {
        type: "string",
        partitionKey: true
      },
      SK: {
        type: "string",
        sortKey: true
      },
      TYPE: {
        type: "string"
      },
      __type: {
        type: "string"
      },
      webinyVersion: {
        type: "string"
      },
      tenant: {
        type: "string"
      },
      entryId: {
        type: "string"
      },
      id: {
        type: "string"
      },
      createdBy: {
        type: "map"
      },
      ownedBy: {
        type: "map"
      },
      createdOn: {
        type: "string"
      },
      savedOn: {
        type: "string"
      },
      modelId: {
        type: "string"
      },
      locale: {
        type: "string"
      },
      publishedOn: {
        type: "string"
      },
      version: {
        type: "number"
      },
      locked: {
        type: "boolean"
      },
      status: {
        type: "string"
      },
      values: {
        type: "map"
      }
    }
  });
};

exports.default = _default;
