// @webiny/api-headless-cms-ddb-es/definitions/model.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _dynamodbToolbox = require("dynamodb-toolbox");

var _default = params => {
  const {
    table
  } = params;
  return new _dynamodbToolbox.Entity({
    name: "ContentModel",
    table,
    attributes: {
      PK: {
        partitionKey: true
      },
      SK: {
        sortKey: true
      },
      TYPE: {
        type: "string"
      },
      webinyVersion: {
        type: "string"
      },
      name: {
        type: "string"
      },
      modelId: {
        type: "string"
      },
      locale: {
        type: "string"
      },
      group: {
        type: "map"
      },
      description: {
        type: "string"
      },
      createdOn: {
        type: "string"
      },
      savedOn: {
        type: "string"
      },
      createdBy: {
        type: "map"
      },
      fields: {
        type: "list"
      },
      layout: {
        type: "list"
      },
      lockedFields: {
        type: "list"
      },
      titleFieldId: {
        type: "string"
      }
    }
  });
};

exports.default = _default;
