// @webiny/app-headless-cms/admin/components/FieldEditor/EditFieldDialog/PredefinedValues.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React, { useCallback, useRef, cloneElement } from "react";
import getValue from "./functions/getValue";
import setValue from "./functions/setValue";
export default function PredefinedValues(_ref) {
  var field = _ref.field,
      fieldPlugin = _ref.fieldPlugin,
      form = _ref.form;
  var memoizedBindComponents = useRef({});
  var BaseFormBind = form.Bind;
  var getBind = useCallback(function () {
    var index = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : -1;
    var memoKey = index;

    if (memoizedBindComponents.current[memoKey]) {
      return memoizedBindComponents.current[memoKey];
    }

    memoizedBindComponents.current[memoKey] = function Bind(_ref2) {
      var children = _ref2.children,
          name = _ref2.name;
      return /*#__PURE__*/React.createElement(BaseFormBind, {
        name: "predefinedValues.values"
      }, function (bind) {
        var props = _objectSpread(_objectSpread({}, bind), {}, {
          value: getValue({
            bind: bind,
            index: index,
            name: name
          }),
          onChange: function onChange(value) {
            setValue({
              value: value,
              bind: bind,
              index: index,
              name: name
            });
          }
        });

        if (typeof children === "function") {
          return children(props);
        }

        return /*#__PURE__*/cloneElement(children, props);
      });
    };

    return memoizedBindComponents.current[memoKey];
  }, []);
  return fieldPlugin.field.renderPredefinedValues({
    field: field,
    getBind: getBind
  });
}
