// @webiny/api-security-admin-users/crud/users.validation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fields = require("@commodo/fields");

var _commodoFieldsObject = require("commodo-fields-object");

var _validation = require("@webiny/validation");

var _UserPlugin = require("../plugins/UserPlugin");

const CreateUserDataModel = (0, _fields.withFields)({
  login: (0, _fields.string)({
    validation: _validation.validation.create("required,minLength:2")
  }),
  firstName: (0, _fields.string)({
    validation: _validation.validation.create("required,minLength:2")
  }),
  lastName: (0, _fields.string)({
    validation: _validation.validation.create("required,minLength:2")
  }),
  avatar: (0, _commodoFieldsObject.object)()
})();
const UpdateUserDataModel = (0, _fields.withFields)({
  avatar: (0, _commodoFieldsObject.object)(),
  firstName: (0, _fields.string)({
    validation: _validation.validation.create("minLength:2")
  }),
  lastName: (0, _fields.string)({
    validation: _validation.validation.create("minLength:2")
  }),
  group: (0, _fields.string)()
})();

var _default = new _UserPlugin.UserPlugin({
  async beforeCreate({
    inputData
  }) {
    const model = await new CreateUserDataModel().populate(inputData);
    await model.validate();
  },

  async beforeUpdate({
    inputData
  }) {
    const model = await new UpdateUserDataModel().populate(inputData);
    await model.validate();
  }

});

exports.default = _default;
