// @webiny/app-headless-cms/admin/views/contentEntries/ContentEntry/RevisionsList.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import React from "react";
import { css } from "emotion";
import { List } from "@webiny/ui/List";
import { Elevation } from "@webiny/ui/Elevation";
import { CircularProgress } from "@webiny/ui/Progress";
import { i18n } from "@webiny/app/i18n";
import RevisionListItem from "./RevisionListItem";
import { useContentEntry } from "../hooks/useContentEntry";
var t = i18n.ns("app-headless-cms/admin/plugins/content-details/content-revisions");
var style = {
  list: /*#__PURE__*/css({
    position: "relative",
    margin: 25,
    display: "flex",
    flexDirection: "column",
    overflow: "auto",
    maxHeight: "calc(100vh - 160px)",
    ".mdc-list .mdc-list-item": {
      borderBottom: "1px solid var(--mdc-theme-on-background)"
    },
    ".mdc-list .mdc-list-item:last-child": {
      borderBottom: "none"
    }
  }, "label:list;"),
  nothingToShow: /*#__PURE__*/css({
    padding: 20,
    margin: 25,
    textAlign: "center"
  }, "label:nothingToShow;")
};

var RevisionsList = function RevisionsList() {
  var _useContentEntry = useContentEntry(),
      entry = _useContentEntry.entry,
      revisions = _useContentEntry.revisions,
      loading = _useContentEntry.loading;

  return /*#__PURE__*/React.createElement(Elevation, {
    className: style.list,
    z: 2
  }, loading && /*#__PURE__*/React.createElement(CircularProgress, null), entry.id && revisions.length ? /*#__PURE__*/React.createElement(List, {
    nonInteractive: true,
    twoLine: true,
    "data-testid": "cms.content-form.revisions"
  }, revisions.map(function (revision) {
    return /*#__PURE__*/React.createElement(RevisionListItem, {
      revision: revision,
      key: revision.id
    });
  })) : /*#__PURE__*/React.createElement("div", {
    className: style.nothingToShow
  }, t(_templateObject || (_templateObject = _taggedTemplateLiteral(["No revisions to show."])))));
};

export default RevisionsList;
