// @webiny/api-file-manager/plugins/crud/utils/createFileModel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fields = require("@commodo/fields");

var _commodoFieldsObject = require("commodo-fields-object");

var _validation = require("@webiny/validation");

var _default = (create = true) => {
  return (0, _fields.withFields)({
    key: (0, _fields.string)({
      validation: _validation.validation.create(`${create ? "required," : ""}maxLength:1000`)
    }),
    name: (0, _fields.string)({
      validation: _validation.validation.create("maxLength:1000")
    }),
    size: (0, _fields.number)(),
    type: (0, _fields.string)({
      validation: _validation.validation.create("maxLength:255")
    }),
    meta: (0, _commodoFieldsObject.object)({
      value: {
        private: false
      }
    }),
    tags: (0, _fields.onSet)(value => {
      if (!Array.isArray(value)) {
        return null;
      }

      return value.map(item => item.toLowerCase());
    })((0, _fields.string)({
      list: true,
      validation: tags => {
        if (!Array.isArray(tags)) {
          return;
        }

        if (tags.length > 15) {
          throw Error("You cannot set more than 15 tags.");
        }

        for (let i = 0; i < tags.length; i++) {
          const tag = tags[i];

          if (typeof tag !== "string") {
            throw Error("Tag must be typeof string.");
          }

          if (tag.length > 50) {
            throw Error(`Tag ${tag} is more than 50 characters long.`);
          }
        }
      }
    }))
  })();
};

exports.default = _default;
