// @webiny/app-admin/ui/views/AdminView/components/Dialog.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _get from "lodash/get";
import React, { useCallback } from "react";
import { useUi } from "@webiny/app/hooks/useUi";
import { Dialog, DialogAccept, DialogTitle, DialogActions, DialogContent } from "@webiny/ui/Dialog";
export var DialogContainer = function DialogContainer() {
  var ui = useUi();

  var message = _get(ui, "dialog.message");

  var _get2 = _get(ui, "dialog.options", {}),
      dataTestId = _get2.dataTestId,
      title = _get2.title,
      _get2$actions = _get2.actions,
      actions = _get2$actions === void 0 ? {
    cancel: null,
    accept: {
      label: "OK"
    }
  } : _get2$actions;

  var hideDialog = useCallback(function () {
    ui.setState(function (ui) {
      return _objectSpread(_objectSpread({}, ui), {}, {
        dialog: null
      });
    });
  }, [ui]);
  return /*#__PURE__*/React.createElement(Dialog, {
    open: !!message,
    onClose: hideDialog,
    "data-testid": dataTestId
  }, title && /*#__PURE__*/React.createElement(DialogTitle, null, title), /*#__PURE__*/React.createElement(DialogContent, null, message), /*#__PURE__*/React.createElement(DialogActions, null, actions.cancel && /*#__PURE__*/React.createElement(DialogAccept, {
    onClick: actions.cancel.onClick
  }, actions.cancel.label), actions.accept && /*#__PURE__*/React.createElement(DialogAccept, {
    onClick: actions.accept.onClick
  }, actions.accept.label)));
};
