// @webiny/app-page-builder/editor/ui/views/PageSettingsView/PageSettingsTabsElementRenderer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { css } from "emotion";
import { UIRenderer } from "@webiny/app-admin/ui/UIRenderer";
import { List } from "@webiny/ui/List";
export var listStyle = /*#__PURE__*/css({
  "&.mdc-list": {
    padding: 0,
    backgroundColor: "var(--mdc-theme-surface)"
  }
}, "label:listStyle;");
export var PageSettingsTabsElementRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(PageSettingsTabsElementRenderer, _UIRenderer);

  var _super = _createSuper(PageSettingsTabsElementRenderer);

  function PageSettingsTabsElementRenderer() {
    _classCallCheck(this, PageSettingsTabsElementRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(PageSettingsTabsElementRenderer, [{
    key: "render",
    value: function render(_ref) {
      var next = _ref.next;
      return /*#__PURE__*/React.createElement(List, {
        twoLine: true,
        className: listStyle
      }, next());
    }
  }]);

  return PageSettingsTabsElementRenderer;
}(UIRenderer);
