// @webiny/app-page-builder/editor/ui/views/PageSettingsView/PageSettingsTabElementRenderer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import styled from "@emotion/styled";
import { css } from "emotion";
import { UIRenderer } from "@webiny/app-admin/ui/UIRenderer";
import { ListItem, ListItemGraphic } from "@webiny/ui/List";
import { Icon } from "@webiny/ui/Icon";
import { Typography } from "@webiny/ui/Typography";
import { PageSettingsView } from "../PageSettingsView";
export var listItem = /*#__PURE__*/css({
  padding: "15px 20px",
  cursor: "pointer",
  borderBottom: "1px solid var(--mdc-theme-background)",
  "&:last-child": {
    borderBottom: "none"
  },
  ".mdc-list-item__graphic": {
    marginRight: 20
  }
}, "label:listItem;");
export var ListItemTitle = /*#__PURE__*/styled("div", {
  target: "ex1xcae0",
  label: "ListItemTitle"
})({
  fontWeight: 600,
  marginBottom: 5
});
export var TitleContent = /*#__PURE__*/styled("div", {
  target: "ex1xcae1",
  label: "TitleContent"
})({
  display: "flex",
  flexDirection: "column"
});
export var PageSettingsTabElementRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(PageSettingsTabElementRenderer, _UIRenderer);

  var _super = _createSuper(PageSettingsTabElementRenderer);

  function PageSettingsTabElementRenderer() {
    _classCallCheck(this, PageSettingsTabElementRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(PageSettingsTabElementRenderer, [{
    key: "render",
    value: function render(_ref) {
      var element = _ref.element;

      var _element$getView$getP = element.getView(PageSettingsView).getPageSettingsHook(),
          setActiveSection = _element$getView$getP.setActiveSection;

      return /*#__PURE__*/React.createElement(ListItem, {
        key: element.id,
        className: listItem,
        onClick: function onClick() {
          return setActiveSection(element.id);
        }
      }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
        icon: element.config.icon
      })), /*#__PURE__*/React.createElement(TitleContent, null, /*#__PURE__*/React.createElement(ListItemTitle, null, element.config.title), /*#__PURE__*/React.createElement(Typography, {
        use: "subtitle2"
      }, element.config.description)));
    }
  }]);

  return PageSettingsTabElementRenderer;
}(UIRenderer);
