// @webiny/app-page-builder/admin/plugins/menuItems/link/LinkForm.js
import { css as _css } from "emotion";
import * as React from "react";
import { Form } from "@webiny/form";
import { Input } from "@webiny/ui/Input";
import { Typography } from "@webiny/ui/Typography";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ButtonSecondary, ButtonPrimary } from "@webiny/ui/Button";
import { Elevation } from "@webiny/ui/Elevation";
import { validation } from "@webiny/validation";
var menuFormStyle = {
  color: "var(--mdc-theme-on-surface)",
  backgroundColor: "var(--mdc-theme-background) !important"
};

var LinkForm = function LinkForm(_ref) {
  var data = _ref.data,
      onSubmit = _ref.onSubmit,
      onCancel = _ref.onCancel;
  return /*#__PURE__*/React.createElement(Elevation, {
    z: 4,
    className: _css(menuFormStyle)
  }, /*#__PURE__*/React.createElement(Form, {
    data: data,
    onSubmit: onSubmit
  }, function (_ref2) {
    var submit = _ref2.submit,
        Bind = _ref2.Bind;
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Typography, {
      use: "overline"
    }, "Link menu item"))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "title",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      label: "Title"
    })))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "url",
      validators: validation.create("required,url:allowRelative:allowHref")
    }, /*#__PURE__*/React.createElement(Input, {
      label: "URL"
    })))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(ButtonSecondary, {
      onClick: onCancel
    }, "Cancel"), /*#__PURE__*/React.createElement(ButtonPrimary, {
      onClick: submit,
      style: {
        float: "right"
      }
    }, "Save menu item"))));
  }));
};

export default LinkForm;
