// @webiny/app-security-admin-users/plugins/installation.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
var _excluded = ["subscribed"];

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import gql from "graphql-tag";
import React, { useState, useCallback } from "react";
import { useApolloClient } from "@apollo/react-hooks";
import { Form } from "@webiny/form";
import { i18n } from "@webiny/app/i18n";
import { Alert } from "@webiny/ui/Alert";
import { ButtonPrimary } from "@webiny/ui/Button";
import { Input } from "@webiny/ui/Input";
import { Checkbox } from "@webiny/ui/Checkbox";
import { Grid, Cell } from "@webiny/ui/Grid";
import { CircularProgress } from "@webiny/ui/Progress";
import { validation } from "@webiny/validation";
import { SimpleForm, SimpleFormHeader, SimpleFormFooter, SimpleFormContent } from "@webiny/app-admin/components/SimpleForm";
import { View } from "@webiny/app/components/View";
var t = i18n.ns("app-security/admin/installation");
var IS_INSTALLED = gql(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    query IsSecurityInstalled {\n        security {\n            version\n        }\n    }\n"])));
var INSTALL = gql(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n    mutation InstallSecurity($data: SecurityInstallInput!) {\n        security {\n            install(data: $data) {\n                data\n                error {\n                    code\n                    message\n                }\n            }\n        }\n    }\n"])));

var Install = function Install(_ref) {
  var onInstalled = _ref.onInstalled;
  var client = useApolloClient();

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      loading = _useState2[0],
      setLoading = _useState2[1];

  var _useState3 = useState(null),
      _useState4 = _slicedToArray(_useState3, 2),
      error = _useState4[0],
      setError = _useState4[1];

  var onSubmit = useCallback( /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(_ref2) {
      var subscribed, form, _yield$client$mutate, res, error;

      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              subscribed = _ref2.subscribed, form = _objectWithoutProperties(_ref2, _excluded);
              setLoading(true);
              setError(null);
              _context.next = 5;
              return client.mutate({
                mutation: INSTALL,
                variables: {
                  data: form
                }
              });

            case 5:
              _yield$client$mutate = _context.sent;
              res = _yield$client$mutate.data;
              setLoading(false);
              error = res.security.install.error;

              if (!error) {
                _context.next = 12;
                break;
              }

              setError(error);
              return _context.abrupt("return");

            case 12:
              if (!subscribed) {
                _context.next = 21;
                break;
              }

              _context.prev = 13;
              _context.next = 16;
              return fetch("https://app.mailerlite.com/webforms/submit/g9f1i1?fields%5Bemail%5D=" + encodeURIComponent(form.email) + "&ml-submit=1&ajax=1", {
                method: "GET",
                headers: {
                  "Content-Type": "application/x-www-form-urlencoded"
                }
              });

            case 16:
              _context.next = 21;
              break;

            case 18:
              _context.prev = 18;
              _context.t0 = _context["catch"](13);
              setError("Unable to subscribe you to the newsletter " + _context.t0);

            case 21:
              onInstalled();

            case 22:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[13, 18]]);
    }));

    return function (_x) {
      return _ref3.apply(this, arguments);
    };
  }(), []);
  var privacyPolicyLink = /*#__PURE__*/React.createElement("a", {
    href: "https://www.webiny.com/privacy-policy"
  }, "privacy policy");
  return /*#__PURE__*/React.createElement(Form, {
    onSubmit: onSubmit,
    submitOnEnter: true
  }, function (_ref4) {
    var data = _ref4.data,
        Bind = _ref4.Bind,
        submit = _ref4.submit;
    return /*#__PURE__*/React.createElement(SimpleForm, null, loading && /*#__PURE__*/React.createElement(CircularProgress, null), /*#__PURE__*/React.createElement(SimpleFormHeader, {
      title: "Install Security"
    }), /*#__PURE__*/React.createElement(SimpleFormContent, null, /*#__PURE__*/React.createElement(Grid, null, error && /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Alert, {
      title: "Something went wrong",
      type: "danger"
    }, error.message)), /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, "Let's create your admin user:"), /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "firstName",
      validators: validation.create("required,minLength:2")
    }, /*#__PURE__*/React.createElement(Input, {
      label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["First Name"])))
    }))), /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "lastName",
      validators: validation.create("required,minLength:2")
    }, /*#__PURE__*/React.createElement(Input, {
      label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Last Name"])))
    }))), /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "login",
      validators: validation.create("required,email"),
      beforeChange: function beforeChange(value, cb) {
        return cb(value.toLowerCase());
      }
    }, /*#__PURE__*/React.createElement(Input, {
      label: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Email"])))
    }))), /*#__PURE__*/React.createElement(View, {
      name: "security.installation.fields",
      props: {
        Bind: Bind,
        data: data
      }
    })), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "subscribed"
    }, /*#__PURE__*/React.createElement(Checkbox, {
      label: /*#__PURE__*/React.createElement("span", null, "I want to receive updates on product improvements and new features. Doing so I accept Webiny's", " ", privacyPolicyLink, ".")
    }))))), /*#__PURE__*/React.createElement(SimpleFormFooter, null, /*#__PURE__*/React.createElement(ButtonPrimary, {
      "data-testid": "install-security-button",
      onClick: submit
    }, "Install security")));
  });
};

export default [{
  name: "admin-installation-security",
  type: "admin-installation",
  secure: false,
  title: "Security",
  getInstalledVersion: function getInstalledVersion(_ref5) {
    return _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
      var client, _yield$client$query, data;

      return _regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              client = _ref5.client;
              _context2.next = 3;
              return client.query({
                query: IS_INSTALLED
              });

            case 3:
              _yield$client$query = _context2.sent;
              data = _yield$client$query.data;
              return _context2.abrupt("return", data.security.version);

            case 6:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }))();
  },
  render: function render(_ref6) {
    var onInstalled = _ref6.onInstalled;
    return /*#__PURE__*/React.createElement(Install, {
      onInstalled: onInstalled
    });
  }
}];
