// @webiny/app-security-admin-users/plugins/routes.js
import React from "react";
import Helmet from "react-helmet";
import { Route } from "@webiny/react-router";
import { AdminLayout } from "@webiny/app-admin/components/AdminLayout";
import { SecureRoute } from "@webiny/app-security/components";
import { UIViewComponent } from "@webiny/app-admin/ui/UIView";
import { Groups } from "../ui/views/Groups";
import { ApiKeys } from "../ui/views/ApiKeys";
import { Account } from "../ui/views/Account";
import { Permission } from "./constants";
import { UsersView } from "../ui/views/Users/UsersView";
var plugins = [{
  name: "route-security-groups",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/security/groups",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: Permission.Groups
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, {
        title: "Security - Groups"
      }), /*#__PURE__*/React.createElement(Groups, null)));
    }
  })
}, {
  name: "route-security-users",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/security/users",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: Permission.Users
      }, /*#__PURE__*/React.createElement(UIViewComponent, {
        view: new UsersView()
      }));
    }
  })
}, {
  name: "route-security-api-keys",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/security/api-keys",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: Permission.ApiKeys
      }, /*#__PURE__*/React.createElement(AdminLayout, {
        title: "Security - API Keys"
      }, /*#__PURE__*/React.createElement(ApiKeys, null)));
    }
  })
}, {
  name: "route-security-account",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/account",
    render: function render() {
      return /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, {
        title: "User Account"
      }), /*#__PURE__*/React.createElement(Account, null));
    }
  })
}];
export default plugins;
