// @webiny/app-page-builder/editor/plugins/elements/utils/textUtils.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import { plugins } from "@webiny/plugins";

var getTypographyFromTheme = function getTypographyFromTheme(type) {
  var _plugins$byType = plugins.byType("pb-theme"),
      _plugins$byType2 = _slicedToArray(_plugins$byType, 1),
      theme = _plugins$byType2[0].theme;

  var themeElement = theme.elements[type];

  if (!themeElement) {
    console.warn("No element of type: \"".concat(type, ": found in theme."));
    return "";
  }

  var types = themeElement.types;

  var _types = _slicedToArray(types, 1),
      defaultType = _types[0];

  return defaultType.className;
};

export var createInitialTextValue = function createInitialTextValue(_ref) {
  var type = _ref.type,
      _ref$alignment = _ref.alignment,
      alignment = _ref$alignment === void 0 ? "left" : _ref$alignment,
      _ref$tag = _ref.tag,
      tag = _ref$tag === void 0 ? "div" : _ref$tag;
  var typography = getTypographyFromTheme(type);
  return {
    type: type,
    typography: typography,
    alignment: alignment,
    tag: tag
  };
};
