// @webiny/app-page-builder/editor/plugins/elements/image/ImageSettings.js
import React, { useCallback } from "react";
import { useRecoilValue } from "recoil";
import { css } from "emotion"; // Components

import Accordion from "../../elementSettings/components/Accordion";
import Wrapper from "../../elementSettings/components/Wrapper";
import InputField from "../../elementSettings/components/InputField";
import SpacingPicker from "../../elementSettings/components/SpacingPicker";
import useUpdateHandlers from "../../elementSettings/useUpdateHandlers";
import { justifySelfEndStyle } from "../../elementSettings/components/StyledComponents";
import { activeElementAtom, elementByIdSelector } from "../../../recoil/modules";
import { WIDTH_UNIT_OPTIONS, HEIGHT_UNIT_OPTIONS } from "../../elementSettings/elementSettingsUtils";
var classes = {
  grid: /*#__PURE__*/css({
    "&.mdc-layout-grid": {
      padding: 0,
      marginBottom: 24
    }
  }, "label:grid;")
};
var spacingPickerStyle = /*#__PURE__*/css({
  width: "120px",
  "& .inner-wrapper": {
    display: "flex"
  }
}, "label:spacingPickerStyle;");

var ImageSettings = function ImageSettings(_ref) {
  var _ref$defaultAccordion = _ref.defaultAccordionValue,
      defaultAccordionValue = _ref$defaultAccordion === void 0 ? false : _ref$defaultAccordion;
  var activeElementId = useRecoilValue(activeElementAtom);
  var element = useRecoilValue(elementByIdSelector(activeElementId));
  var image = element.data.image;

  var _useUpdateHandlers = useUpdateHandlers({
    element: element,
    dataNamespace: "data.image"
  }),
      getUpdateValue = _useUpdateHandlers.getUpdateValue;

  var updateTitle = useCallback(function (value) {
    return getUpdateValue("title")(value);
  }, []);
  var updateWidth = useCallback(function (value) {
    return getUpdateValue("width")(value);
  }, []);
  var updateHeight = useCallback(function (value) {
    return getUpdateValue("height")(value);
  }, []);
  return /*#__PURE__*/React.createElement(Accordion, {
    title: "Image",
    defaultValue: defaultAccordionValue
  }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Wrapper, {
    containerClassName: classes.grid,
    label: "Title"
  }, /*#__PURE__*/React.createElement(InputField, {
    value: (image === null || image === void 0 ? void 0 : image.title) || "",
    onChange: updateTitle
  })), /*#__PURE__*/React.createElement(Wrapper, {
    containerClassName: classes.grid,
    label: "Width",
    rightCellClassName: justifySelfEndStyle
  }, /*#__PURE__*/React.createElement(SpacingPicker, {
    value: (image === null || image === void 0 ? void 0 : image.width) || "",
    onChange: updateWidth,
    options: WIDTH_UNIT_OPTIONS,
    useDefaultStyle: false,
    className: spacingPickerStyle
  })), /*#__PURE__*/React.createElement(Wrapper, {
    containerClassName: classes.grid,
    label: "Height",
    rightCellClassName: justifySelfEndStyle
  }, /*#__PURE__*/React.createElement(SpacingPicker, {
    value: (image === null || image === void 0 ? void 0 : image.height) || "",
    onChange: updateHeight,
    options: HEIGHT_UNIT_OPTIONS,
    useDefaultStyle: false,
    className: spacingPickerStyle
  }))));
};

export default /*#__PURE__*/React.memo(ImageSettings);
