// @webiny/app-headless-cms/admin/plugins/transformers/dateTransformer.js
import WebinyError from "@webiny/error";

var transformDate = function transformDate(value, type) {
  if (type === "date") {
    return value.substr(0, 10);
  }

  return value;
};

var excludeTypesTransformation = ["time", "dateTimeWithTimezone"];
export default (function () {
  return {
    type: "cms-field-value-transformer",
    name: "cms-field-value-transformer-date",
    fieldType: "datetime",
    transform: function transform(value, field) {
      // check types in packages/app-headless-cms/src/admin/plugins/fieldRenderers/dateTime/dateTimeField.tsx
      var type = field.settings.type;

      if (!value) {
        console.log("Field \"".concat(field.fieldId, "\" has no value."));
        return null;
      } else if (excludeTypesTransformation.includes(type) === true) {
        return value;
      } else if (type === "dateTimeWithoutTimezone" && value.includes(" ")) {
        /**
         * Need to replace space and add .000Z to datetime value because it can look like: 2021-02-02 16:34:56
         */
        value = "".concat(value.replace(" ", "T"), ".000Z");
      }

      try {
        var result = new Date(value).toISOString();
        return transformDate(result, type);
      } catch (ex) {
        throw new WebinyError("Could not transform \"".concat(value, "\" to a date."), "TRANSFORM_ERROR", {
          message: ex.message,
          code: ex.code
        });
      }
    }
  };
});
