// @webiny/app-security-admin-users/components/NotAuthorizedError.js
import * as React from "react";
import { Link } from "@webiny/react-router";
import { css } from "emotion";
import styled from "@emotion/styled";
import Helmet from "react-helmet";
import authErrorImg from "../assets/images/SecureRouteError.svg";
import { Typography } from "@webiny/ui/Typography";
var ContentWrapper = /*#__PURE__*/styled("div", {
  target: "ezgbj0z0",
  label: "ContentWrapper"
})({
  display: "block",
  paddingTop: "15%",
  textAlign: "center",
  margin: "auto"
});
var styles = {
  authErrorImgStyle: /*#__PURE__*/css({
    width: "192px",
    paddingBottom: "24px"
  }, "label:authErrorImgStyle;"),
  bodyStyle: /*#__PURE__*/css({
    color: "var(--mdc-theme-text-primary-on-background)",
    display: "block"
  }, "label:bodyStyle;"),
  linkStyle: /*#__PURE__*/css({
    textDecoration: "none",
    "&:hover": {
      textDecoration: "none"
    }
  }, "label:linkStyle;")
};

var NotAuthorizedError = function NotAuthorizedError() {
  return /*#__PURE__*/React.createElement(ContentWrapper, null, /*#__PURE__*/React.createElement(Helmet, {
    title: "Not authorized"
  }), /*#__PURE__*/React.createElement("img", {
    className: styles.authErrorImgStyle,
    src: authErrorImg,
    alt: "Not Authorized"
  }), /*#__PURE__*/React.createElement(Typography, {
    use: "body1",
    className: styles.bodyStyle
  }, "You are not authorized to view this route."), /*#__PURE__*/React.createElement(Typography, {
    use: "body1",
    className: styles.bodyStyle
  }, "Please contact your administrator to request access."), /*#__PURE__*/React.createElement(Link, {
    to: "/",
    className: styles.linkStyle
  }, "Take me back."));
};

export default NotAuthorizedError;
