// @webiny/app-headless-cms/admin/elements/NothingToShowElement.js
import _createClass from "@babel/runtime/helpers/createClass";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { css } from "emotion";
import { NavigationMenuElement } from "@webiny/app-admin/ui/elements/NavigationMenuElement";
import { List, ListItem } from "@webiny/ui/List";
import { UIRenderer } from "@webiny/app-admin/ui/UIRenderer";
var linkStyle = /*#__PURE__*/css({
  color: "var(--mdc-theme-text-primary-on-background)",
  textDecoration: "none",
  display: "flex",
  alignItems: "center",
  width: "100%",
  height: "100%",
  outline: "none",
  paddingLeft: 65,
  "&:hover": {
    textDecoration: "none"
  }
}, "label:linkStyle;");
var submenuItems = /*#__PURE__*/css({
  ".mdc-drawer &.mdc-list-item": {
    paddingLeft: 0
  }
}, "label:submenuItems;");
var submenuList = /*#__PURE__*/css({
  "&.mdc-list": {
    padding: 0
  }
}, "label:submenuList;");
export var NothingToShowElement = /*#__PURE__*/function (_NavigationMenuElemen) {
  _inherits(NothingToShowElement, _NavigationMenuElemen);

  var _super = _createSuper(NothingToShowElement);

  function NothingToShowElement(id, config) {
    var _this;

    _classCallCheck(this, NothingToShowElement);

    _this = _super.call(this, id, config);

    _this.addRenderer(new NothingToShowRenderer());

    return _this;
  }

  return NothingToShowElement;
}(NavigationMenuElement);
export var NothingToShowRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(NothingToShowRenderer, _UIRenderer);

  var _super2 = _createSuper(NothingToShowRenderer);

  function NothingToShowRenderer() {
    _classCallCheck(this, NothingToShowRenderer);

    return _super2.apply(this, arguments);
  }

  _createClass(NothingToShowRenderer, [{
    key: "canRender",
    value: function canRender(element) {
      return element instanceof NothingToShowElement;
    }
  }, {
    key: "render",
    value: function render(_ref) {
      var element = _ref.element;
      return /*#__PURE__*/React.createElement(List, {
        className: submenuList,
        style: {
          opacity: 0.4
        }
      }, /*#__PURE__*/React.createElement(ListItem, {
        className: submenuItems,
        ripple: false,
        disabled: true
      }, /*#__PURE__*/React.createElement("span", {
        className: linkStyle
      }, element.config.label)));
    }
  }]);

  return NothingToShowRenderer;
}(UIRenderer);
