// @webiny/app-headless-cms/admin/plugins/fieldRenderers/ref/refInputs.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import ContentEntriesMultiAutocomplete from "./components/ContentEntriesMultiAutoComplete";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/fields/ref");

var getKey = function getKey(field, bind) {
  return bind.form.state.data.id + "." + field.fieldId;
};

var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-ref-inputs",
  renderer: {
    rendererName: "ref-inputs",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Reference Inputs"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders an auto-complete input, allowing selection of multiple values."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "ref" && field.multipleValues;
    },
    render: function render(props) {
      var Bind = props.getBind();
      return /*#__PURE__*/React.createElement(Bind, null, function (bind) {
        return /*#__PURE__*/React.createElement(ContentEntriesMultiAutocomplete, Object.assign({
          key: getKey(props.field, bind)
        }, props, {
          bind: bind
        }));
      });
    }
  }
};
export default plugin;
