// @webiny/app-headless-cms/admin/views/contentEntries/ContentEntry/header/Header.js
import React from "react";
import { css } from "emotion";
import { renderPlugins } from "@webiny/app/plugins";
import { Grid, Cell } from "@webiny/ui/Grid";
import classNames from "classnames";
var toolbarGrid = /*#__PURE__*/css({
  borderBottom: "1px solid var(--mdc-theme-on-background)"
}, "label:toolbarGrid;");
var headerActions = /*#__PURE__*/css({
  display: "flex",
  alignItems: "center"
}, "label:headerActions;");
var headerActionsLeft = /*#__PURE__*/css({
  justifyContent: "flex-end"
}, "label:headerActionsLeft;");
var headerActionsRight = /*#__PURE__*/css({
  justifyContent: "flex-start"
}, "label:headerActionsRight;");

var Header = function Header(props) {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Grid, {
    className: toolbarGrid
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 4,
    className: classNames(headerActions, headerActionsRight)
  }, renderPlugins("cms-content-details-header-left", props)), /*#__PURE__*/React.createElement(Cell, {
    span: 8,
    className: classNames(headerActions, headerActionsLeft)
  }, renderPlugins("cms-content-details-header-right", props))));
};

export default Header;
