// @webiny/api-file-manager/handlers/transform/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getOptimizedTransformedImageKeyPrefix = exports.getOptimizedImageKeyPrefix = exports.getImageKey = exports.OPTIMIZED_IMAGE_PREFIX = exports.OPTIMIZED_TRANSFORMED_IMAGE_PREFIX = exports.SUPPORTED_TRANSFORMABLE_IMAGES = exports.SUPPORTED_IMAGES = void 0;

var _objectHash = _interopRequireDefault(require("object-hash"));

const SUPPORTED_IMAGES = [".jpg", ".jpeg", ".png", ".svg", ".gif"];
exports.SUPPORTED_IMAGES = SUPPORTED_IMAGES;
const SUPPORTED_TRANSFORMABLE_IMAGES = [".jpg", ".jpeg", ".png"];
exports.SUPPORTED_TRANSFORMABLE_IMAGES = SUPPORTED_TRANSFORMABLE_IMAGES;
const OPTIMIZED_TRANSFORMED_IMAGE_PREFIX = "img-o-t-";
exports.OPTIMIZED_TRANSFORMED_IMAGE_PREFIX = OPTIMIZED_TRANSFORMED_IMAGE_PREFIX;
const OPTIMIZED_IMAGE_PREFIX = "img-o-";
exports.OPTIMIZED_IMAGE_PREFIX = OPTIMIZED_IMAGE_PREFIX;

const getOptimizedImageKeyPrefix = key => {
  return `${OPTIMIZED_IMAGE_PREFIX}${(0, _objectHash.default)(key)}-`;
};

exports.getOptimizedImageKeyPrefix = getOptimizedImageKeyPrefix;

const getOptimizedTransformedImageKeyPrefix = key => {
  return `${OPTIMIZED_TRANSFORMED_IMAGE_PREFIX}${(0, _objectHash.default)(key)}-`;
};

exports.getOptimizedTransformedImageKeyPrefix = getOptimizedTransformedImageKeyPrefix;

const getImageKey = ({
  key,
  transformations
}) => {
  if (!transformations) {
    const prefix = getOptimizedImageKeyPrefix(key);
    return prefix + key;
  }

  const prefix = getOptimizedTransformedImageKeyPrefix(key);
  return `${prefix}${(0, _objectHash.default)(transformations)}-${key}`;
};

exports.getImageKey = getImageKey;
