// @webiny/app-form-builder/admin/plugins/editor/defaultBar/PublishFormButton.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React from "react";
import { useMutation } from "@apollo/react-hooks";
import { useRouter } from "@webiny/react-router";
import { ButtonPrimary } from "@webiny/ui/Button";
import { ConfirmationDialog } from "@webiny/ui/ConfirmationDialog";
import { useSnackbar } from "@webiny/app-admin/hooks/useSnackbar";
import { i18n } from "@webiny/app/i18n";
import { useFormEditor } from "../../../components/FormEditor/Context";
import { PUBLISH_REVISION } from "../../../graphql";
import usePermission from "../../../../hooks/usePermission";
var t = i18n.namespace("FormEditor.PublishPageButton");

var PublishFormButton = function PublishFormButton() {
  var _useFormEditor = useFormEditor(),
      data = _useFormEditor.state.data;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var _useRouter = useRouter(),
      history = _useRouter.history;

  var _useMutation = useMutation(PUBLISH_REVISION),
      _useMutation2 = _slicedToArray(_useMutation, 1),
      publish = _useMutation2[0];

  var _usePermission = usePermission(),
      canPublish = _usePermission.canPublish; // Render nothing if required permission is missing.


  if (!canPublish()) {
    return null;
  }

  return /*#__PURE__*/React.createElement(ConfirmationDialog, {
    title: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Publish form"]))),
    message: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["You are about to publish this form, are you sure want to continue?"]))),
    "data-testid": "fb.editor.default-bar.publish-dialog"
  }, function (_ref) {
    var showConfirmation = _ref.showConfirmation;
    return /*#__PURE__*/React.createElement(ButtonPrimary, {
      "data-testid": "fb.editor.default-bar.publish",
      onClick: /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                showConfirmation( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
                  return _regeneratorRuntime.wrap(function _callee$(_context) {
                    while (1) {
                      switch (_context.prev = _context.next) {
                        case 0:
                          _context.next = 2;
                          return publish({
                            variables: {
                              revision: data.id
                            },
                            update: function update(cache, _ref4) {
                              var data = _ref4.data;

                              var _ref5 = data.formBuilder.publishRevision || {},
                                  revision = _ref5.data,
                                  error = _ref5.error;

                              if (error) {
                                return showSnackbar(error.message);
                              }

                              history.push("/form-builder/forms?id=".concat(encodeURIComponent(revision.id))); // Let's wait a bit, because we are also redirecting the user.

                              // Let's wait a bit, because we are also redirecting the user.
                              setTimeout(function () {
                                showSnackbar(t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Your form was published successfully!"]))));
                              }, 500);
                            }
                          });

                        case 2:
                        case "end":
                          return _context.stop();
                      }
                    }
                  }, _callee);
                })));

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }))
    }, data.version > 1 ? t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Publish changes"]))) : t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Publish"]))));
  });
};

export default PublishFormButton;
