// @webiny/app-form-builder/admin/plugins/editor/defaultBar/Revisions.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import React from "react";
import { css } from "emotion";
import { useRouter } from "@webiny/react-router";
import { Menu, MenuItem } from "@webiny/ui/Menu";
import { ButtonDefault } from "@webiny/ui/Button";
import { Icon } from "@webiny/ui/Icon";
import { Typography } from "@webiny/ui/Typography";
import { ReactComponent as DownButton } from "@svgr/webpack!./icons/round-arrow_drop_down-24px.svg";
import { i18n } from "@webiny/app/i18n";
import { useQuery } from "@apollo/react-hooks";
import { GET_FORM_REVISIONS } from "../../../graphql";
import { useFormEditor } from "../../../components/FormEditor";
var t = i18n.namespace("FormEditor.RevisionsMenu");
var buttonStyle = /*#__PURE__*/css({
  "&.mdc-button": {
    color: "var(--mdc-theme-text-primary-on-background) !important"
  }
}, "label:buttonStyle;");
var menuList = /*#__PURE__*/css({
  ".mdc-list-item": {
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "baseline",
    textAlign: "left"
  }
}, "label:menuList;");

var Revisions = function Revisions() {
  var _useFormEditor = useFormEditor(),
      data = _useFormEditor.state.data;

  var _useRouter = useRouter(),
      history = _useRouter.history;

  var getRevisions = useQuery(GET_FORM_REVISIONS, {
    variables: {
      id: data.id.split("#")[0]
    }
  });
  var revisions = getRevisions.loading ? [] : getRevisions.data.formBuilder.revisions.data;
  return /*#__PURE__*/React.createElement(Menu, {
    className: menuList,
    onSelect: function onSelect(evt) {
      history.push("/form-builder/forms/".concat(encodeURIComponent(revisions[evt.detail.index].id)));
    },
    handle: /*#__PURE__*/React.createElement(ButtonDefault, {
      className: buttonStyle
    }, t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Revisions"]))), " ", /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(DownButton, null)
    }))
  }, revisions.map(function (rev) {
    return /*#__PURE__*/React.createElement(MenuItem, {
      key: rev.id,
      disabled: rev.status !== "draft"
    }, /*#__PURE__*/React.createElement(Typography, {
      use: "body2"
    }, "v", rev.version), /*#__PURE__*/React.createElement(Typography, {
      use: "caption"
    }, "(", rev.status, ")"));
  }));
};

export default /*#__PURE__*/React.memo(Revisions);
