// @webiny/app/contexts/Ui/index.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import { plugins } from "@webiny/plugins";
export var UiContext = /*#__PURE__*/React.createContext({});
export var UiProvider = /*#__PURE__*/function (_React$Component) {
  _inherits(UiProvider, _React$Component);

  var _super = _createSuper(UiProvider);

  function UiProvider() {
    var _this;

    _classCallCheck(this, UiProvider);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      ui: {}
    });

    _defineProperty(_assertThisInitialized(_this), "setData", function (setter) {
      return _this.setState(function (state) {
        return {
          ui: _objectSpread(_objectSpread({}, state.ui), setter(state.ui))
        };
      });
    });

    return _this;
  }

  _createClass(UiProvider, [{
    key: "render",
    value: function render() {
      var value = _objectSpread(_objectSpread({}, this.state.ui), {}, {
        setState: this.setData
      });

      var uiStatePlugins = plugins.byType("ui-state");
      return /*#__PURE__*/React.createElement(UiContext.Provider, {
        value: value
      }, uiStatePlugins.map(function (pl) {
        return /*#__PURE__*/React.cloneElement(pl.render(), {
          key: pl.name
        });
      }), this.props.children);
    }
  }]);

  return UiProvider;
}(React.Component);
export var UiConsumer = function UiConsumer(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/React.createElement(UiContext.Consumer, null, function (ui) {
    return /*#__PURE__*/React.cloneElement(children, {
      ui: ui
    });
  });
};
