// @webiny/app-headless-cms/admin/views/contentEntries/ContentEntry/header/requestChanges/RequestChanges.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React from "react";
import { IconButton } from "@webiny/ui/Button";
import { Tooltip } from "@webiny/ui/Tooltip";
import { useConfirmationDialog } from "@webiny/app-admin/hooks/useConfirmationDialog";
import { i18n } from "@webiny/app/i18n";
import { ReactComponent as RequestChangesIcon } from "@svgr/webpack!./rule-24px.svg";
import usePermission from "../../../../../hooks/usePermission";
import { useRevision } from "../../useRevision";
import { useContentEntry } from "../../../hooks/useContentEntry";
var t = i18n.ns("app-headless-cms/admin/content-details/header/request-review");

var RequestChanges = function RequestChanges() {
  var _useContentEntry = useContentEntry(),
      entry = _useContentEntry.entry;

  var _useRevision = useRevision({
    revision: entry
  }),
      requestChanges = _useRevision.requestChanges;

  var _usePermission = usePermission(),
      canRequestChange = _usePermission.canRequestChange;

  var _useConfirmationDialo = useConfirmationDialog({
    title: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Request Changes"]))),
    message: /*#__PURE__*/React.createElement("p", null, t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["You are about to request changes on this content entry. Are you sure you want to continue?"]))))
  }),
      showConfirmation = _useConfirmationDialo.showConfirmation;

  if (!canRequestChange("cms.contentEntry")) {
    return null;
  }

  var buttonEnabled = entry.id && (entry.meta.status === "draft" || entry.meta.status === "reviewRequested");
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Tooltip, {
    content: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Request Changes"]))),
    placement: "top"
  }, /*#__PURE__*/React.createElement(IconButton, {
    disabled: !buttonEnabled,
    icon: /*#__PURE__*/React.createElement(RequestChangesIcon, null),
    onClick: function onClick() {
      return showConfirmation( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return requestChanges(entry.id);

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      })));
    }
  })));
};

export default RequestChanges;
