// @webiny/api-headless-cms/plugins/graphql/system.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _handlerGraphql = require("@webiny/handler-graphql");

var _default = {
  typeDefs:
  /* GraphQL */
  `
        extend type CmsQuery {
            # Get installed version
            version: String
        }

        extend type CmsMutation {
            # Install CMS
            install: CmsBooleanResponse

            # Upgrade CMS
            upgrade(version: String!): CmsBooleanResponse
        }
    `,
  resolvers: {
    CmsQuery: {
      version: async (_, __, context) => {
        try {
          return context.cms.system.getVersion();
        } catch (e) {
          return new _handlerGraphql.ErrorResponse(e);
        }
      }
    },
    CmsMutation: {
      install: async (_, __, {
        cms
      }) => {
        try {
          const version = await cms.system.getVersion();

          if (version) {
            return new _handlerGraphql.ErrorResponse({
              code: "CMS_INSTALLATION_ERROR",
              message: "CMS is already installed."
            });
          }

          await cms.system.install();
          return new _handlerGraphql.Response(true);
        } catch (e) {
          return new _handlerGraphql.ErrorResponse(e);
        }
      },
      upgrade: async (_, {
        version
      }, {
        cms
      }) => {
        try {
          await cms.system.upgrade(version);
          return new _handlerGraphql.Response(true);
        } catch (e) {
          return new _handlerGraphql.ErrorResponse(e);
        }
      }
    }
  }
};
exports.default = _default;
