// @webiny/app-page-builder/editor/plugins/toolbar/navigator/index.js
import React from "react";
import { css } from "emotion";
import Action from "../Action";
import Navigator from "./Navigator";
import { ReactComponent as NavigatorIcon } from "@svgr/webpack!../../../assets/icons/segment_black_24px.svg";
var drawerClassName = /*#__PURE__*/css({
  "&.mdc-drawer--dismissible": {
    width: "280px !important",
    maxWidth: "280px !important"
  },
  "&.mdc-drawer": {
    height: "calc(100% - 64px)"
  }
}, "label:drawerClassName;");
export default (function () {
  return {
    name: "pb-editor-toolbar-navigator",
    type: "pb-editor-toolbar-top",
    renderAction: function renderAction() {
      return /*#__PURE__*/React.createElement(Action, {
        tooltip: "Navigator",
        plugin: this.name,
        icon: /*#__PURE__*/React.createElement(NavigatorIcon, null),
        closeOtherInGroup: true
      });
    },
    renderDrawer: function renderDrawer() {
      return /*#__PURE__*/React.createElement(Navigator, null);
    },
    toolbar: {
      drawerClassName: drawerClassName
    }
  };
});
