// @webiny/app-page-builder/admin/plugins/permissionRenderer/PageBuilderPermissions/CustomSection.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10;

import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Select } from "@webiny/ui/Select";
import { i18n } from "@webiny/app/i18n";
import { Elevation } from "@webiny/ui/Elevation";
import { Typography } from "@webiny/ui/Typography";
var t = i18n.ns("app-page-builder/admin/plugins/permission-renderer");

var CustomSection = function CustomSection(_ref) {
  var Bind = _ref.Bind,
      data = _ref.data,
      entity = _ref.entity,
      setValue = _ref.setValue,
      title = _ref.title,
      _ref$children = _ref.children,
      children = _ref$children === void 0 ? null : _ref$children;
  var rwdSelectEnabled = ["full"].includes(data["".concat(entity, "AccessScope")]);
  return /*#__PURE__*/React.createElement(Elevation, {
    z: 1,
    style: {
      marginTop: 10
    }
  }, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "overline"
  }, title)), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Grid, {
    style: {
      padding: 0,
      paddingBottom: 24
    }
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "".concat(entity, "AccessScope"),
    beforeChange: function beforeChange(value, cb) {
      if (value === "own") {
        setValue("".concat(entity, "RWD"), "rwd");
      }

      cb(value);
    }
  }, /*#__PURE__*/React.createElement(Select, {
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Access Scope"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["The scope of the content that can be accessed."])))
  }, /*#__PURE__*/React.createElement("option", {
    value: "no"
  }, t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["None"])))), /*#__PURE__*/React.createElement("option", {
    value: "full"
  }, t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["All content"])))), /*#__PURE__*/React.createElement("option", {
    value: "own"
  }, t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Only content created by the user"]))))))), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "".concat(entity, "RWD")
  }, /*#__PURE__*/React.createElement(Select, {
    disabled: !rwdSelectEnabled,
    label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Primary actions"]))),
    description: t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Primary actions that can be performed on the content."])))
  }, /*#__PURE__*/React.createElement("option", {
    value: "r"
  }, t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["Read"])))), /*#__PURE__*/React.createElement("option", {
    value: "rw"
  }, t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Read, write"])))), /*#__PURE__*/React.createElement("option", {
    value: "rwd"
  }, t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["Read, write, delete"]))))))), children))));
};

export default CustomSection;
