// @webiny/app-page-builder/admin/plugins/permissionRenderer/PageBuilderPermissions/index.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import { AccordionItem } from "@webiny/ui/Accordion";
import { ReactComponent as Icon } from "@svgr/webpack!../../../assets/table_chart-24px.svg";
import { PageBuilderPermissions } from "./PageBuilderPermissions";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-page-builder/admin/plugins/permissionRenderer");
export default {
  type: "admin-app-permissions-renderer",
  name: "admin-app-permissions-renderer-page-builder",
  render: function render(props) {
    return /*#__PURE__*/React.createElement(AccordionItem, {
      icon: /*#__PURE__*/React.createElement(Icon, null),
      title: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Page Builder"]))),
      description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Manage Page Builder app access permissions."]))),
      "data-testid": "permission.pb"
    }, /*#__PURE__*/React.createElement(PageBuilderPermissions, props));
  }
};
