// @webiny/handler-graphql/responses.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListResponse = exports.Response = exports.ListErrorResponse = exports.NotFoundResponse = exports.ErrorResponse = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

const defaultParams = {
  code: "",
  message: "",
  data: null
};

class ErrorResponse {
  constructor(params) {
    (0, _defineProperty2.default)(this, "data", void 0);
    (0, _defineProperty2.default)(this, "error", void 0);
    this.data = null;
    this.error = {
      code: params.code || defaultParams.code,
      message: params.message || defaultParams.message,
      data: params.data || defaultParams.data
    };
  }

}

exports.ErrorResponse = ErrorResponse;

class NotFoundResponse extends ErrorResponse {
  constructor(message) {
    super({
      code: "NOT_FOUND",
      message
    });
  }

}

exports.NotFoundResponse = NotFoundResponse;

class ListErrorResponse {
  constructor(params) {
    (0, _defineProperty2.default)(this, "data", void 0);
    (0, _defineProperty2.default)(this, "meta", void 0);
    (0, _defineProperty2.default)(this, "error", void 0);
    this.data = null;
    this.meta = null;
    this.error = {
      code: params.code || defaultParams.code,
      message: params.message || defaultParams.message,
      data: params.data || defaultParams.data
    };
  }

}

exports.ListErrorResponse = ListErrorResponse;

class Response {
  constructor(data) {
    (0, _defineProperty2.default)(this, "data", void 0);
    (0, _defineProperty2.default)(this, "error", void 0);
    this.data = data;
    this.error = null;
  }

}

exports.Response = Response;

class ListResponse {
  constructor(data, meta) {
    (0, _defineProperty2.default)(this, "data", void 0);
    (0, _defineProperty2.default)(this, "meta", void 0);
    (0, _defineProperty2.default)(this, "error", void 0);
    this.data = Array.isArray(data) ? data : [];
    this.meta = meta || {};
    this.error = null;
  }

}

exports.ListResponse = ListResponse;
