// @webiny/app-page-builder/editor/recoil/actions/afterDropElement/action.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import { plugins } from "@webiny/plugins";
var elementPluginType = "pb-editor-page-element";

var getElementTypePlugin = function getElementTypePlugin(type) {
  var pluginsByType = plugins.byType(elementPluginType);
  var plugin = pluginsByType.find(function (pl) {
    return pl.elementType === type;
  });

  if (!plugin) {
    throw new Error("There is no plugin in \"".concat(elementPluginType, "\" for element type ").concat(type));
  }

  return plugin;
};

export var afterDropElementAction = function afterDropElementAction(state, meta, args) {
  var element = args.element;
  var plugin = getElementTypePlugin(element.type);

  if (plugin.onCreate && plugin.onCreate === "open-settings") {
    return {};
  }

  if (plugin.onCreate && plugin.onCreate === "skipElementHighlight") {
    return {};
  }

  return {
    state: _objectSpread(_objectSpread({}, state), {}, {
      activeElement: element.id,
      sidebar: {
        activeTabIndex: 0,
        highlightTab: true
      }
    })
  };
};
