// @webiny/app-page-builder/editor/plugins/elements/media/vimeo/index.js
import React from "react";
import styled from "@emotion/styled";
import kebabCase from "lodash/kebabCase";
import { Typography } from "@webiny/ui/Typography";
import { validation } from "@webiny/validation";
import { createEmbedPlugin, createEmbedSettingsPlugin } from "../../utils/oembed";
import VimeoEmbed from "./VimeoEmbed";
import { ReactComponent as LogoIcon } from "@svgr/webpack!./vimeo-v-brands.svg";
import Accordion from "../../../elementSettings/components/Accordion";
import InputField from "../../../elementSettings/components/InputField";
import { ButtonContainer, SimpleButton } from "../../../elementSettings/components/StyledComponents";
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "e10gswqi0",
  label: "PreviewBox"
})({
  textAlign: "center",
  height: 50,
  svg: {
    height: 50,
    width: 50
  }
});
export default (function () {
  var args = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var elementType = kebabCase(args.elementType || "vimeo");
  var defaultToolbar = {
    title: "Vimeo",
    group: "pb-editor-element-group-media",
    preview: function preview() {
      return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(LogoIcon, null));
    }
  };
  return [createEmbedPlugin({
    type: elementType,
    toolbar: typeof args.toolbar === "function" ? args.toolbar(defaultToolbar) : defaultToolbar,
    create: args.create,
    settings: args.settings,
    oembed: {
      renderEmbed: function renderEmbed(props) {
        return /*#__PURE__*/React.createElement(VimeoEmbed, props);
      }
    },
    renderElementPreview: function renderElementPreview() {
      return /*#__PURE__*/React.createElement("div", null, "Vimeo video");
    }
  }), createEmbedSettingsPlugin({
    type: elementType,
    render: function render(_ref) {
      var Bind = _ref.Bind,
          submit = _ref.submit;
      return /*#__PURE__*/React.createElement(Accordion, {
        title: "Vimeo",
        defaultValue: true
      }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Bind, {
        name: "source.url",
        validators: validation.create("required,url")
      }, /*#__PURE__*/React.createElement(InputField, {
        placeholder: "https://vimeo.com/158050352",
        description: "Enter a video URL"
      })), /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
        onClick: submit
      }, /*#__PURE__*/React.createElement(Typography, {
        use: "caption"
      }, "Save")))));
    }
  })];
});
