// @webiny/api-headless-cms/plugins/graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _merge = _interopRequireDefault(require("lodash/merge"));

var _system = _interopRequireDefault(require("./graphql/system"));

const emptyResolver = () => ({});

var _default = () => [{
  name: "graphql-schema-headless",
  type: "graphql-schema",
  schema: {
    typeDefs:
    /* GraphQL */
    `
                type CmsError {
                    code: String
                    message: String
                    data: JSON
                }

                type CmsCursors {
                    next: String
                    previous: String
                }

                type CmsListMeta {
                    cursor: String
                    hasMoreItems: Boolean
                    totalCount: Int
                }

                type CmsDeleteResponse {
                    data: Boolean
                    error: CmsError
                }

                type CmsBooleanResponse {
                    data: Boolean
                    error: CmsError
                }

                extend type Query {
                    cms: CmsQuery
                }

                extend type Mutation {
                    cms: CmsMutation
                }

                type CmsQuery {
                    _empty: String
                }

                type CmsMutation {
                    _empty: String
                }

                ${_system.default.typeDefs}
            `,
    resolvers: (0, _merge.default)({
      Query: {
        cms: emptyResolver
      },
      Mutation: {
        cms: emptyResolver
      }
    }, _system.default.resolvers)
  }
}];

exports.default = _default;
