// @webiny/app-file-manager-s3/index.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import gql from "graphql-tag";
var GET_PRE_SIGNED_POST_PAYLOAD = gql(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    query getPreSignedPostPayload($data: PreSignedPostPayloadInput!) {\n        fileManager {\n            getPreSignedPostPayload(data: $data) {\n                data {\n                    data\n                    file {\n                        type\n                        name\n                        size\n                        key\n                    }\n                }\n                error {\n                    message\n                }\n            }\n        }\n    }\n"])));
export default (function () {
  return {
    type: "app-file-manager-storage",
    name: "app-file-manager-storage",
    upload: function () {
      var _upload = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(file, _ref) {
        var apolloClient, response, getPreSignedPostPayload;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                apolloClient = _ref.apolloClient;
                _context.next = 3;
                return apolloClient.query({
                  query: GET_PRE_SIGNED_POST_PAYLOAD,
                  fetchPolicy: "no-cache",
                  variables: {
                    data: {
                      size: file.size,
                      name: file.name,
                      type: file.type
                    }
                  }
                });

              case 3:
                response = _context.sent;
                getPreSignedPostPayload = response.data.fileManager.getPreSignedPostPayload;

                if (!getPreSignedPostPayload.error) {
                  _context.next = 8;
                  break;
                }

                console.log(getPreSignedPostPayload); // eslint-disable-line

                return _context.abrupt("return");

              case 8:
                _context.next = 10;
                return new Promise(function (resolve, reject) {
                  var formData = new window.FormData();
                  Object.keys(getPreSignedPostPayload.data.data.fields).forEach(function (key) {
                    formData.append(key, getPreSignedPostPayload.data.data.fields[key]);
                  });
                  formData.append("file", file);
                  var xhr = new window.XMLHttpRequest(); // eslint-disable-line

                  // eslint-disable-line
                  xhr.open("POST", getPreSignedPostPayload.data.data.url, true);
                  xhr.send(formData);

                  xhr.onload = function () {
                    if (this.status === 204) {
                      resolve(getPreSignedPostPayload.data.file);
                      return;
                    }

                    reject(this.responseText);
                  };
                });

              case 10:
                return _context.abrupt("return", _context.sent);

              case 11:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function upload(_x, _x2) {
        return _upload.apply(this, arguments);
      }

      return upload;
    }()
  };
});
