// @webiny/app/plugins/NetworkErrorLinkPlugin/ErrorOverlay.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

import React, { useState } from "react";
import get from "lodash/get";
import { IconButton } from "@webiny/ui/Button";
import { Typography } from "@webiny/ui/Typography";
import { i18n } from "../../i18n";
import { OverlayWrapper, Pre } from "./StyledComponents";
import { ReactComponent as CloseIcon } from "@svgr/webpack!./assets/close_24px.svg";
var t = i18n.ns("app/graphql/error-overlay");
var ENVIRONMENT_VARIABLES_ARTICLE_LINK = "https://www.webiny.com/docs/how-to-guides/environment-variables";

var ErrorOverlay = function ErrorOverlay(_ref) {
  var query = _ref.query,
      networkError = _ref.networkError;

  var _useState = useState(true),
      _useState2 = _slicedToArray(_useState, 2),
      open = _useState2[0],
      setOpen = _useState2[1]; // Log error in browser's developer console for further inspection.


  console.error({
    networkError: networkError
  });

  if (!open) {
    return null;
  }

  var stackTrace = get(networkError, "result.error.stack");
  return /*#__PURE__*/React.createElement(OverlayWrapper, null, /*#__PURE__*/React.createElement("div", {
    className: "inner"
  }, /*#__PURE__*/React.createElement("div", {
    className: "header"
  }, /*#__PURE__*/React.createElement("div", {
    className: "header__title"
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "headline4"
  }, networkError.message)), /*#__PURE__*/React.createElement("div", {
    className: "header__action"
  }, /*#__PURE__*/React.createElement(IconButton, {
    icon: /*#__PURE__*/React.createElement(CloseIcon, null),
    onClick: function onClick() {
      return setOpen(false);
    }
  }))), /*#__PURE__*/React.createElement("div", {
    className: "body"
  }, /*#__PURE__*/React.createElement("div", {
    className: "body__summary"
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "subtitle1"
  }, t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Error occurred while executing operation:"])))), /*#__PURE__*/React.createElement(Pre, null, /*#__PURE__*/React.createElement("code", null, query))), stackTrace && /*#__PURE__*/React.createElement("div", {
    className: "body__description"
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "subtitle1"
  }, t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Complete stack trace:"])))), /*#__PURE__*/React.createElement(Pre, null, /*#__PURE__*/React.createElement("code", null, stackTrace)))), /*#__PURE__*/React.createElement("div", {
    className: "footer"
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "body2"
  }, t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n                        This screen is only visible if {message} environment variable is set.\n                        Read more about it in the {link}. "])))({
    message: /*#__PURE__*/React.createElement("span", {
      className: "highlight"
    }, "`REACT_APP_DEBUG=true`"),
    link: /*#__PURE__*/React.createElement("a", {
      href: ENVIRONMENT_VARIABLES_ARTICLE_LINK,
      target: "_blank",
      rel: "noreferrer noopener"
    }, "environment variables article")
  }), /*#__PURE__*/React.createElement("br", null), t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Open your browser's developer console to further inspect this error."])))))));
};

export default ErrorOverlay;
