// @webiny/app-headless-cms/admin/plugins/fieldValidators/dateLte.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { validation } from "@webiny/validation";
import { createInputField } from "./date/createDateInputField";
import { FormElementMessage } from "@webiny/ui/FormElementMessage";
import { getAvailableValidators } from "./date/availableValidators";
export default (function () {
  return {
    type: "cms-editor-field-validator",
    name: "cms-editor-field-validator-date-lte",
    validator: {
      name: "dateLte",
      label: "Earlier or equal",
      description: "Entered date/time must be equal or earlier compared to the provided date.",
      defaultMessage: "Date/time is later than the provided one.",
      renderSettings: function renderSettings(_ref) {
        var Bind = _ref.Bind,
            field = _ref.field;
        var type = field.settings.type;
        var availableValidators = getAvailableValidators(type).join(",");
        return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
          span: 12
        }, /*#__PURE__*/React.createElement(Bind, {
          name: "settings.type"
        }, function (bind) {
          if (bind.value !== type) {
            bind.onChange(type);
          }

          return /*#__PURE__*/React.createElement(React.Fragment, null);
        }), /*#__PURE__*/React.createElement(Bind, {
          name: "settings.value",
          validators: validation.create(availableValidators)
        }, function (bind) {
          return /*#__PURE__*/React.createElement(React.Fragment, null, createInputField(field, bind), /*#__PURE__*/React.createElement(FormElementMessage, null, "This is the latest date/time that will be allowed."));
        })));
      }
    }
  };
});
