// @webiny/app-headless-cms/admin/plugins/fieldValidators/pattern.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Input } from "@webiny/ui/Input";
import { Select } from "@webiny/ui/Select";
import { plugins } from "@webiny/plugins";
import { validation } from "@webiny/validation";
export default {
  type: "cms-editor-field-validator",
  name: "cms-editor-field-validator-pattern",
  validator: {
    name: "pattern",
    label: "Pattern",
    description: "Entered value must match a specific pattern.",
    defaultMessage: "Invalid value.",
    defaultSettings: {
      preset: "custom"
    },
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.Bind,
          setValue = _ref.setValue,
          setMessage = _ref.setMessage,
          data = _ref.data;
      var inputsDisabled = data.settings.preset !== "custom";
      var presetPlugins = plugins.byType("cms-editor-field-validator-pattern");
      var selectOptions = presetPlugins.map(function (item) {
        return /*#__PURE__*/React.createElement("option", {
          key: item.pattern.name,
          value: item.pattern.name
        }, item.pattern.label);
      });
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 3
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "settings.preset",
        validators: validation.create("required"),
        afterChange: function afterChange(value) {
          if (value === "custom") {
            setMessage("Invalid value.");
            return;
          }

          setValue("settings.regex", null);
          setValue("settings.flags", null);
          var selectedPatternPlugin = presetPlugins.find(function (item) {
            return item.pattern.name === value;
          });
          setMessage(selectedPatternPlugin.pattern.message);
        }
      }, /*#__PURE__*/React.createElement(Select, {
        label: "Preset"
      }, /*#__PURE__*/React.createElement("option", {
        value: "custom"
      }, "Custom"), selectOptions))), /*#__PURE__*/React.createElement(Cell, {
        span: 7
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "settings.regex",
        validators: validation.create("required")
      }, /*#__PURE__*/React.createElement(Input, {
        disabled: inputsDisabled,
        label: "Regex",
        description: "Regex to test the value"
      }))), /*#__PURE__*/React.createElement(Cell, {
        span: 2
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "settings.flags",
        validators: validation.create("required")
      }, /*#__PURE__*/React.createElement(Input, {
        disabled: inputsDisabled,
        label: "Flags",
        description: "Regex flags"
      }))));
    }
  }
};
