// @webiny/app-headless-cms/admin/views/contentModelGroups/ContentModelGroupsDataList.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useCallback, useMemo, useState } from "react";
import get from "lodash/get";
import orderBy from "lodash/orderBy";
import dotProp from "dot-prop-immutable";
import { i18n } from "@webiny/app/i18n";
import { DeleteIcon } from "@webiny/ui/List/DataList/icons";
import { DataList, List, ListItem, ListItemText, ListItemTextSecondary, ListItemMeta, ListActions, DataListModalOverlay, DataListModalOverlayAction } from "@webiny/ui/List";
import { useRouter } from "@webiny/react-router";
import { useSnackbar } from "@webiny/app-admin/hooks/useSnackbar";
import { useQuery, useApolloClient } from "../../hooks";
import { useConfirmationDialog } from "@webiny/app-admin/hooks/useConfirmationDialog";
import * as GQL from "./graphql";
import { ButtonIcon, ButtonSecondary } from "@webiny/ui/Button";
import { Cell, Grid } from "@webiny/ui/Grid";
import { Select } from "@webiny/ui/Select";
import SearchUI from "@webiny/app-admin/components/SearchUI";
import { ReactComponent as AddIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/add-18px.svg";
import { ReactComponent as FilterIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/filter-24px.svg";
import { serializeSorters, deserializeSorters } from "../utils";
import usePermission from "../../hooks/usePermission";
import { Tooltip } from "@webiny/ui/Tooltip";
var t = i18n.ns("app-headless-cms/admin/content-model-groups/data-list");
var SORTERS = [{
  label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Newest to oldest"]))),
  sorters: {
    createdOn: "desc"
  }
}, {
  label: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Oldest to newest"]))),
  sorters: {
    createdOn: "asc"
  }
}, {
  label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Name A-Z"]))),
  sorters: {
    name: "asc"
  }
}, {
  label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Name Z-A"]))),
  sorters: {
    name: "desc"
  }
}];

var ContentModelGroupsDataList = function ContentModelGroupsDataList(_ref) {
  var canCreate = _ref.canCreate;

  var _useState = useState(""),
      _useState2 = _slicedToArray(_useState, 2),
      filter = _useState2[0],
      setFilter = _useState2[1];

  var _useState3 = useState(serializeSorters(SORTERS[0].sorters)),
      _useState4 = _slicedToArray(_useState3, 2),
      sort = _useState4[0],
      setSort = _useState4[1];

  var _useRouter = useRouter(),
      history = _useRouter.history;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var client = useApolloClient();
  var listQuery = useQuery(GQL.LIST_CONTENT_MODEL_GROUPS);

  var _useConfirmationDialo = useConfirmationDialog({
    dataTestId: "cms.contentModelGroup.list-item.delete-dialog"
  }),
      showConfirmation = _useConfirmationDialo.showConfirmation;

  var _usePermission = usePermission(),
      canDelete = _usePermission.canDelete;

  var filterData = useCallback(function (_ref2) {
    var name = _ref2.name;
    return name.toLowerCase().includes(filter);
  }, [filter]);
  var sortData = useCallback(function (list) {
    if (!sort) {
      return list;
    }

    var _Object$entries = Object.entries(deserializeSorters(sort)),
        _Object$entries2 = _slicedToArray(_Object$entries, 1),
        _Object$entries2$ = _slicedToArray(_Object$entries2[0], 2),
        key = _Object$entries2$[0],
        value = _Object$entries2$[1];

    return orderBy(list, [key], [value]);
  }, [sort]);
  var data = listQuery.loading ? [] : get(listQuery, "data.listContentModelGroups.data", []);
  var groupId = new URLSearchParams(location.search).get("id");
  var deleteItem = useCallback(function (group) {
    showConfirmation( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var response, error;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return client.mutate({
                mutation: GQL.DELETE_CONTENT_MODEL_GROUP,
                variables: {
                  id: group.id
                },
                update: function update(cache, _ref4) {
                  var data = _ref4.data;
                  var error = data.deleteContentModelGroup.error;

                  if (error) {
                    return;
                  } // Delete the item from list cache


                  // Delete the item from list cache
                  var gqlParams = {
                    query: GQL.LIST_CONTENT_MODEL_GROUPS
                  };

                  var _cache$readQuery = cache.readQuery(gqlParams),
                      listContentModelGroups = _cache$readQuery.listContentModelGroups;

                  var index = listContentModelGroups.data.findIndex(function (item) {
                    return item.id === group.id;
                  });
                  cache.writeQuery(_objectSpread(_objectSpread({}, gqlParams), {}, {
                    data: {
                      listContentModelGroups: dotProp.delete(listContentModelGroups, "data.".concat(index))
                    }
                  }));
                }
              });

            case 2:
              response = _context.sent;
              error = response.data.deleteContentModelGroup.error;

              if (!error) {
                _context.next = 6;
                break;
              }

              return _context.abrupt("return", showSnackbar(error.message));

            case 6:
              showSnackbar(t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Content model group \"{name}\" deleted."])))({
                name: group.name
              }));

              if (groupId === group.id) {
                history.push("/cms/content-model-groups");
              }

            case 8:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    })));
  }, [groupId]);
  var contentModelGroupsDataListModalOverlay = useMemo(function () {
    return /*#__PURE__*/React.createElement(DataListModalOverlay, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Select, {
      value: sort,
      onChange: setSort,
      label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Sort by"]))),
      description: "Sort pages by"
    }, SORTERS.map(function (_ref5) {
      var label = _ref5.label,
          sorters = _ref5.sorters;
      return /*#__PURE__*/React.createElement("option", {
        key: label,
        value: serializeSorters(sorters)
      }, label);
    })))));
  }, [sort]);
  var filteredData = filter === "" ? data : data.filter(filterData);
  var contentModelGroups = sortData(filteredData);
  return /*#__PURE__*/React.createElement(DataList, {
    loading: listQuery.loading,
    data: contentModelGroups,
    title: t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Content Model Groups"]))),
    actions: canCreate ? /*#__PURE__*/React.createElement(ButtonSecondary, {
      "data-testid": "new-record-button",
      onClick: function onClick() {
        return history.push("/cms/content-model-groups?new=true");
      }
    }, /*#__PURE__*/React.createElement(ButtonIcon, {
      icon: /*#__PURE__*/React.createElement(AddIcon, null)
    }), " ", t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["New Group"])))) : null,
    search: /*#__PURE__*/React.createElement(SearchUI, {
      value: filter,
      onChange: setFilter,
      inputPlaceholder: t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Search content model group"])))
    }),
    modalOverlay: contentModelGroupsDataListModalOverlay,
    modalOverlayAction: /*#__PURE__*/React.createElement(DataListModalOverlayAction, {
      icon: /*#__PURE__*/React.createElement(FilterIcon, null),
      "data-testid": "default-data-list.filter"
    })
  }, function (_ref6) {
    var data = _ref6.data;
    return /*#__PURE__*/React.createElement(List, {
      "data-testid": "default-data-list"
    }, data.map(function (item) {
      return /*#__PURE__*/React.createElement(ListItem, {
        key: item.id,
        selected: item.id === groupId
      }, /*#__PURE__*/React.createElement(ListItemText, {
        onClick: function onClick() {
          return history.push("/cms/content-model-groups?id=".concat(item.id));
        }
      }, item.name, /*#__PURE__*/React.createElement(ListItemTextSecondary, null, item.contentModels.length ? t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["{contentModels|count:1:content model:default:content models}"])))({
        contentModels: item.contentModels.length
      }) : t(_templateObject11 || (_templateObject11 = _taggedTemplateLiteral(["No content models"]))))), canDelete(item, "cms.contentModelGroup") && /*#__PURE__*/React.createElement(ListItemMeta, null, /*#__PURE__*/React.createElement(ListActions, null, item.plugin ? /*#__PURE__*/React.createElement(Tooltip, {
        content: "Content model group is registered via a plugin.",
        placement: "bottom"
      }, /*#__PURE__*/React.createElement(DeleteIcon, {
        disabled: true,
        "data-testid": "cms.contentModelGroup.list-item.delete"
      })) : /*#__PURE__*/React.createElement(DeleteIcon, {
        onClick: function onClick() {
          return deleteItem(item);
        },
        "data-testid": "cms.contentModelGroup.list-item.delete"
      }))));
    }));
  });
};

export default ContentModelGroupsDataList;
