// @webiny/app-page-builder/editor/components/DropZone/Center.js
import { activeElementAtom } from "../../recoil/modules";
import React from "react";
import Droppable from "./../Droppable";
import styled from "@emotion/styled";
import { useRecoilValue } from "recoil";
var Container = /*#__PURE__*/React.memo( /*#__PURE__*/styled("div", {
  target: "e1meegfv0",
  label: "Container"
})(function (_ref) {
  var isOver = _ref.isOver;
  return {
    backgroundColor: "transparent",
    boxSizing: "border-box",
    height: "100%",
    minHeight: 100,
    position: "relative",
    userSelect: "none",
    width: "100%",
    border: isOver ? "1px dashed var(--mdc-theme-primary)" : "1px dashed var(--mdc-theme-secondary)",
    ".addIcon": {
      color: isOver ? "var(--mdc-theme-primary) !important" : "var(--mdc-theme-secondary) !important"
    }
  };
}));
var Add = /*#__PURE__*/styled("div", {
  target: "e1meegfv1",
  label: "Add"
})({
  position: "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%,-50%)",
  margin: 0
});

var isVisible = function isVisible() {
  return true;
};

var Center = function Center(_ref2) {
  var id = _ref2.id,
      type = _ref2.type,
      onDrop = _ref2.onDrop,
      children = _ref2.children,
      isHighlighted = _ref2.isHighlighted;
  var activeElementId = useRecoilValue(activeElementAtom);
  var isActive = activeElementId === id;
  return /*#__PURE__*/React.createElement(Droppable, {
    onDrop: onDrop,
    type: type,
    isVisible: isVisible
  }, function (_ref3) {
    var isOver = _ref3.isOver,
        isDroppable = _ref3.isDroppable,
        drop = _ref3.drop;
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: "100%",
        height: "100%"
      },
      ref: drop
    }, /*#__PURE__*/React.createElement(Container, {
      isOver: isOver && isDroppable || isActive,
      highlight: isHighlighted
    }, /*#__PURE__*/React.createElement(Add, null, children)));
  });
};

export default /*#__PURE__*/React.memo(Center);
