// @webiny/app-page-builder/editor/components/DropZone/Horizontal.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import Droppable from "../Droppable";
import styled from "@emotion/styled";
var InnerDiv = /*#__PURE__*/styled("div", {
  target: "e1u8zmh70",
  label: "InnerDiv"
})({
  height: 5,
  width: "100%",
  //"calc(100% - 50px)",
  zIndex: 3,
  borderRadius: 5,
  boxSizing: "border-box",
  display: "none"
});
var OuterDiv = /*#__PURE__*/React.memo( /*#__PURE__*/styled("div", {
  target: "e1u8zmh71",
  label: "OuterDiv"
})({
  margin: 0,
  padding: 0,
  width: "100%",
  zIndex: 10,
  backgroundColor: "transparent",
  position: "absolute",
  display: "flex",
  justifyContent: "center"
}, function (props) {
  var _ref;

  return _ref = {}, _defineProperty(_ref, props.below ? "bottom" : "top", 0), _defineProperty(_ref, InnerDiv, {
    backgroundColor: props.isOver ? "var(--mdc-theme-primary)" : "var(--mdc-theme-secondary)",
    display: props.isOver ? "block" : "none"
  }), _ref;
}));

var HorizontalComponent = function HorizontalComponent(_ref2) {
  var below = _ref2.below,
      onDrop = _ref2.onDrop,
      isVisible = _ref2.isVisible,
      type = _ref2.type;
  return /*#__PURE__*/React.createElement(Droppable, {
    type: type,
    isVisible: isVisible,
    onDrop: onDrop
  }, function (_ref3) {
    var _ref4;

    var isOver = _ref3.isOver,
        drop = _ref3.drop;
    return /*#__PURE__*/React.createElement("div", {
      ref: drop,
      style: (_ref4 = {
        height: "25px",
        width: "100%",
        position: "absolute"
      }, _defineProperty(_ref4, below ? "bottom" : "top", 0), _defineProperty(_ref4, "left", 0), _defineProperty(_ref4, "zIndex", 1000), _ref4)
    }, /*#__PURE__*/React.createElement(OuterDiv, {
      isOver: isOver,
      below: below
    }, /*#__PURE__*/React.createElement(InnerDiv, null)));
  });
};

export var Horizontal = /*#__PURE__*/React.memo(HorizontalComponent);
