// @webiny/app-page-builder/editor/components/DropZone/Vertical.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import React from "react";
import Droppable from "./../Droppable";
import styled from "@emotion/styled";
var InnerDivVertical = /*#__PURE__*/styled("div", {
  target: "ebloigh0",
  label: "InnerDivVertical"
})({
  position: "absolute",
  width: 5,
  height: "100%",
  zIndex: 3,
  borderRadius: 5,
  display: "none"
});
var OuterDivVertical = /*#__PURE__*/React.memo( /*#__PURE__*/styled("div", {
  target: "ebloigh1",
  label: "OuterDivVertical"
})({
  position: "absolute",
  width: "30%",
  top: 0,
  height: "100%",
  zIndex: 10,
  backgroundColor: "transparent"
}, function (props) {
  var _InnerDivVertical, _ref;

  return _ref = {}, _defineProperty(_ref, props.last ? "right" : "left", -2), _defineProperty(_ref, "textAlign", props.last ? "right" : "left"), _defineProperty(_ref, InnerDivVertical, (_InnerDivVertical = {
    backgroundColor: props.isOver ? "var(--mdc-theme-primary)" : "var(--mdc-theme-secondary)"
  }, _defineProperty(_InnerDivVertical, props.last ? "right" : "left", -2), _defineProperty(_InnerDivVertical, "display", props.isOver ? "block" : "none"), _InnerDivVertical)), _ref;
}));

var VerticalComponent = function VerticalComponent(_ref2) {
  var last = _ref2.last,
      onDrop = _ref2.onDrop,
      isVisible = _ref2.isVisible,
      type = _ref2.type;
  return /*#__PURE__*/React.createElement(Droppable, {
    type: type,
    isVisible: isVisible,
    onDrop: onDrop
  }, function (_ref3) {
    var _ref4;

    var isOver = _ref3.isOver,
        drop = _ref3.drop;
    return /*#__PURE__*/React.createElement("div", {
      ref: drop,
      style: (_ref4 = {
        width: "30%",
        maxWidth: "100px",
        height: "100%",
        position: "absolute",
        top: 0
      }, _defineProperty(_ref4, last ? "right" : "left", 0), _defineProperty(_ref4, "zIndex", 1000), _ref4)
    }, /*#__PURE__*/React.createElement(OuterDivVertical, {
      isOver: isOver,
      last: last
    }, /*#__PURE__*/React.createElement(InnerDivVertical, null)));
  });
};

export var Vertical = /*#__PURE__*/React.memo(VerticalComponent);
