// @webiny/app-graphql-playground/plugins/index.js
import React from "react";
import Helmet from "react-helmet";
import { Route } from "@webiny/react-router";
import { AdminLayout } from "@webiny/app-admin/components/AdminLayout";
import { ReactComponent as InfoIcon } from "@svgr/webpack!./graphql.svg";
import Playground from "./Playground";
// @ts-ignore
import placeholder from "!!raw-loader!./placeholder.graphql";
import { NavigationMenuElement } from "@webiny/app-admin/ui/elements/NavigationMenuElement";
import { UIViewPlugin } from "@webiny/app-admin/ui/UIView";
import { NavigationView } from "@webiny/app-admin/ui/views/NavigationView";
// @ts-ignore
export default (function (options) {
  return [new UIViewPlugin(NavigationView, function (view) {
    view.getFooterElement().addMenuElement(new NavigationMenuElement("apiPlayground", {
      label: "API Playground",
      icon: /*#__PURE__*/React.createElement(InfoIcon, null),
      path: "/api-playground",
      onClick: function onClick() {
        view.getNavigationHook().hideMenu();
      }
    }));
  }), {
    name: "route-api-playground",
    type: "route",
    route: /*#__PURE__*/React.createElement(Route, {
      exact: true,
      path: "/api-playground",
      render: function render() {
        return /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, null, /*#__PURE__*/React.createElement("title", null, "API Playground")), /*#__PURE__*/React.createElement(Playground, {
          createApolloClient: options.createApolloClient
        }));
      }
    })
  }, {
    type: "graphql-playground-tab",
    tab: function tab() {
      return {
        name: "Main API",
        endpoint: process.env.REACT_APP_API_URL + "/graphql",
        headers: {},
        query: placeholder
      };
    }
  }];
});
