// @webiny/app-form-builder/admin/plugins/editor/formSettings/index.js
import * as React from "react";
import GeneralSettings from "./components/GeneralSettings";
import TermsOfServiceSettings from "./components/TermsOfServiceSettings";
import ReCaptchaSettings from "./components/ReCaptchaSettings";
import { ReactComponent as SettingsIcon } from "@svgr/webpack!./icons/round-settings-24px.svg";
import { ReactComponent as TermsOfServiceIcon } from "@svgr/webpack!./icons/round-receipt-24px.svg";
import { ReactComponent as ReCaptchaIcon } from "@svgr/webpack!./icons/round-vpn_lock-24px.svg";
export default [{
  name: "form-editor-form-settings-general",
  type: "form-editor-form-settings",
  title: "General settings",
  description: "Manage things like submit success messages and form layout.",
  icon: /*#__PURE__*/React.createElement(SettingsIcon, null),
  render: function render(props) {
    return /*#__PURE__*/React.createElement(GeneralSettings, props);
  }
}, {
  name: "form-editor-form-settings-tos",
  type: "form-editor-form-settings",
  title: "Terms of service",
  description: "Manage terms of service messaging.",
  icon: /*#__PURE__*/React.createElement(TermsOfServiceIcon, null),
  render: function render(props) {
    return /*#__PURE__*/React.createElement(TermsOfServiceSettings, props);
  }
}, {
  name: "form-editor-form-settings-recaptcha",
  type: "form-editor-form-settings",
  title: "ReCAPTCHA",
  description: "Enable reCAPTCHA to prevent spam and abuse.",
  icon: /*#__PURE__*/React.createElement(ReCaptchaIcon, null),
  render: function render(props) {
    return /*#__PURE__*/React.createElement(ReCaptchaSettings, props);
  }
}];
