// @webiny/app-form-builder/page-builder/admin/plugins/index.js
import React from "react";
import styled from "@emotion/styled";
import FormElementAdvancedSettings from "./components/FormElementAdvancedSettings";
import formElement from "../../../page-builder/render/plugins/formElement";
import FormElement from "./components/FormElement";
import { ReactComponent as FormLogo } from "@svgr/webpack!./components/icons/round-description-24px.svg";
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "e1yc4yzg0",
  label: "PreviewBox"
})({
  textAlign: "center",
  height: 80,
  svg: {
    height: 80,
    width: "auto"
  }
});
export default (function () {
  return [formElement, {
    name: "pb-page-element-form",
    type: "pb-editor-page-element",
    elementType: "form",
    toolbar: {
      title: "Form",
      group: "pb-editor-element-group-form",
      preview: function preview() {
        return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(FormLogo, null));
      }
    },
    settings: ["pb-editor-page-element-settings-delete", "pb-editor-page-element-settings-height"],
    target: ["block", "cell"],
    onCreate: "open-settings",
    render: function render(props) {
      return /*#__PURE__*/React.createElement(FormElement, props);
    },
    create: function create() {
      return {
        type: "form",
        elements: [],
        data: {},
        settings: {}
      };
    }
  }, {
    name: "pb-element-advanced-settings-form",
    type: "pb-editor-page-element-advanced-settings",
    elementType: "form",
    render: function render(props) {
      return /*#__PURE__*/React.createElement(FormElementAdvancedSettings, props);
    }
  }];
});
