// @webiny/app-i18n-content/admin/contentPermissions/index.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import { AccordionItem } from "@webiny/ui/Accordion";
import { ReactComponent as Icon } from "@svgr/webpack!./icons/how_to_vote-24px.svg";
import { ContentPermissions } from "./ContentPermissions";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-i18n/admin/plugins/permissionRenderer");
export default {
  type: "admin-app-permissions-renderer",
  name: "admin-app-permissions-renderer-content-locales",
  system: true,
  render: function render(props) {
    return /*#__PURE__*/React.createElement(AccordionItem, {
      open: true,
      icon: /*#__PURE__*/React.createElement(Icon, null),
      title: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Content"]))),
      description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Per-locale content access permissions management."]))),
      "data-testid": "permission.content"
    }, /*#__PURE__*/React.createElement(ContentPermissions, props));
  }
};
