// @webiny/app-headless-cms/admin/plugins/fieldRenderers/radioButtons.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import get from "lodash/get";
import { i18n } from "@webiny/app/i18n";
import { Radio, RadioGroup } from "@webiny/ui/Radio";
var t = i18n.ns("app-headless-cms/admin/fields/text");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-radio-buttons",
  renderer: {
    rendererName: "radio-buttons",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Radio Buttons"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders radio buttons, allowing selection of a single value."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return !field.multipleValues && get(field, "predefinedValues.enabled");
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind;
      var Bind = getBind();
      var options = field.predefinedValues.values;
      return /*#__PURE__*/React.createElement(Bind, null, /*#__PURE__*/React.createElement(RadioGroup, {
        label: field.label,
        description: field.helpText
      }, function (_ref3) {
        var onChange = _ref3.onChange,
            getValue = _ref3.getValue;
        return /*#__PURE__*/React.createElement(React.Fragment, null, options.map(function (option, index) {
          return /*#__PURE__*/React.createElement("div", {
            key: option.value + index
          }, /*#__PURE__*/React.createElement(Radio, {
            label: option.label,
            value: getValue(option.value),
            onChange: onChange(option.value)
          }));
        }));
      }));
    }
  }
};
export default plugin;
