// @webiny/app-headless-cms/admin/plugins/fieldRenderers/select.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import get from "lodash/get";
import { i18n } from "@webiny/app/i18n";
import { Select } from "@webiny/ui/Select";
var t = i18n.ns("app-headless-cms/admin/fields/text");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-select-box",
  renderer: {
    rendererName: "select-box",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Select Box"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders a select box, allowing selection of a single value."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return !field.multipleValues && get(field, "predefinedValues.enabled");
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind;
      var Bind = getBind();
      var options = field.predefinedValues.values;
      return /*#__PURE__*/React.createElement(Bind, null, /*#__PURE__*/React.createElement(Select, {
        label: field.label,
        description: field.helpText,
        options: options
      }));
    }
  }
};
export default plugin;
