// @webiny/api-page-builder/plugins/PagePlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PagePlugin = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _plugins = require("@webiny/plugins");

class PagePlugin extends _plugins.Plugin {
  constructor(config) {
    super();
    (0, _defineProperty2.default)(this, "_config", void 0);
    this._config = config || {};
  }

  beforeCreate(params) {
    return this._execute("beforeCreate", params);
  }

  afterCreate(params) {
    return this._execute("afterCreate", params);
  }

  beforeUpdate(params) {
    return this._execute("beforeUpdate", params);
  }

  afterUpdate(params) {
    return this._execute("afterUpdate", params);
  }

  afterDelete(params) {
    return this._execute("afterDelete", params);
  }

  beforeDelete(params) {
    return this._execute("beforeDelete", params);
  }

  beforePublish(params) {
    return this._execute("beforePublish", params);
  }

  afterPublish(params) {
    return this._execute("afterPublish", params);
  }

  beforeUnpublish(params) {
    return this._execute("beforeUnpublish", params);
  }

  afterUnpublish(params) {
    return this._execute("afterUnpublish", params);
  }

  notFound(params) {
    return this._execute("notFound", params);
  }

  _execute(callback, params) {
    if (typeof this._config[callback] === "function") {
      return this._config[callback](params);
    }
  }

}

exports.PagePlugin = PagePlugin;
(0, _defineProperty2.default)(PagePlugin, "type", "pb.page");
