// @webiny/app-headless-cms/admin/plugins/apiInformation/index.js
var manageQuery = "#  Webiny Headless CMS Manage API\n#\n#  This API controls all the operations within the Headless CMS module.\n#  Use it to manage content, create and modify content models and more.\n#  Explore the schema for details.\n#\n#  Note: to use the API outside of the playground, you will need to provide an API key via the Authorization header.\n#\n#  Example query - list content model groups:\n{\n  listContentModelGroups {\n    data {\n      name\n      icon\n      slug\n    }\n  }\n}\n";
var readQuery = "#  Webiny Headless CMS Read API\n#\n#  This is the API to use in your 3rd party apps to read your content.\n#  It returns only content that has been published.\n#\n#  Note: to use the API outside of the playground, you will need to provide an API key via the Authorization header.\n#\n#  Example query - list content models:\n{\n  listContentModels {\n    data {\n      name\n      modelId\n    }\n  }\n}\n";
var previewQuery = "#  Webiny Headless CMS Preview API\n#\n#  This is the API to use in your 3rd party apps to preview content.\n#  It returns the latest content revision, regardless of whether it was published or not.\n#\n#  Note: to use the API outside of the playground, you will need to provide an API key via the Authorization header.\n#\n#  Example query - list content models:\n{\n  listContentModels {\n    data {\n      name\n    }\n  }\n}\n";
var plugins = [{
  type: "graphql-playground-tab",
  name: "graphql-playground-tab-manage",
  tab: function tab(_ref) {
    var locale = _ref.locale,
        identity = _ref.identity;

    if (!identity.getPermission("cms.endpoint.manage")) {
      return null;
    }

    return {
      name: "Headless CMS - Manage API",
      endpoint: process.env.REACT_APP_API_URL + "/cms/manage/" + locale,
      headers: {},
      query: manageQuery
    };
  }
}, {
  type: "graphql-playground-tab",
  name: "graphql-playground-tab-read",
  tab: function tab(_ref2) {
    var locale = _ref2.locale,
        identity = _ref2.identity;

    if (!identity.getPermission("cms.endpoint.read")) {
      return null;
    }

    return {
      name: "Headless CMS - Read API",
      endpoint: process.env.REACT_APP_API_URL + "/cms/read/" + locale,
      headers: {},
      query: readQuery
    };
  }
}, {
  type: "graphql-playground-tab",
  name: "graphql-playground-tab-preview",
  tab: function tab(_ref3) {
    var locale = _ref3.locale,
        identity = _ref3.identity;

    if (!identity.getPermission("cms.endpoint.preview")) {
      return null;
    }

    return {
      name: "Headless CMS - Preview API",
      endpoint: process.env.REACT_APP_API_URL + "/cms/preview/" + locale,
      headers: {},
      query: previewQuery
    };
  }
}];
export default plugins;
