// @webiny/ui/List/DataList/DataListModalOverlay/DataListModalOverlay.js
import React, { useContext } from "react";
import styled from "@emotion/styled";
import { DataListModalOverlayContext } from "./DataListModalOverlayContext";
export var SimpleOverlay = /*#__PURE__*/styled("div", {
  target: "e1dhql4g0",
  label: "SimpleOverlay"
})(function (_ref) {
  var showOverlay = _ref.showOverlay;
  return {
    width: "100%",
    height: "100%",
    position: "absolute",
    top: 0,
    left: 0,
    backgroundColor: showOverlay ? "rgba(0, 0, 0, 0.32)" : "transparent",
    transition: "150ms all ease-in-out",
    pointerEvents: showOverlay ? "auto" : "none",
    zIndex: showOverlay ? 5 : -1,
    opacity: showOverlay ? 1 : 0
  };
});
export var DataListModalWrapper = /*#__PURE__*/styled("div", {
  target: "e1dhql4g1",
  label: "DataListModalWrapper"
})(function (_ref2) {
  var showOverlay = _ref2.showOverlay;
  return {
    position: "absolute",
    top: "0px",
    left: "50%",
    width: "90%",
    height: "auto",
    maxWidth: "426px",
    maxHeight: "520px",
    overflow: "auto",
    backgroundColor: "var(--mdc-theme-surface)",
    boxShadow: "0px 1px 5px rgba(0, 0, 0, 0.12)",
    borderRadius: "0px 0px 4px 4px",
    transition: "150ms opacity ease-in-out",
    pointerEvents: showOverlay ? "auto" : "none",
    zIndex: showOverlay ? 5 : -1,
    opacity: showOverlay ? 1 : 0,
    transform: showOverlay ? "translateX(-50%)" : "translateX(-1000px)",
    "& .input-wrapper": {
      paddingBottom: "16px"
    },
    "& .datalist-modal__footer": {
      padding: "24px",
      display: "flex",
      justifyContent: "space-between",
      "& .mdc-button": {
        borderRadius: 4,
        padding: "0px 32px"
      },
      "& .datalist-modal__footer__action--secondary": {
        color: "var(--mdc-theme-text-primary-on-background)",
        border: "1px solid var(--mdc-theme-on-background)"
      },
      "& .datalist-modal__footer__action--primary": {
        backgroundColor: "var(--mdc-theme-secondary)"
      }
    }
  };
});
export var DataListModalOverlay = function DataListModalOverlay(_ref3) {
  var onDismiss = _ref3.onDismiss,
      children = _ref3.children;

  var _useContext = useContext(DataListModalOverlayContext),
      isOpen = _useContext.isOpen,
      setIsOpen = _useContext.setIsOpen;

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(SimpleOverlay, {
    onClick: function onClick(e) {
      e.stopPropagation(); // Close the modal.

      setIsOpen(false);

      if (typeof onDismiss === "function") {
        onDismiss();
      }
    },
    showOverlay: isOpen
  }), /*#__PURE__*/React.createElement(DataListModalWrapper, {
    showOverlay: isOpen
  }, children));
};
