// @webiny/app-admin/plugins/userMenu/UserMenuRenderer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { UIRenderer } from "../../ui/UIRenderer";
import { css } from "emotion";
import { Menu } from "@webiny/ui/Menu";
import { TopAppBarActionItem } from "@webiny/ui/TopAppBar";
import { List } from "@webiny/ui/List";
var menuDialog = /*#__PURE__*/css({
  "&.mdc-menu": {
    minWidth: 300
  }
}, "label:menuDialog;");
export var UserMenuRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(UserMenuRenderer, _UIRenderer);

  var _super = _createSuper(UserMenuRenderer);

  function UserMenuRenderer() {
    _classCallCheck(this, UserMenuRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(UserMenuRenderer, [{
    key: "render",
    value: function render(_ref) {
      var element = _ref.element,
          props = _ref.props,
          next = _ref.next;
      var handle = element.getMenuHandleElement();
      return /*#__PURE__*/React.createElement(Menu, {
        className: menuDialog,
        anchor: "topEnd",
        handle: /*#__PURE__*/React.createElement(TopAppBarActionItem, {
          icon: /*#__PURE__*/React.createElement("div", null, handle ? handle.render(props) : null)
        })
      }, /*#__PURE__*/React.createElement(List, {
        "data-testid": "logged-in-user-menu-list"
      }, next()));
    }
  }]);

  return UserMenuRenderer;
}(UIRenderer);
