// @webiny/ui/Progress/CircularProgress.js
// This is just to center the spinner
import * as React from "react";
import styled from "@emotion/styled";
import Spinner from "react-spinner-material";
import { Typography } from "../Typography";
var SpinnerWrapper = /*#__PURE__*/styled("div", {
  target: "e1s1fi5k0",
  label: "SpinnerWrapper"
})({
  width: "100%",
  height: "100%",
  position: "absolute",
  background: "var(--mdc-theme-surface)",
  opacity: 0.92,
  top: 0,
  left: 0,
  zIndex: 10,
  "& .spinner__inner-wrapper": {
    position: "absolute",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center"
  }
});
var Label = /*#__PURE__*/styled("div", {
  target: "e1s1fi5k1",
  label: "Label"
})({
  marginTop: 15
});

var CircularProgress = function CircularProgress(_ref) {
  var label = _ref.label,
      size = _ref.size,
      spinnerWidth = _ref.spinnerWidth,
      spinnerColor = _ref.spinnerColor,
      visible = _ref.visible,
      style = _ref.style;
  return /*#__PURE__*/React.createElement(SpinnerWrapper, {
    style: style
  }, /*#__PURE__*/React.createElement("div", {
    className: "spinner__inner-wrapper"
  }, /*#__PURE__*/React.createElement(Spinner, {
    size: size,
    spinnerColor: spinnerColor,
    spinnerWidth: spinnerWidth,
    visible: visible
  }), label && /*#__PURE__*/React.createElement(Label, null, /*#__PURE__*/React.createElement(Typography, {
    use: "overline"
  }, label))));
};

CircularProgress.defaultProps = {
  size: 45,
  spinnerColor: "#fa5723",
  spinnerWidth: 4,
  visible: true
};
export default CircularProgress;
