// @webiny/app-page-builder/render/plugins/elements/embeds/vimeo/VimeoEmbed.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import * as React from "react";
import { css } from "emotion";
import { get } from "dot-prop-immutable";
var outerWrapper = /*#__PURE__*/css({
  boxSizing: "border-box",
  width: "100%"
}, "label:outerWrapper;");
var innerWrapper = /*#__PURE__*/css({
  left: 0,
  width: "100%",
  height: "auto",
  position: "relative",
  paddingBottom: 0
}, "label:innerWrapper;");
var scaleableVideo = /*#__PURE__*/css(_defineProperty({
  position: "relative",
  height: 0,
  paddingTop: "56.25%",
  width: "100%",
  backgroundColor: "#000",
  marginBottom: "1.2rem"
}, "iframe", {
  maxWidth: "100%",
  border: "none",
  display: "block",
  height: "100%",
  margin: 0,
  padding: 0,
  position: "absolute",
  top: 0,
  width: "100%"
}), "label:scaleableVideo;");

var VimeoEmbed = function VimeoEmbed(props) {
  var element = props.element;
  return /*#__PURE__*/React.createElement("div", {
    className: outerWrapper
  }, /*#__PURE__*/React.createElement("div", {
    className: innerWrapper
  }, /*#__PURE__*/React.createElement("div", {
    id: element.id,
    className: scaleableVideo,
    dangerouslySetInnerHTML: {
      __html: get(element, "data.oembed.html") || ""
    }
  })));
};

export default VimeoEmbed;
