// @webiny/app-headless-cms/admin/components/DropZone/Center.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React from "react";
import styled from "@emotion/styled";
import Droppable from "./../Droppable";
// @ts-ignore
var Container = /*#__PURE__*/styled("div", {
  target: "e14zhyp10",
  label: "Container"
})(function (_ref) {
  var isOver = _ref.isOver;
  return {
    backgroundColor: "transparent",
    boxSizing: "border-box",
    height: "100%",
    minHeight: 100,
    position: "relative",
    userSelect: "none",
    width: "100%",
    border: isOver ? "2px dashed var(--mdc-theme-primary)" : "2px dashed var(--mdc-theme-secondary)",
    opacity: 1
  };
}); // @ts-ignore

var Add = /*#__PURE__*/styled("div", {
  target: "e14zhyp11",
  label: "Add"
})(function (_ref2) {
  var isOver = _ref2.isOver;
  return {
    position: "absolute",
    top: "50%",
    left: "50%",
    transform: "translate(-50%,-50%)",
    margin: 0,
    color: isOver ? "var(--mdc-theme-primary)" : "var(--mdc-theme-secondary)"
  };
});
export default function Center(_ref3) {
  var onDrop = _ref3.onDrop,
      children = _ref3.children,
      style = _ref3.style;
  return /*#__PURE__*/React.createElement(Droppable, {
    onDrop: onDrop
  }, function (_ref4) {
    var isOver = _ref4.isOver,
        drop = _ref4.drop;
    return /*#__PURE__*/React.createElement("div", {
      ref: drop,
      style: _objectSpread({
        width: "100%",
        height: "100%"
      }, style),
      "data-testid": "cms-editor-first-field-area"
    }, /*#__PURE__*/React.createElement(Container, {
      isOver: isOver
    }, /*#__PURE__*/React.createElement(Add, {
      isOver: isOver
    }, children)));
  });
}
