// @webiny/app-page-builder/editor/plugins/elementSettings/shadow/ShadowSettings.js
import React from "react";
import get from "lodash/get";
import { useRecoilValue } from "recoil";
import { activeElementAtom, elementWithChildrenByIdSelector } from "../../../recoil/modules";
import useUpdateHandlers from "../useUpdateHandlers"; // Components

import InputField from "../components/InputField";
import Wrapper from "../components/Wrapper";
import { BaseColorPickerComponent } from "../components/ColorPicker";
import { ContentWrapper, classes } from "../components/StyledComponents";
import Accordion from "../components/Accordion";
var DATA_NAMESPACE = "data.settings.shadow";

var Settings = function Settings() {
  var activeElementId = useRecoilValue(activeElementAtom);
  var element = useRecoilValue(elementWithChildrenByIdSelector(activeElementId));

  var _useUpdateHandlers = useUpdateHandlers({
    element: element,
    dataNamespace: DATA_NAMESPACE
  }),
      getUpdateValue = _useUpdateHandlers.getUpdateValue,
      getUpdatePreview = _useUpdateHandlers.getUpdatePreview;

  return /*#__PURE__*/React.createElement(Accordion, {
    title: "Shadow"
  }, /*#__PURE__*/React.createElement(ContentWrapper, {
    direction: "column"
  }, /*#__PURE__*/React.createElement(Wrapper, {
    label: "Color",
    containerClassName: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(BaseColorPickerComponent, {
    label: "Color",
    valueKey: DATA_NAMESPACE + ".color",
    updateValue: getUpdateValue("color"),
    updatePreview: getUpdatePreview("color")
  })), /*#__PURE__*/React.createElement(Wrapper, {
    containerClassName: classes.simpleGrid,
    label: "Horizontal offset",
    leftCellSpan: 8,
    rightCellSpan: 4
  }, /*#__PURE__*/React.createElement(InputField, {
    value: get(element, DATA_NAMESPACE + ".horizontal", 0),
    onChange: getUpdateValue("horizontal")
  })), /*#__PURE__*/React.createElement(Wrapper, {
    containerClassName: classes.simpleGrid,
    label: "Vertical offset",
    leftCellSpan: 8,
    rightCellSpan: 4
  }, /*#__PURE__*/React.createElement(InputField, {
    value: get(element, DATA_NAMESPACE + ".vertical", 0),
    onChange: getUpdateValue("vertical")
  })), /*#__PURE__*/React.createElement(Wrapper, {
    containerClassName: classes.simpleGrid,
    label: "Blur",
    leftCellSpan: 8,
    rightCellSpan: 4
  }, /*#__PURE__*/React.createElement(InputField, {
    value: get(element, DATA_NAMESPACE + ".blur", 0),
    onChange: getUpdateValue("blur")
  })), /*#__PURE__*/React.createElement(Wrapper, {
    containerClassName: classes.simpleGrid,
    label: "Spread",
    leftCellSpan: 8,
    rightCellSpan: 4
  }, /*#__PURE__*/React.createElement(InputField, {
    value: get(element, DATA_NAMESPACE + ".spread", 0),
    onChange: getUpdateValue("spread")
  }))));
};

export default /*#__PURE__*/React.memo(Settings);
