// @webiny/app-headless-cms/admin/plugins/fieldRenderers/number/numberInput.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import get from "lodash/get";
import { Input } from "@webiny/ui/Input";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-headless-cms/admin/fields/number");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-number",
  renderer: {
    rendererName: "number-input",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Number Input"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders a simple input with its type set to \"number\"."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "number" && !field.multipleValues && !get(field, "predefinedValues.enabled");
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind;
      var Bind = getBind();
      return /*#__PURE__*/React.createElement(Bind, null, function (bindProps) {
        return /*#__PURE__*/React.createElement(Input, Object.assign({}, bindProps, {
          onChange: function onChange(value) {
            return bindProps.onChange(value);
          },
          label: field.label,
          placeholder: field.placeholderText,
          description: field.helpText,
          type: "number"
        }));
      });
    }
  }
};
export default plugin;
