// @webiny/app-page-builder/admin/plugins/pageDetails/previewContent/PagePreview.js
import React from "react";
import { css } from "emotion";
import styled from "@emotion/styled";
import classNames from "classnames";
import { Typography } from "@webiny/ui/Typography";
import { Select } from "@webiny/ui/Select";
import RenderElement from "../../../../render/components/Element";
import useResponsiveClassName from "../../../../hooks/useResponsiveClassName";
import Zoom from "./Zoom";
var pageInnerWrapper = /*#__PURE__*/css({
  overflowY: "scroll",
  overflowX: "hidden",
  height: "calc(100vh - 230px)",
  position: "relative",
  ".webiny-pb-page-document": {
    transform: "scale(var(--webiny-pb-page-preview-scale))",
    transition: "transform 0.5s ease-in-out",
    transformOrigin: "top center"
  }
}, "label:pageInnerWrapper;");
var PagePreviewToolbar = /*#__PURE__*/styled("div", {
  target: "efu4gb0",
  label: "PagePreviewToolbar"
})({
  position: "sticky",
  bottom: 0,
  height: 30,
  paddingLeft: 15,
  color: "var(--mdc-theme-text-secondary-on-background)",
  borderTop: "1px solid var(--mdc-theme-on-background)",
  backgroundColor: "var(--mdc-theme-background)",
  width: "100%",
  transform: "translateZ(0)",
  display: "flex",
  overflow: "hidden",
  ".webiny-ui-select": {
    color: "var(--mdc-theme-text-secondary-on-background)",
    backgroundColor: "transparent !important",
    width: "120px !important",
    ".mdc-select__dropdown-icon": {
      display: "none"
    },
    select: {
      fontSize: 14,
      border: "none",
      height: 30,
      padding: 0,
      backgroundColor: "transparent !important"
    }
  }
});

var PagePreview = function PagePreview(_ref) {
  var page = _ref.page;

  var _useResponsiveClassNa = useResponsiveClassName(),
      pageElementRef = _useResponsiveClassNa.pageElementRef,
      responsiveClassName = _useResponsiveClassNa.responsiveClassName;

  return /*#__PURE__*/React.createElement(Zoom, null, function (_ref2) {
    var zoom = _ref2.zoom,
        setZoom = _ref2.setZoom;
    return /*#__PURE__*/React.createElement("div", {
      ref: pageElementRef,
      className: classNames(pageInnerWrapper, responsiveClassName),
      style: {
        "--webiny-pb-page-preview-scale": zoom
      }
    }, /*#__PURE__*/React.createElement(RenderElement, {
      key: page.id,
      element: page.content
    }), /*#__PURE__*/React.createElement(PagePreviewToolbar, null, /*#__PURE__*/React.createElement("span", null, /*#__PURE__*/React.createElement(Typography, {
      use: "overline"
    }, "Zoom:\xA0")), /*#__PURE__*/React.createElement(Select, {
      value: zoom.toString(),
      onChange: setZoom
    }, /*#__PURE__*/React.createElement("option", {
      value: "1"
    }, "100%"), /*#__PURE__*/React.createElement("option", {
      value: "0.75"
    }, "75%"), /*#__PURE__*/React.createElement("option", {
      value: "0.5"
    }, "50%"))));
  });
};

export default PagePreview;
