// @webiny/ui/ImageUpload/ImageEditorDialog.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var _excluded = ["src", "options", "onAccept", "open", "dialogZIndex"];
import * as React from "react";
import { ImageEditor } from "../ImageEditor";
import { Tooltip } from "../Tooltip";
import { css } from "emotion";
import { Dialog, DialogAccept, DialogCancel, DialogActions, DialogContent } from "../Dialog";
var imageEditorDialog = /*#__PURE__*/css({
  width: "100vw",
  height: "100vh",
  ".mdc-dialog__surface": {
    maxWidth: "100% !important",
    maxHeight: "100% !important",
    ".webiny-ui-dialog__content": {
      maxWidth: "100% !important",
      maxHeight: "100% !important",
      width: "100vw",
      height: "100vh",
      paddingTop: "0 !important"
    }
  }
}, "label:imageEditorDialog;");

var ImageEditorDialog = /*#__PURE__*/function (_React$Component) {
  _inherits(ImageEditorDialog, _React$Component);

  var _super = _createSuper(ImageEditorDialog);

  function ImageEditorDialog() {
    var _this;

    _classCallCheck(this, ImageEditorDialog);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "imageEditor", /*#__PURE__*/React.createRef());

    return _this;
  }

  _createClass(ImageEditorDialog, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          src = _this$props.src,
          options = _this$props.options,
          onAccept = _this$props.onAccept,
          open = _this$props.open,
          dialogZIndex = _this$props.dialogZIndex,
          dialogProps = _objectWithoutProperties(_this$props, _excluded);

      return /*#__PURE__*/React.createElement(Dialog, Object.assign({
        className: imageEditorDialog,
        style: {
          zIndex: dialogZIndex
        },
        open: open
      }, dialogProps), open && /*#__PURE__*/React.createElement(ImageEditor, {
        ref: this.imageEditor,
        src: src,
        options: options
      }, function (_ref) {
        var render = _ref.render,
            activeTool = _ref.activeTool;
        return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(DialogContent, null, render()), /*#__PURE__*/React.createElement(DialogActions, null, /*#__PURE__*/React.createElement(DialogCancel, null, "Cancel"), activeTool ? /*#__PURE__*/React.createElement(Tooltip, {
          content: "Please close currently active tool.",
          placement: "top"
        }, /*#__PURE__*/React.createElement(DialogAccept, {
          disabled: true
        }, "Save")) : /*#__PURE__*/React.createElement(DialogAccept, {
          onClick: function onClick() {
            return onAccept(_this2.imageEditor.current.getCanvasDataUrl());
          }
        }, "Save")));
      }));
    }
  }]);

  return ImageEditorDialog;
}(React.Component);

export default ImageEditorDialog;
