// @webiny/ui/ImageUpload/styled.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

var _hover, _styled;

// TODO remove
// @ts-nocheck
import styled from "@emotion/styled";
export var AddImageIconWrapper = /*#__PURE__*/styled("div", {
  target: "e1engwcq0",
  label: "AddImageIconWrapper"
})({
  color: "var(--mdc-theme-text-secondary-on-background)",
  ">svg": {
    width: "100%",
    height: "100%",
    maxWidth: 50,
    maxHeight: 50,
    display: "block",
    opacity: 0.5,
    margin: "0 auto"
  }
});
export var AddImageWrapper = /*#__PURE__*/styled("div", {
  target: "e1engwcq1",
  label: "AddImageWrapper"
})({
  width: "100%",
  height: "100%",
  minHeight: 50,
  minWidth: 50,
  textAlign: "center",
  backgroundColor: "var(--mdc-theme-on-background)",
  borderRadius: 0,
  borderBottom: "1px solid var(--mdc-theme-text-hint-on-background)",
  display: "flex",
  alignItems: "center",
  justifyContent: "space-around",
  flexDirection: "column",
  transition: "color 0.2s",
  cursor: "pointer",
  padding: 10,
  boxSizing: "border-box",
  "&:hover": {
    backgroundColor: "rgba(0,0,0, 0.5)",
    borderBottom: "1px solid var(--mdc-theme-on-surface)"
  }
});
export var AddImageWrapperRound = /*#__PURE__*/styled(AddImageWrapper, {
  target: "e1engwcq2",
  label: "AddImageWrapperRound"
})({
  margin: "auto",
  width: "150px",
  height: "150px",
  borderRadius: "50%"
});
export var RemoveImage = /*#__PURE__*/styled("div", {
  target: "e1engwcq3",
  label: "RemoveImage"
})({
  position: "absolute",
  cursor: "pointer",
  top: 10,
  right: 10,
  display: "none",
  color: "white",
  opacity: 0.5,
  "&:hover": {
    opacity: 0.75
  }
});
export var EditImage = /*#__PURE__*/styled("div", {
  target: "e1engwcq4",
  label: "EditImage"
})({
  position: "absolute",
  cursor: "pointer",
  top: 10,
  left: 10,
  display: "none",
  color: "white",
  opacity: 0.5,
  "&:hover": {
    opacity: 0.75
  }
});
export var ImagePreviewWrapper = /*#__PURE__*/styled("div", {
  target: "e1engwcq5",
  label: "ImagePreviewWrapper"
})((_styled = {
  width: "100%",
  height: "100%",
  minHeight: 50,
  minWidth: 50,
  display: "flex",
  alignItems: "center",
  justifyContent: "space-around",
  flexDirection: "column",
  boxSizing: "border-box",
  position: "relative"
}, _defineProperty(_styled, AddImageWrapper, _defineProperty({
  position: "absolute",
  display: "none",
  top: 0,
  height: "100%",
  zIndex: 1,
  backgroundColor: "rgba(0,0,0, 0.75)"
}, AddImageIconWrapper, {
  top: "50%",
  left: "50%",
  transform: "translateY(-50%) translateX(-50%)",
  position: "absolute",
  color: "white"
})), _defineProperty(_styled, "&:hover", (_hover = {}, _defineProperty(_hover, AddImageWrapper, {
  display: "block"
}), _defineProperty(_hover, RemoveImage, {
  display: "block",
  zIndex: 2
}), _defineProperty(_hover, EditImage, {
  display: "block",
  zIndex: 2
}), _hover)), _styled));
