// @webiny/app-page-builder/contexts/PageBuilder/PageBuilderContext.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import * as React from "react";
import { plugins } from "@webiny/plugins";
import { DisplayMode } from "../../types";
export var PageBuilderContext = /*#__PURE__*/React.createContext(null);
export var PageBuilderProvider = function PageBuilderProvider(_ref) {
  var children = _ref.children;

  var _React$useState = React.useState(DisplayMode.DESKTOP),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      displayMode = _React$useState2[0],
      setDisplayMode = _React$useState2[1];

  var value = React.useMemo(function () {
    var theme = Object.assign.apply(Object, [{}].concat(_toConsumableArray(plugins.byType("pb-theme").map(function (pl) {
      return pl.theme;
    }))));
    return {
      theme: theme
    };
  }, []);
  return /*#__PURE__*/React.createElement(PageBuilderContext.Provider, {
    value: _objectSpread(_objectSpread({}, value), {}, {
      responsiveDisplayMode: {
        displayMode: displayMode,
        setDisplayMode: setDisplayMode
      }
    })
  }, children);
};
export var PageBuilderConsumer = function PageBuilderConsumer(_ref2) {
  var children = _ref2.children;
  return /*#__PURE__*/React.createElement(PageBuilderContext.Consumer, null, function (props) {
    return /*#__PURE__*/React.cloneElement(children, {
      pageBuilder: props
    });
  });
};
