// @webiny/app-page-builder/admin/utils/components/EditElementDialog.js
import React from "react";
import { css } from "emotion";
import { plugins } from "@webiny/plugins";
import { Dialog, DialogTitle, DialogContent, DialogActions, DialogButton, DialogCancel } from "@webiny/ui/Dialog";
import { Input } from "@webiny/ui/Input";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Form } from "@webiny/form";
import styled from "@emotion/styled";
import { validation } from "@webiny/validation";
var narrowDialog = /*#__PURE__*/css({
  ".mdc-dialog__surface": {
    width: 600,
    minWidth: 600
  }
}, "label:narrowDialog;");
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "ed1qryb0",
  label: "PreviewBox"
})({
  width: 500,
  minHeight: 250,
  border: "1px solid var(--mdc-theme-on-background)",
  backgroundColor: "#fff",
  // this must always be white
  display: "flex",
  justifyContent: "center",
  alignItems: "center",
  img: {
    maxHeight: 500,
    maxWidth: 500
  }
});
var EditElementDialog = /*#__PURE__*/React.memo(function (props) {
  var open = props.open,
      onClose = props.onClose,
      onSubmit = props.onSubmit,
      pluginName = props.plugin;
  var plugin = plugins.byName(pluginName);
  return /*#__PURE__*/React.createElement(Dialog, {
    open: open,
    onClose: onClose,
    className: narrowDialog
  }, plugin && /*#__PURE__*/React.createElement(Form, {
    onSubmit: onSubmit,
    data: plugin
  }, function (_ref) {
    var submit = _ref.submit,
        Bind = _ref.Bind;
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(DialogTitle, null, "Update ", plugin.title), /*#__PURE__*/React.createElement(DialogContent, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "title",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      label: "Name",
      autoFocus: true
    })))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(PreviewBox, null, plugin.toolbar.preview())))), /*#__PURE__*/React.createElement(DialogActions, null, /*#__PURE__*/React.createElement(DialogCancel, null, "Cancel"), /*#__PURE__*/React.createElement(DialogButton, {
      onClick: submit
    }, "Save")));
  }));
}, function (props, nextProps) {
  return props.open === nextProps.open;
});
EditElementDialog.displayName = "EditElementDialog";
export default EditElementDialog;
