// @webiny/app-admin/ui/views/FormView/FormContainerElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _get from "@babel/runtime/helpers/get";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import classNames from "classnames";
import styled from "@emotion/styled";
import { Elevation } from "@webiny/ui/Elevation";
import { UIElement } from "../../UIElement";
var FormContainerWrapper = /*#__PURE__*/styled("div", {
  target: "eyimbyl0",
  label: "FormContainerWrapper"
})({
  position: "relative",
  margin: "17px 50px"
});
export var FormContainerElement = /*#__PURE__*/function (_UIElement) {
  _inherits(FormContainerElement, _UIElement);

  var _super = _createSuper(FormContainerElement);

  function FormContainerElement(id, config) {
    var _this;

    _classCallCheck(this, FormContainerElement);

    _this = _super.call(this, id, config);

    _this.useGrid(false);

    return _this;
  }

  _createClass(FormContainerElement, [{
    key: "render",
    value: function render(props) {
      var children = _get(_getPrototypeOf(FormContainerElement.prototype), "render", this).call(this, props);

      return /*#__PURE__*/React.createElement(FormContainerWrapper, {
        className: classNames("webiny-data-list", this.config.className),
        "data-testid": this.config.testId
      }, this.config.noElevation ? children : /*#__PURE__*/React.createElement(Elevation, {
        z: 1
      }, children));
    }
  }]);

  return FormContainerElement;
}(UIElement);
