// @webiny/app-admin/ui/views/FormView/FormFooterElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _get from "@babel/runtime/helpers/get";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import styled from "@emotion/styled";
import { UIElement } from "../../UIElement";
var ButtonWrapper = /*#__PURE__*/styled("div", {
  target: "es96np90",
  label: "ButtonWrapper"
})({
  display: "flex",
  justifyContent: "space-between",
  borderTop: "1px solid var(--mdc-theme-on-background)",
  color: "var(--mdc-theme-text-primary-on-background)",
  textAlign: "right",
  padding: 25,
  ">:last-child": {
    marginLeft: "auto"
  }
});
export var FormFooterElement = /*#__PURE__*/function (_UIElement) {
  _inherits(FormFooterElement, _UIElement);

  var _super = _createSuper(FormFooterElement);

  function FormFooterElement(id) {
    var _this;

    _classCallCheck(this, FormFooterElement);

    _this = _super.call(this, id);

    _this.useGrid(false);

    return _this;
  }

  _createClass(FormFooterElement, [{
    key: "render",
    value: function render(props) {
      return /*#__PURE__*/React.createElement(ButtonWrapper, null, _get(_getPrototypeOf(FormFooterElement.prototype), "render", this).call(this, props));
    }
  }]);

  return FormFooterElement;
}(UIElement);
