// @webiny/app-page-builder/admin/plugins/installation.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11, _templateObject12, _templateObject13, _templateObject14;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useState, useCallback, lazy } from "react";
import gql from "graphql-tag";
import { useApolloClient } from "@apollo/react-hooks";
import { i18n } from "@webiny/app/i18n";
import { Form } from "@webiny/form";
import { Alert } from "@webiny/ui/Alert";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ButtonPrimary } from "@webiny/ui/Button";
import { CircularProgress } from "@webiny/ui/Progress";
import { Input } from "@webiny/ui/Input";
import { validation } from "@webiny/validation";
import { SimpleForm, SimpleFormHeader, SimpleFormFooter, SimpleFormContent } from "@webiny/app-admin/components/SimpleForm";
var t = i18n.ns("api-page-builder/admin/installation");
var IS_INSTALLED = gql(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    query IsPageBuilderInstalled {\n        pageBuilder {\n            version\n        }\n    }\n"])));
var INSTALL = gql(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n    mutation InstallPageBuilder($data: PbInstallInput!) {\n        pageBuilder {\n            install(data: $data) {\n                data\n                error {\n                    code\n                    message\n                }\n            }\n        }\n    }\n"]))); // eslint-disable-next-line

var installationSteps = {
  1: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Creating page categories..."]))),
  2: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Creating page blocks..."]))),
  3: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Creating pages..."]))),
  4: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Creating menus..."]))),
  5: t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Finalizing..."])))
};

var PBInstaller = function PBInstaller(_ref) {
  var onInstalled = _ref.onInstalled;
  var client = useApolloClient();

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      loading = _useState2[0],
      setLoading = _useState2[1];

  var _useState3 = useState(null),
      _useState4 = _slicedToArray(_useState3, 2),
      error = _useState4[0],
      setError = _useState4[1];

  var onSubmit = useCallback( /*#__PURE__*/function () {
    var _ref2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(form) {
      return _regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              setLoading(true);
              setError(null);
              setTimeout( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
                var _yield$client$mutate, res, error;

                return _regeneratorRuntime.wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        _context.next = 2;
                        return client.mutate({
                          mutation: INSTALL,
                          variables: {
                            data: form
                          }
                        });

                      case 2:
                        _yield$client$mutate = _context.sent;
                        res = _yield$client$mutate.data;
                        setLoading(false);
                        error = res.pageBuilder.install.error;

                        if (!error) {
                          _context.next = 9;
                          break;
                        }

                        setError(error.message);
                        return _context.abrupt("return");

                      case 9:
                        onInstalled();

                      case 10:
                      case "end":
                        return _context.stop();
                    }
                  }
                }, _callee);
              })), 10000);

            case 3:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }));

    return function (_x) {
      return _ref2.apply(this, arguments);
    };
  }(), []);
  var label = error ? /*#__PURE__*/React.createElement(Alert, {
    title: t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["Something went wrong"]))),
    type: "danger"
  }, error) : t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Installing Page Builder..."])));
  return /*#__PURE__*/React.createElement(Form, {
    onSubmit: onSubmit,
    submitOnEnter: true
  }, function (_ref4) {
    var Bind = _ref4.Bind,
        submit = _ref4.submit;
    return /*#__PURE__*/React.createElement(SimpleForm, null, loading && /*#__PURE__*/React.createElement(CircularProgress, {
      label: label
    }), /*#__PURE__*/React.createElement(SimpleFormHeader, {
      title: t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["Install Page Builder"])))
    }), /*#__PURE__*/React.createElement(SimpleFormContent, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "name",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      label: t(_templateObject11 || (_templateObject11 = _taggedTemplateLiteral(["Site Name"]))),
      description: t(_templateObject12 || (_templateObject12 = _taggedTemplateLiteral(["Name of your website, eg: \"My Site\""])))
    }))))), /*#__PURE__*/React.createElement(SimpleFormFooter, null, /*#__PURE__*/React.createElement(ButtonPrimary, {
      "data-testid": "install-pb-button",
      onClick: submit
    }, t(_templateObject13 || (_templateObject13 = _taggedTemplateLiteral(["Install Page Builder"]))))));
  });
};

export default {
  name: "admin-installation-pb",
  type: "admin-installation",
  title: t(_templateObject14 || (_templateObject14 = _taggedTemplateLiteral(["Page Builder"]))),
  dependencies: [],
  secure: true,
  getInstalledVersion: function getInstalledVersion(_ref5) {
    return _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3() {
      var client, _yield$client$query, data;

      return _regeneratorRuntime.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              client = _ref5.client;
              _context3.next = 3;
              return client.query({
                query: IS_INSTALLED
              });

            case 3:
              _yield$client$query = _context3.sent;
              data = _yield$client$query.data;
              return _context3.abrupt("return", data.pageBuilder.version);

            case 6:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3);
    }))();
  },
  render: function render(_ref6) {
    var onInstalled = _ref6.onInstalled;
    return /*#__PURE__*/React.createElement(PBInstaller, {
      onInstalled: onInstalled
    });
  },
  upgrades: [{
    version: "5.0.0",
    getComponent: function getComponent() {
      return /*#__PURE__*/lazy(function () {
        return import("./upgrades/v5.0.0");
      });
    }
  }]
};
