// @webiny/app-page-builder/admin/plugins/menus.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

import React from "react";
import { ReactComponent as PagesIcon } from "@svgr/webpack!../assets/table_chart-24px.svg";
import { i18n } from "@webiny/app/i18n";
import { useSecurity } from "@webiny/app-security";
var t = i18n.ns("app-form-builder/admin/menus");

var PageBuilderMenu = function PageBuilderMenu(_ref) {
  var Menu = _ref.Menu,
      Section = _ref.Section,
      Item = _ref.Item;

  var _useSecurity = useSecurity(),
      identity = _useSecurity.identity;

  var menus = identity.getPermission("pb.menu");
  var categories = identity.getPermission("pb.category");
  var pages = identity.getPermission("pb.page");

  if (!menus && !categories && !pages) {
    return null;
  }

  return /*#__PURE__*/React.createElement(Menu, {
    name: "app-page-builder",
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Page Builder"]))),
    icon: /*#__PURE__*/React.createElement(PagesIcon, null)
  }, /*#__PURE__*/React.createElement(Section, {
    label: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Pages"])))
  }, categories && /*#__PURE__*/React.createElement(Item, {
    label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Categories"]))),
    path: "/page-builder/categories"
  }), pages && /*#__PURE__*/React.createElement(Item, {
    label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Pages"]))),
    path: "/page-builder/pages"
  }), menus && /*#__PURE__*/React.createElement(Item, {
    label: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Menus"]))),
    path: "/page-builder/menus"
  })));
};

var plugin = {
  type: "admin-menu",
  name: "admin-menu-page-builder",
  render: function render(props) {
    return /*#__PURE__*/React.createElement(PageBuilderMenu, props);
  }
};
export default plugin;
