// @webiny/app-page-builder/admin/plugins/routes.js
import React from "react";
import { Route } from "@webiny/react-router";
import Helmet from "react-helmet";
import { AdminLayout } from "@webiny/app-admin/components/AdminLayout";
import { SecureRoute } from "@webiny/app-security/components";
import { EditorPluginsLoader } from "../components/EditorPluginsLoader";
import Categories from "../views/Categories/Categories";
import Menus from "../views/Menus/Menus";
import Pages from "../views/Pages/Pages";
import Editor from "../views/Pages/Editor";
var ROLE_PB_CATEGORY = "pb.category";
var ROLE_PB_MENUS = "pb.menu";
var ROLE_PB_PAGES = "pb.page";
var plugins = [{
  name: "route-pb-categories",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/page-builder/categories",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: ROLE_PB_CATEGORY
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, {
        title: "Page Builder - Categories"
      }), /*#__PURE__*/React.createElement(Categories, null)));
    }
  })
}, {
  name: "route-pb-menus",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/page-builder/menus",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: ROLE_PB_MENUS
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, {
        title: "Page Builder - Menus"
      }), /*#__PURE__*/React.createElement(Menus, null)));
    }
  })
}, {
  name: "route-pb-pages",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/page-builder/pages",
    render: function render(_ref) {
      var location = _ref.location;
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: ROLE_PB_PAGES
      }, /*#__PURE__*/React.createElement(EditorPluginsLoader, {
        location: location
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, {
        title: "Page Builder - Pages"
      }), /*#__PURE__*/React.createElement(Pages, null))));
    }
  })
}, {
  name: "route-pb-editor",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/page-builder/editor/:id",
    render: function render(_ref2) {
      var location = _ref2.location;
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: ROLE_PB_PAGES
      }, /*#__PURE__*/React.createElement(EditorPluginsLoader, {
        location: location
      }, /*#__PURE__*/React.createElement(Helmet, {
        title: "Page Builder - Edit page"
      }), /*#__PURE__*/React.createElement(Editor, null)));
    }
  })
}];
export default plugins;
