// @webiny/app-headless-cms/admin/plugins/fieldRenderers/text/textInputs.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import React from "react";
import get from "lodash/get";
import { i18n } from "@webiny/app/i18n";
import { ReactComponent as DeleteIcon } from "@svgr/webpack!../../../icons/close.svg";
import DynamicSection from "../DynamicSection";
import { Input } from "@webiny/ui/Input";
var t = i18n.ns("app-headless-cms/admin/fields/text");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-text-inputs",
  renderer: {
    rendererName: "text-inputs",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Text Inputs"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders a simple list of text inputs."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "text" && field.multipleValues && !get(field, "predefinedValues.enabled");
    },
    render: function render(props) {
      return /*#__PURE__*/React.createElement(DynamicSection, props, function (_ref2) {
        var bind = _ref2.bind,
            index = _ref2.index;
        return /*#__PURE__*/React.createElement(Input, Object.assign({}, bind.index, {
          autoFocus: true,
          onEnter: function onEnter() {
            return bind.field.appendValue("");
          },
          label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Value {number}"])))({
            number: index + 1
          }),
          trailingIcon: index > 0 && {
            icon: /*#__PURE__*/React.createElement(DeleteIcon, null),
            onClick: function onClick() {
              return bind.field.removeValue(index);
            }
          }
        }));
      });
    }
  }
};
export default plugin;
