// @webiny/app-headless-cms/admin/plugins/fieldRenderers/text/textInput.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import { i18n } from "@webiny/app/i18n";
import { Input } from "@webiny/ui/Input";
import get from "lodash/get";
var t = i18n.ns("app-headless-cms/admin/fields/text");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-text",
  renderer: {
    rendererName: "text-input",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Text Input"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Renders a simple input with its type set to \"text\"."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "text" && !field.multipleValues && !get(field, "predefinedValues.enabled");
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind;
      var Bind = getBind();
      return /*#__PURE__*/React.createElement(Bind, null, function (bind) {
        return /*#__PURE__*/React.createElement(Input, Object.assign({}, bind, {
          label: field.label,
          placeholder: field.placeholderText,
          description: field.helpText
        }));
      });
    }
  }
};
export default plugin;
