// @webiny/app-admin/ui/elements/form/FileManagerElement/EmptyStateElementRenderer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { UIRenderer } from "../../../UIRenderer";
import * as Ui from "@webiny/ui/ImageUpload";
import { Image } from "@webiny/app/components";
import { AddImageIconWrapper, AddImageWrapper, FilePreviewWrapper, RemoveImage } from "./styled";
import { ReactComponent as AddFileIcon } from "@svgr/webpack!../../../../assets/icons/attach_file_black_24dp.svg";
import { Typography } from "@webiny/ui/Typography";
import { ReactComponent as RemoveImageIcon } from "@svgr/webpack!@webiny/ui/ImageUpload/icons/round-close-24px.svg";
export var EmptyStateElementRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(EmptyStateElementRenderer, _UIRenderer);

  var _super = _createSuper(EmptyStateElementRenderer);

  function EmptyStateElementRenderer() {
    _classCallCheck(this, EmptyStateElementRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(EmptyStateElementRenderer, [{
    key: "render",
    value: function render(_ref) {
      var props = _ref.props;
      var accept = props.fileManagerElement.getAccept();
      var acceptsOnlyImages = !accept || accept.every(function (el) {
        return el.startsWith("image/");
      });

      if (acceptsOnlyImages) {
        return /*#__PURE__*/React.createElement(Ui.Image, {
          placeholder: "Select an image",
          renderImagePreview: function renderImagePreview(renderImageProps) {
            return /*#__PURE__*/React.createElement(Image, renderImageProps);
          },
          style: {
            width: "100%",
            height: "auto"
          },
          value: props.value,
          uploadImage: props.showFileManager,
          removeImage: props.onChange
        });
      }

      if (props.value) {
        return /*#__PURE__*/React.createElement(FilePreviewWrapper, null, /*#__PURE__*/React.createElement("span", null, "Attached file: ", /*#__PURE__*/React.createElement("strong", null, props.value.src.split("/").pop())), /*#__PURE__*/React.createElement(RemoveImage, {
          onClick: function onClick() {
            return props.onChange(null);
          }
        }, /*#__PURE__*/React.createElement(RemoveImageIcon, null)), /*#__PURE__*/React.createElement(AddImageWrapper, {
          "data-role": "select-file",
          onClick: props.showFileManager
        }, /*#__PURE__*/React.createElement(AddImageIconWrapper, null, /*#__PURE__*/React.createElement(AddFileIcon, null), /*#__PURE__*/React.createElement(Typography, {
          use: "caption"
        }, "Select a file"))));
      }

      return /*#__PURE__*/React.createElement(AddImageWrapper, {
        "data-role": "select-file",
        onClick: props.showFileManager
      }, /*#__PURE__*/React.createElement(AddImageIconWrapper, null, /*#__PURE__*/React.createElement(AddFileIcon, null), /*#__PURE__*/React.createElement(Typography, {
        use: "caption"
      }, "Select a file")));
    }
  }]);

  return EmptyStateElementRenderer;
}(UIRenderer);
