// @webiny/app-headless-cms/admin/plugins/editor/defaultBar/Name/NameStyled.js
import styled from "@emotion/styled";
import { css } from "emotion";
export var NameInputWrapper = /*#__PURE__*/styled("div", {
  target: "e1go0so10",
  label: "NameInputWrapper"
})({
  width: "100%",
  display: "flex",
  alignItems: "center",
  position: "relative",
  "> .mdc-text-field--upgraded": {
    height: 35,
    marginTop: "0 !important",
    paddingLeft: 10,
    paddingRight: 40
  }
});
export var NameWrapper = /*#__PURE__*/styled("div", {
  target: "e1go0so11",
  label: "NameWrapper"
})({
  display: "flex",
  alignItems: "baseline",
  justifyContent: "flex-center",
  flexDirection: "column",
  color: "var(--mdc-theme-text-primary-on-background)",
  position: "relative",
  width: "100%",
  marginLeft: 10
});
export var FormName = /*#__PURE__*/styled("div", {
  target: "e1go0so12",
  label: "FormName"
})({
  border: "1px solid transparent",
  fontSize: 20,
  lineHeight: 1.25,
  whiteSpace: "nowrap",
  overflow: "hidden",
  textOverflow: "ellipsis",
  width: "100%",
  "&:hover": {
    border: "1px solid var(--mdc-theme-on-background)"
  }
});
export var formNameWrapper = /*#__PURE__*/css({
  maxWidth: "calc(100% - 50px)"
}, "label:formNameWrapper;");
