// @webiny/app-page-builder/render/plugins/elements/image/Image.js
import React from "react";
import { ElementRoot } from "../../../components/ElementRoot";
import { Image as WebinyImage } from "@webiny/app/components";
import { Link as RouterLink } from "@webiny/react-router";

var Link = function Link(_ref) {
  var link = _ref.link,
      children = _ref.children;

  if (!link || !link.href) {
    return children;
  }

  return /*#__PURE__*/React.createElement(RouterLink, {
    to: link.href,
    target: link.newTab ? "_blank" : "_self"
  }, children);
};

var Image = function Image(_ref2) {
  var element = _ref2.element;

  var _ref3 = element.data || {},
      _ref3$image = _ref3.image,
      image = _ref3$image === void 0 ? {} : _ref3$image,
      _ref3$link = _ref3.link,
      link = _ref3$link === void 0 ? {} : _ref3$link;

  if (!image || !image.file) {
    return null;
  }

  var width = image.width,
      height = image.height,
      title = image.title;
  var style = {
    width: width,
    height: height
  };
  return /*#__PURE__*/React.createElement(ElementRoot, {
    element: element,
    style: {
      display: "flex"
    },
    className: "webiny-pb-base-page-element-style webiny-pb-page-element-image"
  }, /*#__PURE__*/React.createElement(Link, {
    link: link
  }, /*#__PURE__*/React.createElement(WebinyImage, {
    title: title,
    alt: title,
    style: style,
    src: image.file.src,
    srcSet: "auto"
  })));
};

export default Image;
