// @webiny/app-admin/ui/views/OverlayView/HeaderTitleElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { UIElement } from "../../UIElement";
import { Typography } from "@webiny/ui/Typography";
import { UIRenderer } from "../../UIRenderer";
export var HeaderTitleElementRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(HeaderTitleElementRenderer, _UIRenderer);

  var _super = _createSuper(HeaderTitleElementRenderer);

  function HeaderTitleElementRenderer() {
    _classCallCheck(this, HeaderTitleElementRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(HeaderTitleElementRenderer, [{
    key: "render",
    value: function render(_ref) {
      var element = _ref.element;
      return /*#__PURE__*/React.createElement(Typography, {
        style: {
          margin: "0 auto",
          color: "var(--mdc-theme-on-surface)"
        },
        use: "headline6"
      }, element.config.title());
    }
  }]);

  return HeaderTitleElementRenderer;
}(UIRenderer);
export var HeaderTitleElement = /*#__PURE__*/function (_UIElement) {
  _inherits(HeaderTitleElement, _UIElement);

  var _super2 = _createSuper(HeaderTitleElement);

  function HeaderTitleElement(id, config) {
    var _this;

    _classCallCheck(this, HeaderTitleElement);

    _this = _super2.call(this, id, config);

    _this.addRenderer(new HeaderTitleElementRenderer());

    return _this;
  }

  return HeaderTitleElement;
}(UIElement);
