// @webiny/app-admin/ui/views/OverlayView/useOverlayView.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import { useEffect, useState } from "react";
import { css } from "emotion";
import { OverlayView } from "../OverlayView";
var noScroll = /*#__PURE__*/css({
  overflow: "hidden",
  height: "100vh"
}, "label:noScroll;");
export function useOverlayView() {
  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isVisible = _useState2[0],
      setIsVisible = _useState2[1];

  useEffect(function () {
    if (isVisible) {
      OverlayView.openedViews++;
      document.body.classList.add(noScroll);
    } else if (!OverlayView.openedViews) {
      document.body.classList.remove(noScroll);
    }
  }, [isVisible]);
  useEffect(function () {
    return function () {
      if (OverlayView.openedViews > 0) {
        OverlayView.openedViews--;
      }

      if (!OverlayView.openedViews) {
        document.body.classList.remove(noScroll);
      }
    };
  }, []);
  return {
    isVisible: isVisible,
    setIsVisible: setIsVisible
  };
}
