// @webiny/app-page-builder/editor/plugins/elementSettings/components/Footer.js
import * as React from "react";
import { Cell, Grid } from "@webiny/ui/Grid";
import { Typography } from "@webiny/ui/Typography";
import { Switch } from "@webiny/ui/Switch";
import { Footer as FooterStyled } from "./StyledComponents";
import { css } from "emotion";
var switchStyle = /*#__PURE__*/css({
  "&.webiny-ui-switch": {
    marginRight: "0"
  }
}, "label:switchStyle;");

var Footer = function Footer(_ref) {
  var advanced = _ref.advanced,
      toggleAdvanced = _ref.toggleAdvanced;
  return /*#__PURE__*/React.createElement(FooterStyled, null, /*#__PURE__*/React.createElement(Grid, {
    className: "no-bottom-padding"
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 8
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "subtitle2"
  }, "Show advanced options")), /*#__PURE__*/React.createElement(Cell, {
    span: 4
  }, /*#__PURE__*/React.createElement(Switch, {
    className: switchStyle,
    value: advanced,
    onChange: toggleAdvanced
  }))));
};

export default /*#__PURE__*/React.memo(Footer);
