// @webiny/app-page-builder/editor/plugins/elementSettings/components/SliderWithInput.js
import React from "react";
import { elementWithChildrenByIdSelector, activeElementAtom } from "../../../recoil/modules";
import lodashGet from "lodash/get";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Icon } from "@webiny/ui/Icon";
import { Slider } from "@webiny/ui/Slider";
import { useRecoilValue } from "recoil";
import InputField from "./InputField";

var SliderWithInput = function SliderWithInput(_ref) {
  var icon = _ref.icon,
      placeholder = _ref.placeholder,
      updateValue = _ref.updateValue,
      updatePreview = _ref.updatePreview,
      className = _ref.className,
      valueKey = _ref.valueKey,
      _ref$max = _ref.max,
      max = _ref$max === void 0 ? 100 : _ref$max,
      _ref$step = _ref.step,
      step = _ref$step === void 0 ? 1 : _ref$step;
  var activeElementId = useRecoilValue(activeElementAtom);
  var element = useRecoilValue(elementWithChildrenByIdSelector(activeElementId));
  var value = lodashGet(element, valueKey, 0);
  return /*#__PURE__*/React.createElement(Grid, {
    className: className
  }, /*#__PURE__*/React.createElement(Cell, {
    align: "middle",
    span: 2
  }, /*#__PURE__*/React.createElement(Icon, {
    icon: icon
  })), /*#__PURE__*/React.createElement(Cell, {
    align: "middle",
    span: 6
  }, /*#__PURE__*/React.createElement(Slider, {
    value: value,
    onChange: updateValue,
    onInput: updatePreview,
    min: 0,
    max: max,
    step: step
  })), /*#__PURE__*/React.createElement(Cell, {
    align: "middle",
    span: 4
  }, /*#__PURE__*/React.createElement(InputField, {
    placeholder: placeholder || "px",
    value: value,
    onChange: updateValue
  })));
};

export default /*#__PURE__*/React.memo(SliderWithInput);
