// @webiny/app-page-builder/editor/plugins/elementSettings/components/StyledComponents.js
import styled from "@emotion/styled";
import { css } from "emotion";
export var classes = {
  simpleGrid: /*#__PURE__*/css({
    "&.mdc-layout-grid": {
      padding: 0,
      margin: "0px 0px 16px"
    }
  }, "label:simpleGrid;"),
  grid: /*#__PURE__*/css({
    "& .mdc-layout-grid": {
      padding: 0,
      margin: "0px 0px 16px"
    }
  }, "label:grid;")
};
export var Footer = /*#__PURE__*/styled("div", {
  target: "e12ge6nj0",
  label: "Footer"
})({
  backgroundColor: "var(--mdc-theme-background)",
  paddingBottom: 10,
  margin: "0 -15px -15px -15px",
  ".mdc-layout-grid": {
    padding: "15px 10px 10px 15px",
    ".mdc-layout-grid__cell.mdc-layout-grid__cell--span-4": {
      paddingRight: 10
    }
  }
});
export var InputContainer = /*#__PURE__*/styled("div", {
  target: "e12ge6nj1",
  label: "InputContainer"
})(function (props) {
  return {
    "> .mdc-text-field.mdc-text-field--upgraded": {
      height: "30px !important",
      width: props.width || 50,
      margin: props.hasOwnProperty("margin") ? props.margin : "0 0 0 18px",
      ".mdc-text-field__input": {
        paddingTop: 16
      }
    }
  };
});
export var ContentWrapper = /*#__PURE__*/styled("div", {
  target: "e12ge6nj2",
  label: "ContentWrapper"
})(function (props) {
  return {
    display: "flex",
    flexDirection: props.direction || "row"
  };
});
export var COLORS = {
  lightGray: "hsla(0, 0%, 97%, 1)",
  gray: "hsla(300, 2%, 92%, 1)",
  darkGray: "hsla(0, 0%, 70%, 1)",
  darkestGray: "hsla(0, 0%, 20%, 1)",
  black: "hsla(208, 100%, 5%, 1)"
};
export var TopLeft = /*#__PURE__*/styled("div", {
  target: "e12ge6nj3",
  label: "TopLeft"
})({
  gridArea: "topLeft"
});
export var Top = /*#__PURE__*/styled("div", {
  target: "e12ge6nj4",
  label: "Top"
})({
  gridArea: "top"
});
export var TopRight = /*#__PURE__*/styled("div", {
  target: "e12ge6nj5",
  label: "TopRight"
})({
  gridArea: "topRight"
});
export var Left = /*#__PURE__*/styled("div", {
  target: "e12ge6nj6",
  label: "Left"
})({
  gridArea: "left"
});
export var Center = /*#__PURE__*/styled("div", {
  target: "e12ge6nj7",
  label: "Center"
})({
  gridArea: "center",
  backgroundColor: "rgb(204,229,255)",
  border: "1px dashed rgb(0,64,133)"
});
export var Right = /*#__PURE__*/styled("div", {
  target: "e12ge6nj8",
  label: "Right"
})({
  gridArea: "right"
});
export var BottomLeft = /*#__PURE__*/styled("div", {
  target: "e12ge6nj9",
  label: "BottomLeft"
})({
  gridArea: "bottomLeft"
});
export var Bottom = /*#__PURE__*/styled("div", {
  target: "e12ge6nj10",
  label: "Bottom"
})({
  gridArea: "bottom"
});
export var BottomRight = /*#__PURE__*/styled("div", {
  target: "e12ge6nj11",
  label: "BottomRight"
})({
  gridArea: "bottomRight"
});
export var SpacingGrid = /*#__PURE__*/styled("div", {
  target: "e12ge6nj12",
  label: "SpacingGrid"
})({
  display: "grid",
  gridTemplateColumns: "1fr 2fr 1fr",
  gridTemplateRows: "1fr 1fr 1fr",
  gap: "0px 0px",
  gridTemplateAreas: '"topLeft top topRight"' + '"left center right"' + '"bottomLeft bottom bottomRight"',
  border: "1px dashed rgb(21,87,36)",
  backgroundColor: COLORS.lightGray,
  "& .text": {
    fontSize: 11,
    padding: "4px 8px"
  },
  "& .mono": {
    fontFamily: "monospace"
  },
  "& .align-center": {
    display: "flex",
    justifyContent: "center"
  }
});
export var SimpleButton = /*#__PURE__*/styled("button", {
  target: "e12ge6nj13",
  label: "SimpleButton"
})({
  boxSizing: "border-box",
  border: "1px solid var(--mdc-theme-on-background)",
  borderRadius: 1,
  backgroundColor: "transparent",
  padding: "8px 16px",
  cursor: "pointer"
});
export var ButtonContainer = /*#__PURE__*/styled("div", {
  target: "e12ge6nj14",
  label: "ButtonContainer"
})({
  marginTop: 16
});
export var justifySelfEndStyle = /*#__PURE__*/css({
  justifySelf: "end"
}, "label:justifySelfEndStyle;");
