// @webiny/app-file-manager/admin/plugins/permissionRenderer/index.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import { AccordionItem } from "@webiny/ui/Accordion";
import { ReactComponent as FileManagerIcon } from "@svgr/webpack!../../assets/icons/folder-open.svg";
import { FileManagerPermissions } from "./FileManagerPermissions";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-file-manager/admin/plugins/permissionRenderer");
export default {
  type: "admin-app-permissions-renderer",
  name: "admin-app-permissions-renderer-file-manager",
  render: function render(props) {
    return /*#__PURE__*/React.createElement(AccordionItem, {
      icon: /*#__PURE__*/React.createElement(FileManagerIcon, null),
      title: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["File Manager"]))),
      description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Manage File manager app access permissions."]))),
      "data-testid": "permission.fm"
    }, /*#__PURE__*/React.createElement(FileManagerPermissions, props));
  }
};
