// @webiny/api-prerendering-service/client/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _error = _interopRequireDefault(require("@webiny/error"));

var _default = configuration => {
  return [{
    type: "context",

    apply(context) {
      if (!context.handlerClient) {
        throw new _error.default(`Prerendering service's client relies on the "context.handlerClient", defined by the "@webiny/handler-client" package. Please add it to your handler and try again.`);
      }

      if (context.prerenderingServiceClient) {
        return;
      }

      context.prerenderingServiceClient = {
        async render(args) {
          await context.handlerClient.invoke({
            name: configuration.handlers.render,
            await: false,
            payload: args
          });
        },

        async flush(args) {
          await context.handlerClient.invoke({
            name: configuration.handlers.flush,
            await: false,
            payload: args
          });
        },

        queue: {
          async add(args) {
            await context.handlerClient.invoke({
              name: configuration.handlers.queue.add,
              await: false,
              payload: args
            });
          },

          async process() {
            return context.handlerClient.invoke({
              name: configuration.handlers.queue.process,
              await: false
            });
          }

        }
      };
    }

  }];
};

exports.default = _default;
