// @webiny/api-headless-cms/content/plugins/modelManager/DefaultContentModelManager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultContentModelManager = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

class DefaultContentModelManager {
  constructor(context, model) {
    (0, _defineProperty2.default)(this, "_context", void 0);
    (0, _defineProperty2.default)(this, "_model", void 0);
    this._context = context;
    this._model = model;
  }

  async create(data) {
    return this._context.cms.entries.create(this._model, data);
  }

  async delete(id) {
    if (id.includes("#")) {
      return this._context.cms.entries.deleteRevision(this._model, id);
    }

    return this._context.cms.entries.deleteEntry(this._model, id);
  }

  async get(args) {
    return this._context.cms.entries.get(this._model, args);
  }

  async list(args) {
    return this._context.cms.entries.list(this._model, args);
  }

  async listPublished(args) {
    return this._context.cms.entries.listPublished(this._model, args);
  }

  async listLatest(args) {
    return this._context.cms.entries.listLatest(this._model, args);
  }

  async getPublishedByIds(ids) {
    return this._context.cms.entries.getPublishedByIds(this._model, ids);
  }

  async getLatestByIds(ids) {
    return this._context.cms.entries.getLatestByIds(this._model, ids);
  }

  async update(id, data) {
    return this._context.cms.entries.update(this._model, id, data);
  }

}

exports.DefaultContentModelManager = DefaultContentModelManager;
