// @webiny/app-admin/plugins/menu/fileManager.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { css } from "emotion";
import { ListItem, ListItemGraphic } from "@webiny/ui/List";
import { Icon } from "@webiny/ui/Icon";
import { ReactComponent as FileIcon } from "@svgr/webpack!../../assets/icons/insert_drive_file-24px.svg";
import { FileManager } from "../../components";
import { NavigationMenuElement } from "../../ui/elements/NavigationMenuElement";
import { UIRenderer } from "../../ui/UIRenderer";
import { UIViewPlugin } from "../../ui/UIView";
import { NavigationView } from "../../ui/views/NavigationView";
var listItemStyle = /*#__PURE__*/css({
  ".mdc-list &.mdc-list-item:hover": {
    cursor: "pointer",
    backgroundColor: "var(--mdc-theme-background)"
  }
}, "label:listItemStyle;");

var FileManagerMenuItemRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(FileManagerMenuItemRenderer, _UIRenderer);

  var _super = _createSuper(FileManagerMenuItemRenderer);

  function FileManagerMenuItemRenderer() {
    _classCallCheck(this, FileManagerMenuItemRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(FileManagerMenuItemRenderer, [{
    key: "render",
    value: function render(_ref) {
      var element = _ref.element;
      return /*#__PURE__*/React.createElement(FileManager, null, function (_ref2) {
        var showFileManager = _ref2.showFileManager;
        return /*#__PURE__*/React.createElement(ListItem, {
          ripple: false,
          onClick: showFileManager,
          className: listItemStyle,
          "data-testid": element.config.testId
        }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
          icon: element.config.icon
        })), element.config.label);
      });
    }
  }]);

  return FileManagerMenuItemRenderer;
}(UIRenderer);

export default (function () {
  return [new UIViewPlugin(NavigationView, function (view) {
    var element = new NavigationMenuElement("fileManager", {
      label: "File Manager",
      icon: /*#__PURE__*/React.createElement(FileIcon, null),
      testId: "admin-drawer-footer-menu-file-manager"
    });
    element.addRenderer(new FileManagerMenuItemRenderer());
    view.getFooterElement().addMenuElement(element);
  })];
});
