// @webiny/api-headless-cms/content/plugins/graphqlFields/boolean.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

const createListFilters = ({
  field
}) => {
  return `
        ${field.fieldId}: Boolean
        ${field.fieldId}_not: Boolean
    `;
};

const plugin = {
  name: "cms-model-field-to-graphql-boolean",
  type: "cms-model-field-to-graphql",
  fieldType: "boolean",
  isSortable: true,
  isSearchable: true,
  read: {
    createListFilters,

    createGetFilters({
      field
    }) {
      return `${field.fieldId}: Boolean`;
    },

    createTypeField({
      field
    }) {
      if (field.multipleValues) {
        return `${field.fieldId}: [Boolean]`;
      }

      return `${field.fieldId}: Boolean`;
    }

  },
  manage: {
    createListFilters,

    createTypeField({
      field
    }) {
      if (field.multipleValues) {
        return field.fieldId + ": [Boolean]";
      }

      return field.fieldId + ": Boolean";
    },

    createInputField({
      field
    }) {
      if (field.multipleValues) {
        return field.fieldId + ": [Boolean]";
      }

      return field.fieldId + ": Boolean";
    }

  }
};
var _default = plugin;
exports.default = _default;
