// @webiny/api-headless-cms/content/plugins/graphqlFields/text.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

const createListFilters = ({
  field
}) => {
  return `
        ${field.fieldId}: String
        ${field.fieldId}_not: String
        ${field.fieldId}_in: [String]
        ${field.fieldId}_not_in: [String]
        ${field.fieldId}_contains: String
        ${field.fieldId}_not_contains: String
    `;
};

const plugin = {
  name: "cms-model-field-to-graphql-text",
  type: "cms-model-field-to-graphql",
  fieldType: "text",
  isSortable: true,
  isSearchable: true,
  read: {
    createTypeField({
      field
    }) {
      if (field.multipleValues) {
        return `${field.fieldId}: [String]`;
      }

      return `${field.fieldId}: String`;
    },

    createGetFilters({
      field
    }) {
      return `${field.fieldId}: String`;
    },

    createListFilters
  },
  manage: {
    createListFilters,

    createTypeField({
      field
    }) {
      if (field.multipleValues) {
        return `${field.fieldId}: [String]`;
      }

      return `${field.fieldId}: String`;
    },

    createInputField({
      field
    }) {
      if (field.multipleValues) {
        return field.fieldId + ": [String]";
      }

      return field.fieldId + ": String";
    }

  }
};
var _default = plugin;
exports.default = _default;
