// @webiny/ui/List/DataList/Loader.js
import * as React from "react";
import styled from "@emotion/styled";
import Skeleton from "react-loading-skeleton";
var LoaderUl = /*#__PURE__*/styled("ul", {
  target: "e1nb7uzc0",
  label: "LoaderUl"
})({
  listStyle: "none",
  padding: "10px 20px",
  "li > div": {
    display: "inline-block",
    verticalAlign: "middle",
    ".react-loading-skeleton": {
      backgroundColor: "var(--mdc-theme-background)",
      backgroundImage: "linear-gradient(90deg, var(--mdc-theme-background), var(--mdc-theme-surface), var(--mdc-theme-background))"
    }
  },
  ".graphic": {
    fontSize: 36,
    width: 36
  },
  ".data": {
    width: "calc(-36px + 75%)",
    marginLeft: 10
  },
  ".actions": {
    width: "calc(-36px + 25%)",
    marginLeft: 10,
    textAlign: "right",
    "> div": {
      display: "inline-block",
      fontSize: 24,
      marginLeft: 10,
      width: 24
    }
  }
});

var Loader = function Loader() {
  return /*#__PURE__*/React.createElement(LoaderUl, null, [1, 2, 3, 4, 5].map(function (item) {
    return /*#__PURE__*/React.createElement("li", {
      key: "list-" + item
    }, /*#__PURE__*/React.createElement("div", {
      className: "graphic"
    }, /*#__PURE__*/React.createElement(Skeleton, null)), /*#__PURE__*/React.createElement("div", {
      className: "data"
    }, /*#__PURE__*/React.createElement(Skeleton, {
      count: 2
    })), /*#__PURE__*/React.createElement("div", {
      className: "actions"
    }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Skeleton, null)), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Skeleton, null))));
  }));
};

export default Loader;
