// @webiny/app-form-builder/admin/plugins/permissionRenderer/index.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2;

import React from "react";
import { i18n } from "@webiny/app/i18n";
import { AccordionItem } from "@webiny/ui/Accordion";
import { ReactComponent as FormBuilderIcon } from "@svgr/webpack!../../icons/round-ballot-24px.svg";
import { FormBuilderPermissions } from "./FormBuilderPermissions";
var t = i18n.ns("app-form-builder/admin/plugins/permissionRenderer");
export default (function () {
  return {
    type: "admin-app-permissions-renderer",
    name: "admin-app-permissions-renderer-form-builder",
    render: function render(props) {
      return /*#__PURE__*/React.createElement(AccordionItem, {
        icon: /*#__PURE__*/React.createElement(FormBuilderIcon, null),
        title: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Form Builder"]))),
        description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Manage Form Builder app access permissions."]))),
        "data-testid": "permission.fb"
      }, /*#__PURE__*/React.createElement(FormBuilderPermissions, props));
    }
  };
});
