// @webiny/app-headless-cms/admin/plugins/fieldRenderers/file/fileField.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import React from "react";
import { i18n } from "@webiny/app/i18n";
import { Cell, Grid } from "@webiny/ui/Grid";
import { FileManager } from "@webiny/app-admin/components";
import File from "./File";
var t = i18n.ns("app-headless-cms/admin/fields/file");
var plugin = {
  type: "cms-editor-field-renderer",
  name: "cms-editor-field-renderer-file",
  renderer: {
    rendererName: "file-input",
    name: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["File Input"]))),
    description: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Enables selecting a single file via File Manager."]))),
    canUse: function canUse(_ref) {
      var field = _ref.field;
      return field.type === "file" && !field.multipleValues;
    },
    render: function render(_ref2) {
      var field = _ref2.field,
          getBind = _ref2.getBind,
          Label = _ref2.Label;
      var Bind = getBind();
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Label, null, field.label), /*#__PURE__*/React.createElement(Bind, null, function (_ref3) {
        var value = _ref3.value,
            onChange = _ref3.onChange;
        return /*#__PURE__*/React.createElement(FileManager, {
          multiple: false
        }, function (_ref4) {
          var showFileManager = _ref4.showFileManager;

          var selectFile = function selectFile() {
            showFileManager(function (newFile) {
              if (newFile !== null) {
                onChange(newFile.src);
              }
            });
          };

          return /*#__PURE__*/React.createElement(File, {
            url: value,
            onRemove: function onRemove() {
              return onChange(null);
            },
            showFileManager: selectFile,
            placeholder: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Select a file\""])))
          });
        });
      })));
    }
  }
};
export default plugin;
