// @webiny/app-headless-cms/admin/plugins/fieldRenderers/file/utils.js
import * as React from "react";
import { css } from "emotion";
import { ReactComponent as FileIcon } from "@svgr/webpack!../../fields/icons/round_insert_drive_file-24px.svg";
export var imageExtensions = [".jpg", ".jpeg", ".gif", ".png", ".svg"];
export var imageWrapperStyles = /*#__PURE__*/css({
  height: "auto"
}, "label:imageWrapperStyles;");
var fileLabel = /*#__PURE__*/css({
  padding: "10px",
  whiteSpace: "nowrap",
  overflow: "hidden",
  textOverflow: "ellipsis",
  fontSize: "0.8rem",
  color: "var(--mdc-theme-on-surface)",
  backgroundColor: "var(--mdc-theme-on-background)"
}, "label:fileLabel;");
var IconWrapperStyle = /*#__PURE__*/css({
  display: "flex",
  justifyContent: "center",
  alignItems: "center",
  height: 200
}, "label:IconWrapperStyle;");
var IconStyle = /*#__PURE__*/css({
  width: 64,
  height: 64
}, "label:IconStyle;");
export var createRenderImagePreview = function createRenderImagePreview(_ref) {
  var value = _ref.value,
      imagePreviewProps = _ref.imagePreviewProps;
  var fileName = getFileName(value);

  var renderImagePreview = function renderImagePreview(renderImageProps) {
    return /*#__PURE__*/React.createElement("div", Object.assign({}, renderImageProps, imagePreviewProps), /*#__PURE__*/React.createElement("div", {
      className: IconWrapperStyle
    }, /*#__PURE__*/React.createElement(FileIcon, {
      className: IconStyle
    })), /*#__PURE__*/React.createElement("div", {
      className: fileLabel
    }, fileName));
  };

  return renderImagePreview;
};

var getFileName = function getFileName(value) {
  if (!value) {
    return "";
  }

  return value.split("/").pop();
};
