// @webiny/app-file-manager/admin/plugins/installation.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useState, useEffect, lazy } from "react";
import gql from "graphql-tag";
import { useApolloClient } from "@apollo/react-hooks";
import { i18n } from "@webiny/app/i18n";
import { Alert } from "@webiny/ui/Alert";
import { CircularProgress } from "@webiny/ui/Progress";
import { SimpleForm, SimpleFormContent } from "@webiny/app-admin/components/SimpleForm";
import styled from "@emotion/styled";
var SimpleFormPlaceholder = /*#__PURE__*/styled("div", {
  target: "eratgeh0",
  label: "SimpleFormPlaceholder"
})({
  minHeight: 300,
  minWidth: 400
});
var t = i18n.ns("app-file-manager/admin/installation");
var IS_INSTALLED = gql(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    query IsFileManagerInstalled {\n        fileManager {\n            version\n        }\n    }\n"])));
var INSTALL = gql(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n    mutation InstallFileManager($srcPrefix: String) {\n        fileManager {\n            install(srcPrefix: $srcPrefix) {\n                data\n                error {\n                    code\n                    message\n                }\n            }\n        }\n    }\n"])));

var FMInstaller = function FMInstaller(_ref) {
  var onInstalled = _ref.onInstalled;
  var client = useApolloClient();

  var _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      error = _useState2[0],
      setError = _useState2[1];

  useEffect(function () {
    client.mutate({
      mutation: INSTALL,
      variables: {
        srcPrefix: process.env.REACT_APP_API_URL + "/files"
      }
    }).then(function (_ref2) {
      var data = _ref2.data;
      var error = data.fileManager.install.error;

      if (error) {
        setError(error.message);
        return;
      } // Just so the user sees the actual message.


      setTimeout(onInstalled, 3000);
    });
  }, []);
  var label = error ? /*#__PURE__*/React.createElement(Alert, {
    title: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Something went wrong"]))),
    type: "danger"
  }, error) : t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Installing File Manager..."])));
  return /*#__PURE__*/React.createElement(SimpleForm, null, /*#__PURE__*/React.createElement(CircularProgress, {
    label: label
  }), /*#__PURE__*/React.createElement(SimpleFormContent, null, /*#__PURE__*/React.createElement(SimpleFormPlaceholder, null)));
};

var plugin = {
  name: "admin-installation-fm",
  type: "admin-installation",
  title: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["File Manager"]))),
  dependencies: ["admin-installation-security", "admin-installation-i18n"],
  secure: true,
  getInstalledVersion: function getInstalledVersion(_ref3) {
    return _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var client, _yield$client$query, data;

      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              client = _ref3.client;
              _context.next = 3;
              return client.query({
                query: IS_INSTALLED
              });

            case 3:
              _yield$client$query = _context.sent;
              data = _yield$client$query.data;
              return _context.abrupt("return", data.fileManager.version);

            case 6:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }))();
  },
  render: function render(_ref4) {
    var onInstalled = _ref4.onInstalled;
    return /*#__PURE__*/React.createElement(FMInstaller, {
      onInstalled: onInstalled
    });
  },
  upgrades: [{
    version: "5.0.0",
    getComponent: function getComponent() {
      return /*#__PURE__*/lazy(function () {
        return import("./upgrades/v5.0.0");
      });
    }
  }]
};
export default plugin;
