// @webiny/api-elasticsearch/plugins/operator/notContains.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ElasticsearchQueryBuilderOperatorNotContainsPlugin = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ElasticsearchQueryBuilderOperatorPlugin = require("../definition/ElasticsearchQueryBuilderOperatorPlugin");

var _normalize = require("../../normalize");

class ElasticsearchQueryBuilderOperatorNotContainsPlugin extends _ElasticsearchQueryBuilderOperatorPlugin.ElasticsearchQueryBuilderOperatorPlugin {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "name", "elasticsearch.queryBuilder.operator.notContains.default");
  }

  getOperator() {
    return "not_contains";
  }

  apply(query, params) {
    const {
      value,
      basePath
    } = params;
    query.must_not.push({
      query_string: {
        allow_leading_wildcard: true,
        fields: [basePath],
        query: (0, _normalize.normalizeValue)(value),
        default_operator: "and"
      }
    });
  }

}

exports.ElasticsearchQueryBuilderOperatorNotContainsPlugin = ElasticsearchQueryBuilderOperatorNotContainsPlugin;
