// @webiny/app-form-builder/admin/plugins/formDetails/formSubmissions/Block.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["children", "title"];
import * as React from "react";
import styled from "@emotion/styled";
import { css } from "emotion";
import { Typography } from "@webiny/ui/Typography";
import { Elevation } from "@webiny/ui/Elevation";
var BlockWrapper = /*#__PURE__*/styled("div", {
  target: "ef49oxh0",
  label: "BlockWrapper"
})({
  padding: "0 0 25px 0"
});
var titleStyle = /*#__PURE__*/css({
  display: "flex",
  alignItems: "center",
  color: "var(--mdc-theme-on-surface)",
  ".tooltip-content-wrapper": {
    lineHeight: "100%",
    svg: {
      height: 13,
      color: "var(--mdc-theme-on-surface)"
    }
  }
}, "label:titleStyle;");
var ElevationContent = /*#__PURE__*/styled("div", {
  target: "ef49oxh1",
  label: "ElevationContent"
})({
  padding: 20
});

var Block = function Block(_ref) {
  var children = _ref.children,
      title = _ref.title,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(BlockWrapper, props, /*#__PURE__*/React.createElement("h4", null, /*#__PURE__*/React.createElement(Typography, {
    className: titleStyle,
    use: "overline"
  }, title)), /*#__PURE__*/React.createElement(Elevation, {
    z: 2
  }, /*#__PURE__*/React.createElement(ElevationContent, null, children)));
};

export default Block;
