// @webiny/app-form-builder/admin/plugins/formDetails/formSubmissions/FormSubmissionsOverview.js
import * as React from "react";
import Block from "./Block";
import { Typography } from "@webiny/ui/Typography";
import styled from "@emotion/styled";
var StatBox = /*#__PURE__*/styled("div", {
  target: "e1v7hbwf0",
  label: "StatBox"
})({
  width: "33.33%",
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
  flexDirection: "column",
  color: "var(--mdc-theme-on-surface)"
});
var ContentWrapper = /*#__PURE__*/styled("div", {
  target: "e1v7hbwf1",
  label: "ContentWrapper"
})({
  flexDirection: "row",
  display: "flex",
  width: "100%",
  boxSizing: "border-box"
});
export var FormSubmissionsOverview = function FormSubmissionsOverview(_ref) {
  var form = _ref.form;
  return /*#__PURE__*/React.createElement(Block, {
    title: "Overview"
  }, /*#__PURE__*/React.createElement(ContentWrapper, null, /*#__PURE__*/React.createElement(StatBox, null, /*#__PURE__*/React.createElement(Typography, {
    use: "headline2"
  }, form.overallStats.submissions), /*#__PURE__*/React.createElement(Typography, {
    use: "overline"
  }, "Submissions")), /*#__PURE__*/React.createElement(StatBox, null, /*#__PURE__*/React.createElement(Typography, {
    use: "headline2"
  }, form.overallStats.views), /*#__PURE__*/React.createElement(Typography, {
    use: "overline"
  }, "Views")), /*#__PURE__*/React.createElement(StatBox, null, /*#__PURE__*/React.createElement(Typography, {
    use: "headline2"
  }, form.overallStats.conversionRate, "%"), /*#__PURE__*/React.createElement(Typography, {
    use: "overline"
  }, "Conversion Rate"))));
};
