// @webiny/app-form-builder/admin/plugins/formDetails/formSubmissions/index.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import * as React from "react";
import { renderPlugins } from "@webiny/app/plugins";
import { Tab } from "@webiny/ui/Tabs";
import styled from "@emotion/styled";
import { CircularProgress } from "@webiny/ui/Progress";
import { i18n } from "@webiny/app/i18n";
import { FormSubmissionsOverview } from "./FormSubmissionsOverview";
import { FormSubmissionsList } from "./FormSubmissionsList";
var t = i18n.namespace("FormsApp.FormDetails.PreviewContent");
var RenderBlock = /*#__PURE__*/styled("div", {
  target: "e1tsi7oi0",
  label: "RenderBlock"
})({
  padding: 25
});
export default [{
  name: "forms-form-details-revision-content-submissions",
  type: "forms-form-details-revision-content",
  render: function render(_ref) {
    var form = _ref.form,
        loading = _ref.loading,
        security = _ref.security;

    var _security$identity$ge = security.identity.getPermission("fb.form"),
        submissions = _security$identity$ge.submissions;

    if (typeof submissions !== "undefined" && submissions !== true) {
      return null;
    }

    return /*#__PURE__*/React.createElement(Tab, {
      label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Submissions"]))),
      disabled: loading,
      "data-testid": "fb.form-details.tab.submissions"
    }, /*#__PURE__*/React.createElement(RenderBlock, null, /*#__PURE__*/React.createElement("div", {
      style: {
        position: "relative"
      }
    }, loading && /*#__PURE__*/React.createElement(CircularProgress, null), form && renderPlugins("forms-form-details-submissions", {
      form: form
    }))));
  }
}, {
  name: "forms-form-details-submissions-overview",
  type: "forms-form-details-submissions",
  render: function render(_ref2) {
    var form = _ref2.form;
    return /*#__PURE__*/React.createElement(FormSubmissionsOverview, {
      form: form
    });
  }
}, {
  name: "forms-form-details-submissions-list",
  type: "forms-form-details-submissions",
  render: function render(_ref3) {
    var form = _ref3.form;
    return /*#__PURE__*/React.createElement(FormSubmissionsList, {
      form: form
    });
  }
}];
