// @webiny/ui/Accordion/AccordionItem.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import React, { useState, useCallback, useEffect } from "react";
import { ListItem, ListItemGraphic, ListItemMeta } from "../List";
import Transition from "react-transition-group/Transition";
import { Icon } from "../Icon";
import styled from "@emotion/styled";
import { css } from "emotion";
import { Typography } from "../Typography";
import { ReactComponent as UpArrow } from "@svgr/webpack!./icons/round-keyboard_arrow_up-24px.svg";
import { ReactComponent as DownArrow } from "@svgr/webpack!./icons/round-keyboard_arrow_down-24px.svg";
import classNames from "classnames";
var Content = /*#__PURE__*/styled("div", {
  target: "eypjd5u0",
  label: "Content"
})({
  width: "100%",
  borderRight: "1px solid var(--mdc-theme-background)",
  borderBottom: "1px solid var(--mdc-theme-background)",
  borderLeft: "1px solid var(--mdc-theme-background)",
  boxSizing: "border-box"
});
var listItem = /*#__PURE__*/css({
  padding: "15px 20px",
  cursor: "pointer",
  borderBottom: "1px solid var(--mdc-theme-background)",
  "&:last-child": {
    borderBottom: "none"
  },
  ".mdc-list-item__graphic": {
    marginRight: 20
  }
}, "label:listItem;");
var ListItemTitle = /*#__PURE__*/styled("div", {
  target: "eypjd5u1",
  label: "ListItemTitle"
})({
  fontWeight: 600,
  marginBottom: 5
});
var ListItemDescription = /*#__PURE__*/styled("div", {
  target: "eypjd5u2",
  label: "ListItemDescription"
})({});
var TitleContent = /*#__PURE__*/styled("div", {
  target: "eypjd5u3",
  label: "TitleContent"
})({
  display: "flex",
  flexDirection: "column"
});
var openedState = /*#__PURE__*/css({
  backgroundColor: "var(--mdc-theme-on-background)"
}, "label:openedState;");
var duration = 150;
var defaultStyle = {
  transition: "all ".concat(duration, "ms ease-in-out"),
  opacity: 0,
  height: 0,
  pointerEvents: "none",
  overflow: "hidden"
};
var transitionStyles = {
  entering: {
    opacity: 0,
    height: 0,
    padding: "20px 20px 20px 65px",
    pointerEvents: "auto",
    overflow: "initial"
  },
  entered: {
    opacity: 1,
    height: "auto",
    padding: "20px 20px 20px 65px",
    pointerEvents: "auto",
    overflow: "initial"
  },
  exiting: {
    height: "auto",
    padding: "20px 20px 20px 65px",
    pointerEvents: "auto",
    overflow: "initial"
  }
};

var AccordionItem = function AccordionItem(props) {
  var _useState = useState(props.open ? props.open : false),
      _useState2 = _slicedToArray(_useState, 2),
      open = _useState2[0],
      setState = _useState2[1];

  var toggleState = useCallback(function () {
    setState(!open);
  }, [open]);
  useEffect(function () {
    setState(props.open);
  }, [props.open]);
  return /*#__PURE__*/React.createElement("div", {
    className: classNames("webiny-ui-accordion-item", props.className)
  }, /*#__PURE__*/React.createElement(ListItem, {
    className: classNames(listItem, _defineProperty({}, openedState, open), "webiny-ui-accordion-item__list-item"),
    onClick: toggleState,
    "data-testid": props["data-testid"]
  }, props.icon && /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
    icon: props.icon
  })), /*#__PURE__*/React.createElement(TitleContent, {
    className: "webiny-ui-accordion-item__title"
  }, /*#__PURE__*/React.createElement(ListItemTitle, null, props.title), props.description && /*#__PURE__*/React.createElement(ListItemDescription, null, /*#__PURE__*/React.createElement(Typography, {
    use: "subtitle2"
  }, props.description))), /*#__PURE__*/React.createElement(ListItemMeta, null, /*#__PURE__*/React.createElement(Icon, {
    icon: !open ? /*#__PURE__*/React.createElement(DownArrow, null) : /*#__PURE__*/React.createElement(UpArrow, null)
  }))), /*#__PURE__*/React.createElement(Transition, {
    in: open,
    timeout: duration
  }, function (state) {
    return /*#__PURE__*/React.createElement(Content, {
      style: _objectSpread(_objectSpread({}, defaultStyle), transitionStyles[state]),
      className: "webiny-ui-accordion-item__content"
    }, props.children);
  }));
};

export { AccordionItem };
