// @webiny/app-headless-cms/admin/components/ContentModelEditor/Editor.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import React, { useRef, useState } from "react";
import { Prompt } from "@webiny/react-router";
import styled from "@emotion/styled";
import { css } from "emotion";
import { i18n } from "@webiny/app/i18n";
import { CircularProgress } from "@webiny/ui/Progress";
import { LeftPanel, RightPanel, SplitView } from "@webiny/app-admin/components/SplitView";
import { Icon } from "@webiny/ui/Icon";
import { Typography } from "@webiny/ui/Typography";
import { Tab, Tabs } from "@webiny/ui/Tabs";
import { ReactComponent as FormIcon } from "@svgr/webpack!./icons/round-assignment-24px.svg";
import { FieldsSidebar } from "./FieldsSidebar";
import { FieldEditor } from "../FieldEditor";
import { PreviewTab } from "./PreviewTab";
import Header from "./Header";
import DragPreview from "../DragPreview";
import { useContentModelEditor } from "./useContentModelEditor";
var t = i18n.ns("app-headless-cms/admin/editor");
var prompt = t(_templateObject || (_templateObject = _taggedTemplateLiteral(["There are some unsaved changes! Are you sure you want to navigate away and discard all changes?"])));
var ContentContainer = /*#__PURE__*/styled("div", {
  target: "epn2fgg0",
  label: "ContentContainer"
})({
  paddingTop: 65
});
export var EditContainer = /*#__PURE__*/styled("div", {
  target: "epn2fgg1",
  label: "EditContainer"
})({
  padding: 40,
  position: "relative"
});
var LeftBarTitle = /*#__PURE__*/styled("div", {
  target: "epn2fgg2",
  label: "LeftBarTitle"
})({
  borderBottom: "1px solid var(--mdc-theme-on-background)",
  display: "flex",
  alignItems: "center",
  padding: 25,
  color: "var(--mdc-theme-on-surface)"
});
var titleIcon = /*#__PURE__*/css({
  height: 24,
  marginRight: 15,
  color: "var(--mdc-theme-primary)"
}, "label:titleIcon;");
var LeftBarFieldList = /*#__PURE__*/styled("div", {
  target: "epn2fgg3",
  label: "LeftBarFieldList"
})({
  padding: 40,
  overflow: "auto",
  height: "calc(100vh - 250px)"
});
var formTabs = /*#__PURE__*/css({
  "&.webiny-ui-tabs": {
    ".webiny-ui-tabs__tab-bar": {
      backgroundColor: "var(--mdc-theme-surface)"
    }
  }
}, "label:formTabs;");
export var Editor = function Editor() {
  var _useContentModelEdito = useContentModelEditor(),
      data = _useContentModelEdito.data,
      setData = _useContentModelEdito.setData,
      isPristine = _useContentModelEdito.isPristine;

  var tabsRef = useRef(null);

  var _useState = useState(0),
      _useState2 = _slicedToArray(_useState, 2),
      activeTabIndex = _useState2[0],
      setActiveTabIndex = _useState2[1];

  var onChange = function onChange(_ref) {
    var fields = _ref.fields,
        layout = _ref.layout;
    setData(function (data) {
      return _objectSpread(_objectSpread({}, data), {}, {
        fields: fields,
        layout: layout
      });
    });
  };

  if (!data) {
    return /*#__PURE__*/React.createElement(CircularProgress, {
      label: "Loading content model..."
    });
  }

  return /*#__PURE__*/React.createElement("div", {
    className: "content-model-editor"
  }, /*#__PURE__*/React.createElement(Prompt, {
    when: !isPristine,
    message: prompt
  }), /*#__PURE__*/React.createElement(Header, null), /*#__PURE__*/React.createElement(ContentContainer, null, /*#__PURE__*/React.createElement(SplitView, null, /*#__PURE__*/React.createElement(LeftPanel, {
    span: 4
  }, /*#__PURE__*/React.createElement(LeftBarTitle, null, /*#__PURE__*/React.createElement(Icon, {
    className: titleIcon,
    icon: /*#__PURE__*/React.createElement(FormIcon, null)
  }), /*#__PURE__*/React.createElement(Typography, {
    use: "headline6"
  }, "Fields")), /*#__PURE__*/React.createElement(LeftBarFieldList, null, /*#__PURE__*/React.createElement(FieldsSidebar, {
    onFieldDragStart: function onFieldDragStart() {
      if (tabsRef.current) {
        tabsRef.current.switchTab(0);
      }
    }
  }))), /*#__PURE__*/React.createElement(RightPanel, {
    span: 8
  }, /*#__PURE__*/React.createElement(Tabs, {
    className: formTabs,
    ref: tabsRef,
    onActivate: function onActivate(e) {
      return setActiveTabIndex(e);
    }
  }, /*#__PURE__*/React.createElement(Tab, {
    label: "Edit",
    "data-testid": "cms.editor.tab.edit"
  }, /*#__PURE__*/React.createElement(EditContainer, null, /*#__PURE__*/React.createElement(FieldEditor, {
    fields: data.fields,
    layout: data.layout,
    onChange: onChange
  }))), /*#__PURE__*/React.createElement(Tab, {
    label: "Preview",
    "data-testid": "cms.editor.tab.preview"
  }, /*#__PURE__*/React.createElement(PreviewTab, {
    activeTab: activeTabIndex === 1
  })))))), /*#__PURE__*/React.createElement(DragPreview, null));
};
