// @webiny/app-headless-cms/admin/components/ContentModelEditor/PreviewTab.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import React from "react";
import { css } from "emotion";
import { Elevation } from "@webiny/ui/Elevation";
import { i18n } from "@webiny/app/i18n";
import { useContentModelEditor } from "./useContentModelEditor";
import { ContentEntryFormPreview } from "../ContentEntryForm/ContentEntryFormPreview";
var t = i18n.ns("app-headless-cms/admin/components/editor/tabs/preview");
var formPreviewWrapper = /*#__PURE__*/css({
  padding: 40,
  margin: 40,
  boxSizing: "border-box"
}, "label:formPreviewWrapper;");
var style = {
  noFieldsMessage: /*#__PURE__*/css({
    textAlign: "center"
  }, "label:noFieldsMessage;")
};
export var PreviewTab = function PreviewTab(_ref) {
  var activeTab = _ref.activeTab;

  var _useContentModelEdito = useContentModelEditor(),
      data = _useContentModelEdito.data;

  return /*#__PURE__*/React.createElement(Elevation, {
    z: 1,
    className: formPreviewWrapper
  }, data.fields && data.fields.length && activeTab ? /*#__PURE__*/React.createElement(ContentEntryFormPreview, {
    contentModel: data
  }) : /*#__PURE__*/React.createElement("div", {
    className: style.noFieldsMessage
  }, t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Before previewing the form, please add at least one field to the content model."])))));
};
