// @webiny/api-page-builder/graphql/graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _menus = _interopRequireDefault(require("./graphql/menus.gql"));

var _pages = _interopRequireDefault(require("./graphql/pages.gql"));

var _pageElements = _interopRequireDefault(require("./graphql/pageElements.gql"));

var _categories = _interopRequireDefault(require("./graphql/categories.gql"));

var _settings = _interopRequireDefault(require("./graphql/settings.gql"));

var _install = _interopRequireDefault(require("./graphql/install.gql"));

const emptyResolver = () => ({});

var _default = [{
  type: "graphql-schema",
  schema: {
    typeDefs:
    /* GraphQL */
    `
                input PbFileInput {
                    id: ID!
                    src: String!
                }

                type PbFile {
                    id: ID!
                    src: String!
                }

                type PbCreatedBy {
                    id: ID
                    displayName: String
                }

                type PbError {
                    code: String
                    message: String
                    data: JSON
                }

                type PbDeleteResponse {
                    data: Boolean
                    error: PbError
                }

                type PbQuery {
                    pageBuilder: PbQuery
                }

                type PbMutation {
                    pageBuilder: PbMutation
                }

                extend type Query {
                    pageBuilder: PbQuery
                }

                extend type Mutation {
                    pageBuilder: PbMutation
                }
            `,
    resolvers: {
      Query: {
        pageBuilder: emptyResolver
      },
      Mutation: {
        pageBuilder: emptyResolver
      }
    }
  }
}, _menus.default, _categories.default, _pages.default, _pageElements.default, _settings.default, _install.default];
exports.default = _default;
