// @webiny/app-admin/ui/elements/form/HiddenElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { FormFieldElement } from "./FormFieldElement";
export var HiddenElement = /*#__PURE__*/function (_FormFieldElement) {
  _inherits(HiddenElement, _FormFieldElement);

  var _super = _createSuper(HiddenElement);

  function HiddenElement(id, config) {
    var _this;

    _classCallCheck(this, HiddenElement);

    _this = _super.call(this, id, config);

    _this.applyPlugins(HiddenElement);

    return _this;
  }

  _createClass(HiddenElement, [{
    key: "render",
    value: function render(props) {
      var _this2 = this;

      if (!props.formProps) {
        throw Error("HiddenElement must be placed inside of a FormElement.");
      }

      var Bind = props.formProps.Bind;
      return /*#__PURE__*/React.createElement(Bind, {
        name: this.getName(),
        validators: this.getValidators(props),
        defaultValue: this.getDefaultValue(props),
        beforeChange: function beforeChange(value, cb) {
          return _this2.onBeforeChange(value, cb);
        },
        afterChange: function afterChange(value, form) {
          return _this2.onAfterChange(value, form);
        }
      });
    }
  }]);

  return HiddenElement;
}(FormFieldElement);
