// @webiny/app-admin/ui/elements/form/PasswordElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { InputElement } from "./InputElement";
import { Input } from "@webiny/ui/Input";
export var PasswordElement = /*#__PURE__*/function (_InputElement) {
  _inherits(PasswordElement, _InputElement);

  var _super = _createSuper(PasswordElement);

  function PasswordElement() {
    _classCallCheck(this, PasswordElement);

    return _super.apply(this, arguments);
  }

  _createClass(PasswordElement, [{
    key: "render",
    value: function render(props) {
      var _this = this;

      if (!props.formProps) {
        throw Error("PasswordElement must be placed inside of a FormElement.");
      }

      var Bind = props.formProps.Bind;
      return /*#__PURE__*/React.createElement(Bind, {
        name: this.getName(),
        validators: this.getValidators(props),
        defaultValue: this.getDefaultValue(props),
        beforeChange: function beforeChange(value, cb) {
          return _this.onBeforeChange(value, cb);
        },
        afterChange: function afterChange(value, form) {
          return _this.onAfterChange(value, form);
        }
      }, /*#__PURE__*/React.createElement(Input, {
        type: "password",
        label: this.getLabel(props),
        placeholder: this.getPlaceholder(props),
        disabled: this.getIsDisabled(props),
        description: this.getDescription(props)
      }));
    }
  }]);

  return PasswordElement;
}(InputElement);
