// @webiny/app-form-builder/admin/plugins/menus.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3;

import React from "react";
import { ReactComponent as PagesIcon } from "@svgr/webpack!@webiny/app-page-builder/admin/assets/round-ballot-24px.svg";
import { i18n } from "@webiny/app/i18n";
import { useSecurity } from "@webiny/app-security";
var t = i18n.ns("app-form-builder/admin/menus");

var FormBuilderMenu = function FormBuilderMenu(_ref) {
  var Menu = _ref.Menu,
      Section = _ref.Section,
      Item = _ref.Item;

  var _useSecurity = useSecurity(),
      identity = _useSecurity.identity;

  if (!identity.getPermission("fb.form")) {
    return null;
  }

  return /*#__PURE__*/React.createElement(Menu, {
    name: "app-form-builder",
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Form Builder"]))),
    icon: /*#__PURE__*/React.createElement(PagesIcon, null)
  }, /*#__PURE__*/React.createElement(Section, {
    label: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Forms"])))
  }, /*#__PURE__*/React.createElement(Item, {
    label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Forms"]))),
    path: "/form-builder/forms"
  })));
};

var plugin = {
  type: "admin-menu",
  name: "admin-menu-form-builder",
  render: function render(props) {
    return /*#__PURE__*/React.createElement(FormBuilderMenu, props);
  }
};
export default plugin;
