// @webiny/app-form-builder/admin/plugins/routes.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["children", "label"];
import React, { Suspense, lazy } from "react";
import Helmet from "react-helmet";
import { SecureRoute } from "@webiny/app-security/components";
import { CircularProgress } from "@webiny/ui/Progress";
import { Route } from "@webiny/react-router";
import { AdminLayout } from "@webiny/app-admin/components/AdminLayout";
import { Permission } from "./constants";

var Loader = function Loader(_ref) {
  var children = _ref.children,
      label = _ref.label,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(Suspense, {
    fallback: /*#__PURE__*/React.createElement(CircularProgress, {
      label: label
    })
  }, /*#__PURE__*/React.cloneElement(children, props));
};

var FormEditor = /*#__PURE__*/lazy(function () {
  return import("../views/Editor");
});
var Forms = /*#__PURE__*/lazy(function () {
  return import("../views/Forms/Forms");
});
var plugins = [{
  name: "route-form-editors-editor",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/form-builder/forms/:id",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: Permission.Form
      }, /*#__PURE__*/React.createElement(Helmet, null, /*#__PURE__*/React.createElement("title", null, "Edit form")), /*#__PURE__*/React.createElement(Loader, {
        label: "Loading editor..."
      }, /*#__PURE__*/React.createElement(FormEditor, null)));
    }
  })
}, {
  name: "route-form-editors",
  type: "route",
  route: /*#__PURE__*/React.createElement(Route, {
    exact: true,
    path: "/form-builder/forms",
    render: function render() {
      return /*#__PURE__*/React.createElement(SecureRoute, {
        permission: Permission.Form
      }, /*#__PURE__*/React.createElement(AdminLayout, null, /*#__PURE__*/React.createElement(Helmet, {
        title: "Form Builder"
      }), /*#__PURE__*/React.createElement(Loader, {
        label: "Loading view..."
      }, /*#__PURE__*/React.createElement(Forms, null))));
    }
  })
}];
export default plugins;
