// @webiny/app-headless-cms/admin/plugins/permissionRenderer/components/ContentModelList.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React from "react";
import { css } from "emotion";
import groupBy from "lodash/groupBy";
import get from "lodash/get";
import { Typography } from "@webiny/ui/Typography";
import { Checkbox } from "@webiny/ui/Checkbox";
var groupStyles = /*#__PURE__*/css({
  marginLeft: 20
}, "label:groupStyles;");
var labelStyles = /*#__PURE__*/css({
  display: "block",
  color: "var(--mdc-theme-text-secondary-on-light)",
  fontWeight: "bold"
}, "label:labelStyles;");

var ContentModelList = function ContentModelList(_ref) {
  var items = _ref.items,
      onChange = _ref.onChange,
      getValue = _ref.getValue;
  var list = Object.entries(groupBy(items.map(function (i) {
    return _objectSpread(_objectSpread({}, i), {}, {
      groupName: get(i, "group.label")
    });
  }), "groupName"));
  return /*#__PURE__*/React.createElement(React.Fragment, null, list.map(function (_ref2) {
    var _ref3 = _slicedToArray(_ref2, 2),
        key = _ref3[0],
        value = _ref3[1];

    return /*#__PURE__*/React.createElement("div", {
      key: key,
      className: groupStyles
    }, /*#__PURE__*/React.createElement(Typography, {
      use: "caption",
      className: labelStyles
    }, key), value.map(function (_ref4) {
      var id = _ref4.id,
          label = _ref4.label;
      return /*#__PURE__*/React.createElement("div", {
        key: id
      }, /*#__PURE__*/React.createElement(Checkbox, {
        key: id,
        label: label,
        value: getValue(id),
        onChange: onChange(id)
      }));
    }));
  }));
};

export default ContentModelList;
