// @webiny/app-headless-cms/admin/plugins/permissionRenderer/components/ContentModelGroupPermission.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8;

import React, { useCallback, useEffect } from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Select } from "@webiny/ui/Select";
import { i18n } from "@webiny/app/i18n";
import { Elevation } from "@webiny/ui/Elevation";
import { Typography } from "@webiny/ui/Typography";
import { PermissionSelector, PermissionSelectorWrapper } from "./PermissionSelector";
import { useCmsData } from "./useCmsData";
var t = i18n.ns("app-headless-cms/admin/plugins/permissionRenderer");

var ContentModelGroupPermission = function ContentModelGroupPermission(_ref) {
  var Bind = _ref.Bind,
      data = _ref.data,
      entity = _ref.entity,
      title = _ref.title,
      locales = _ref.locales,
      setValue = _ref.setValue,
      form = _ref.form;
  var modelsGroups = useCmsData(locales);
  var getItems = useCallback(function (code) {
    return modelsGroups[code]["groups"];
  }, [modelsGroups]);
  useEffect(function () {
    // Let's set default values for "accessScopes"
    if (data.endpoints.length > 0 && !data["".concat(entity, "AccessScope")] && form.onChangeFns["".concat(entity, "AccessScope")]) {
      setValue("".concat(entity, "AccessScope"), "full");
    }
  }, [data, setValue, form]);
  var disabledPrimaryActions = [undefined, "own", "no"].includes(data["".concat(entity, "AccessScope")]) || !data.endpoints.includes("manage");
  return /*#__PURE__*/React.createElement(Elevation, {
    z: 1,
    style: {
      marginTop: 10
    }
  }, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "overline"
  }, title)), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Grid, {
    style: {
      padding: 0,
      paddingBottom: 24
    }
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "".concat(entity, "AccessScope")
  }, /*#__PURE__*/React.createElement(Select, {
    label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Access Scope"])))
  }, /*#__PURE__*/React.createElement("option", {
    value: "full"
  }, t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["All groups"])))), /*#__PURE__*/React.createElement("option", {
    value: "groups"
  }, t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Only specific groups"])))), data.endpoints.includes("manage") && /*#__PURE__*/React.createElement("option", {
    value: "own"
  }, t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Only groups created by the user"]))))))), data["".concat(entity, "AccessScope")] === "groups" && /*#__PURE__*/React.createElement(PermissionSelectorWrapper, null, /*#__PURE__*/React.createElement(PermissionSelector, {
    locales: locales,
    Bind: Bind,
    entity: entity,
    selectorKey: "groups",
    getItems: getItems
  })), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "".concat(entity, "RWD")
  }, /*#__PURE__*/React.createElement(Select, {
    label: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Primary Actions"]))),
    disabled: disabledPrimaryActions
  }, /*#__PURE__*/React.createElement("option", {
    value: "r"
  }, t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Read"])))), data.endpoints.includes("manage") ? /*#__PURE__*/React.createElement("option", {
    value: "rw"
  }, t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Read, write"])))) : null, data.endpoints.includes("manage") ? /*#__PURE__*/React.createElement("option", {
    value: "rwd"
  }, t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["Read, write, delete"])))) : null)))))));
};

export default ContentModelGroupPermission;
