// @webiny/app-form-builder/admin/plugins/formDetails/formRevisions/RevisionsList.js
import React from "react";
import { List } from "@webiny/ui/List";
import Revision from "./Revision";
import { Elevation } from "@webiny/ui/Elevation";
import { css } from "emotion";
import { CircularProgress } from "@webiny/ui/Progress";
var listWrapper = /*#__PURE__*/css({
  margin: 25,
  display: "flex",
  flexDirection: "column",
  overflow: "auto",
  maxHeight: "calc(100vh - 160px)",
  ".mdc-list .mdc-list-item": {
    borderBottom: "1px solid var(--mdc-theme-on-background)"
  },
  ".mdc-list .mdc-list-item:last-child": {
    borderBottom: "none"
  }
}, "label:listWrapper;");

var RevisionsList = function RevisionsList(_ref) {
  var form = _ref.form,
      revisions = _ref.revisions,
      loading = _ref.loading;
  return form && /*#__PURE__*/React.createElement(Elevation, {
    className: listWrapper,
    z: 2
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      position: "relative"
    }
  }, loading && /*#__PURE__*/React.createElement(CircularProgress, null), /*#__PURE__*/React.createElement(List, {
    nonInteractive: true,
    twoLine: true,
    "data-testid": "fb.form-details.tab.revisions.content-list"
  }, Array.isArray(revisions) ? revisions.map(function (rev) {
    return /*#__PURE__*/React.createElement(Revision, {
      form: form,
      revision: rev,
      key: rev.id
    });
  }) : null)));
};

export default RevisionsList;
