// @webiny/app-security-admin-users/ui/views/ApiKeys/ApiKeyForm.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useCallback } from "react";
import { useMutation, useQuery } from "@apollo/react-hooks";
import get from "lodash/get";
import { useRouter } from "@webiny/react-router";
import { i18n } from "@webiny/app/i18n";
import { Form } from "@webiny/form";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Input } from "@webiny/ui/Input";
import { ButtonDefault, ButtonIcon, ButtonPrimary, CopyButton } from "@webiny/ui/Button";
import { CircularProgress } from "@webiny/ui/Progress";
import { FormElementMessage } from "@webiny/ui/FormElementMessage";
import { Permissions } from "@webiny/app-admin/components/Permissions";
import { validation } from "@webiny/validation";
import { SimpleForm, SimpleFormFooter, SimpleFormContent, SimpleFormHeader } from "@webiny/app-admin/components/SimpleForm";
import { Typography } from "@webiny/ui/Typography";
import { useSnackbar } from "@webiny/app-admin/hooks/useSnackbar";
import { pickDataForAPI } from "./utils";
import * as GQL from "./graphql";
import { SnackbarAction } from "@webiny/ui/Snackbar";
import isEmpty from "lodash/isEmpty";
import EmptyView from "@webiny/app-admin/components/EmptyView";
import { ReactComponent as AddIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/add-18px.svg";
import styled from "@emotion/styled";
var t = i18n.ns("app-security-admin-users/admin/api-keys/form");
var ButtonWrapper = /*#__PURE__*/styled("div", {
  target: "ehh5xcd0",
  label: "ButtonWrapper"
})({
  display: "flex",
  justifyContent: "space-between"
});

var ApiKeyForm = function ApiKeyForm() {
  var _useRouter = useRouter(),
      location = _useRouter.location,
      history = _useRouter.history;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var newEntry = new URLSearchParams(location.search).get("new") === "true";
  var id = new URLSearchParams(location.search).get("id");
  var getQuery = useQuery(GQL.READ_API_KEY, {
    variables: {
      id: id
    },
    skip: !id,
    onCompleted: function onCompleted(data) {
      if (!data) {
        return;
      }

      var error = data.security.apiKey.error;

      if (error) {
        history.push("/security/api-keys");
        showSnackbar(error.message);
      }
    }
  });

  var _useMutation = useMutation(GQL.CREATE_API_KEY, {
    refetchQueries: [{
      query: GQL.LIST_API_KEYS
    }]
  }),
      _useMutation2 = _slicedToArray(_useMutation, 2),
      create = _useMutation2[0],
      createMutation = _useMutation2[1];

  var _useMutation3 = useMutation(GQL.UPDATE_API_KEY, {
    refetchQueries: [{
      query: GQL.LIST_API_KEYS
    }]
  }),
      _useMutation4 = _slicedToArray(_useMutation3, 2),
      update = _useMutation4[0],
      updateMutation = _useMutation4[1];

  var loading = [getQuery, createMutation, updateMutation].find(function (item) {
    return item.loading;
  });
  var onSubmit = useCallback( /*#__PURE__*/function () {
    var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(data) {
      var isUpdate, _ref2, _ref3, operation, args, response, error, id;

      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              if (!(!data.permissions || !data.permissions.length)) {
                _context.next = 3;
                break;
              }

              showSnackbar(t(_templateObject || (_templateObject = _taggedTemplateLiteral(["You must configure permissions before saving!"]))), {
                timeout: 60000,
                dismissesOnAction: true,
                action: /*#__PURE__*/React.createElement(SnackbarAction, {
                  label: "OK"
                })
              });
              return _context.abrupt("return");

            case 3:
              isUpdate = data.createdOn;
              _ref2 = isUpdate ? [update, {
                variables: {
                  id: data.id,
                  data: pickDataForAPI(data)
                }
              }] : [create, {
                variables: {
                  data: pickDataForAPI(data)
                }
              }], _ref3 = _slicedToArray(_ref2, 2), operation = _ref3[0], args = _ref3[1];
              _context.next = 7;
              return operation(args);

            case 7:
              response = _context.sent;
              error = response.data.security.apiKey.error;

              if (!error) {
                _context.next = 11;
                break;
              }

              return _context.abrupt("return", showSnackbar(error.message));

            case 11:
              id = response.data.security.apiKey.data.id;
              !isUpdate && history.push("/security/api-keys?id=".concat(id));
              showSnackbar(t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["API key saved successfully."]))));

            case 14:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function (_x) {
      return _ref.apply(this, arguments);
    };
  }(), [id]);
  var data = get(getQuery, "data.security.apiKey.data", {});
  var showEmptyView = !newEntry && !loading && isEmpty(data); // Render "No content" selected view.

  if (showEmptyView) {
    return /*#__PURE__*/React.createElement(EmptyView, {
      title: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Click on the left side list to display API key details or create a..."]))),
      action: /*#__PURE__*/React.createElement(ButtonDefault, {
        "data-testid": "new-record-button",
        onClick: function onClick() {
          return history.push("/security/api-keys?new=true");
        }
      }, /*#__PURE__*/React.createElement(ButtonIcon, {
        icon: /*#__PURE__*/React.createElement(AddIcon, null)
      }), " ", t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["New API Key"]))))
    });
  }

  return /*#__PURE__*/React.createElement(Form, {
    data: data,
    onSubmit: onSubmit
  }, function (_ref4) {
    var data = _ref4.data,
        form = _ref4.form,
        Bind = _ref4.Bind;
    return /*#__PURE__*/React.createElement(SimpleForm, null, loading && /*#__PURE__*/React.createElement(CircularProgress, null), /*#__PURE__*/React.createElement(SimpleFormHeader, {
      title: data.name ? data.name : "Untitled"
    }), /*#__PURE__*/React.createElement(SimpleFormContent, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "name",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      label: t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Name"])))
    })))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "description",
      validators: validation.create("required")
    }, /*#__PURE__*/React.createElement(Input, {
      label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Description"]))),
      rows: 4
    })))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(Typography, {
      use: "subtitle1"
    }, t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Token"])))), data.token ? /*#__PURE__*/React.createElement("div", {
      style: {
        background: "var(--mdc-theme-background)",
        padding: "8px",
        paddingLeft: "16px"
      }
    }, /*#__PURE__*/React.createElement("span", {
      style: {
        lineHeight: "48px",
        verticalAlign: "middle"
      }
    }, data.token), /*#__PURE__*/React.createElement("span", {
      style: {
        position: "absolute",
        right: "32px"
      }
    }, /*#__PURE__*/React.createElement(CopyButton, {
      value: data.token,
      onCopy: function onCopy() {
        return showSnackbar("Successfully copied!");
      }
    }))) : /*#__PURE__*/React.createElement(FormElementMessage, null, "Your token will be shown once you submit the form.")))), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Typography, {
      use: "subtitle1"
    }, t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["Permissions"]))))), /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Bind, {
      name: "permissions",
      defaultValue: []
    }, function (bind) {
      return /*#__PURE__*/React.createElement(Permissions, Object.assign({
        id: data.id || "new"
      }, bind));
    })))), /*#__PURE__*/React.createElement(SimpleFormFooter, null, /*#__PURE__*/React.createElement(ButtonWrapper, null, /*#__PURE__*/React.createElement(ButtonDefault, {
      onClick: function onClick() {
        return history.push("/security/api-keys");
      }
    }, t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Cancel"])))), /*#__PURE__*/React.createElement(ButtonPrimary, {
      onClick: form.submit
    }, t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["Save API key"])))))));
  });
};

export default ApiKeyForm;
