// @webiny/app-headless-cms/admin/components/FieldEditor/Styled.js
import { css } from "emotion";
import styled from "@emotion/styled";
export var RowContainer = /*#__PURE__*/styled("div", {
  target: "e1pk1rsi0",
  label: "RowContainer"
})({
  position: "relative",
  display: "flex",
  flexDirection: "column",
  marginBottom: 20,
  borderRadius: 2,
  backgroundColor: "var(--mdc-theme-surface)",
  border: "1px solid var(--mdc-theme-on-background)",
  boxShadow: "var(--mdc-theme-on-background) 1px 1px 1px, var(--mdc-theme-on-background) 1px 1px 2px"
});
export var Row = /*#__PURE__*/styled("div", {
  target: "e1pk1rsi1",
  label: "Row"
})({
  display: "flex",
  flexDirection: "row",
  backgroundColor: "var(--mdc-theme-surface)",
  paddingLeft: 40,
  paddingRight: 10,
  position: "relative"
});
export var fieldContainer = /*#__PURE__*/css({
  position: "relative",
  flex: "1 100%",
  backgroundColor: "var(--mdc-theme-background)",
  padding: "0 15px",
  margin: 10,
  borderRadius: 2,
  border: "1px solid var(--mdc-theme-on-background)",
  transition: "box-shadow 225ms",
  color: "var(--mdc-theme-on-surface)",
  cursor: "grab",
  "&:hover": {
    boxShadow: "var(--mdc-theme-on-background) 1px 1px 1px, var(--mdc-theme-on-background) 1px 1px 2px"
  }
}, "label:fieldContainer;");
export var rowHandle = /*#__PURE__*/css({
  width: 30,
  cursor: "grab",
  position: "absolute",
  left: 10,
  top: "50%",
  transform: "translateY(-50%)",
  zIndex: 1,
  color: "var(--mdc-theme-on-surface)"
}, "label:rowHandle;");
export var fieldHandle = /*#__PURE__*/css({
  cursor: "grab"
}, "label:fieldHandle;");
