// @webiny/app-page-builder/editor/plugins/elements/imagesList/File.js
import React from "react";
import { css } from "emotion";
import { Image } from "@webiny/app/components";
import { ReactComponent as RemoveIcon } from "@svgr/webpack!../../../../admin/assets/round-close-24px.svg";
var COMPONENT_WIDTH = 176;
var COMPONENT_HEIGHT = 176;
var styles = /*#__PURE__*/css({
  cursor: "move",
  display: "inline-block",
  position: "relative",
  zIndex: 1,
  margin: 10,
  width: "100%",
  maxWidth: COMPONENT_WIDTH,
  border: "1px solid var(--mdc-theme-on-background)",
  borderRadius: 2,
  "> .body": {
    width: COMPONENT_WIDTH,
    height: COMPONENT_HEIGHT,
    overflow: "hidden",
    "--icon-color": "var(--mdc-theme-on-background)",
    "&:hover": {
      "--icon-color": "var(--mdc-theme-secondary)"
    },
    ".infoIcon": {
      color: "var(--icon-color)",
      position: "absolute",
      top: 4,
      right: 4,
      zIndex: 10
    },
    ".filePreview": {
      textAlign: "center",
      position: "relative",
      backgroundColor: "#fff",
      width: "100%",
      height: "100%",
      ".clickableArea": {
        position: "absolute",
        top: 30,
        left: 0,
        width: "100%",
        height: 170,
        zIndex: 2
      }
    }
  },
  "> .label": {
    padding: "15px 10px",
    whiteSpace: "nowrap",
    overflow: "hidden",
    textOverflow: "ellipsis",
    fontSize: "0.8rem",
    color: "var(--mdc-theme-on-surface)",
    backgroundColor: "var(--mdc-theme-on-background)"
  }
}, "label:styles;");
var imageStyles = /*#__PURE__*/css({
  maxHeight: COMPONENT_HEIGHT,
  maxWidth: COMPONENT_WIDTH,
  width: "auto",
  position: "absolute",
  top: "50%",
  left: "50%",
  transform: "translateX(-50%) translateY(-50%)"
}, "label:imageStyles;");
export default function File(props) {
  var file = props.file,
      onSelect = props.onSelect,
      onRemove = props.onRemove;
  return /*#__PURE__*/React.createElement("div", {
    className: styles
  }, /*#__PURE__*/React.createElement("div", {
    className: "body"
  }, /*#__PURE__*/React.createElement("div", {
    className: "infoIcon"
  }, /*#__PURE__*/React.createElement(RemoveIcon, {
    onClick: onRemove
  })), /*#__PURE__*/React.createElement("div", {
    className: "filePreview"
  }, /*#__PURE__*/React.createElement("div", {
    className: "clickableArea",
    onClick: onSelect
  }), /*#__PURE__*/React.createElement(Image, {
    className: imageStyles,
    src: file.src,
    alt: file.name,
    transform: {
      width: 300
    }
  }))), /*#__PURE__*/React.createElement("div", {
    className: "label"
  }, file.name));
}
