// @webiny/app-page-builder/editor/plugins/elements/imagesList/ImagesListDesignSettings.js
import * as React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { plugins } from "@webiny/plugins";
import Accordion from "../../elementSettings/components/Accordion";
import Wrapper from "../../elementSettings/components/Wrapper";
import SelectField from "../../elementSettings/components/SelectField";
import { ButtonContainer, classes, SimpleButton } from "../../elementSettings/components/StyledComponents";

var ImagesListDesignSettings = function ImagesListDesignSettings(_ref) {
  var Bind = _ref.Bind,
      submit = _ref.submit;
  var components = plugins.byType("pb-page-element-images-list-component");
  return /*#__PURE__*/React.createElement(Accordion, {
    title: "Design",
    defaultValue: true
  }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Wrapper, {
    label: "Component",
    containerClassName: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "component",
    defaultValue: components[0] ? components[0].componentName : null
  }, /*#__PURE__*/React.createElement(SelectField, {
    label: "Design",
    description: "Select a component to render the list"
  }, components.map(function (cmp) {
    return /*#__PURE__*/React.createElement("option", {
      key: cmp.name,
      value: cmp.componentName
    }, cmp.title);
  })))), /*#__PURE__*/React.createElement(Grid, {
    className: classes.simpleGrid
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
    onClick: submit
  }, "Save"))))));
};

export default ImagesListDesignSettings;
