// @webiny/app-headless-cms/admin/views/contentModels/ContentModelEditor.js
import React from "react";
import HTML5Backend from "react-dnd-html5-backend";
import { DndProvider } from "react-dnd";
import { Editor } from "../../components/ContentModelEditor/Editor";
import { useRouter } from "@webiny/react-router";
import { useCms } from "../../hooks";
import { ContentModelEditorProvider } from "../../components/ContentModelEditor/Context";
import Snackbar from "@webiny/app-admin/ui/views/AdminView/components/Snackbar";
export default function ContentModelEditorView() {
  var _useRouter = useRouter(),
      match = _useRouter.match;

  var _useCms = useCms(),
      apolloClient = _useCms.apolloClient;

  var _ref = match.params,
      modelId = _ref.modelId;

  if (!apolloClient) {
    return null;
  }

  return /*#__PURE__*/React.createElement(ContentModelEditorProvider, {
    key: modelId,
    apolloClient: apolloClient,
    modelId: modelId
  }, /*#__PURE__*/React.createElement(DndProvider, {
    backend: HTML5Backend
  }, /*#__PURE__*/React.createElement(Editor, null), /*#__PURE__*/React.createElement("div", {
    style: {
      zIndex: 10,
      position: "absolute"
    }
  }, /*#__PURE__*/React.createElement(Snackbar, null))));
}
