// @webiny/app-security-admin-users/ui/views/Groups/GroupsDataList.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useCallback, useMemo, useState } from "react";
import orderBy from "lodash/orderBy";
import { i18n } from "@webiny/app/i18n";
import { DataList, ScrollList, ListItem, ListItemText, ListItemTextSecondary, ListItemMeta, ListActions, DataListModalOverlayAction, DataListModalOverlay } from "@webiny/ui/List";
import { DeleteIcon } from "@webiny/ui/List/DataList/icons";
import { useRouter } from "@webiny/react-router";
import { useSnackbar } from "@webiny/app-admin/hooks/useSnackbar";
import { useQuery, useMutation } from "@apollo/react-hooks";
import { useConfirmationDialog } from "@webiny/app-admin/hooks/useConfirmationDialog";
import { LIST_GROUPS, DELETE_GROUP } from "./graphql";
import { Tooltip } from "@webiny/ui/Tooltip";
import { ButtonIcon, ButtonSecondary } from "@webiny/ui/Button";
import { Cell, Grid } from "@webiny/ui/Grid";
import { Select } from "@webiny/ui/Select";
import SearchUI from "@webiny/app-admin/components/SearchUI";
import { ReactComponent as AddIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/add-18px.svg";
import { ReactComponent as FilterIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/filter-24px.svg";
import { deserializeSorters, serializeSorters } from "../utils";
var t = i18n.ns("app-security/admin/groups/data-list");
var SORTERS = [{
  label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Newest to oldest"]))),
  sorters: {
    createdOn: "desc"
  }
}, {
  label: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Oldest to newest"]))),
  sorters: {
    createdOn: "asc"
  }
}, {
  label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Name A-Z"]))),
  sorters: {
    name: "asc"
  }
}, {
  label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Name Z-A"]))),
  sorters: {
    name: "desc"
  }
}];

var GroupsDataList = function GroupsDataList() {
  var _useState = useState(""),
      _useState2 = _slicedToArray(_useState, 2),
      filter = _useState2[0],
      setFilter = _useState2[1];

  var _useState3 = useState(serializeSorters(SORTERS[0].sorters)),
      _useState4 = _slicedToArray(_useState3, 2),
      sort = _useState4[0],
      setSort = _useState4[1];

  var _useRouter = useRouter(),
      history = _useRouter.history,
      location = _useRouter.location;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var _useConfirmationDialo = useConfirmationDialog({
    dataTestId: "default-data-list.delete-dialog"
  }),
      showConfirmation = _useConfirmationDialo.showConfirmation;

  var _useQuery = useQuery(LIST_GROUPS),
      listResponse = _useQuery.data,
      listLoading = _useQuery.loading;

  var _useMutation = useMutation(DELETE_GROUP, {
    refetchQueries: [{
      query: LIST_GROUPS
    }]
  }),
      _useMutation2 = _slicedToArray(_useMutation, 2),
      deleteIt = _useMutation2[0],
      deleteLoading = _useMutation2[1].loading;

  var data = listLoading && !listResponse ? [] : listResponse.security.groups.data;
  var slug = new URLSearchParams(location.search).get("slug");
  var filterGroup = useCallback(function (_ref) {
    var name = _ref.name,
        slug = _ref.slug,
        description = _ref.description;
    return name.toLowerCase().includes(filter) || slug.toLowerCase().includes(filter) || description && description.toLowerCase().includes(filter);
  }, [filter]);
  var sortGroups = useCallback(function (groups) {
    if (!sort) {
      return groups;
    }

    var _Object$entries = Object.entries(deserializeSorters(sort)),
        _Object$entries2 = _slicedToArray(_Object$entries, 1),
        _Object$entries2$ = _slicedToArray(_Object$entries2[0], 2),
        key = _Object$entries2$[0],
        value = _Object$entries2$[1];

    return orderBy(groups, [key], [value]);
  }, [sort]);
  var deleteItem = useCallback(function (item) {
    showConfirmation( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var _yield$deleteIt, data, error;

      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return deleteIt({
                variables: item
              });

            case 2:
              _yield$deleteIt = _context.sent;
              data = _yield$deleteIt.data;
              error = data.security.deleteGroup.error;

              if (!error) {
                _context.next = 7;
                break;
              }

              return _context.abrupt("return", showSnackbar(error.message));

            case 7:
              showSnackbar(t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["Group \"{slug}\" deleted."])))({
                slug: item.slug
              }));

              if (slug === item.slug) {
                history.push("/security/groups");
              }

            case 9:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    })));
  }, [slug]);
  var groupsDataListModalOverlay = useMemo(function () {
    return /*#__PURE__*/React.createElement(DataListModalOverlay, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Select, {
      value: sort,
      onChange: setSort,
      label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Sort by"])))
    }, SORTERS.map(function (_ref3) {
      var label = _ref3.label,
          sorters = _ref3.sorters;
      return /*#__PURE__*/React.createElement("option", {
        key: label,
        value: serializeSorters(sorters)
      }, label);
    })))));
  }, [sort]);
  var filteredData = filter === "" ? data : data.filter(filterGroup);
  var groupList = sortGroups(filteredData);
  return /*#__PURE__*/React.createElement(DataList, {
    title: t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Security Groups"]))),
    actions: /*#__PURE__*/React.createElement(ButtonSecondary, {
      "data-testid": "new-record-button",
      onClick: function onClick() {
        return history.push("/security/groups?new=true");
      }
    }, /*#__PURE__*/React.createElement(ButtonIcon, {
      icon: /*#__PURE__*/React.createElement(AddIcon, null)
    }), " ", t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["New Group"])))),
    data: groupList,
    loading: listLoading || deleteLoading,
    search: /*#__PURE__*/React.createElement(SearchUI, {
      value: filter,
      onChange: setFilter,
      inputPlaceholder: t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Search Groups"])))
    }),
    modalOverlay: groupsDataListModalOverlay,
    modalOverlayAction: /*#__PURE__*/React.createElement(DataListModalOverlayAction, {
      icon: /*#__PURE__*/React.createElement(FilterIcon, null),
      "data-testid": "default-data-list.filter"
    })
  }, function (_ref4) {
    var data = _ref4.data;
    return /*#__PURE__*/React.createElement(ScrollList, {
      "data-testid": "default-data-list"
    }, data.map(function (item) {
      return /*#__PURE__*/React.createElement(ListItem, {
        key: item.slug,
        selected: item.slug === slug
      }, /*#__PURE__*/React.createElement(ListItemText, {
        onClick: function onClick() {
          return history.push("/security/groups?slug=".concat(item.slug));
        }
      }, item.name, /*#__PURE__*/React.createElement(ListItemTextSecondary, null, item.description)), /*#__PURE__*/React.createElement(ListItemMeta, null, /*#__PURE__*/React.createElement(ListActions, null, item.slug !== "full-access" ? /*#__PURE__*/React.createElement(DeleteIcon, {
        onClick: function onClick() {
          return deleteItem(item);
        },
        "data-testid": "default-data-list.delete"
      }) : /*#__PURE__*/React.createElement(Tooltip, {
        placement: "bottom",
        content: /*#__PURE__*/React.createElement("span", null, t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["You can't delete this group."]))))
      }, /*#__PURE__*/React.createElement(DeleteIcon, {
        disabled: true
      })))));
    }));
  });
};

export default GroupsDataList;
