// @webiny/app-admin/ui/views/AdminView/ContentElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import styled from "@emotion/styled";
import { UIElement } from "../../UIElement";
import { UIRenderer } from "../../UIRenderer";
var ContentWrapper = /*#__PURE__*/styled("div", {
  target: "enpnxwx0",
  label: "ContentWrapper"
})({
  width: "100%",
  paddingTop: 67
});

var ContentElementRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(ContentElementRenderer, _UIRenderer);

  var _super = _createSuper(ContentElementRenderer);

  function ContentElementRenderer() {
    _classCallCheck(this, ContentElementRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(ContentElementRenderer, [{
    key: "render",
    value: function render(_ref) {
      var next = _ref.next,
          props = _ref.props;
      return /*#__PURE__*/React.createElement(ContentWrapper, null, props.children ? props.children : next());
    }
  }]);

  return ContentElementRenderer;
}(UIRenderer);

export var ContentElement = /*#__PURE__*/function (_UIElement) {
  _inherits(ContentElement, _UIElement);

  var _super2 = _createSuper(ContentElement);

  function ContentElement(id) {
    var _this;

    _classCallCheck(this, ContentElement);

    _this = _super2.call(this, id);

    _this.useGrid(false);

    _this.addRenderer(new ContentElementRenderer());

    _this.applyPlugins(ContentElement);

    return _this;
  }

  return ContentElement;
}(UIElement);
