// @webiny/app-admin/ui/views/AdminView/HeaderElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { UIElement } from "../../UIElement";
import { TopAppBarPrimary, TopAppBarTitle } from "@webiny/ui/TopAppBar";
import { GenericElement } from "../../elements/GenericElement";
import { PlaceholderElement } from "../../elements/PlaceholderElement";
import { HeaderSectionLeftElement } from "./HeaderSectionLeftElement";
import { HeaderSectionCenterElement } from "./HeaderSectionCenterElement";
import { HeaderSectionRightElement } from "./HeaderSectionRightElement";
import Hamburger from "./components/Hamburger";
import { UIRenderer } from "../../UIRenderer";
var ElementID;

(function (ElementID) {
  ElementID["MenuButton"] = "headerMenuButton";
  ElementID["Logo"] = "headerLogo";
})(ElementID || (ElementID = {}));

var HeaderElementRenderer = /*#__PURE__*/function (_UIRenderer) {
  _inherits(HeaderElementRenderer, _UIRenderer);

  var _super = _createSuper(HeaderElementRenderer);

  function HeaderElementRenderer() {
    _classCallCheck(this, HeaderElementRenderer);

    return _super.apply(this, arguments);
  }

  _createClass(HeaderElementRenderer, [{
    key: "render",
    value: function render(_ref) {
      var next = _ref.next;
      return /*#__PURE__*/React.createElement(TopAppBarPrimary, {
        fixed: true
      }, next());
    }
  }]);

  return HeaderElementRenderer;
}(UIRenderer);

export var HeaderElement = /*#__PURE__*/function (_UIElement) {
  _inherits(HeaderElement, _UIElement);

  var _super2 = _createSuper(HeaderElement);

  function HeaderElement(id) {
    var _this;

    _classCallCheck(this, HeaderElement);

    _this = _super2.call(this, id);

    _this.useGrid(false);

    var leftSection = _this.addElement(new HeaderSectionLeftElement("headerLeft"));

    leftSection.addElement(new PlaceholderElement(ElementID.MenuButton));
    leftSection.addElement(new PlaceholderElement(ElementID.Logo));

    _this.addElement(new HeaderSectionCenterElement("headerCenter"));

    _this.addElement(new HeaderSectionRightElement("headerRight"));

    _this.addRenderer(new HeaderElementRenderer());

    _this.setMenuButton( /*#__PURE__*/React.createElement(Hamburger, null));

    _this.applyPlugins(HeaderElement);

    return _this;
  }

  _createClass(HeaderElement, [{
    key: "getLeftSection",
    value: function getLeftSection() {
      return this.getElement("headerLeft");
    }
  }, {
    key: "getCenterSection",
    value: function getCenterSection() {
      return this.getElement("headerCenter");
    }
  }, {
    key: "getRightSection",
    value: function getRightSection() {
      return this.getElement("headerRight");
    }
  }, {
    key: "setLogo",
    value: function setLogo(logo) {
      // TODO: extract into a `LogoElement` class
      this.getElement(ElementID.Logo).replaceWith(new GenericElement(ElementID.Logo, function () {
        return /*#__PURE__*/React.createElement(TopAppBarTitle, null, logo);
      }));
    }
  }, {
    key: "setMenuButton",
    value: function setMenuButton(menuButton) {
      this.getElement(ElementID.MenuButton).replaceWith(new GenericElement(ElementID.MenuButton, function () {
        return menuButton;
      }));
    }
  }]);

  return HeaderElement;
}(UIElement);
