// @webiny/app-security-admin-users/ui/views/Users/UsersView.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { UIView } from "@webiny/app-admin/ui/UIView";
import { AdminView } from "@webiny/app-admin/ui/views/AdminView";
import { SplitView } from "@webiny/app-admin/ui/views/SplitView";
import { ViewElement } from "@webiny/app-admin/ui/elements/ViewElement";
import { GenericElement } from "@webiny/app-admin/ui/elements/GenericElement";
import { UsersFormView } from "./UsersFormView";
import UsersDataList from "./UsersDataList";
export var UsersView = /*#__PURE__*/function (_UIView) {
  _inherits(UsersView, _UIView);

  var _super = _createSuper(UsersView);

  function UsersView() {
    var _this;

    _classCallCheck(this, UsersView);

    _this = _super.call(this, "UsersView");

    _this.addElements();

    return _this;
  }

  _createClass(UsersView, [{
    key: "setTitle",
    value: function setTitle(title) {
      this.getElement("AdminView").setTitle(title);
    }
  }, {
    key: "addElements",
    value: function addElements() {
      var adminView = this.addElement(new AdminView());
      adminView.setTitle("Security - Users");
      var splitView = new SplitView("adminUsers", {
        leftPanel: new GenericElement("adminUsersLis", function () {
          return /*#__PURE__*/React.createElement(UsersDataList, null);
        }),
        rightPanel: new ViewElement("adminUsersForm", {
          view: new UsersFormView()
        })
      });
      adminView.setContentElement(new ViewElement("adminUsersSplitView", {
        view: splitView
      }));
    }
  }]);

  return UsersView;
}(UIView);
