// @webiny/app-security-admin-users/ui/views/Users/UsersDataList.js
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11;

import _regeneratorRuntime from "@babel/runtime/regenerator";
import React, { useCallback, useMemo, useState } from "react";
import { useMutation, useQuery } from "@apollo/react-hooks";
import orderBy from "lodash/orderBy";
import { i18n } from "@webiny/app/i18n";
import { useSecurity } from "@webiny/app-security";
import { Tooltip } from "@webiny/ui/Tooltip";
import { Image } from "@webiny/app/components";
import { DataList, ScrollList, ListItem, ListItemText, ListItemTextSecondary, ListItemMeta, ListActions, ListItemGraphic, DataListModalOverlayAction, DataListModalOverlay } from "@webiny/ui/List";
import { ButtonIcon, ButtonSecondary } from "@webiny/ui/Button";
import { DeleteIcon } from "@webiny/ui/List/DataList/icons";
import { Avatar } from "@webiny/ui/Avatar";
import { Cell, Grid } from "@webiny/ui/Grid";
import { Select } from "@webiny/ui/Select";
import { useRouter } from "@webiny/react-router";
import { useSnackbar } from "@webiny/app-admin/hooks/useSnackbar";
import { useConfirmationDialog } from "@webiny/app-admin/hooks/useConfirmationDialog";
import SearchUI from "@webiny/app-admin/components/SearchUI";
import { ReactComponent as AddIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/add-18px.svg";
import { ReactComponent as FilterIcon } from "@svgr/webpack!@webiny/app-admin/assets/icons/filter-24px.svg";
import { DELETE_USER, LIST_USERS } from "./graphql";
import { serializeSorters, deserializeSorters } from "../utils";
var t = i18n.ns("app-identity/admin/users/data-list");
var SORTERS = [{
  label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Newest to oldest"]))),
  sorters: {
    createdOn: "desc"
  }
}, {
  label: t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Oldest to newest"]))),
  sorters: {
    createdOn: "asc"
  }
}, {
  label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Login A-Z"]))),
  sorters: {
    login: "asc"
  }
}, {
  label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Login Z-A"]))),
  sorters: {
    login: "desc"
  }
}];

var UsersDataList = function UsersDataList() {
  var _useState = useState(""),
      _useState2 = _slicedToArray(_useState, 2),
      filter = _useState2[0],
      setFilter = _useState2[1];

  var _useState3 = useState(serializeSorters(SORTERS[0].sorters)),
      _useState4 = _slicedToArray(_useState3, 2),
      sort = _useState4[0],
      setSort = _useState4[1];

  var _useSecurity = useSecurity(),
      identity = _useSecurity.identity;

  var _useRouter = useRouter(),
      history = _useRouter.history;

  var _useSnackbar = useSnackbar(),
      showSnackbar = _useSnackbar.showSnackbar;

  var _useConfirmationDialo = useConfirmationDialog({
    dataTestId: "default-data-list.delete-dialog"
  }),
      showConfirmation = _useConfirmationDialo.showConfirmation;

  var filterUsers = useCallback(function (_ref) {
    var login = _ref.login,
        firstName = _ref.firstName,
        lastName = _ref.lastName;
    return login.toLowerCase().includes(filter) || firstName.toLowerCase().includes(filter) || lastName.toLowerCase().includes(filter);
  }, [filter]);
  var sortUsers = useCallback(function (users) {
    if (!sort) {
      return users;
    }

    var _Object$entries = Object.entries(deserializeSorters(sort)),
        _Object$entries2 = _slicedToArray(_Object$entries, 1),
        _Object$entries2$ = _slicedToArray(_Object$entries2[0], 2),
        key = _Object$entries2$[0],
        value = _Object$entries2$[1];

    return orderBy(users, [key], [value]);
  }, [sort]);

  var _useQuery = useQuery(LIST_USERS),
      listUsers = _useQuery.data,
      usersLoading = _useQuery.loading;

  var _useMutation = useMutation(DELETE_USER, {
    refetchQueries: [{
      query: LIST_USERS
    }]
  }),
      _useMutation2 = _slicedToArray(_useMutation, 2),
      deleteIt = _useMutation2[0],
      deleteLoading = _useMutation2[1].loading;

  var data = usersLoading && !listUsers ? [] : listUsers.security.users.data || [];
  var filteredData = filter === "" ? data : data.filter(filterUsers);
  var userList = sortUsers(filteredData);
  var login = new URLSearchParams(location.search).get("login");
  var deleteItem = useCallback(function (item) {
    showConfirmation( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
      var _response$data, _response$data$securi, _response$data$securi2;

      var response, error;
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return deleteIt({
                variables: item
              });

            case 2:
              response = _context.sent;
              error = response === null || response === void 0 ? void 0 : (_response$data = response.data) === null || _response$data === void 0 ? void 0 : (_response$data$securi = _response$data.security) === null || _response$data$securi === void 0 ? void 0 : (_response$data$securi2 = _response$data$securi.deleteUser) === null || _response$data$securi2 === void 0 ? void 0 : _response$data$securi2.error;

              if (!error) {
                _context.next = 6;
                break;
              }

              return _context.abrupt("return", showSnackbar(error.message));

            case 6:
              showSnackbar(t(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["User \"{login}\" deleted."])))({
                login: item.login
              }));

              if (login === item.login) {
                history.push("/security/users");
              }

            case 8:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    })));
  }, [login]);
  var usersDataListModalOverlay = useMemo(function () {
    return /*#__PURE__*/React.createElement(DataListModalOverlay, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
      span: 12
    }, /*#__PURE__*/React.createElement(Select, {
      value: sort,
      onChange: setSort,
      label: t(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["Sort by"])))
    }, SORTERS.map(function (_ref3) {
      var label = _ref3.label,
          sorters = _ref3.sorters;
      return /*#__PURE__*/React.createElement("option", {
        key: label,
        value: serializeSorters(sorters)
      }, label);
    })))));
  }, [sort]);
  var loading = usersLoading || deleteLoading;
  return /*#__PURE__*/React.createElement(DataList, {
    title: t(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["Security Users"]))),
    actions: /*#__PURE__*/React.createElement(ButtonSecondary, {
      "data-testid": "new-record-button",
      onClick: function onClick() {
        return history.push("/security/users?new=true");
      }
    }, /*#__PURE__*/React.createElement(ButtonIcon, {
      icon: /*#__PURE__*/React.createElement(AddIcon, null)
    }), " ", t(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["New User"])))),
    data: userList,
    loading: loading,
    search: /*#__PURE__*/React.createElement(SearchUI, {
      value: filter,
      onChange: setFilter,
      inputPlaceholder: t(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["Search users"])))
    }),
    modalOverlay: usersDataListModalOverlay,
    modalOverlayAction: /*#__PURE__*/React.createElement(DataListModalOverlayAction, {
      icon: /*#__PURE__*/React.createElement(FilterIcon, null),
      "data-testid": "default-data-list.filter"
    })
  }, function (_ref4) {
    var data = _ref4.data;
    return /*#__PURE__*/React.createElement(ScrollList, {
      twoLine: true,
      avatarList: true,
      "data-testid": "default-data-list"
    }, data.map(function (item) {
      return /*#__PURE__*/React.createElement(ListItem, {
        key: item.login,
        selected: item.login === login
      }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Avatar, {
        renderImage: function renderImage(props) {
          return /*#__PURE__*/React.createElement(Image, Object.assign({}, props, {
            transform: {
              width: 100
            }
          }));
        },
        src: item.avatar ? item.avatar.src : item.gravatar,
        fallbackText: item.firstName,
        alt: t(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["User's avatar."])))
      })), /*#__PURE__*/React.createElement(ListItemText, {
        onClick: function onClick() {
          return history.push("/security/users?login=".concat(encodeURIComponent(item.login)));
        }
      }, item.firstName, " ", item.lastName, /*#__PURE__*/React.createElement(ListItemTextSecondary, null, item.login)), /*#__PURE__*/React.createElement(ListItemMeta, null, /*#__PURE__*/React.createElement(ListActions, null, identity && identity.login !== item.login ? /*#__PURE__*/React.createElement(DeleteIcon, {
        onClick: function onClick() {
          return deleteItem(item);
        },
        "data-testid": "default-data-list.delete"
      }) : /*#__PURE__*/React.createElement(Tooltip, {
        placement: "bottom",
        content: /*#__PURE__*/React.createElement("span", null, t(_templateObject11 || (_templateObject11 = _taggedTemplateLiteral(["You can't delete your own user account."]))))
      }, /*#__PURE__*/React.createElement(DeleteIcon, {
        disabled: true
      })))));
    }));
  });
};

export default UsersDataList;
