// @webiny/app-form-builder/admin/plugins/formDetails/previewContent/FormPreview.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import * as React from "react";
import { css } from "emotion";
import { Form } from "../../../../components/Form";
import { DATA_FIELDS } from "../../../../components/Form/graphql";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import CircularProgress from "@webiny/ui/Progress/CircularProgress";
var GET_FORM = gql(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    query FbGetForm($revision: ID!) {\n        formBuilder {\n            getForm(revision: $revision) {\n                data {\n                    ", "\n                }\n                error {\n                    message\n                }\n            }\n        }\n    }\n"])), DATA_FIELDS);
var pageInnerWrapper = /*#__PURE__*/css({
  overflowY: "scroll",
  overflowX: "hidden",
  maxHeight: "calc(100vh - 290px)",
  position: "relative",
  padding: 25,
  backgroundColor: "var(--webiny-theme-color-surface, #fff) !important"
}, "label:pageInnerWrapper;");

var FormPreview = function FormPreview(_ref) {
  var _data$formBuilder, _data$formBuilder$get;

  var revision = _ref.revision;

  var _useQuery = useQuery(GET_FORM, {
    variables: {
      revision: revision.id
    }
  }),
      data = _useQuery.data,
      loading = _useQuery.loading;

  if (loading) {
    return /*#__PURE__*/React.createElement(CircularProgress, null);
  }

  return /*#__PURE__*/React.createElement("div", {
    className: pageInnerWrapper
  }, revision && /*#__PURE__*/React.createElement(Form, {
    preview: true,
    data: data === null || data === void 0 ? void 0 : (_data$formBuilder = data.formBuilder) === null || _data$formBuilder === void 0 ? void 0 : (_data$formBuilder$get = _data$formBuilder.getForm) === null || _data$formBuilder$get === void 0 ? void 0 : _data$formBuilder$get.data
  }));
};

export default FormPreview;
