// @webiny/app-form-builder/admin/plugins/formDetails/previewContent/Header.js
import React from "react";
import { css } from "emotion";
import { Typography } from "@webiny/ui/Typography";
import { Grid, Cell } from "@webiny/ui/Grid";
import { PublishRevision, DeleteRevision, EditRevision, RevisionSelector } from "./HeaderComponents";
var headerTitle = /*#__PURE__*/css({
  "&.mdc-layout-grid": {
    borderBottom: "1px solid var(--mdc-theme-on-background)",
    color: "var(--mdc-theme-text-primary-on-background)",
    background: "var(--mdc-theme-surface)",
    paddingTop: 10,
    paddingBottom: 9,
    ".mdc-layout-grid__inner": {
      alignItems: "center"
    }
  }
}, "label:headerTitle;");
var pageTitle = /*#__PURE__*/css({
  whiteSpace: "nowrap",
  overflow: "hidden",
  textOverflow: "ellipsis"
}, "label:pageTitle;");
var headerActions = /*#__PURE__*/css({
  justifyContent: "flex-end",
  marginRight: "-15px",
  display: "flex",
  alignItems: "center"
}, "label:headerActions;");

var Header = function Header(props) {
  var revision = props.revision;
  return /*#__PURE__*/React.createElement(React.Fragment, null, revision && /*#__PURE__*/React.createElement(Grid, {
    className: headerTitle
  }, /*#__PURE__*/React.createElement(Cell, {
    span: 8,
    className: pageTitle
  }, /*#__PURE__*/React.createElement(Typography, {
    use: "headline5"
  }, revision.name)), /*#__PURE__*/React.createElement(Cell, {
    span: 4,
    className: headerActions
  }, /*#__PURE__*/React.createElement(RevisionSelector, props), /*#__PURE__*/React.createElement(EditRevision, props), /*#__PURE__*/React.createElement(PublishRevision, props), /*#__PURE__*/React.createElement(DeleteRevision, props))));
};

export default Header;
