// @webiny/app-form-builder/admin/plugins/formDetails/previewContent/index.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";

var _templateObject;

import React, { useEffect, useState } from "react";
import { Tab } from "@webiny/ui/Tabs";
import styled from "@emotion/styled";
import { Elevation } from "@webiny/ui/Elevation";
import FormPreview from "./FormPreview";
import Header from "./Header";
import { CircularProgress } from "@webiny/ui/Progress";
import { i18n } from "@webiny/app/i18n";
var t = i18n.namespace("FormsApp.FormDetails.PreviewContent");
var RenderBlock = /*#__PURE__*/styled("div", {
  target: "e1itow6s0",
  label: "RenderBlock"
})({
  position: "relative",
  zIndex: 0,
  backgroundColor: "var(--mdc-theme-background)",
  height: "100%",
  overflow: "auto",
  padding: 25
});

var PreviewContentTab = function PreviewContentTab(props) {
  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      revisionId = _useState2[0],
      setRevisionId = _useState2[1];

  useEffect(function () {
    if (!props.revisions.length) {
      return;
    }

    setRevisionId(props.revisions[0].id);
  }, [props.form.id, props.revisions.length]);
  var revision = props.revisions.find(function (item) {
    return item.id === revisionId;
  });

  if (!revision) {
    return null;
  }

  return /*#__PURE__*/React.createElement(RenderBlock, null, /*#__PURE__*/React.createElement(Elevation, {
    z: 2
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      position: "relative"
    }
  }, props.loading && /*#__PURE__*/React.createElement(CircularProgress, null), /*#__PURE__*/React.createElement(Header, Object.assign({}, props, {
    revision: revision,
    selectRevision: function selectRevision(revision) {
      return setRevisionId(revision.id);
    }
  })), /*#__PURE__*/React.createElement(FormPreview, {
    revision: revision,
    form: props.form
  }))));
};

export default [{
  name: "forms-form-details-revision-content-preview",
  type: "forms-form-details-revision-content",
  render: function render(props) {
    return /*#__PURE__*/React.createElement(Tab, {
      label: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Form preview"]))),
      disabled: props.loading,
      "data-testid": "fb.form-details.tab.form-preview"
    }, /*#__PURE__*/React.createElement(PreviewContentTab, props));
  }
}];
