// @webiny/api-form-builder/plugins/crud/settings.models.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UpdateDataModel = exports.CreateDataModel = void 0;

var _validation = require("@webiny/validation");

var _fields = require("@commodo/fields");

const CreateDataModel = (0, _fields.withFields)({
  installed: (0, _fields.boolean)({
    value: false
  }),
  domain: (0, _fields.string)(),
  reCaptcha: (0, _fields.fields)({
    value: {},
    instanceOf: (0, _fields.withFields)({
      enabled: (0, _fields.boolean)(),
      siteKey: (0, _fields.string)({
        validation: _validation.validation.create("maxLength:100")
      }),
      secretKey: (0, _fields.string)({
        validation: _validation.validation.create("maxLength:100")
      })
    })()
  })
})();
exports.CreateDataModel = CreateDataModel;
const UpdateDataModel = (0, _fields.withFields)({
  domain: (0, _fields.string)(),
  reCaptcha: (0, _fields.fields)({
    value: {},
    instanceOf: (0, _fields.withFields)({
      enabled: (0, _fields.boolean)(),
      siteKey: (0, _fields.string)({
        validation: _validation.validation.create("maxLength:100")
      }),
      secretKey: (0, _fields.string)({
        validation: _validation.validation.create("maxLength:100")
      })
    })()
  })
})();
exports.UpdateDataModel = UpdateDataModel;
