// @webiny/app-page-builder/editor/plugins/toolbar/undoRedo/index.js
import React from "react";
import platform from "platform";
import { useSetRecoilState } from "recoil";
import { ReactComponent as UndoIcon } from "@svgr/webpack!../../../assets/icons/undo-icon.svg";
import { ReactComponent as RedoIcon } from "@svgr/webpack!../../../assets/icons/redo-icon.svg";
import { activeElementAtom } from "../../../recoil/modules";
import Action from "../Action";
import { useEventActionHandler } from "../../../hooks/useEventActionHandler";
var metaKey = platform.os.family === "OS X" ? "CMD" : "CTRL";
export var undo = {
  name: "pb-editor-toolbar-undo",
  type: "pb-editor-toolbar-bottom",
  renderAction: function renderAction() {
    var _useEventActionHandle = useEventActionHandler(),
        undo = _useEventActionHandle.undo;

    var setActiveElementAtomValue = useSetRecoilState(activeElementAtom);

    var _onClick = function onClick() {
      undo();
      setActiveElementAtomValue(null);
    };

    return /*#__PURE__*/React.createElement(Action, {
      id: "action-undo",
      tooltip: "Undo (".concat(metaKey, "+Z)"),
      onClick: function onClick() {
        return _onClick();
      },
      icon: /*#__PURE__*/React.createElement(UndoIcon, null)
    });
  }
};
export var redo = {
  name: "pb-editor-toolbar-redo",
  type: "pb-editor-toolbar-bottom",
  renderAction: function renderAction() {
    var _useEventActionHandle2 = useEventActionHandler(),
        redo = _useEventActionHandle2.redo;

    var setActiveElementAtomValue = useSetRecoilState(activeElementAtom);

    var _onClick2 = function onClick() {
      setActiveElementAtomValue(null);
      redo();
    };

    return /*#__PURE__*/React.createElement(Action, {
      id: "action-redo",
      tooltip: "Redo (".concat(metaKey, "+SHIFT+Z)"),
      onClick: function onClick() {
        return _onClick2();
      },
      icon: /*#__PURE__*/React.createElement(RedoIcon, null)
    });
  }
};
