// @webiny/app-page-builder/editor/plugins/elements/media/youtube/index.js
import React from "react";
import styled from "@emotion/styled";
import { validation } from "@webiny/validation";
import { Typography } from "@webiny/ui/Typography";
import { createEmbedSettingsPlugin, createEmbedPlugin } from "../../utils/oembed";
import YoutubeEmbed from "./YoutubeEmbed";
import placeholder from "./placeholder.png";
import { ReactComponent as LogoIcon } from "@svgr/webpack!./youtube-brands.svg";
import Accordion from "../../../elementSettings/components/Accordion";
import InputField from "../../../elementSettings/components/InputField";
import { SimpleButton } from "../../../elementSettings/components/StyledComponents";
import kebabCase from "lodash/kebabCase";
var PreviewBox = /*#__PURE__*/styled("div", {
  target: "ewlif860",
  label: "PreviewBox"
})({
  textAlign: "center",
  height: 50,
  svg: {
    height: 50,
    width: 50
  }
});
var ButtonContainer = /*#__PURE__*/styled("div", {
  target: "ewlif861",
  label: "ButtonContainer"
})({
  marginTop: 16
});
export default (function () {
  var args = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var elementType = kebabCase(args.elementType || "youtube");
  var defaultToolbar = {
    title: "Youtube",
    group: "pb-editor-element-group-media",
    preview: function preview() {
      return /*#__PURE__*/React.createElement(PreviewBox, null, /*#__PURE__*/React.createElement(LogoIcon, null));
    }
  };
  return [createEmbedPlugin({
    type: elementType,
    toolbar: typeof args.toolbar === "function" ? args.toolbar(defaultToolbar) : defaultToolbar,
    create: args.create,
    settings: args.settings,
    onCreate: "open-settings",
    oembed: {
      renderEmbed: function renderEmbed(props) {
        return /*#__PURE__*/React.createElement(YoutubeEmbed, props);
      }
    },
    renderElementPreview: function renderElementPreview(_ref) {
      var width = _ref.width,
          height = _ref.height;
      return /*#__PURE__*/React.createElement("img", {
        style: {
          width: width,
          height: height
        },
        src: placeholder,
        alt: "Youtube"
      });
    }
  }), createEmbedSettingsPlugin({
    type: elementType,
    render: function render(_ref2) {
      var Bind = _ref2.Bind,
          submit = _ref2.submit;
      return /*#__PURE__*/React.createElement(Accordion, {
        title: "YouTube",
        defaultValue: true
      }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Bind, {
        name: "source.url",
        validators: validation.create("required,url")
      }, /*#__PURE__*/React.createElement(InputField, {
        placeholder: "https://youtube.com/watch?v=4qcDLzu8kVM",
        description: "Enter a video URL"
      })), /*#__PURE__*/React.createElement(ButtonContainer, null, /*#__PURE__*/React.createElement(SimpleButton, {
        onClick: submit
      }, /*#__PURE__*/React.createElement(Typography, {
        use: "caption"
      }, "Save")))));
    }
  })];
});
