// @webiny/api-page-builder/graphql/crud/utils/defaults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _default = {
  db: {
    table: process.env.DB_TABLE_PAGE_BUILDER,
    keys: [{
      primary: true,
      unique: true,
      name: "primary",
      fields: [{
        name: "PK"
      }, {
        name: "SK"
      }]
    }]
  },
  esDb: {
    table: process.env.DB_TABLE_ELASTICSEARCH,
    keys: [{
      primary: true,
      unique: true,
      name: "primary",
      fields: [{
        name: "PK"
      }, {
        name: "SK"
      }]
    }]
  },

  es(context) {
    const tenant = context.tenancy.getCurrentTenant();

    if (!tenant) {
      throw new Error("Tenant missing.");
    }

    const sharedIndex = process.env.ELASTICSEARCH_SHARED_INDEXES === "true";
    const index = `${sharedIndex ? "root" : tenant.id}-page-builder`;
    const prefix = process.env.ELASTIC_SEARCH_INDEX_PREFIX;

    if (prefix) {
      return {
        index: prefix + index
      };
    }

    return {
      index
    };
  }

};
exports.default = _default;
