// @webiny/app-page-builder/admin/utils/createElementPlugin.js
import React from "react";
import cloneDeep from "lodash/cloneDeep";
import { plugins } from "@webiny/plugins";
import Title from "./components/Title";
export default (function (el) {
  var elementPlugins = plugins.byType("pb-editor-page-element");
  var rootPlugin = elementPlugins.find(function (pl) {
    return pl.elementType === el.content.type;
  });

  if (!rootPlugin) {
    return;
  }

  var name = "saved-element-" + el.id;
  plugins.register({
    name: name,
    title: el.name,
    type: "pb-editor-page-element",
    elementType: name,
    target: rootPlugin.target,
    toolbar: {
      title: function title(_ref) {
        var refresh = _ref.refresh;
        return /*#__PURE__*/React.createElement(Title, {
          plugin: name,
          title: el.name,
          id: el.id,
          refresh: refresh
        });
      },
      group: "pb-editor-element-group-saved",
      preview: function preview() {
        return /*#__PURE__*/React.createElement("img", {
          src: el.preview.src,
          alt: el.name,
          style: {
            width: 227,
            height: "auto",
            backgroundColor: "#fff"
          }
        });
      }
    },
    onCreate: "skip",
    settings: rootPlugin ? rootPlugin.settings : [],
    create: function create() {
      return cloneDeep(el.content);
    }
  });
});
