// @webiny/app-page-builder/editor/plugins/blocksCategories/index.js
import React from "react";
import { ReactComponent as GeneralIcon } from "@svgr/webpack!./icons/round-gesture-24px.svg";
import { ReactComponent as CtaIcon } from "@svgr/webpack!./icons/round-notifications_active-24px.svg";
import { ReactComponent as ContentIcon } from "@svgr/webpack!./icons/round-view_quilt-24px.svg";
import { ReactComponent as FeaturesIcon } from "@svgr/webpack!./icons/round-stars-24px.svg";
import { ReactComponent as HeaderIcon } from "@svgr/webpack!./icons/round-home-24px.svg";
import { ReactComponent as TeamIcon } from "@svgr/webpack!./icons/round-group_work-24px.svg";
import { ReactComponent as TestimonialIcon } from "@svgr/webpack!./icons/round-record_voice_over-24px.svg";
var plugins = [{
  type: "pb-editor-block-category",
  name: "pb-editor-block-category-cta",
  title: "Call To Action",
  categoryName: "cta",
  description: "Call to action blocks.",
  icon: /*#__PURE__*/React.createElement(CtaIcon, null)
}, {
  type: "pb-editor-block-category",
  name: "pb-editor-block-category-content",
  title: "Content",
  categoryName: "content",
  description: "Pre-formatted content blocks.",
  icon: /*#__PURE__*/React.createElement(ContentIcon, null)
}, {
  type: "pb-editor-block-category",
  name: "pb-editor-block-category-features",
  title: "Features",
  categoryName: "features",
  description: "Blocks for listing features and benefits.",
  icon: /*#__PURE__*/React.createElement(FeaturesIcon, null)
}, {
  type: "pb-editor-block-category",
  name: "pb-editor-block-category-general",
  title: "General",
  categoryName: "general",
  description: "List of general purpose blocks.",
  icon: /*#__PURE__*/React.createElement(GeneralIcon, null)
}, {
  type: "pb-editor-block-category",
  name: "pb-editor-block-category-header",
  title: "Headers",
  categoryName: "header",
  description: "Page headers.",
  icon: /*#__PURE__*/React.createElement(HeaderIcon, null)
}, {
  type: "pb-editor-block-category",
  name: "pb-editor-block-category-team",
  title: "Team",
  categoryName: "team",
  description: "Blocks to list out your team members.",
  icon: /*#__PURE__*/React.createElement(TeamIcon, null)
}, {
  type: "pb-editor-block-category",
  name: "pb-editor-block-category-testimonial",
  title: "Testimonial",
  categoryName: "testimonial",
  description: "Display comments and user feedback.",
  icon: /*#__PURE__*/React.createElement(TestimonialIcon, null)
}];
export default plugins;
