// @webiny/app-admin/components/SplitView/SplitView.js
import _clone from "lodash/clone";
import * as React from "react";
import classSet from "classnames";
import { Cell, Grid, GridInner } from "@webiny/ui/Grid";
import { css } from "emotion";
import styled from "@emotion/styled";
import { getClasses } from "@webiny/ui/Helpers";
var grid = /*#__PURE__*/css({
  "&.mdc-layout-grid": {
    padding: 0,
    backgroundColor: "var(--mdc-theme-background)"
  }
}, "label:grid;");
var gridInner = /*#__PURE__*/css({
  "&.mdc-layout-grid__inner": {
    gridGap: 0
  }
}, "label:gridInner;");
var RightPanelWrapper = /*#__PURE__*/styled("div", {
  target: "em0s1980",
  label: "RightPanelWrapper"
})({
  backgroundColor: "var(--mdc-theme-background)",
  overflow: "auto",
  height: "calc(100vh - 70px)"
});
var leftPanel = /*#__PURE__*/css({
  backgroundColor: "var(--mdc-theme-surface)",
  ">.webiny-data-list": {
    display: "flex",
    flexDirection: "column",
    height: "calc(100vh - 70px)",
    ".mdc-list": {
      overflow: "auto"
    }
  },
  ">.mdc-list": {
    display: "flex",
    flexDirection: "column",
    maxHeight: "calc(100vh - 70px)",
    overflow: "auto"
  }
}, "label:leftPanel;");

var SplitView = function SplitView(props) {
  return /*#__PURE__*/React.createElement(Grid, {
    className: classSet(grid, props.className, "webiny-split-view")
  }, /*#__PURE__*/React.createElement(GridInner, {
    className: gridInner + " webiny-split-view__inner"
  }, props.children));
};

var LeftPanel = function LeftPanel(props) {
  var propList = _clone(props);

  if (!propList.hasOwnProperty("span")) {
    propList.span = 5;
  }

  return /*#__PURE__*/React.createElement(Cell, getClasses(propList, classSet(leftPanel, props.className, "webiny-split-view__left-panel")), propList.children);
};

var RightPanel = function RightPanel(props) {
  var propList = _clone(props);

  if (!propList.hasOwnProperty("span")) {
    propList.span = 7;
  }

  return /*#__PURE__*/React.createElement(Cell, getClasses(propList, "webiny-split-view__right-panel"), /*#__PURE__*/React.createElement(RightPanelWrapper, {
    className: "webiny-split-view__right-panel-wrapper",
    id: "webiny-split-view-right-panel"
  }, propList.children));
};

export { SplitView, LeftPanel, RightPanel };
