// @webiny/ui/Radio/RadioGroup.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import * as React from "react";
import { webinyRadioTitle } from "./Radio.styles";
import { FormElementMessage } from "../FormElementMessage";

var RadioGroup = /*#__PURE__*/function (_React$Component) {
  _inherits(RadioGroup, _React$Component);

  var _super = _createSuper(RadioGroup);

  function RadioGroup() {
    _classCallCheck(this, RadioGroup);

    return _super.apply(this, arguments);
  }

  _createClass(RadioGroup, [{
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          description = _this$props.description,
          label = _this$props.label,
          validation = _this$props.validation;
      return /*#__PURE__*/React.createElement(React.Fragment, null, label && /*#__PURE__*/React.createElement("div", {
        className: "mdc-text-field-helper-text mdc-text-field-helper-text--persistent " + webinyRadioTitle
      }, label), this.props.children({
        onChange: function onChange(value) {
          return function () {
            return _this.props.onChange && _this.props.onChange(value);
          };
        },
        getValue: function getValue(id) {
          return _this.props.value === id;
        }
      }), validation.isValid === false && /*#__PURE__*/React.createElement(FormElementMessage, {
        error: true
      }, validation.message), validation.isValid !== false && description && /*#__PURE__*/React.createElement(FormElementMessage, null, description));
    }
  }]);

  return RadioGroup;
}(React.Component);

_defineProperty(RadioGroup, "defaultProps", {
  validation: {
    isValid: null
  }
});

export default RadioGroup;
