// @webiny/app-page-builder/editor/plugins/elements/block/Block.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import React from "react";
import BlockContainer from "./BlockContainer";
import ElementAnimation from "../../../../render/components/ElementAnimation";
import styled from "@emotion/styled";
import { ElementRoot } from "../../../../render/components/ElementRoot";
var BlockStyle = /*#__PURE__*/styled("div", {
  target: "ejk4xf50",
  label: "BlockStyle"
})({
  position: "relative",
  color: "#666",
  padding: 5,
  boxSizing: "border-box"
});

var Block = function Block(_ref) {
  var element = _ref.element;
  var id = element.id;
  return /*#__PURE__*/React.createElement(BlockStyle, {
    id: id,
    style: {
      position: "relative"
    }
  }, /*#__PURE__*/React.createElement(ElementAnimation, null, /*#__PURE__*/React.createElement(ElementRoot, {
    element: element
  }, function (_ref2) {
    var elementStyle = _ref2.elementStyle,
        elementAttributes = _ref2.elementAttributes,
        customClasses = _ref2.customClasses,
        combineClassNames = _ref2.combineClassNames;
    return /*#__PURE__*/React.createElement(BlockContainer, {
      elementId: id,
      elementStyle: elementStyle,
      elementAttributes: elementAttributes,
      customClasses: ["webiny-pb-layout-block webiny-pb-base-page-element-style"].concat(_toConsumableArray(customClasses)),
      combineClassNames: combineClassNames
    });
  })));
};

export default /*#__PURE__*/React.memo(Block);
