// @webiny/app-page-builder/editor/plugins/defaultBar/components/Revisions.js
import React from "react";
import { revisionsAtom } from "../../../recoil/modules";
import { css } from "emotion";
import { useRouter } from "@webiny/react-router";
import { Menu, MenuItem } from "@webiny/ui/Menu";
import { ButtonDefault } from "@webiny/ui/Button";
import { Icon } from "@webiny/ui/Icon";
import { Typography } from "@webiny/ui/Typography";
import { useRecoilValue } from "recoil";
import { ReactComponent as DownButton } from "@svgr/webpack!./icons/round-arrow_drop_down-24px.svg";
var buttonStyle = /*#__PURE__*/css({
  "&.mdc-button": {
    color: "var(--mdc-theme-text-primary-on-background) !important"
  }
}, "label:buttonStyle;");
var menuList = /*#__PURE__*/css({
  ".mdc-list-item": {
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "baseline",
    textAlign: "left"
  }
}, "label:menuList;");
var RevisionStatusEnum;

(function (RevisionStatusEnum) {
  RevisionStatusEnum["PUBLISHED"] = "published";
  RevisionStatusEnum["LOCKED"] = "locked";
  RevisionStatusEnum["DRAFT"] = "draft";
})(RevisionStatusEnum || (RevisionStatusEnum = {}));

var getStatus = function getStatus(revision) {
  if (revision.published) {
    return RevisionStatusEnum.PUBLISHED;
  } else if (revision.locked && !revision.published) {
    return RevisionStatusEnum.LOCKED;
  }

  return RevisionStatusEnum.DRAFT;
};

var Revisions = function Revisions() {
  var revisions = useRecoilValue(revisionsAtom);

  var _useRouter = useRouter(),
      history = _useRouter.history;

  return /*#__PURE__*/React.createElement(Menu, {
    className: menuList,
    onSelect: function onSelect(evt) {
      history.push("/page-builder/editor/".concat(revisions[evt.detail.index].id));
    },
    handle: /*#__PURE__*/React.createElement(ButtonDefault, {
      className: buttonStyle
    }, "Revisions ", /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(DownButton, null)
    }))
  }, revisions.map(function (rev) {
    var status = getStatus(rev);
    return /*#__PURE__*/React.createElement(MenuItem, {
      key: rev.id,
      disabled: status !== RevisionStatusEnum.DRAFT
    }, /*#__PURE__*/React.createElement(Typography, {
      use: "body2"
    }, "v", rev.version), /*#__PURE__*/React.createElement(Typography, {
      use: "caption"
    }, "(", status, ") "));
  }));
};

export default Revisions;
