// @webiny/app-page-builder/editor/plugins/defaultBar/components/Styled.js
import styled from "@emotion/styled";
import { css } from "emotion";
export var TitleInputWrapper = /*#__PURE__*/styled("div", {
  target: "e1wv2ted0",
  label: "TitleInputWrapper"
})({
  width: "100%",
  display: "flex",
  alignItems: "center",
  position: "relative",
  "> .mdc-text-field--upgraded": {
    height: 35,
    marginTop: "0 !important",
    paddingLeft: 10,
    paddingRight: 40
  }
});
export var TitleWrapper = /*#__PURE__*/styled("div", {
  target: "e1wv2ted1",
  label: "TitleWrapper"
})({
  height: 50,
  display: "flex",
  alignItems: "baseline",
  justifyContent: "flex-start",
  flexDirection: "column",
  color: "var(--mdc-theme-text-primary-on-background)",
  position: "relative",
  width: "100%",
  marginLeft: 10
});
export var PageTitle = /*#__PURE__*/styled("div", {
  target: "e1wv2ted2",
  label: "PageTitle"
})({
  border: "1px solid transparent",
  fontSize: 20,
  whiteSpace: "nowrap",
  overflow: "hidden",
  textOverflow: "ellipsis",
  width: "100%",
  lineHeight: "120%",
  "&:hover": {
    border: "1px solid var(--mdc-theme-on-background)"
  }
});
export var pageTitleWrapper = /*#__PURE__*/css({
  maxWidth: "calc(100% - 50px)"
}, "label:pageTitleWrapper;");
export var PageVersion = /*#__PURE__*/styled("span", {
  target: "e1wv2ted3",
  label: "PageVersion"
})({
  fontSize: 20,
  color: "var(--mdc-theme-text-secondary-on-background)",
  marginLeft: 5,
  lineHeight: "120%"
});
export var PageMeta = /*#__PURE__*/styled("div", {
  target: "e1wv2ted4",
  label: "PageMeta"
})({
  height: 20,
  margin: "-2px 2px 2px 2px"
});
