// @webiny/api-i18n/plugins/LocalePlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LocalePlugin = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _plugins = require("@webiny/plugins");

class LocalePlugin extends _plugins.Plugin {
  constructor(config) {
    super();
    (0, _defineProperty2.default)(this, "_config", void 0);
    this._config = config;
  }

  async beforeCreate(params) {
    await this.execute("beforeCreate", params);
  }

  async afterCreate(params) {
    await this.execute("afterCreate", params);
  }

  async beforeUpdate(params) {
    await this.execute("beforeUpdate", params);
  }

  async afterUpdate(params) {
    await this.execute("afterUpdate", params);
  }

  async beforeDelete(params) {
    await this.execute("beforeDelete", params);
  }

  async afterDelete(params) {
    await this.execute("afterDelete", params);
  }

  async execute(event, params) {
    if (!this._config[event]) {
      return;
    }

    await this._config[event](params);
  }

}

exports.LocalePlugin = LocalePlugin;
(0, _defineProperty2.default)(LocalePlugin, "type", "i18n.locale");
