// @webiny/ui/Alert/Alert.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["title", "type", "children"];
import * as React from "react";
import { css } from "emotion";
var alertStyles = /*#__PURE__*/css({
  borderLeft: "3px solid red",
  margin: "5px 0 15px 0",
  padding: "2px 0 2px 10px",
  ".webiny-ui-alert__title": {
    fontWeight: 600,
    lineHeight: "150%",
    color: "var(--mdc-theme-on-surface)",
    marginBottom: 5
  },
  ".webiny-ui-alert__message": {
    lineHeight: "100%",
    fontSize: 14,
    color: "var(--mdc-theme-on-surface)"
  },
  "&.webiny-ui-alert--success": {
    borderColor: "#21CEAF"
  },
  "&.webiny-ui-alert--info": {
    borderColor: "#29A9DB"
  },
  "&.webiny-ui-alert--warning": {
    borderColor: "#F45C3C"
  },
  "&.webiny-ui-alert--danger": {
    borderColor: "#E4495C"
  }
}, "label:alertStyles;");

/**
 * Use Alert component to display user's avatar.
 */
var Alert = function Alert(props) {
  var title = props.title,
      type = props.type,
      children = props.children,
      rest = _objectWithoutProperties(props, _excluded);

  return /*#__PURE__*/React.createElement("div", Object.assign({}, rest, {
    className: alertStyles + " webiny-ui-alert webiny-ui-alert--" + type
  }), /*#__PURE__*/React.createElement("p", {
    className: "webiny-ui-alert__title"
  }, title), /*#__PURE__*/React.createElement("p", {
    className: "webiny-ui-alert__message"
  }, children));
};

export { Alert };
