// @webiny/app-form-builder/admin/plugins/editor/defaultBar/FormSettings/FormSettingsStyled.js
import * as React from "react";
import { css } from "emotion";
import styled from "@emotion/styled";
import { Typography } from "@webiny/ui/Typography";
export var listStyle = /*#__PURE__*/css({
  "&.mdc-list": {
    padding: 0,
    backgroundColor: "var(--mdc-theme-surface)"
  }
}, "label:listStyle;");
export var listItem = /*#__PURE__*/css({
  padding: "15px 20px",
  cursor: "pointer",
  borderBottom: "1px solid var(--mdc-theme-background)",
  "&:last-child": {
    borderBottom: "none"
  },
  ".mdc-list-item__graphic": {
    marginRight: 20
  }
}, "label:listItem;");
export var ListItemTitle = /*#__PURE__*/styled("div", {
  target: "ex0qbqf0",
  label: "ListItemTitle"
})({
  fontWeight: 600,
  marginBottom: 5
});
export var Title = /*#__PURE__*/React.createElement(Typography, {
  style: {
    margin: "0 auto",
    color: "var(--mdc-theme-on-surface)"
  },
  use: "headline6"
}, "Form Settings");
export var TitleContent = /*#__PURE__*/styled("div", {
  target: "ex0qbqf1",
  label: "TitleContent"
})({
  display: "flex",
  flexDirection: "column"
});
