// @webiny/app-page-builder/editor/ui/elements/TagsMultiAutocompleteElement.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _createSuper from "@babel/runtime/helpers/createSuper";
import React from "react";
import { FormFieldElement } from "@webiny/app-admin/ui/elements/form/FormFieldElement";
import { TagsMultiAutocomplete } from "../../../admin/components/TagsMultiAutocomplete";
export var TagsMultiAutocompleteElement = /*#__PURE__*/function (_FormFieldElement) {
  _inherits(TagsMultiAutocompleteElement, _FormFieldElement);

  var _super = _createSuper(TagsMultiAutocompleteElement);

  function TagsMultiAutocompleteElement(id, config) {
    var _this;

    _classCallCheck(this, TagsMultiAutocompleteElement);

    _this = _super.call(this, id, config);

    _this.applyPlugins(TagsMultiAutocompleteElement);

    return _this;
  }

  _createClass(TagsMultiAutocompleteElement, [{
    key: "render",
    value: function render(props) {
      var _this2 = this;

      if (!props.formProps) {
        throw Error("TagsMultiAutocompleteElement must be placed inside of a FormElement.");
      }

      var Bind = props.formProps.Bind;
      return /*#__PURE__*/React.createElement(Bind, {
        name: this.getName(),
        validators: this.getValidators(props),
        defaultValue: this.getDefaultValue(props),
        beforeChange: function beforeChange(value, cb) {
          return _this2.onBeforeChange(value, cb);
        },
        afterChange: function afterChange(value, form) {
          return _this2.onAfterChange(value, form);
        }
      }, /*#__PURE__*/React.createElement(TagsMultiAutocomplete, {
        label: this.getLabel(props),
        description: this.getDescription(props)
      }));
    }
  }]);

  return TagsMultiAutocompleteElement;
}(FormFieldElement);
