// @webiny/api-security-admin-users/plugins/UserPlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserPlugin = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _plugins = require("@webiny/plugins");

class UserPlugin extends _plugins.Plugin {
  constructor(config) {
    super();
    (0, _defineProperty2.default)(this, "_config", void 0);
    this._config = config || {};
  }

  beforeCreate(params) {
    return this._execute("beforeCreate", params);
  }

  afterCreate(params) {
    return this._execute("afterCreate", params);
  }

  afterUpdate(params) {
    return this._execute("afterUpdate", params);
  }

  beforeUpdate(params) {
    return this._execute("beforeUpdate", params);
  }

  beforeDelete(params) {
    return this._execute("beforeDelete", params);
  }

  afterDelete(params) {
    return this._execute("afterDelete", params);
  }

  onLogin(params) {
    return this._execute("onLogin", params);
  }

  _execute(callback, params) {
    if (typeof this._config[callback] === "function") {
      return this._config[callback](params);
    }
  }

}

exports.UserPlugin = UserPlugin;
(0, _defineProperty2.default)(UserPlugin, "type", "security.user");
