// @webiny/app-page-builder/admin/plugins/pageDetails/pageRevisions/RevisionsList.js
import React from "react";
import { css } from "emotion";
import { List } from "@webiny/ui/List";
import { Elevation } from "@webiny/ui/Elevation";
import { CircularProgress } from "@webiny/ui/Progress";
import Revision from "./Revision";
var listWrapper = /*#__PURE__*/css({
  margin: 25,
  display: "flex",
  flexDirection: "column",
  overflow: "auto",
  maxHeight: "calc(100vh - 160px)",
  ".mdc-list .mdc-list-item": {
    borderBottom: "1px solid var(--mdc-theme-on-background)"
  },
  ".mdc-list .mdc-list-item:last-child": {
    borderBottom: "none"
  }
}, "label:listWrapper;");

var RevisionsList = function RevisionsList(props) {
  var page = props.page,
      getPageQuery = props.getPageQuery;
  var _page$revisions = page.revisions,
      revisions = _page$revisions === void 0 ? [] : _page$revisions;
  return /*#__PURE__*/React.createElement(Elevation, {
    className: listWrapper,
    z: 2
  }, /*#__PURE__*/React.createElement("div", {
    style: {
      position: "relative"
    }
  }, getPageQuery.loading && /*#__PURE__*/React.createElement(CircularProgress, null), /*#__PURE__*/React.createElement(List, {
    nonInteractive: true,
    twoLine: true
  }, revisions.map(function (revision) {
    return /*#__PURE__*/React.createElement(Revision, Object.assign({}, props, {
      revision: revision,
      key: revision.id
    }));
  }))));
};

export default RevisionsList;
