// @webiny/app-page-builder/admin/plugins/pageDetails/pageRevisions/Revision.js
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import React from "react";
import { css } from "emotion";
import TimeAgo from "timeago-react";
import { ListItem, ListItemText, ListItemTextPrimary, ListItemTextSecondary, ListItemGraphic, ListItemMeta } from "@webiny/ui/List";
import { IconButton } from "@webiny/ui/Button";
import { Icon } from "@webiny/ui/Icon";
import { MenuItem, Menu, MenuDivider } from "@webiny/ui/Menu";
import { ConfirmationDialog } from "@webiny/ui/ConfirmationDialog";
import { Tooltip } from "@webiny/ui/Tooltip";
import { ReactComponent as MoreVerticalIcon } from "@svgr/webpack!../../../assets/more_vert.svg";
import { ReactComponent as LockIcon } from "@svgr/webpack!../../../assets/lock.svg";
import { ReactComponent as BeenHereIcon } from "@svgr/webpack!../../../assets/beenhere.svg";
import { ReactComponent as GestureIcon } from "@svgr/webpack!../../../assets/gesture.svg";
import { useRevisionHandlers } from "./useRevisionHandlers";
import { useConfigureWebsiteUrlDialog } from "../../../hooks/useConfigureWebsiteUrl";
import { usePageBuilderSettings } from "../../../hooks/usePageBuilderSettings";
import { useSiteStatus } from "../../../hooks/useSiteStatus";
import { ReactComponent as AddIcon } from "@svgr/webpack!../../../assets/add.svg";
import { ReactComponent as EditIcon } from "@svgr/webpack!../../../assets/edit.svg";
import { ReactComponent as PublishIcon } from "@svgr/webpack!../../../assets/round-publish-24px.svg";
import { ReactComponent as UnpublishIcon } from "@svgr/webpack!../../../assets/unpublish.svg";
import { ReactComponent as DeleteIcon } from "@svgr/webpack!../../../assets/delete.svg";
import { ReactComponent as PreviewIcon } from "@svgr/webpack!../../../assets/visibility.svg";
import usePermission from "../../../../hooks/usePermission";
var primaryColor = /*#__PURE__*/css({
  color: "var(--mdc-theme-primary)"
}, "label:primaryColor;");
var revisionsMenu = /*#__PURE__*/css({
  width: 250,
  right: -105,
  left: "auto !important"
}, "label:revisionsMenu;");

var getIcon = function getIcon(rev) {
  var published = rev.status === "published";

  switch (true) {
    case rev.locked && !published:
      return {
        icon: /*#__PURE__*/React.createElement(Icon, {
          icon: /*#__PURE__*/React.createElement(LockIcon, null)
        }),
        text: "This revision is locked (it has already been published)"
      };

    case published:
      return {
        icon: /*#__PURE__*/React.createElement(Icon, {
          icon: /*#__PURE__*/React.createElement(BeenHereIcon, null),
          className: primaryColor
        }),
        text: "This revision is currently published!"
      };

    default:
      return {
        icon: /*#__PURE__*/React.createElement(Icon, {
          icon: /*#__PURE__*/React.createElement(GestureIcon, null)
        }),
        text: "This is a draft"
      };
  }
};

var Div = function Div(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/React.createElement("div", null, children);
};

var Revision = function Revision(_ref2) {
  var revision = _ref2.revision,
      page = _ref2.page;

  var _getIcon = getIcon(revision),
      icon = _getIcon.icon,
      tooltipText = _getIcon.text;

  var _usePageBuilderSettin = usePageBuilderSettings(),
      getWebsiteUrl = _usePageBuilderSettin.getWebsiteUrl,
      getPageUrl = _usePageBuilderSettin.getPageUrl;

  var _useSiteStatus = useSiteStatus(getWebsiteUrl()),
      _useSiteStatus2 = _slicedToArray(_useSiteStatus, 2),
      isSiteRunning = _useSiteStatus2[0],
      refreshSiteStatus = _useSiteStatus2[1];

  var _useRevisionHandlers = useRevisionHandlers({
    revision: revision,
    page: page
  }),
      deleteRevision = _useRevisionHandlers.deleteRevision,
      createRevision = _useRevisionHandlers.createRevision,
      publishRevision = _useRevisionHandlers.publishRevision,
      unpublishRevision = _useRevisionHandlers.unpublishRevision,
      editRevision = _useRevisionHandlers.editRevision;

  var _useConfigureWebsiteU = useConfigureWebsiteUrlDialog(getWebsiteUrl(), refreshSiteStatus),
      showConfigureWebsiteUrlDialog = _useConfigureWebsiteU.showConfigureWebsiteUrlDialog;

  var _usePermission = usePermission(),
      canPublish = _usePermission.canPublish,
      canUnpublish = _usePermission.canUnpublish,
      canDelete = _usePermission.canDelete;

  return /*#__PURE__*/React.createElement(ConfirmationDialog, {
    title: "Confirmation required!",
    message: /*#__PURE__*/React.createElement("span", null, "Are you sure you want to delete this revision?")
  }, function (_ref3) {
    var showConfirmation = _ref3.showConfirmation;
    return /*#__PURE__*/React.createElement(ListItem, null, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Tooltip, {
      content: tooltipText,
      placement: "bottom"
    }, icon)), /*#__PURE__*/React.createElement(ListItemText, null, /*#__PURE__*/React.createElement(ListItemTextPrimary, null, revision.title), /*#__PURE__*/React.createElement(ListItemTextSecondary, null, "Last modified ", /*#__PURE__*/React.createElement(TimeAgo, {
      datetime: revision.savedOn
    }), "(#", revision.version, ")")), /*#__PURE__*/React.createElement(ListItemMeta, null, /*#__PURE__*/React.createElement(Menu, {
      handle: /*#__PURE__*/React.createElement(IconButton, {
        icon: /*#__PURE__*/React.createElement(MoreVerticalIcon, null)
      }),
      className: revisionsMenu
    }, /*#__PURE__*/React.createElement(MenuItem, {
      onClick: createRevision
    }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(AddIcon, null)
    })), "New from current"), !revision.locked && /*#__PURE__*/React.createElement(MenuItem, {
      onClick: editRevision
    }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(EditIcon, null)
    })), "Edit"), revision.status !== "published" && canPublish() && /*#__PURE__*/React.createElement(MenuItem, {
      onClick: function onClick() {
        return publishRevision(revision);
      }
    }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(PublishIcon, null)
    })), "Publish"), revision.status === "published" && canUnpublish() && /*#__PURE__*/React.createElement(MenuItem, {
      onClick: function onClick() {
        return unpublishRevision(revision);
      }
    }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(UnpublishIcon, null)
    })), "Unpublish"), /*#__PURE__*/React.createElement(MenuItem, {
      onClick: function onClick() {
        if (isSiteRunning) {
          window.open(getPageUrl(_objectSpread(_objectSpread({}, revision), {}, {
            path: page.path
          })), "_blank", "noopener");
        } else {
          showConfigureWebsiteUrlDialog();
        }
      }
    }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(PreviewIcon, null)
    })), "Preview"), canDelete(page) && !revision.locked && /*#__PURE__*/React.createElement(Div, null, /*#__PURE__*/React.createElement(MenuDivider, null), /*#__PURE__*/React.createElement(MenuItem, {
      onClick: function onClick() {
        return showConfirmation(deleteRevision);
      }
    }, /*#__PURE__*/React.createElement(ListItemGraphic, null, /*#__PURE__*/React.createElement(Icon, {
      icon: /*#__PURE__*/React.createElement(DeleteIcon, null)
    })), "Delete")))));
  });
};

export default Revision;
