// @webiny/app-admin/components/AppInstaller/Sidebar.js
import React, { Fragment } from "react";
import styled from "@emotion/styled";
import { Typography } from "@webiny/ui/Typography";
import classSet from "classnames";
import webinyLogo from "../../assets/images/webiny-orange-logo.svg";
import signInDivider from "./assets/sign-in-divider.svg";
var SidebarWrapper = /*#__PURE__*/styled("div", {
  target: "eu74ttw0",
  label: "SidebarWrapper"
})({});
var Logo = /*#__PURE__*/styled("div", {
  target: "eu74ttw1",
  label: "Logo"
})({
  padding: 15,
  borderBottom: "1px solid var(--mdc-theme-background)",
  img: {
    width: "100px",
    height: "auto"
  }
});
var List = /*#__PURE__*/styled("ul", {
  target: "eu74ttw2",
  label: "List"
})({
  li: {
    display: "flex",
    alignItems: "center",
    padding: "15px 15px",
    ".status": {
      borderRadius: "50%",
      display: "inline-block",
      height: 20,
      width: 20,
      marginRight: 15
    },
    "&.sign-in": {
      display: "block",
      textAlign: "center",
      ".note": {
        display: "inline-block",
        lineHeight: "150%",
        fontSize: 10,
        paddingTop: 10,
        color: "var(--mdc-theme-text-icon-on-background)"
      },
      img: {
        width: "100%"
      }
    },
    "&.active": {
      backgroundColor: "var(--mdc-theme-background)",
      fontWeight: 600,
      ".status": {
        backgroundColor: "var(--mdc-theme-primary)"
      }
    },
    "&.pending": {
      ".status": {
        backgroundColor: "var(--mdc-theme-on-surface)"
      }
    },
    "&.installed": {
      ".status": {
        backgroundColor: "var(--mdc-theme-secondary)"
      }
    }
  }
});
var Note = /*#__PURE__*/styled("div", {
  target: "eu74ttw3",
  label: "Note"
})({
  padding: 15
});

var Installations = function Installations(_ref) {
  var title = _ref.title,
      allInstallers = _ref.allInstallers,
      installer = _ref.installer,
      showLogin = _ref.showLogin;

  var renderList = function renderList() {
    var loginItem = /*#__PURE__*/React.createElement("li", {
      key: "login",
      className: "sign-in"
    }, /*#__PURE__*/React.createElement("img", {
      src: signInDivider
    }), /*#__PURE__*/React.createElement(Typography, {
      use: "overline",
      className: "note"
    }, "TO CONTINUE THE INSTALLATION AFTER THIS POINT YOU\u2019LL NEED TO SIGN IN"));
    var items = [];

    for (var i = 0; i < allInstallers.length; i++) {
      var _allInstallers$i = allInstallers[i],
          name = _allInstallers$i.name,
          _title = _allInstallers$i.title,
          secure = _allInstallers$i.secure,
          installed = _allInstallers$i.installed;
      var prev = i === 0 ? null : allInstallers[i - 1];

      if ((!prev || !prev.secure) && secure && showLogin) {
        items.push(loginItem);
      }

      var active = installer && !showLogin && name === installer.name;
      items.push( /*#__PURE__*/React.createElement("li", {
        key: name,
        className: classSet({
          installed: installed
        }, {
          active: active
        }, {
          pending: !installed && !active
        })
      }, /*#__PURE__*/React.createElement("span", {
        className: "status"
      }), /*#__PURE__*/React.createElement("span", {
        className: "title"
      }, _title)));
    }

    return items;
  };

  return /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement(Note, null, /*#__PURE__*/React.createElement(Typography, {
    use: "body1"
  }, title)), /*#__PURE__*/React.createElement(List, null, renderList()));
};

var Sidebar = function Sidebar(_ref2) {
  var allInstallers = _ref2.allInstallers,
      installer = _ref2.installer,
      showLogin = _ref2.showLogin;
  var upgrades = allInstallers.filter(function (installer) {
    return installer.type === "upgrade";
  });
  var installations = allInstallers.filter(function (installer) {
    return installer.type === "install";
  });
  return /*#__PURE__*/React.createElement(SidebarWrapper, null, /*#__PURE__*/React.createElement(Logo, null, /*#__PURE__*/React.createElement("img", {
    src: webinyLogo,
    alt: "Webiny CMS"
  })), upgrades.length > 0 ? /*#__PURE__*/React.createElement(Installations, {
    title: /*#__PURE__*/React.createElement("span", null, "The following apps will be upgraded to", " ", /*#__PURE__*/React.createElement("strong", null, process.env.REACT_APP_WEBINY_VERSION), ":"),
    allInstallers: upgrades,
    installer: installer,
    showLogin: showLogin
  }) : null, installations.length > 0 && /*#__PURE__*/React.createElement(Installations, {
    title: "The following apps will be installed and configured:",
    allInstallers: installations,
    installer: installer,
    showLogin: upgrades.length > 0 ? false : showLogin
  }));
};

export default Sidebar;
