// @webiny/app-headless-cms/admin/plugins/editor/formSettings/components/GeneralSettings.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { Input } from "@webiny/ui/Input";
import { validation } from "@webiny/validation";
import GroupSelect from "./GroupSelect";

var GeneralSettings = function GeneralSettings(_ref) {
  var Bind = _ref.Bind;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "name"
  }, /*#__PURE__*/React.createElement(Input, {
    label: "Content model name"
  }))), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "modelId"
  }, /*#__PURE__*/React.createElement(Input, {
    disabled: true,
    label: "Content model ID"
  }))), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "description"
  }, /*#__PURE__*/React.createElement(Input, {
    rows: 5,
    label: "Content model description"
  }))), /*#__PURE__*/React.createElement(Cell, {
    span: 12
  }, /*#__PURE__*/React.createElement(Bind, {
    name: "group",
    validators: validation.create("required")
  }, /*#__PURE__*/React.createElement(GroupSelect, null)))));
};

export default GeneralSettings;
