// @webiny/app-headless-cms/admin/contexts/Cms/index.js
import React from "react";
import { useI18N } from "@webiny/app-i18n/hooks/useI18N";
import { CircularProgress } from "@webiny/ui/Progress";
export var CmsContext = /*#__PURE__*/React.createContext(null);
var apolloClientsCache = {};
export function CmsProvider(props) {
  var _useI18N = useI18N(),
      getCurrentLocale = _useI18N.getCurrentLocale;

  var currentLocale = getCurrentLocale("content"); // TODO: not sure why this was necessary :thinking:
  // const hasPermission = identity.getPermission("cms.endpoint.manage");
  // if (!hasPermission) {
  //     return <CmsContext.Provider value={{}} {...props} />;
  // }

  if (currentLocale && !apolloClientsCache[currentLocale]) {
    apolloClientsCache[currentLocale] = props.createApolloClient({
      uri: "".concat(process.env.REACT_APP_API_URL, "/cms/manage/").concat(currentLocale)
    });
  }

  var value = {
    getApolloClient: function getApolloClient(locale) {
      if (!apolloClientsCache[locale]) {
        apolloClientsCache[locale] = props.createApolloClient({
          uri: "".concat(process.env.REACT_APP_API_URL, "/cms/manage/").concat(locale)
        });
      }

      return apolloClientsCache[locale];
    },
    createApolloClient: props.createApolloClient,
    apolloClient: apolloClientsCache[currentLocale]
  };

  if (!currentLocale) {
    return /*#__PURE__*/React.createElement(CircularProgress, null);
  }

  return /*#__PURE__*/React.createElement(CmsContext.Provider, Object.assign({
    value: value
  }, props));
}
