// @webiny/app-form-builder/admin/plugins/editor/triggers/webhook.js
import React from "react";
import { Grid, Cell } from "@webiny/ui/Grid";
import { ButtonPrimary } from "@webiny/ui/Button";
import { ReactComponent as LinkIcon } from "@svgr/webpack!./icons/round-link-24px.svg";
import WebhooksRequestsDynamicFieldset from "./components/WebhooksRequestsDynamicFieldset";
export default {
  type: "form-editor-trigger",
  name: "form-editor-trigger-webhook",
  trigger: {
    id: "webhook",
    title: "Webhook",
    description: "Do a POST HTTP request to a specific URL.",
    icon: /*#__PURE__*/React.createElement(LinkIcon, null),
    renderSettings: function renderSettings(_ref) {
      var Bind = _ref.Bind,
          submit = _ref.submit;
      return /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement(Cell, {
        span: 12
      }, /*#__PURE__*/React.createElement(Bind, {
        name: "urls"
      }, /*#__PURE__*/React.createElement(WebhooksRequestsDynamicFieldset, {
        Bind: Bind,
        title: "Webhook URLs",
        inputLabel: "URL",
        addButtonLabel: "+ Add URL"
      }))), /*#__PURE__*/React.createElement(Cell, null, /*#__PURE__*/React.createElement(ButtonPrimary, {
        onClick: submit
      }, "Save")));
    }
  }
};
