// @webiny/app-page-builder/editor/components/Editor/Sidebar/ElementSettingsTabContent.js
import React from "react";
import styled from "@emotion/styled";
import useElementSettings from "../../../plugins/elementSettings/hooks/useElementSettings";
import AdvancedSettings from "../../../plugins/elementSettings/advanced/AdvancedSettings";
import { COLORS } from "../../../plugins/elementSettings/components/StyledComponents";
import NoActiveElement from "./NoActiveElement";
import { ReactComponent as TouchIcon } from "@svgr/webpack!./icons/touch_app.svg";
var RootElement = /*#__PURE__*/styled("div", {
  target: "e1etcgjt0",
  label: "RootElement"
})({
  height: "calc(100vh - 65px - 48px)",
  // Subtract top-bar and tab-header height
  overflowY: "auto",
  // Style scrollbar
  "&::-webkit-scrollbar": {
    width: 1
  },
  "&::-webkit-scrollbar-track": {
    boxShadow: "inset 0 0 6px rgba(0, 0, 0, 0.3)"
  },
  "&::-webkit-scrollbar-thumb": {
    backgroundColor: "darkgrey",
    outline: "1px solid slategrey"
  }
});
var SidebarActionsWrapper = /*#__PURE__*/styled("div", {
  target: "e1etcgjt1",
  label: "SidebarActionsWrapper"
})({
  display: "flex",
  flexWrap: "wrap",
  borderBottom: "1px solid ".concat(COLORS.gray),
  backgroundColor: COLORS.lightGray,
  borderTop: "1px solid ".concat(COLORS.gray),
  justifyContent: "center"
});

var ElementSettingsTabContent = function ElementSettingsTabContent(_ref) {
  var element = _ref.element;
  var elementSettings = useElementSettings();

  if (!element) {
    return /*#__PURE__*/React.createElement(NoActiveElement, {
      icon: /*#__PURE__*/React.createElement(TouchIcon, null),
      message: "Select an element on the canvas to activate this panel."
    });
  }

  return /*#__PURE__*/React.createElement(RootElement, null, /*#__PURE__*/React.createElement(SidebarActionsWrapper, null, elementSettings.map(function (_ref2, index) {
    var plugin = _ref2.plugin,
        options = _ref2.options;
    return /*#__PURE__*/React.createElement("div", {
      key: plugin.name + "-" + index
    }, typeof plugin.renderAction === "function" && plugin.renderAction({
      options: options
    }));
  })), /*#__PURE__*/React.createElement(AdvancedSettings, null));
};

export default /*#__PURE__*/React.memo(ElementSettingsTabContent);
