// @webiny/app-page-builder/editor/components/Editor/Sidebar/NoActiveElement.js
import React from "react";
import { css } from "emotion";
import { Typography } from "@webiny/ui/Typography";
var noActiveElementWrapper = /*#__PURE__*/css({
  padding: 16,
  display: "flex",
  flexDirection: "column",
  justifyContent: "center",
  alignItems: "center",
  margin: "48px 16px",
  backgroundColor: "var(--mdc-theme-background)",
  color: "var(--mdc-theme-text-primary-on-background)",
  maxHeight: 150,
  "& .icon": {
    fill: "var(--mdc-theme-text-icon-on-background)",
    width: 36,
    height: 36
  },
  "& .text": {
    marginTop: 16,
    color: "var(--mdc-theme-text-on-background)",
    textAlign: "center"
  }
}, "label:noActiveElementWrapper;");

var NoActiveElement = function NoActiveElement(_ref) {
  var message = _ref.message,
      icon = _ref.icon;
  return /*#__PURE__*/React.createElement("div", {
    className: noActiveElementWrapper
  }, icon && /*#__PURE__*/React.cloneElement(icon, {
    className: "icon"
  }), /*#__PURE__*/React.createElement(Typography, {
    use: "subtitle1",
    className: "text"
  }, message));
};

export default /*#__PURE__*/React.memo(NoActiveElement);
