// @webiny/app-page-builder/editor/components/Editor/Sidebar/StyleSettingsTabContent.js
import React from "react";
import styled from "@emotion/styled";
import useElementStyleSettings from "../../../plugins/elementSettings/hooks/useElementStyleSettings";
import NoActiveElement from "./NoActiveElement";
import { ReactComponent as TouchIcon } from "@svgr/webpack!./icons/touch_app.svg";
import { ReactComponent as WarningIcon } from "@svgr/webpack!./icons/warning-black.svg";
var RootElement = /*#__PURE__*/styled("div", {
  target: "e13xon620",
  label: "RootElement"
})({
  height: "calc(100vh - 65px - 48px)",
  // Subtract top-bar and tab-header height
  overflowY: "auto",
  // Style scrollbar
  "&::-webkit-scrollbar": {
    width: 1
  },
  "&::-webkit-scrollbar-track": {
    boxShadow: "inset 0 0 6px rgba(0, 0, 0, 0.3)"
  },
  "&::-webkit-scrollbar-thumb": {
    backgroundColor: "darkgrey",
    outline: "1px solid slategrey"
  }
});

var StyleSettingsTabContent = function StyleSettingsTabContent(_ref) {
  var element = _ref.element;
  var elementStyleSettings = useElementStyleSettings();

  if (!element) {
    return /*#__PURE__*/React.createElement(NoActiveElement, {
      icon: /*#__PURE__*/React.createElement(TouchIcon, null),
      message: "Select an element on the canvas to activate this panel."
    });
  }

  return /*#__PURE__*/React.createElement(RootElement, null, elementStyleSettings.length ? elementStyleSettings.map(function (_ref2, index) {
    var plugin = _ref2.plugin,
        options = _ref2.options;
    return /*#__PURE__*/React.cloneElement(plugin.render({
      options: options
    }), {
      key: index,
      defaultAccordionValue: index === 0
    });
  }) : /*#__PURE__*/React.createElement(NoActiveElement, {
    icon: /*#__PURE__*/React.createElement(WarningIcon, null),
    message: "No style settings found for selected element."
  }));
};

export default /*#__PURE__*/React.memo(StyleSettingsTabContent);
