// @webiny/api-file-manager/plugins/definitions/FilePlugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilePlugin = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _plugins = require("@webiny/plugins");

class FilePlugin extends _plugins.Plugin {
  constructor(params) {
    super();
    (0, _defineProperty2.default)(this, "_params", void 0);
    this._params = params || {};
  }

  async beforeCreate(params) {
    await this._execute("beforeCreate", params);
  }

  async afterCreate(params) {
    await this._execute("afterCreate", params);
  }

  async beforeUpdate(params) {
    await this._execute("beforeUpdate", params);
  }

  async afterUpdate(params) {
    await this._execute("afterUpdate", params);
  }

  async beforeBatchCreate(params) {
    await this._execute("beforeBatchCreate", params);
  }

  async afterBatchCreate(params) {
    await this._execute("afterBatchCreate", params);
  }

  async beforeDelete(params) {
    await this._execute("beforeDelete", params);
  }

  async afterDelete(params) {
    await this._execute("afterDelete", params);
  }

  async _execute(callback, params) {
    if (typeof this._params[callback] !== "function") {
      return;
    }

    await this._params[callback](params);
  }

}

exports.FilePlugin = FilePlugin;
(0, _defineProperty2.default)(FilePlugin, "type", "fm.file");
