// @webiny/app-headless-cms/admin/plugins/editor/defaultBar/CreateContentButton.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

var _templateObject;

import React from "react";
import get from "lodash/get";
import { useRouter } from "@webiny/react-router";
import { i18n } from "@webiny/app/i18n";
import { IconButton } from "@webiny/ui/Button";
import { Tooltip } from "@webiny/ui/Tooltip";
import { ReactComponent as ViewListIcon } from "@svgr/webpack!../../../icons/view_list.svg";
import { GET_CONTENT_MODEL } from "../../../graphql/contentModels";
import { useContentModelEditor } from "../../../components/ContentModelEditor/useContentModelEditor";
var t = i18n.namespace("app-headless-cms/admin/editor/top-bar/save-button");

var CreateContentButton = function CreateContentButton() {
  var router = useRouter();

  var _useContentModelEdito = useContentModelEditor(),
      data = _useContentModelEdito.data,
      apolloClient = _useContentModelEdito.apolloClient;

  var getQuery = apolloClient.readQuery({
    query: GET_CONTENT_MODEL,
    variables: {
      modelId: data.modelId
    }
  });
  var fields = get(getQuery, "getContentModel.data.fields", []);
  var disableViewContent = fields.length === 0;
  var message = disableViewContent ? "To view the content, you first need to add a field and save the form" : "View content";
  return /*#__PURE__*/React.createElement(Tooltip, {
    content: t(_templateObject || (_templateObject = _taggedTemplateLiteral(["{message}"])))({
      message: message
    }),
    placement: "bottom"
  }, /*#__PURE__*/React.createElement(IconButton, {
    icon: /*#__PURE__*/React.createElement(ViewListIcon, null),
    onClick: function onClick() {
      return router.history.push("/cms/content-entries/".concat(data.modelId));
    },
    disabled: disableViewContent
  }));
};

export default CreateContentButton;
