// @webiny/app-page-builder/editor/plugins/elementSettings/background/BackgroundPositionSelector.js
import * as React from "react";
import classNames from "classnames";
import styled from "@emotion/styled";
import { Tooltip } from "@webiny/ui/Tooltip";
import { Typography } from "@webiny/ui/Typography";
var positions = ["top left", "top", "top right", "center left", "center", "center right", "bottom left", "bottom center", "bottom right"];
var Ul = /*#__PURE__*/styled("ul", {
  target: "e1m6p8f20",
  label: "Ul"
})({
  listStyle: "none",
  margin: "0 auto",
  width: 100,
  textAlign: "center",
  "&.disabled": {
    opacity: 0.5,
    pointerEvents: "none"
  },
  li: {
    display: "inline-block",
    border: "1px solid lightgray",
    width: 25,
    height: 25,
    margin: 2,
    cursor: "pointer",
    "&.active": {
      border: "1px solid gray",
      backgroundColor: "var(--mdc-theme-primary)"
    }
  }
});
var PositionWrapper = /*#__PURE__*/styled("div", {
  target: "e1m6p8f21",
  label: "PositionWrapper"
})({
  width: "100%",
  display: "flex"
});

var BackgroundPositionSelector = function BackgroundPositionSelector(props) {
  return /*#__PURE__*/React.createElement(PositionWrapper, null, /*#__PURE__*/React.createElement(Typography, {
    style: {
      width: "60%"
    },
    use: "subtitle2"
  }, "Position"), /*#__PURE__*/React.createElement(Ul, {
    className: classNames({
      disabled: props.disabled
    })
  }, positions.map(function (position) {
    return /*#__PURE__*/React.createElement(Tooltip, {
      key: position,
      content: /*#__PURE__*/React.createElement("span", null, position),
      placement: "top"
    }, /*#__PURE__*/React.createElement("li", {
      onClick: function onClick() {
        return props.onChange(position);
      },
      className: classNames({
        active: props.value === position
      })
    }));
  })));
};

export default BackgroundPositionSelector;
