// @webiny/api-file-manager-ddb-es/upgrades/v5.0.0/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _utils = require("../utils");

var _configurations = _interopRequireDefault(require("../../operations/configurations"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

const plugin = {
  name: "api-upgrade-file-manager",
  type: "api-upgrade",
  app: "file-manager",
  version: "5.0.0",

  async apply({
    elasticsearch,
    fileManager,
    db
  }) {
    const limit = 1000;
    let hasMoreItems = true;
    let after = undefined;
    let esItems = [];

    while (hasMoreItems) {
      const response = await elasticsearch.search({
        index: "root-file-manager",
        body: {
          sort: {
            createdOn: {
              order: "asc",
              // eslint-disable-next-line
              unmapped_type: "date"
            }
          },
          size: limit + 1,
          after
        }
      });
      const {
        hits
      } = response.body.hits;
      hasMoreItems = hits.length > limit;
      after = hasMoreItems ? hits[limit - 1].sort : undefined;
      esItems = [...esItems, ...hits.filter(item => !item._id.includes("T#root#"))];
    }

    console.log(`Fetched ${esItems.length} items from Elasticsearch`);

    if (esItems.length === 0) {
      return;
    } // Store a backup of old items


    const esJSON = JSON.stringify(esItems);
    const {
      file
    } = await fileManager.storage.storagePlugin.upload({
      name: "upgrade-file-manager-es-5.0.0.json",
      type: "application/json",
      size: esJSON.length,
      buffer: Buffer.from(esJSON)
    });
    console.log(`Stored backup of Elasticsearch items to ${file.key}`); // Store items to ES DDB table

    await (0, _utils.paginateBatch)(esItems, 25, async items => {
      await db.batch().create(...items.map(item => _objectSpread(_objectSpread({}, _configurations.default.esDb), {}, {
        data: {
          PK: `T#root#L#${item._source.locale}#FM#F#${item._source.id}`,
          SK: "A",
          index: item._index,
          data: item._source,
          savedOn: new Date().toISOString(),
          version: "5.0.0"
        }
      }))).execute();
    });
    console.log(`Inserted items into Elasticsearch DynamoDB table.`); // Delete original items from ES index

    const operations = esItems.map(item => {
      return {
        delete: {
          _id: item._id,
          _index: item._index
        }
      };
    });
    const {
      body: {
        items,
        errors
      }
    } = await elasticsearch.bulk({
      body: operations,
      // eslint-disable-next-line
      filter_path: "errors,items.*.error"
    });
    console.log(`Deleted old Elasticsearch items from "root-file-manager" index.`);

    if (errors) {
      console.warn("These items were not deleted", items);
    }
  }

};
var _default = plugin; // Target _id: T#root#L#en-US#FM#F#603e248212ee4400089d16eb:A
// const record = {
//     _index: "root-file-manager",
//     _type: "_doc",
//     _id: "6040a6e2a6180e00085d168a",
//     _score: 1.0,
//     _source: {
//         id: "6040a6e2a6180e00085d168a",
//         createdOn: "2021-03-04T09:22:42.029Z",
//         key: "8klunupd0-004.jpg",
//         size: 412963,
//         type: "image/jpeg",
//         name: "8klunupd0-004.jpg",
//         tags: null,
//         createdBy: {
//             id: "admin@webiny.com",
//             displayName: "Pavel Denisjuk",
//             type: "admin"
//         },
//         meta: {
//             private: false
//         },
//         locale: "en-US"
//     }
// };

exports.default = _default;
