// @webiny/app-admin/hooks/useConfirmationDialog.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";

var _templateObject, _templateObject2, _templateObject3, _templateObject4;

var _excluded = ["title", "message"];
import { useUi } from "@webiny/app/hooks/useUi";
import { i18n } from "@webiny/app/i18n";
var t = i18n.ns("app-admin/hooks/use-confirmation-dialog");

var useConfirmationDialog = function useConfirmationDialog() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var title = _ref.title,
      message = _ref.message,
      options = _objectWithoutProperties(_ref, _excluded);

  var ui = useUi();
  return {
    showConfirmation: function showConfirmation(onAccept) {
      var onCancel = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
      ui.setState(function (ui) {
        return _objectSpread(_objectSpread({}, ui), {}, {
          dialog: {
            message: message || t(_templateObject || (_templateObject = _taggedTemplateLiteral(["Are you sure you want to continue?"]))),
            options: _objectSpread(_objectSpread({}, options), {}, {
              title: title || t(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["Confirmation"]))),
              actions: {
                accept: {
                  label: t(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["Confirm"]))),
                  onClick: onAccept
                },
                cancel: {
                  label: t(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["Cancel"]))),
                  onClick: onCancel
                }
              }
            })
          }
        });
      });
    }
  };
};

export { useConfirmationDialog };
