// @webiny/api-page-builder/graphql/graphql/install.gql.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _responses = require("@webiny/handler-graphql/responses");

const plugin = {
  type: "graphql-schema",
  schema: {
    typeDefs:
    /* GraphQL */
    `
            input PbInstallInput {
                websiteUrl: String
                name: String!
            }

            type PbInstallResponse {
                data: Boolean
                error: PbError
            }

            type PbUpgradeResponse {
                data: Boolean
                error: PbError
            }

            extend type PbQuery {
                # Get installed version
                version: String
            }

            extend type PbMutation {
                # Install Page Builder (there are x steps because the process takes a long time).
                install(data: PbInstallInput!): PbInstallResponse

                # Upgrade Page Builder to a newer version
                upgrade(version: String!): PbUpgradeResponse
            }
        `,
    resolvers: {
      PbQuery: {
        version: async (_, args, context) => {
          const {
            tenancy,
            pageBuilder
          } = context;

          if (!tenancy.getCurrentTenant()) {
            return null;
          }

          return pageBuilder.system.getVersion();
        }
      },
      PbMutation: {
        install: async (_, args, context) => {
          try {
            await context.pageBuilder.system.install({
              name: args.data.name,
              insertDemoData: true
            });
            return new _responses.Response(true);
          } catch (e) {
            return new _responses.ErrorResponse(e);
          }
        },
        upgrade: async (_, args, context) => {
          try {
            await context.pageBuilder.system.upgrade(args.version);
            return new _responses.Response(true);
          } catch (e) {
            return new _responses.ErrorResponse(e);
          }
        }
      }
    }
  }
};
var _default = plugin;
exports.default = _default;
